{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import glob, os, yaml, subprocess, itertools, sparse, vcf\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9daf0d89-bc7e-4576-b62e-bcc3282d7f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_models_population_structure_correction(pop_corr, no_pop_corr, plot=False, alpha=0.05, diff_thresh=0):\n",
    "    '''\n",
    "    Comparison to determine if population structure determination is necessary. Things to compare:\n",
    "    \n",
    "    1. Do all the same variants have non-zero coefficients (the returned dataframes are only non-zero coefficients)?\n",
    "    2. Compute Pearson correlation between coefficients.\n",
    "    3. Do features have p-values that are on the same side of the threshold (0.05 or 0.01)?\n",
    "    '''\n",
    "    \n",
    "    # merge outer to include anything (in case some features have 0 or non-zero coefficients, depending on whether PCs are included)\n",
    "    plot_df = pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\"]].merge(no_pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\"]], how=\"outer\", on=\"orig_variant\")    \n",
    "    \n",
    "    # then drop the principal components\n",
    "    plot_df = plot_df.loc[~plot_df[\"orig_variant\"].str.contains(\"PC\")]\n",
    "    \n",
    "    # plot_df_missing = plot_df.loc[(pd.isnull(plot_df[\"coef_x\"])) | (pd.isnull(plot_df[\"coef_y\"]))]\n",
    "    # print(f\"{len(plot_df_missing)} coefficients are zero in one model and non-zero in the other\")\n",
    "    # plot_df_nominal_sig = plot_df.query(\"pval_x < @alpha & pval_y < @alpha\")\n",
    "    # print(\"   \", len(plot_df_nominal_sig.query(\"(coef_x < 0 & coef_y > 0) | (coef_x > 0 & coef_y < 0)\")), \"nominally significant features have conflicting coefficients\")\n",
    "    \n",
    "    plot_df_sig = plot_df.query(\"BH_pval_x < @alpha & BH_pval_y < @alpha\")\n",
    "    print(\"   \", len(plot_df_sig.query(\"(coef_x < 0 & coef_y > 0) | (coef_x > 0 & coef_y < 0)\")), \"significant features have conflicting coefficients\")    \n",
    "    \n",
    "    # NaNs are 0 coefficients, fill them in for plotting and correlation computations\n",
    "    plot_df[[\"coef_x\", \"coef_y\"]] = plot_df[[\"coef_x\", \"coef_y\"]].fillna(0)\n",
    "    \n",
    "    # coefficient without population structure correction - coefficient with correction\n",
    "    plot_df[\"diff\"] = plot_df[\"coef_y\"] - plot_df[\"coef_x\"]\n",
    "    plot_df[\"norm_diff\"] = plot_df[\"diff\"] / plot_df[\"coef_x\"]\n",
    "    \n",
    "    pearson = st.pearsonr(plot_df['coef_x'], plot_df['coef_y'])[0]\n",
    "\n",
    "    if plot:\n",
    "        fig, ax = plt.subplots(1, 2, figsize=(9, 3))\n",
    "        sns.scatterplot(data=plot_df, x=\"coef_x\", y=\"coef_y\", ax=ax[0], alpha=0.5)\n",
    "        ax[0].set_xlabel(\"Coef, Population Structure Correction\")\n",
    "        ax[0].set_ylabel(\"Coef, No Correction\")\n",
    "        ax[0].set_title(np.round(pearson, 4), \"Pearson R\")\n",
    "\n",
    "        sns.scatterplot(data=plot_df, x=\"coef_x\", y=\"norm_diff\", ax=ax[1], alpha=0.5)\n",
    "        ax[1].set_xlabel(\"Coef, Population Structure Correction\")\n",
    "        ax[1].set_ylabel(\"\")\n",
    "        ax[1].set_title(\"Norm. Difference in Coefs.\")\n",
    "\n",
    "        sns.despine()\n",
    "        plt.show()\n",
    "    else:\n",
    "        print(\"   \", np.round(pearson, 4), \"Pearson R\")\n",
    "    \n",
    "    # get features whose coefficients differ by more than some threshold (0 to 100) for percent difference\n",
    "    # (difference between coefficients without and with correction is greater than the coefficient with population structure correction\n",
    "    #large_diff = plot_df.query(\"percent_diff >= @diff_thresh | percent_diff <= -@diff_thresh\")\n",
    "    \n",
    "    # get the number of features with confidence intervals that lie on different sides of 0\n",
    "    print('   ', len(plot_df.query(\"(coef_LB_x > 0 & coef_UB_x > 0 & coef_LB_y < 0 & coef_UB_y < 0) | (coef_LB_x < 0 & coef_UB_x < 0 & coef_LB_y > 0 & coef_UB_y > 0)\")), 'features have confidence intervals on different sides of 0')\n",
    "    #print(f'{len(large_diff.query(\"coef_LB_x > 0 & coef_UB_x > 0 & coef_LB_y < 0 & coef_UB_y < 0\"))} features with very different coefficients have confidence intervals on different sides of 0')\n",
    "    \n",
    "    # TO-DO: COMPARISON OF CONFIDENCE INTERVALS\n",
    "    return pearson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "37b0349c-af23-4a72-b246-9b637512536b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 drugs with phenotypes and genotypes\n"
     ]
    }
   ],
   "source": [
    "genos_dir = '/n/data1/hms/dbmi/farhat/ye12/who/full_genotypes'\n",
    "phenos_dir = '/n/data1/hms/dbmi/farhat/ye12/who/phenotypes'\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/ye12/who/analysis'\n",
    "\n",
    "pheno_drugs = os.listdir(phenos_dir)\n",
    "geno_drugs = os.listdir(genos_dir)\n",
    "\n",
    "drugs_for_analysis = list(set(geno_drugs).intersection(set(pheno_drugs)))\n",
    "print(len(drugs_for_analysis), \"drugs with phenotypes and genotypes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "34fa06c4-f8d0-4f15-893b-12db753c7cee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Delamanid\n",
      "    5 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9787 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Kanamycin\n",
      "    3 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9992 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Ethionamide\n",
      "    1 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9989 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Rifampicin\n",
      "    2 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9995 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Amikacin\n",
      "    2 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9998 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Isoniazid\n",
      "    3 principal components have non-zero coefficients\n",
      "    1 significant features have conflicting coefficients\n",
      "    0.9938 Pearson R\n",
      "    1 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Ethambutol\n",
      "    2 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9921 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Streptomycin\n",
      "    3 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9987 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Capreomycin\n",
      "    2 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9998 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Clofazimine\n",
      "    2 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9945 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Pyrazinamide\n",
      "    3 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9952 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Moxifloxacin\n",
      "    1 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9985 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Levofloxacin\n",
      "    3 principal components have non-zero coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9998 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pearson_df = pd.DataFrame(columns={\"Drug\", \"R\"})\n",
    "for drug in drugs_for_analysis:\n",
    "    \n",
    "    drug = drug.split(\"=\")[1]\n",
    "    out_dir = os.path.join(analysis_dir, drug, \"tiers=1/phenos=WHO\")\n",
    "    corr_prefix = \"dropAF_noSyn\"\n",
    "    no_corr_prefix = \"dropAF_noSyn_noPopCorr\"\n",
    "    \n",
    "    if os.path.isfile(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\")) and os.path.isfile(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\")):\n",
    "        \n",
    "        print(drug)\n",
    "        pop_corr = pd.read_pickle(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\"))\n",
    "        no_pop_corr = pd.read_pickle(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\"))\n",
    "\n",
    "        # number of principal components with positive coefficients (OR > 1)\n",
    "        print(\"   \", len(pop_corr.query(\"coef > 0\").loc[pop_corr.query(\"coef > 0\").orig_variant.str.contains('PC')]), \"principal components have non-zero coefficients\")\n",
    "        \n",
    "        # variants in moxi_res that are not in the results without population structure correction\n",
    "        overlap_variants = set(pop_corr.loc[~pop_corr[\"orig_variant\"].str.contains(\"PC\")][\"orig_variant\"]).intersection(set(no_pop_corr[\"orig_variant\"]))\n",
    "        if len(overlap_variants) == len(pop_corr.loc[~pop_corr[\"orig_variant\"].str.contains(\"PC\")]) == len(no_pop_corr):\n",
    "            pearson = compare_models_population_structure_correction(pop_corr, no_pop_corr, plot=False, alpha=0.05, diff_thresh=0)\n",
    "            pearson_df = pd.concat([pearson_df, pd.DataFrame({\"Drug\": drug, \"R\": pearson}, index=[0])], axis=0)\n",
    "        else:\n",
    "            print(\"    Differing non-zero coefficients\")        \n",
    "        print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2e590a4f-0b52-4338-844c-3522cb5660c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#who_variants.loc[(who_variants.drug == 'PZA') & (who_variants.confidence.str.contains(\"|\".join([\"1\"])))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8a536f2b-7d48-4eec-8b1d-f4a119409765",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Capreomycin</td>\n",
       "      <td>0.9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Amikacin</td>\n",
       "      <td>0.9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Levofloxacin</td>\n",
       "      <td>0.9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>0.9995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Kanamycin</td>\n",
       "      <td>0.9992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ethionamide</td>\n",
       "      <td>0.9989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Streptomycin</td>\n",
       "      <td>0.9987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Moxifloxacin</td>\n",
       "      <td>0.9985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pyrazinamide</td>\n",
       "      <td>0.9952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clofazimine</td>\n",
       "      <td>0.9945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Isoniazid</td>\n",
       "      <td>0.9938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>0.9921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delamanid</td>\n",
       "      <td>0.9787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Drug      R\n",
       "0   Capreomycin 0.9998\n",
       "0      Amikacin 0.9998\n",
       "0  Levofloxacin 0.9998\n",
       "0    Rifampicin 0.9995\n",
       "0     Kanamycin 0.9992\n",
       "0   Ethionamide 0.9989\n",
       "0  Streptomycin 0.9987\n",
       "0  Moxifloxacin 0.9985\n",
       "0  Pyrazinamide 0.9952\n",
       "0   Clofazimine 0.9945\n",
       "0     Isoniazid 0.9938\n",
       "0    Ethambutol 0.9921\n",
       "0     Delamanid 0.9787"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pearson_df.sort_values(\"R\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7726ce8-605b-4e29-aa53-ecdd1f260e37",
   "metadata": {},
   "source": [
    "# 1 variant in INH has positive or negative coefficients in the 2 different models\n",
    "\n",
    "## It's a category 5 mutation, and the confidence intervals are significant\n",
    "\n",
    "## The other 2 variants do not have significant p-values, and the confidence intervals are not fully above or below 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7205f478-ec9e-4d70-a175-c2cf982bde88",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Isoniazid\"\n",
    "out_dir = os.path.join(analysis_dir, drug, \"tiers=1/phenos=WHO\")\n",
    "corr_prefix = \"dropAF_noSyn\"\n",
    "no_corr_prefix = \"dropAF_noSyn_noPopCorr\"\n",
    "\n",
    "pop_corr = pd.read_pickle(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\"))\n",
    "no_pop_corr = pd.read_pickle(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\"))\n",
    "\n",
    "plot_df = pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\", \"confidence_WHO_2021\"]].merge(no_pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\", \"confidence_WHO_2021\"]], how=\"outer\", on=\"orig_variant\")    \n",
    "    \n",
    "# then drop the principal components\n",
    "plot_df = plot_df.loc[~plot_df[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "613bdb38-1bc4-4f12-8ab9-3c1ccb53be54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef_x</th>\n",
       "      <th>BH_pval_x</th>\n",
       "      <th>coef_LB_x</th>\n",
       "      <th>coef_UB_x</th>\n",
       "      <th>coef_y</th>\n",
       "      <th>BH_pval_y</th>\n",
       "      <th>coef_LB_y</th>\n",
       "      <th>coef_UB_y</th>\n",
       "      <th>confidence_WHO_2021_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>katG_p.Arg463Leu</td>\n",
       "      <td>-0.1853</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>-0.2719</td>\n",
       "      <td>-0.0975</td>\n",
       "      <td>0.1827</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1364</td>\n",
       "      <td>0.2290</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          orig_variant  coef_x  BH_pval_x  coef_LB_x  coef_UB_x  coef_y  \\\n",
       "1317  katG_p.Arg463Leu -0.1853     0.0005    -0.2719    -0.0975  0.1827   \n",
       "\n",
       "      BH_pval_y  coef_LB_y  coef_UB_y confidence_WHO_2021_y  \n",
       "1317     0.0000     0.1364     0.2290      5) Not assoc w R  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_cols = [\"orig_variant\", \"coef_x\", \"BH_pval_x\", \"coef_LB_x\", \"coef_UB_x\", \"coef_y\", \"BH_pval_y\", \"coef_LB_y\", \"coef_UB_y\", \"confidence_WHO_2021_y\"]\n",
    "plot_df.query(\"(coef_LB_x > 0 & coef_UB_x > 0 & coef_LB_y < 0 & coef_UB_y < 0) | (coef_LB_x < 0 & coef_UB_x < 0 & coef_LB_y > 0 & coef_UB_y > 0)\")[show_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d51ec463-379d-4747-9408-a1c2245927ad",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
