{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3098ae9b-710b-45e3-9706-922cdc10d3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 150\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "\n",
    "import glob, os, yaml, subprocess, itertools, sparse, vcf\n",
    "from sklearn.model_selection import train_test_split, KFold, GridSearchCV\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "import sklearn.metrics\n",
    "from sklearn.decomposition import PCA\n",
    "import timeit\n",
    "import scipy.stats as st\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ffa238f-4cba-422b-a5ee-af5ce6a40e38",
   "metadata": {},
   "source": [
    "# Numbers of Isolates and Features dropped at different thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5149aab3-19a0-4d35-bd47-501b5dfb70ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17756, 1242)\n",
      "(17613, 1242)\n"
     ]
    }
   ],
   "source": [
    "drug = \"Pyrazinamide\"\n",
    "model_prefix = \"tiers=1_drop_HET_noPopCorrect\"\n",
    "\n",
    "prefilt = pd.read_pickle(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Drugs/{drug}/{model_prefix}/prefilt_matrix.pkl\")\n",
    "filt = pd.read_pickle(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Drugs/{drug}/{model_prefix}/filt_matrix.pkl\")\n",
    "print(prefilt.shape)\n",
    "print(filt.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8bfa7ce6-fc7e-40b4-9bee-62d698371b90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_threshold_summaries(prefilt, missing_thresh=None):\n",
    "    \n",
    "    # drop all isolates with more than 1 missing feature\n",
    "    if missing_thresh is None:\n",
    "        filt_isolates = prefilt.dropna(axis=0, thresh=prefilt.shape[1]-1)\n",
    "    else:\n",
    "        filt_isolates = prefilt.dropna(axis=0, thresh=(1-missing_thresh)*prefilt.shape[1])\n",
    "        \n",
    "    # remove all features with ANYTHING missing\n",
    "    filt_feat = filt_isolates.dropna(axis=1)\n",
    "    print(filt_feat.shape)\n",
    "    print(f\"Dropped {prefilt.shape[0] - filt_isolates.shape[0]} out of {prefilt.shape[0]} isolates and {prefilt.shape[1] - filt_feat.shape[1]} out of {prefilt.shape[1]} features\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "068e1c69-60d9-4920-9590-4f9f68001d6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tiers for this model: ['1']\n",
      "22743 samples with phenotypes for Pyrazinamide\n",
      "Working on dataframe 1/3\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/full_genotypes/drug_name=Pyrazinamide/tier=1/run-1660066839545-part-r-00036\n",
      "Working on dataframe 2/3\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/full_genotypes/drug_name=Pyrazinamide/tier=1/run-1660066839545-part-r-00055\n",
      "Working on dataframe 3/3\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/full_genotypes/drug_name=Pyrazinamide/tier=1/run-1660066839545-part-r-00046\n"
     ]
    }
   ],
   "source": [
    "kwargs = yaml.safe_load(open(\"config.yaml\"))\n",
    "\n",
    "tiers_lst = kwargs[\"tiers_lst\"]\n",
    "drug = kwargs[\"drug\"]\n",
    "out_dir = kwargs[\"out_dir\"]\n",
    "model_prefix = kwargs[\"model_prefix\"]\n",
    "missing_thresh = kwargs[\"missing_thresh\"]\n",
    "het_mode = kwargs[\"het_mode\"]\n",
    "af_thresh = kwargs[\"AF_thresh\"]\n",
    "include_synonymous = kwargs[\"include_synonymous\"]\n",
    "\n",
    "print(\"Tiers for this model:\", tiers_lst)\n",
    "\n",
    "if not os.path.isdir(os.path.join(out_dir, drug)):\n",
    "    print(f\"Creating output directory {os.path.join(out_dir, drug)}\")\n",
    "    os.mkdir(os.path.join(out_dir, drug))\n",
    "    \n",
    "if not os.path.isdir(os.path.join(out_dir, drug, model_prefix)):\n",
    "    print(f\"Creating output directory {os.path.join(out_dir, drug, model_prefix)}\")\n",
    "    os.mkdir(os.path.join(out_dir, drug, model_prefix))\n",
    "\n",
    "genos_dir = '/n/data1/hms/dbmi/farhat/ye12/who/full_genotypes'\n",
    "phenos_dir = '/n/data1/hms/dbmi/farhat/ye12/who/phenotypes'\n",
    "phenos_dir = os.path.join(phenos_dir, f\"drug_name={drug}\")\n",
    "\n",
    "\n",
    "############# STEP 1: GET ALL AVAILABLE PHENOTYPES #############\n",
    "\n",
    "\n",
    "dfs_list_phenos = []\n",
    "\n",
    "for fName in os.listdir(phenos_dir):\n",
    "    dfs_list_phenos.append(pd.read_csv(os.path.join(phenos_dir, fName)))\n",
    "\n",
    "df_phenos = pd.concat(dfs_list_phenos)\n",
    "\n",
    "# check that there are no duplicated phenotypes\n",
    "assert len(df_phenos) == len(df_phenos.sample_id.unique())\n",
    "\n",
    "# check that there is resistance data for all samples\n",
    "assert sum(pd.isnull(df_phenos.phenotype)) == 0\n",
    "    \n",
    "print(f\"{len(df_phenos)} samples with phenotypes for {drug}\")\n",
    "\n",
    "\n",
    "############# STEP 2: GET ALL AVAILABLE GENOTYPES #############\n",
    "\n",
    "\n",
    "# first get all the genotype files associated with the drug\n",
    "geno_files = []\n",
    "\n",
    "for subdir in os.listdir(os.path.join(genos_dir, f\"drug_name={drug}\")):\n",
    "    \n",
    "    # subdirectory (tiers)\n",
    "    full_subdir = os.path.join(genos_dir, f\"drug_name={drug}\", subdir)\n",
    "\n",
    "    # the last character is the tier number\n",
    "    if subdir[-1] in tiers_lst:\n",
    "        \n",
    "        for fName in os.listdir(full_subdir):\n",
    "            \n",
    "            # some hidden files (i.e. Git files) are present, so ignore them\n",
    "            if fName[0] != \".\":\n",
    "                geno_files.append(os.path.join(full_subdir, fName))\n",
    "          \n",
    "        \n",
    "dfs_lst = []\n",
    "for i, fName in enumerate(geno_files):\n",
    "        \n",
    "    print(f\"Working on dataframe {i+1}/{len(geno_files)}\")\n",
    "    print(fName)\n",
    "\n",
    "    # read in the dataframe\n",
    "    df = pd.read_csv(fName)\n",
    "\n",
    "    df_avail_isolates = df.loc[df.sample_id.isin(df_phenos.sample_id)]\n",
    "    \n",
    "    # keep all variants\n",
    "    if include_synonymous:\n",
    "        dfs_lst.append(df_avail_isolates)\n",
    "    else:\n",
    "        # keep only 1) noncoding variants and 2) non-synonymous variants in coding regions. \n",
    "        # P = coding variants, C = synonymous or upstream variants (get only upstream variants by getting only negative positions), and N = non-coding variants on rrs/rrl\n",
    "        dfs_lst.append(df_avail_isolates.loc[((df_avail_isolates.category.astype(str).str[0] == 'c') & (df_avail_isolates.category.str.contains('-'))) | \n",
    "                                             (df_avail_isolates.category.astype(str).str[0].isin(['n', 'p']))])\n",
    "        \n",
    "df_model = pd.concat(dfs_lst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "421217f0-9a8e-4f8c-9553-22b3daab9c42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_allele_frequency</th>\n",
       "      <th>variant_binary_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57075</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Val130Met</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64839</td>\n",
       "      <td>pncA</td>\n",
       "      <td>c.-363_-37del</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57075</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.His57Gln</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>164473</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Val155Leu</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>64839</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.His137Arg</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19307760</th>\n",
       "      <td>139790</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Val180Phe</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19307761</th>\n",
       "      <td>55356</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Arg148Cys</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19307763</th>\n",
       "      <td>55356</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Asn11Tyr</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19307764</th>\n",
       "      <td>139790</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.His82Arg</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19307765</th>\n",
       "      <td>55356</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Arg154Trp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9537291 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample_id gene_symbol       category  variant_allele_frequency  \\\n",
       "1             57075        pncA    p.Val130Met                       0.0   \n",
       "2             64839        pncA  c.-363_-37del                       0.0   \n",
       "3             57075        pncA     p.His57Gln                       0.0   \n",
       "4            164473        pncA    p.Val155Leu                       0.0   \n",
       "6             64839        pncA    p.His137Arg                       0.0   \n",
       "...             ...         ...            ...                       ...   \n",
       "19307760     139790        pncA    p.Val180Phe                       0.0   \n",
       "19307761      55356        pncA    p.Arg148Cys                       0.0   \n",
       "19307763      55356        pncA     p.Asn11Tyr                       0.0   \n",
       "19307764     139790        pncA     p.His82Arg                       0.0   \n",
       "19307765      55356        pncA    p.Arg154Trp                       0.0   \n",
       "\n",
       "          variant_binary_status  \n",
       "1                           0.0  \n",
       "2                           0.0  \n",
       "3                           0.0  \n",
       "4                           0.0  \n",
       "6                           0.0  \n",
       "...                         ...  \n",
       "19307760                    0.0  \n",
       "19307761                    0.0  \n",
       "19307763                    0.0  \n",
       "19307764                    0.0  \n",
       "19307765                    0.0  \n",
       "\n",
       "[9537291 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_lst[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "950ece64-e698-4a58-bb72-867454574c93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['WHO_current'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_phenos.category.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "23f94b88-9345-4f44-a8b6-42cf901d2cb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_allele_frequency</th>\n",
       "      <th>variant_binary_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57075</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Val130Met</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64839</td>\n",
       "      <td>pncA</td>\n",
       "      <td>c.-363_-37del</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57075</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.His57Gln</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>164473</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.Val155Leu</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>64839</td>\n",
       "      <td>pncA</td>\n",
       "      <td>p.His137Arg</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5605235</th>\n",
       "      <td>164459</td>\n",
       "      <td>panD</td>\n",
       "      <td>c.-400_-35del</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5605237</th>\n",
       "      <td>164459</td>\n",
       "      <td>panD</td>\n",
       "      <td>c.-588_-25del</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5605239</th>\n",
       "      <td>164459</td>\n",
       "      <td>panD</td>\n",
       "      <td>p.Thr92Met</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5605243</th>\n",
       "      <td>164459</td>\n",
       "      <td>panD</td>\n",
       "      <td>p.Lys7_Val106del</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5605245</th>\n",
       "      <td>164459</td>\n",
       "      <td>panD</td>\n",
       "      <td>p.Thr14Ala</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22307562 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_id gene_symbol          category  variant_allele_frequency  \\\n",
       "1            57075        pncA       p.Val130Met                       0.0   \n",
       "2            64839        pncA     c.-363_-37del                       0.0   \n",
       "3            57075        pncA        p.His57Gln                       0.0   \n",
       "4           164473        pncA       p.Val155Leu                       0.0   \n",
       "6            64839        pncA       p.His137Arg                       0.0   \n",
       "...            ...         ...               ...                       ...   \n",
       "5605235     164459        panD     c.-400_-35del                       0.0   \n",
       "5605237     164459        panD     c.-588_-25del                       0.0   \n",
       "5605239     164459        panD        p.Thr92Met                       0.0   \n",
       "5605243     164459        panD  p.Lys7_Val106del                       0.0   \n",
       "5605245     164459        panD        p.Thr14Ala                       0.0   \n",
       "\n",
       "         variant_binary_status  \n",
       "1                          0.0  \n",
       "2                          0.0  \n",
       "3                          0.0  \n",
       "4                          0.0  \n",
       "6                          0.0  \n",
       "...                        ...  \n",
       "5605235                    0.0  \n",
       "5605237                    0.0  \n",
       "5605239                    0.0  \n",
       "5605243                    0.0  \n",
       "5605245                    0.0  \n",
       "\n",
       "[22307562 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "90b8605e-d691-4a8d-9c16-24e7a04dd925",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17961"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_model.sample_id.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "90a28443-4bf8-443a-a8cc-a28cc39a23bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_df = pd.DataFrame(variants).rename(columns={0:\"category\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "827d0e72-190b-4622-afc4-c9ccb6fc78d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.Asp129_Val131del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>p.Asp12_Phe13insAlaSerGluProProSerGlnLysSer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>p.Asp129_Val131dup</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>p.Gly24_Ile133del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>p.Thr22_Thr160del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131</th>\n",
       "      <td>p.Glu522_Thr660del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1135</th>\n",
       "      <td>p.Ser546_Pro772del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1191</th>\n",
       "      <td>p.Lys7_Val106del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>p.Arg99_Thr100insSer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229</th>\n",
       "      <td>p.Leu2_Ala110delinsPhe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>227 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         category\n",
       "12                             p.Asp129_Val131del\n",
       "16    p.Asp12_Phe13insAlaSerGluProProSerGlnLysSer\n",
       "20                             p.Asp129_Val131dup\n",
       "45                              p.Gly24_Ile133del\n",
       "51                              p.Thr22_Thr160del\n",
       "...                                           ...\n",
       "1131                           p.Glu522_Thr660del\n",
       "1135                           p.Ser546_Pro772del\n",
       "1191                             p.Lys7_Val106del\n",
       "1201                         p.Arg99_Thr100insSer\n",
       "1229                       p.Leu2_Ala110delinsPhe\n",
       "\n",
       "[227 rows x 1 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_df.loc[(variants_df.category.astype(str).str[0] == 'p') & (variants_df.category.str.contains(\"_\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3546ef8-db00-4576-bb67-83169167af57",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "480ed603-4eb3-436c-ab51-ba91311d06e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = df_model.category.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0e4d90f0-9f36-4402-a078-acea443bb75e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_allele_frequency</th>\n",
       "      <th>variant_binary_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample_id, gene_symbol, category, variant_allele_frequency, variant_binary_status]\n",
       "Index: []"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model.loc[df_model.gene_symbol.str.contains(\"lof\", case=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f382abbd-e797-4edf-8459-8280c13c3f28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_allele_frequency</th>\n",
       "      <th>variant_binary_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample_id, gene_symbol, category, variant_allele_frequency, variant_binary_status]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model.loc[df_model.category.str.contains(\"fs\", case=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6356f6d-87dd-4ddc-8ace-8ac186a4c645",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c0d3e237-c695-42d8-a905-0de89311c460",
   "metadata": {},
   "source": [
    "# Unused Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e78e87f9-cafc-432f-830b-78b2de47ad7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_null_regression(var_idx, num_bootstrap=100):\n",
    "        \n",
    "    # make variant of interest all 0 so that its coefficient will be 0\n",
    "    df = model_inputs.copy()\n",
    "    if \"sample_id\" in df.columns:\n",
    "        df = df.set_index(\"sample_id\")\n",
    "    df.iloc[:, var_idx] = 0\n",
    "    \n",
    "    # concatenate the eigenvectors to the matrix\n",
    "    X = np.concatenate([df.values, eigenvec_df.values], axis=1)\n",
    "\n",
    "    # scale inputs\n",
    "    scaler = StandardScaler()\n",
    "    X = scaler.fit_transform(X)\n",
    "    assert len(y) == X.shape[0]\n",
    "    \n",
    "    coef_lst = []\n",
    "    for _ in range(num_bootstrap):\n",
    "        null_model = LogisticRegressionCV(Cs=1/np.logspace(-4, 4, 9), \n",
    "                                 cv=5,\n",
    "                                 penalty='l2', \n",
    "                                 max_iter=10000, \n",
    "                                 multi_class='ovr',\n",
    "                                 scoring='neg_log_loss'\n",
    "                                )\n",
    "\n",
    "        null_model.fit(X, y)\n",
    "        coef_lst.append(np.squeeze(null_model.coef_)[var_idx])\n",
    "    return coef_lst\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1405a411-cafb-4842-b754-afa8c8f5ec06",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pvalues_add_ci(baseline_df, bs_df, col, ci=95):\n",
    "    \n",
    "    alpha = (100-ci)/100\n",
    "    baseline_df[col] = baseline_df[col].astype(str)\n",
    "    pvals = []\n",
    "    for i, pos in enumerate(baseline_df[col].values):\n",
    "        \n",
    "        # get the percentile of 0\n",
    "        ecdf = ECDF(bs_df[pos].values)\n",
    "        zero_percentile = ecdf(0)\n",
    "        \n",
    "        # if the percentile is greater than 0.5, then the hypothesis test is for probability of lying in the upper tail (i.e. > 97.5 for a 95% CI)\n",
    "        # if the percentile i less than 0.5, then the hypothesis test is for probability of lying in the lower tail (i.e. < 2.5 for a 95% CI)\n",
    "        if zero_percentile > 1 or zero_percentile < 0:\n",
    "            print(pos)\n",
    "        \n",
    "        if zero_percentile < 0.5:\n",
    "            pvals.append(2 * zero_percentile)\n",
    "        else:\n",
    "            pvals.append(2 * (1 - zero_percentile))\n",
    "            \n",
    "        # add confidence intervals\n",
    "        diff = (100-ci)/2\n",
    "        lower, upper = np.percentile(bs_df[pos].values, q=(diff, 100-diff))\n",
    "        baseline_df.loc[i, \"Lower_CI\"] = lower\n",
    "        baseline_df.loc[i, \"Upper_CI\"] = upper\n",
    "        \n",
    "    pvals = np.array(pvals)\n",
    "    return pvals, alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f72ee2d3-4baa-4dfc-8be7-2996974753d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def volcano_plot(baseline_df, bs_df, col, pvals, ci=95, exclude_lst=[], label_thresh=0, title_suffix=\"\", saveFig=None):\n",
    "        \n",
    "    # replace 0s so that log can be taken for plotting\n",
    "    if len(np.unique(pvals)) > 1:\n",
    "        min_non_zero = np.sort(np.unique(pvals))[1]\n",
    "    else:\n",
    "        min_non_zero = 1e-3\n",
    "    pvals[pvals == 0] = min_non_zero\n",
    "    \n",
    "    baseline_df[\"neg_log_pval\"] = -np.log(pvals)\n",
    "    baseline_df[\"pval_significant\"] = (baseline_df[\"pval\"] < alpha).astype(int)\n",
    "    \n",
    "    # verify input order\n",
    "    assert len(bs_df.columns) == len(set(bs_df.columns).intersection(baseline_df[col].values))\n",
    "    \n",
    "    # only plot features with nonzero coefficients\n",
    "    bs_df_positive = baseline_df.query(\"coef > 0\")\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "    sns.scatterplot(data=baseline_df.loc[~baseline_df[col].isin(exclude_lst)], x=\"coef\", y=\"neg_log_pval\", \n",
    "                    alpha=0.75,\n",
    "                    hue=\"pval_significant\", \n",
    "                    s=30,\n",
    "                    palette={1:\"darkred\", 0:\"lightgray\"},\n",
    "                    ax=ax)\n",
    "\n",
    "    ax.legend().set_visible(False)\n",
    "    #plt.xscale(\"log\")\n",
    "    plt.title(\"Ridge Regression Coefficients and Bootstrapped\\np-values\" + title_suffix + f\", α = {alpha}\")\n",
    "    sns.despine()\n",
    "    \n",
    "    if saveFig is not None:\n",
    "        plt.savefig(saveFig, dpi=300, bbox_inches=\"tight\")\n",
    "    else:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74141da4-cced-488e-b452-46eb6c7869cd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
