{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "86d79dbc-86b7-4a6d-b26c-3052aaeb12a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 150\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "\n",
    "import glob, os, yaml, subprocess, itertools, sparse, vcf\n",
    "from sklearn.model_selection import train_test_split, KFold, GridSearchCV\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "import sklearn.metrics\n",
    "from sklearn.decomposition import PCA\n",
    "import timeit\n",
    "import scipy.stats as st\n",
    "import statsmodels.api as sm\n",
    "import pickle, yaml, tracemalloc\n",
    "from scipy.stats import binomtest\n",
    "\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue'\n",
    "input_data_dir = '/n/data1/hms/dbmi/farhat/ye12/who'\n",
    "who_variants_combined = pd.read_csv(\"who_confidence_2021.csv\")\n",
    "\n",
    "import itertools\n",
    "from stats_utils import *\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07de7321-bf6a-40ca-9cd5-51a5df1b16c0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# RIFAMPICIN TEST CASE\n",
    "\n",
    "# 0. Original Analysis: 137 new significant resistance-associated variants\n",
    "\n",
    "## But, there are 152 with OR > 1 and BH p-val < 0.01 that are in tier 2 genes (rpoB is the only tier 1)\n",
    "## These 152 mutations were studied in the next two analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d46b8cbd-3e49-4b4d-94fc-fb3f944f4ab4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "152 significant tier 2 mutations associated with WHO resistance\n",
      "(137, 40)\n"
     ]
    }
   ],
   "source": [
    "folder = \"BINARY\"\n",
    "phenos_name = \"WHO\"\n",
    "drug = \"Rifampicin\"\n",
    "drug_WHO_abbr = \"RIF\"\n",
    "\n",
    "# get the number of high confidence resistance-associated \n",
    "high_Conf_variants = who_variants_combined.loc[(who_variants_combined[\"drug\"]==drug_WHO_abbr) & \n",
    "                                               (who_variants_combined[\"confidence\"].str.contains(\"|\".join([\"1\", \"2\"])))\n",
    "                                              ].mutation.unique()\n",
    "\n",
    "# these have Odds_Ratio > 1, BH pval < 0.01, and are not in the Tier 1 analysis (for RIF, it means they are not in rpoB)\n",
    "# DIDN'T THRESHOLD USING THE UNIVARIATE STATS BECAUSE THOSE WILL BE COMPUTED AT THE VERY END AND SHOULDN'T BE INVOLVED IN MAKING DECISIONS ABOUT THE REGRESSION ANALYSIS\n",
    "tier2_mutations_of_interest = get_tier2_mutations_of_interest(analysis_dir, drug, phenos_name)\n",
    "\n",
    "og_model_analysis = pd.read_excel(\"../results/BINARY/Rifampicin.xlsx\", sheet_name=\"Model_7\")\n",
    "print(og_model_analysis.query(\"Odds_Ratio > 1 & BH_pval < 0.01 & PPV_LB > 0.25 & TP >= 5 & ~mutation.str.contains('rpoB')\").shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b973a2c2-a4a3-4b8f-bfeb-e48bb769d19b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>Total_Isolates</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>TN</th>\n",
       "      <th>FN</th>\n",
       "      <th>PPV_LB</th>\n",
       "      <th>PPV</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>NPV</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>1.215757</td>\n",
       "      <td>1.142315</td>\n",
       "      <td>1.29766</td>\n",
       "      <td>1.500060e-09</td>\n",
       "      <td>2.883625e-08</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>2987</td>\n",
       "      <td>35401</td>\n",
       "      <td>1892</td>\n",
       "      <td>1095</td>\n",
       "      <td>22769</td>\n",
       "      <td>9645</td>\n",
       "      <td>0.615839</td>\n",
       "      <td>0.633411</td>\n",
       "      <td>0.650722</td>\n",
       "      <td>0.702443</td>\n",
       "      <td>0.163994</td>\n",
       "      <td>0.954115</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation  Odds_Ratio     OR_LB    OR_UB          pval  \\\n",
       "16  rpoC_p.Glu1092Asp    1.215757  1.142315  1.29766  1.500060e-09   \n",
       "\n",
       "         BH_pval  Bonferroni_pval  Num_Isolates  Total_Isolates    TP    FP  \\\n",
       "16  2.883625e-08         0.000005          2987           35401  1892  1095   \n",
       "\n",
       "       TN    FN    PPV_LB       PPV    PPV_UB       NPV      Sens      Spec  \n",
       "16  22769  9645  0.615839  0.633411  0.650722  0.702443  0.163994  0.954115  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_model_analysis.query(\"mutation=='rpoC_p.Glu1092Asp'\")[['mutation', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval',\n",
    "       'Bonferroni_pval', 'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN',\n",
    "       'FN', 'PPV_LB', 'PPV', 'PPV_UB', 'NPV', 'Sens', 'Spec']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "398cff6c-2ad5-4976-a441-74e8df7b7760",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3609, 40)"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_model_analysis.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5ac0441-6a40-4c56-98fd-a46ae44e6e8e",
   "metadata": {},
   "source": [
    "# 1. Excluded Samples Analysis: 54 new significant resistance-associated variants\n",
    "\n",
    "Removed ~5,000 samples that contain both high confidence rpoB mutations (Category 1 or 2 in 2021 catalogue) and any of the ~150 significant tier2 mutations. \n",
    "\n",
    "An L2-penalized regression was fit to determine the effect sizes of tier 2 mutations, independent of resistance-associated mutations they may occur with. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ee479d3c-82e4-4390-8956-620fefc7b84e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_df = pd.read_csv(os.path.join(analysis_dir, f\"{drug}/{folder}/exclude_comutation/WHOphenos_univariate_stats.csv\"))\n",
    "# exclude_df = exclude_df.loc[~exclude_df[\"mutation\"].str.contains(\"PC\")]\n",
    "# exclude_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']] = exclude_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']].astype(int)\n",
    "\n",
    "# exclude_df.query(\"Odds_Ratio > 1 & BH_pval < 0.01 & TP >= 5 & PPV_LB >= 0.25 & ~mutation.str.contains('rpoB')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "1ce0aaa4-a55c-4658-ac8c-3836f6a67f53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_df.query(\"mutation=='rpoC_p.Glu1092Asp'\")[['mutation', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval',\n",
    "#        'Bonferroni_pval', 'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN',\n",
    "#        'FN', 'PPV_LB', 'PPV', 'PPV_UB', 'NPV', 'Sens', 'Spec']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "bc772389-498a-4304-aec9-9ed041bc63c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# og_model_analysis.query(\"mutation=='rpoC_p.Glu1092Asp'\")[['mutation', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval',\n",
    "#        'Bonferroni_pval', 'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN',\n",
    "#        'FN', 'PPV_LB', 'PPV', 'PPV_UB', 'NPV', 'Sens', 'Spec']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b8b21b1-b6a4-4a9d-a822-5806d1916379",
   "metadata": {},
   "source": [
    "# 2. Likelihood Ratio Test: 5/152 of the Above Tier 2 Mutations are Significant\n",
    "\n",
    "152 tier 2 variants were found significant in the first analysis. We then performed a likelihood ratio test:\n",
    "\n",
    "A mutation is removed from the original model input matrix, and then a new model is fit. \n",
    "\n",
    "We then compare the log-likelihoods of the original model and the model with 1 less mutation. If the p-value is significant, it means that removing the mutation significantly changes the fit of the model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "f3798b73-118c-40a6-9edd-77fd1490f1b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  mutation     log_like  chi_stat  pval       AUC      Sens      Spec  \\\n",
      "0     FULL -3386.116022       NaN   NaN  0.986142  0.951108  0.981836   \n",
      "\n",
      "   accuracy  \n",
      "0  0.971856  \n",
      "(3609, 10)\n"
     ]
    }
   ],
   "source": [
    "LRT_res = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_results_tiers1+2.csv\"))\n",
    "print(LRT_res.head(1))\n",
    "\n",
    "LRT_res = add_pval_corrections(LRT_res.iloc[1:, ], col=\"pval\")\n",
    "print(LRT_res.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "812717c0-1676-43a9-88fb-043173f0fadc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.022368999999999972"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "0.986142-0.963773"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "9f63596d-b00e-41fa-b5fb-2de1f56d42aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>log_like</th>\n",
       "      <th>chi_stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>-6132.979553</td>\n",
       "      <td>5493.727061</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.963773</td>\n",
       "      <td>0.860876</td>\n",
       "      <td>0.979829</td>\n",
       "      <td>0.941195</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Rv2752c_p.Asn30Ser</td>\n",
       "      <td>-3399.923547</td>\n",
       "      <td>27.615050</td>\n",
       "      <td>1.480226e-07</td>\n",
       "      <td>0.985922</td>\n",
       "      <td>0.950711</td>\n",
       "      <td>0.981884</td>\n",
       "      <td>0.971760</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>-3398.741570</td>\n",
       "      <td>25.251095</td>\n",
       "      <td>5.033098e-07</td>\n",
       "      <td>0.985492</td>\n",
       "      <td>0.951108</td>\n",
       "      <td>0.981789</td>\n",
       "      <td>0.971824</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.001816</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>rpoC_p.Ile491Thr</td>\n",
       "      <td>-3394.700163</td>\n",
       "      <td>17.168281</td>\n",
       "      <td>3.421011e-05</td>\n",
       "      <td>0.986078</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.005144</td>\n",
       "      <td>0.123464</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>rpoC_p.Asn698Ser</td>\n",
       "      <td>-3394.005705</td>\n",
       "      <td>15.779366</td>\n",
       "      <td>7.117452e-05</td>\n",
       "      <td>0.986091</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.009514</td>\n",
       "      <td>0.256869</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "      <td>-3393.225162</td>\n",
       "      <td>14.218279</td>\n",
       "      <td>1.627815e-04</td>\n",
       "      <td>0.986078</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.017802</td>\n",
       "      <td>0.587479</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              mutation     log_like     chi_stat          pval       AUC  \\\n",
       "0     rpoB_p.Ser450Leu -6132.979553  5493.727061  0.000000e+00  0.963773   \n",
       "26  Rv2752c_p.Asn30Ser -3399.923547    27.615050  1.480226e-07  0.985922   \n",
       "30   rpoC_p.Glu1092Asp -3398.741570    25.251095  5.033098e-07  0.985492   \n",
       "38    rpoC_p.Ile491Thr -3394.700163    17.168281  3.421011e-05  0.986078   \n",
       "41    rpoC_p.Asn698Ser -3394.005705    15.779366  7.117452e-05  0.986091   \n",
       "47   rpoC_p.Pro1040Arg -3393.225162    14.218279  1.627815e-04  0.986078   \n",
       "\n",
       "        Sens      Spec  accuracy   BH_pval  Bonferroni_pval  Significant  \n",
       "0   0.860876  0.979829  0.941195  0.000000         0.000000            1  \n",
       "26  0.950711  0.981884  0.971760  0.000045         0.000534            1  \n",
       "30  0.951108  0.981789  0.971824  0.000114         0.001816            1  \n",
       "38  0.950909  0.981932  0.971856  0.005144         0.123464            1  \n",
       "41  0.950909  0.981932  0.971856  0.009514         0.256869            1  \n",
       "47  0.950909  0.981932  0.971856  0.017802         0.587479            0  "
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_res.query(\"mutation in @LRT_bootstrap_StatsDiff.mutation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "ea849591-06c1-4e5b-aad7-e9a469563d70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mutation\n",
       "Rv2752c_p.Asn30Ser    0.000199\n",
       "rpoB_p.Ser450Leu      0.021720\n",
       "rpoC_p.Asn698Ser      0.000055\n",
       "rpoC_p.Glu1092Asp     0.000479\n",
       "rpoC_p.Ile491Thr      0.000079\n",
       "rpoC_p.Pro1040Arg     0.000068\n",
       "Name: AUC, dtype: float64"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_bootstrap_StatsDiff.groupby(\"mutation\")[\"AUC\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67375bbf-13f9-4e93-8e8f-fc31a1fde062",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "5727be1c-b61e-4bce-88ab-ac2f5d62ef42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "350 mutations are significantly associated with resistance\n",
      "167 mutations are Tier 2\n"
     ]
    }
   ],
   "source": [
    "og_model_analysis[\"Significant\"] = og_model_analysis[\"Significant\"].fillna(0).astype(int)\n",
    "print(f\"{len(og_model_analysis.query('Odds_Ratio > 1 & Significant == 1'))} mutations are significantly associated with resistance\")\n",
    "print(f\"{len(og_model_analysis.query('Odds_Ratio > 1 & Significant == 1').loc[~og_model_analysis['mutation'].str.contains('rpoB')])} mutations are Tier 2\")\n",
    "\n",
    "for i, row in og_model_analysis.iterrows():\n",
    "    \n",
    "    if \"rpoB\" in row[\"mutation\"]:\n",
    "        if row[\"BH_pval\"] < 0.05:\n",
    "            og_model_analysis.loc[i, \"Significant\"] = 1\n",
    "    else:\n",
    "        if row[\"BH_pval\"] < 0.01:\n",
    "            og_model_analysis.loc[i, \"Significant\"] = 1\n",
    "\n",
    "\n",
    "og_summary = pd.DataFrame(og_model_analysis.query(\"Odds_Ratio > 1 & Significant == 1\").confidence.value_counts(dropna=False)).reset_index().rename(columns={\"index\": \"confidence\", \n",
    "                                                                                                                                             \"confidence\": \"orig_significant_count\"}).sort_values(\"confidence\")\n",
    "\n",
    "og_summary = og_summary.merge(pd.DataFrame(who_variants_combined.query(\"drug=='RIF'\").confidence.value_counts()).reset_index().rename(columns={\"index\": \"confidence\", \n",
    "                                                                                                                                                 \"confidence\": \"who2021_count\"}).sort_values(\"confidence\"), \n",
    "                  on=\"confidence\", how=\"outer\")\n",
    "\n",
    "og_summary[\"who2021_count\"] = og_summary[\"who2021_count\"].fillna(0)\n",
    "og_summary[\"who2021_count\"] = og_summary[\"who2021_count\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "93fc7047-8761-4ce3-b227-973406d7d09c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>confidence</th>\n",
       "      <th>who2021_count</th>\n",
       "      <th>orig_significant_count</th>\n",
       "      <th>LRT_significant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>93</td>\n",
       "      <td>33</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1515</td>\n",
       "      <td>117</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  confidence  who2021_count  orig_significant_count  \\\n",
       "0               1) Assoc w R             24                      24   \n",
       "1     2) Assoc w R - Interim             93                      33   \n",
       "2  3) Uncertain significance           1515                     117   \n",
       "3           5) Not assoc w R             28                       1   \n",
       "4                        NaN              0                     175   \n",
       "\n",
       "   LRT_significant_count  \n",
       "0                     24  \n",
       "1                     15  \n",
       "2                      5  \n",
       "3                      1  \n",
       "4                      4  "
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final = og_summary.merge(lrt_summary, on=\"confidence\")\n",
    "final[[\"confidence\", \"who2021_count\", \"orig_significant_count\", \"LRT_significant_count\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "28ec30bb-8c31-4eae-9d4c-040875083a21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "350"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "24 + 33 + 118 + 175"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "f65b98c9-13c3-46c9-8e08-1ad3909a5a18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "39 + 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "ca9d512e-fd21-44f2-a781-fc1b505d480f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>confidence</th>\n",
       "      <th>who2021_count</th>\n",
       "      <th>orig_significant_count</th>\n",
       "      <th>LRT_significant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>93</td>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1515</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  confidence  who2021_count  orig_significant_count  \\\n",
       "0               1) Assoc w R             24                      24   \n",
       "1     2) Assoc w R - Interim             93                      18   \n",
       "2  3) Uncertain significance           1515                      81   \n",
       "3           5) Not assoc w R             28                       1   \n",
       "4                        NaN              0                     161   \n",
       "\n",
       "   LRT_significant_count  \n",
       "0                     24  \n",
       "1                     11  \n",
       "2                      2  \n",
       "3                      1  \n",
       "4                      4  "
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final = og_summary.merge(lrt_summary, on=\"confidence\")\n",
    "final[[\"confidence\", \"who2021_count\", \"orig_significant_count\", \"LRT_significant_count\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "30c9398d-667c-40e5-89e5-4b898c5a4546",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(350, 40)"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_model_analysis.query('Odds_Ratio > 1 & Significant == 1').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "862baec3-b490-4c6c-9580-87d889f6d9dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Rv1129c_p.Asp214Asn', 'glpK_p.Ile255Leu'}"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(LRT_res.query('Significant == 1 & mutation in @mut_assoc_res')[\"mutation\"]) - set(og_model_analysis.query('Odds_Ratio > 1 & Significant == 1')[\"mutation\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "0dcc9c5d-e015-4ea2-ab8a-4aee87d5c3a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(350, 40)"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_model_analysis.query('Odds_Ratio > 1 & Significant == 1').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "142c1f44-0374-4d57-96fb-1c1973a29c58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>confidence</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>...</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Phenos</th>\n",
       "      <th>pool_type</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>HET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Rv1129c_p.Asp214Asn</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1253895</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.067306</td>\n",
       "      <td>0.993585</td>\n",
       "      <td>1.079323</td>\n",
       "      <td>0.004564</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.99999</td>\n",
       "      <td>0.291409</td>\n",
       "      <td>14.682404</td>\n",
       "      <td>0.999644</td>\n",
       "      <td>1.000177</td>\n",
       "      <td>2</td>\n",
       "      <td>WHO</td>\n",
       "      <td>unpooled</td>\n",
       "      <td>0</td>\n",
       "      <td>DROP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>glpK_p.Ile255Leu</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>4138993</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.062619</td>\n",
       "      <td>1.000009</td>\n",
       "      <td>1.070424</td>\n",
       "      <td>0.001785</td>\n",
       "      <td>0.011241</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.999586</td>\n",
       "      <td>1.000067</td>\n",
       "      <td>2</td>\n",
       "      <td>WHO</td>\n",
       "      <td>unpooled</td>\n",
       "      <td>0</td>\n",
       "      <td>DROP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                mutation  predicted_effect position confidence  Odds_Ratio  \\\n",
       "94   Rv1129c_p.Asp214Asn  missense_variant  1253895        NaN    1.067306   \n",
       "106     glpK_p.Ile255Leu  missense_variant  4138993        NaN    1.062619   \n",
       "\n",
       "        OR_LB     OR_UB      pval   BH_pval  Bonferroni_pval  ...  Spec_UB  \\\n",
       "94   0.993585  1.079323  0.004564  0.022627              1.0  ...  0.99999   \n",
       "106  1.000009  1.070424  0.001785  0.011241              1.0  ...  1.00000   \n",
       "\n",
       "       LR+_LB     LR+_UB    LR-_LB    LR-_UB  Tier  Phenos  pool_type  \\\n",
       "94   0.291409  14.682404  0.999644  1.000177     2     WHO   unpooled   \n",
       "106       NaN        inf  0.999586  1.000067     2     WHO   unpooled   \n",
       "\n",
       "     synonymous   HET  \n",
       "94            0  DROP  \n",
       "106           0  DROP  \n",
       "\n",
       "[2 rows x 40 columns]"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_model_analysis.query(\"Odds_Ratio > 1 & mutation in ['Rv1129c_p.Asp214Asn', 'glpK_p.Ile255Leu']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "9c86a86c-d08b-4a13-af86-8e5cd9895167",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>log_like</th>\n",
       "      <th>chi_stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>glpK_p.Ile255Leu</td>\n",
       "      <td>-3395.126391</td>\n",
       "      <td>18.020738</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.986016</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.003585</td>\n",
       "      <td>0.078861</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Rv1129c_p.Asp214Asn</td>\n",
       "      <td>-3394.187504</td>\n",
       "      <td>16.142963</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.986032</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981836</td>\n",
       "      <td>0.971792</td>\n",
       "      <td>0.008153</td>\n",
       "      <td>0.211980</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               mutation     log_like   chi_stat      pval       AUC      Sens  \\\n",
       "36     glpK_p.Ile255Leu -3395.126391  18.020738  0.000022  0.986016  0.950909   \n",
       "40  Rv1129c_p.Asp214Asn -3394.187504  16.142963  0.000059  0.986032  0.950909   \n",
       "\n",
       "        Spec  accuracy   BH_pval  Bonferroni_pval  Significant  \n",
       "36  0.981932  0.971856  0.003585         0.078861            1  \n",
       "40  0.981836  0.971792  0.008153         0.211980            1  "
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_res.query(\"mutation in ['Rv1129c_p.Asp214Asn', 'glpK_p.Ile255Leu']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "5039b5ab-89dc-4b6f-8c38-0d6bb8017ea7",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'X' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [253]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mX\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X' is not defined"
     ]
    }
   ],
   "source": [
    "np.array([1, 2, 3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "fe4fbd2e-3e3c-41f3-bb79-fea30d24852c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49 mutations are significantly associated with resistance\n",
      "6 mutations are Tier 2\n"
     ]
    }
   ],
   "source": [
    "mut_assoc_res = og_model_analysis.query(\"Odds_Ratio > 1\")[\"mutation\"].values\n",
    "\n",
    "for i, row in LRT_res.iterrows():\n",
    "    \n",
    "    if \"rpoB\" in row[\"mutation\"]:\n",
    "        if row[\"BH_pval\"] < 0.05:\n",
    "            LRT_res.loc[i, \"Significant\"] = 1\n",
    "    else:\n",
    "        if row[\"BH_pval\"] < 0.01:\n",
    "            LRT_res.loc[i, \"Significant\"] = 1\n",
    "\n",
    "LRT_res[\"Significant\"] = LRT_res[\"Significant\"].fillna(0).astype(int)\n",
    "\n",
    "print(f\"{len(LRT_res.query('Significant == 1 & mutation in @mut_assoc_res'))} mutations are significantly associated with resistance\")\n",
    "print(f\"{len(LRT_res.query('Significant == 1 & mutation in @mut_assoc_res').loc[~LRT_res['mutation'].str.contains('rpoB')])} mutations are Tier 2\")\n",
    "\n",
    "\n",
    "lrt_summary = LRT_res.query(\"Significant == 1 & mutation in @mut_assoc_res\").merge(who_variants_combined.query(\"drug=='RIF'\"), \n",
    "                                      on=\"mutation\", how=\"left\"\n",
    "                                     ).confidence.value_counts(dropna=False).reset_index().rename(columns={\"index\": \"confidence\", \"confidence\": \"LRT_significant_count\"}).sort_values(\"confidence\")\n",
    "\n",
    "# lrt_summary = lrt_summary.merge(pd.DataFrame(who_variants_combined.query(\"drug=='RIF'\").confidence.value_counts()).reset_index().rename(columns={\"index\": \"confidence\", \n",
    "#                                                                                                                                                  \"confidence\": \"who2021_count\"}).sort_values(\"confidence\"), \n",
    "#                   on=\"confidence\", how=\"outer\")\n",
    "\n",
    "# lrt_summary[\"who2021_count\"] = lrt_summary[\"who2021_count\"].fillna(0)\n",
    "# lrt_summary[\"who2021_count\"] = lrt_summary[\"who2021_count\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "ddf477bd-a31a-4ef4-b86e-9854a3acdf51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_significant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  confidence  LRT_significant_count\n",
       "0               1) Assoc w R                     24\n",
       "1     2) Assoc w R - Interim                     15\n",
       "2  3) Uncertain significance                      5\n",
       "4           5) Not assoc w R                      1\n",
       "3                        NaN                      4"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lrt_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a52a9f-b6fc-4dbb-924f-dfbdcffcbc8d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "820ac536-9828-49aa-9111-788504f4f40c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rpoB'], dtype='<U4')"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique([gene.split(\"_\")[0] for gene in who_variants_combined.query(\"drug=='RIF' & confidence == '1) Assoc w R'\")[\"mutation\"].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "9e959b2b-a044-48df-af2b-a807fb5c2bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rpoB'], dtype='<U4')"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique([gene.split(\"_\")[0] for gene in who_variants_combined.query(\"drug=='RIF' & confidence == '2) Assoc w R - Interim'\")[\"mutation\"].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "334a8e28-1117-42f2-b6ca-26d9549fdab8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Rv2752c', 'rpoA', 'rpoB', 'rpoC'], dtype='<U7')"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique([gene.split(\"_\")[0] for gene in who_variants_combined.query(\"drug=='RIF' & confidence == '3) Uncertain significance'\")[\"mutation\"].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "840636a9-29d8-4934-99a5-c80059e899f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Rv2752c', 'rpoA', 'rpoB', 'rpoC'], dtype='<U7')"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique([gene.split(\"_\")[0] for gene in who_variants_combined.query(\"drug=='RIF' & confidence == '5) Not assoc w R'\")[\"mutation\"].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca0766f3-a829-4cb5-9fe4-ce21bc76bcc4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97e42c4c-7761-4fd4-8a50-13898778d4b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "3b9ffd95-b277-49cb-8990-6f5925f6475f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42, 10)"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_res.query(\"BH_pval < 0.01 & mutation in @mut_assoc_res\").shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32ab6924-375e-4c39-91b4-d61259e73854",
   "metadata": {},
   "outputs": [],
   "source": [
    "#.query(\"confidence not in ['1) Assoc w R', '2) Assoc w R - Interim']\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "651dd7e7-b428-4b87-bc0f-5946578a8e19",
   "metadata": {},
   "source": [
    "# Perform bootstrapping for all 152 mutations to see how much AUC, Sens, Spec, and Accuracy Change\n",
    "\n",
    "If the confidence interval lies above 0, it means that removing the feature SIGNIFICANTLY DECREASES THE PREDICTIVE POWER OF THE MODEL\n",
    "\n",
    "CI > 0 = FEATURE IS IMPORTANT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "8b9d72b2-6e0e-4bc7-b8ad-39cf938671da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "# bootstrapped the difference: full model stat - (model - 1) stat\n",
    "LRT_bootstrap_StatsDiff = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_bootstrap_StatsDiff.csv.gz\"), compression=\"gzip\")\n",
    "print(len(LRT_bootstrap_StatsDiff.mutation.unique()))\n",
    "LRT_bootstrap_StatsDiff2 = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_bootstrap_StatsDiff_2.csv.gz\"), compression=\"gzip\")\n",
    "print(len(LRT_bootstrap_StatsDiff2.mutation.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "efbe6b73-21bd-4aa1-8da0-92eb976b2f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "LRT_bootstrap_StatsDiff = pd.concat([LRT_bootstrap_StatsDiff, LRT_bootstrap_StatsDiff2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4aa436c7-0826-4920-836b-fb4c6ef88055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.0, 0.010000000000000009, 0.010000000000000009,\n",
       "       0.09999999999999998, 0.020000000000000018, 0.09999999999999998],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pval_df[\"AUC\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "abe8faf2-c819-4c5c-9023-1446757d3e8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.946493e-02</td>\n",
       "      <td>0.080225</td>\n",
       "      <td>0.002110</td>\n",
       "      <td>0.027659</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.122932e-02</td>\n",
       "      <td>0.088078</td>\n",
       "      <td>0.002849</td>\n",
       "      <td>0.030145</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.066901e-02</td>\n",
       "      <td>0.081655</td>\n",
       "      <td>0.005795</td>\n",
       "      <td>0.030951</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.224834e-02</td>\n",
       "      <td>0.097551</td>\n",
       "      <td>-0.001723</td>\n",
       "      <td>0.030596</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.095228e-02</td>\n",
       "      <td>0.085680</td>\n",
       "      <td>0.004442</td>\n",
       "      <td>0.030790</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>1.091928e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Rv1129c_p.Ser362Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>1.616239e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Rv1129c_p.Ser362Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>5.641436e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Rv1129c_p.Ser362Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>6.643447e-08</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Rv1129c_p.Ser362Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>9.216352e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Rv1129c_p.Ser362Thr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1100 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              AUC      Sens      Spec  accuracy             mutation\n",
       "0    1.946493e-02  0.080225  0.002110  0.027659     rpoB_p.Ser450Leu\n",
       "1    2.122932e-02  0.088078  0.002849  0.030145     rpoB_p.Ser450Leu\n",
       "2    2.066901e-02  0.081655  0.005795  0.030951     rpoB_p.Ser450Leu\n",
       "3    2.224834e-02  0.097551 -0.001723  0.030596     rpoB_p.Ser450Leu\n",
       "4    2.095228e-02  0.085680  0.004442  0.030790     rpoB_p.Ser450Leu\n",
       "..            ...       ...       ...       ...                  ...\n",
       "495  1.091928e-07  0.000000  0.000000  0.000000  Rv1129c_p.Ser362Thr\n",
       "496  1.616239e-07  0.000000  0.000000  0.000000  Rv1129c_p.Ser362Thr\n",
       "497  5.641436e-07  0.000000  0.000000  0.000000  Rv1129c_p.Ser362Thr\n",
       "498  6.643447e-08  0.000000  0.000000  0.000000  Rv1129c_p.Ser362Thr\n",
       "499  9.216352e-07  0.000000  0.000000  0.000000  Rv1129c_p.Ser362Thr\n",
       "\n",
       "[1100 rows x 5 columns]"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_bootstrap_StatsDiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "02aa8be2-8717-4e60-ab1a-059962996b45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rpoB_p.Ser450Leu', 'Rv2752c_p.Asn30Ser', 'rpoC_p.Glu1092Asp',\n",
       "       'rpoC_p.Ile491Thr', 'rpoC_p.Asn698Ser', 'rpoC_p.Pro1040Arg',\n",
       "       'glpK_p.Leu152Arg', 'Rv2752c_p.Val396Gly', 'lpqB_p.Asp370Glu',\n",
       "       'rpoA_c.-316G>A', 'Rv1129c_p.Ser362Thr'], dtype=object)"
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_bootstrap_StatsDiff.mutation.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "ac1d8646-8b6c-4b9c-908b-3ee43f4123ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rpoB_p.Ser450Leu</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv2752c_p.Asn30Ser</th>\n",
       "      <td>0.01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC_p.Glu1092Asp</th>\n",
       "      <td>0.01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC_p.Ile491Thr</th>\n",
       "      <td>0.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC_p.Asn698Ser</th>\n",
       "      <td>0.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC_p.Pro1040Arg</th>\n",
       "      <td>0.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glpK_p.Leu152Arg</th>\n",
       "      <td>0.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv2752c_p.Val396Gly</th>\n",
       "      <td>0.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lpqB_p.Asp370Glu</th>\n",
       "      <td>0.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoA_c.-316G&gt;A</th>\n",
       "      <td>0.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv1129c_p.Ser362Thr</th>\n",
       "      <td>0.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      AUC Sens Spec accuracy\n",
       "rpoB_p.Ser450Leu      0.0  NaN  NaN      NaN\n",
       "Rv2752c_p.Asn30Ser   0.01  NaN  NaN      NaN\n",
       "rpoC_p.Glu1092Asp    0.01  NaN  NaN      NaN\n",
       "rpoC_p.Ile491Thr      0.1  NaN  NaN      NaN\n",
       "rpoC_p.Asn698Ser     0.02  NaN  NaN      NaN\n",
       "rpoC_p.Pro1040Arg     0.1  NaN  NaN      NaN\n",
       "glpK_p.Leu152Arg      0.5  NaN  NaN      NaN\n",
       "Rv2752c_p.Val396Gly  0.27  NaN  NaN      NaN\n",
       "lpqB_p.Asp370Glu     0.31  NaN  NaN      NaN\n",
       "rpoA_c.-316G>A       0.46  NaN  NaN      NaN\n",
       "Rv1129c_p.Ser362Thr   0.3  NaN  NaN      NaN"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pval_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "0924c0cd-55e4-495c-8686-d492251baff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SpearmanrResult(correlation=0.9338806280906226, pvalue=2.6137422278744596e-05)\n"
     ]
    }
   ],
   "source": [
    "for metric in ['AUC']:\n",
    "\n",
    "    print(st.spearmanr(LRT_res.query(\"mutation in @LRT_bootstrap_StatsDiff.mutation.unique()\")[\"BH_pval\"], \n",
    "                 pval_df[metric]\n",
    "                ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "f4451bac-538f-4944-8807-7a2823193781",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "rpoB_p.Ser450Leu\n",
      "AUC: 0.0\n",
      "\n",
      "Rv2752c_p.Asn30Ser\n",
      "AUC: 0.01\n",
      "\n",
      "rpoC_p.Glu1092Asp\n",
      "AUC: 0.01\n",
      "\n",
      "rpoC_p.Ile491Thr\n",
      "\n",
      "rpoC_p.Asn698Ser\n",
      "AUC: 0.02\n",
      "\n",
      "rpoC_p.Pro1040Arg\n",
      "\n",
      "glpK_p.Leu152Arg\n",
      "\n",
      "Rv2752c_p.Val396Gly\n",
      "\n",
      "lpqB_p.Asp370Glu\n",
      "\n",
      "rpoA_c.-316G>A\n",
      "\n",
      "Rv1129c_p.Ser362Thr\n"
     ]
    }
   ],
   "source": [
    "pval_df = pd.DataFrame(columns=['AUC', 'Sens', 'Spec', 'accuracy'])\n",
    "\n",
    "for mut in LRT_bootstrap_StatsDiff.mutation.unique():\n",
    "    \n",
    "    print(\"\\n\" + mut)\n",
    "    \n",
    "    for metric in ['AUC']:\n",
    "    \n",
    "#         lower, upper = np.percentile(LRT_bootstrap_StatsDiff.query(\"mutation==@mut\")[metric], q=[2.5, 97.5])\n",
    "    \n",
    "#         if lower > 0 and upper > 0:\n",
    "#             print(f\"    {metric}: {round(lower, 4)}, {round(upper, 4)}\")\n",
    "\n",
    "        pval = 1 - np.mean(LRT_bootstrap_StatsDiff.query('mutation==@mut')[metric].values > 0)\n",
    "        \n",
    "        if pval < 0.05:\n",
    "            print(f\"{metric}: {round(pval, 2)}\")\n",
    "            \n",
    "        pval_df.loc[mut, metric] = pval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "bd41e669-0efa-4997-b931-861222083f77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.019465</td>\n",
       "      <td>0.080225</td>\n",
       "      <td>0.002110</td>\n",
       "      <td>0.027659</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.021229</td>\n",
       "      <td>0.088078</td>\n",
       "      <td>0.002849</td>\n",
       "      <td>0.030145</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.020669</td>\n",
       "      <td>0.081655</td>\n",
       "      <td>0.005795</td>\n",
       "      <td>0.030951</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.022248</td>\n",
       "      <td>0.097551</td>\n",
       "      <td>-0.001723</td>\n",
       "      <td>0.030596</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.020952</td>\n",
       "      <td>0.085680</td>\n",
       "      <td>0.004442</td>\n",
       "      <td>0.030790</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>-0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000048</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>-0.000337</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>-0.000048</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>600 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AUC      Sens      Spec  accuracy           mutation\n",
       "0    0.019465  0.080225  0.002110  0.027659   rpoB_p.Ser450Leu\n",
       "1    0.021229  0.088078  0.002849  0.030145   rpoB_p.Ser450Leu\n",
       "2    0.020669  0.081655  0.005795  0.030951   rpoB_p.Ser450Leu\n",
       "3    0.022248  0.097551 -0.001723  0.030596   rpoB_p.Ser450Leu\n",
       "4    0.020952  0.085680  0.004442  0.030790   rpoB_p.Ser450Leu\n",
       "..        ...       ...       ...       ...                ...\n",
       "595  0.000246  0.000302  0.000000  0.000097  rpoC_p.Pro1040Arg\n",
       "596  0.000006  0.000000  0.000000  0.000000  rpoC_p.Pro1040Arg\n",
       "597 -0.000002  0.000000 -0.000048 -0.000032  rpoC_p.Pro1040Arg\n",
       "598  0.000222  0.000685 -0.000337  0.000000  rpoC_p.Pro1040Arg\n",
       "599  0.000014  0.000098 -0.000048  0.000000  rpoC_p.Pro1040Arg\n",
       "\n",
       "[600 rows x 5 columns]"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_bootstrap_StatsDiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57621f28-c480-4a8a-9e7d-8859adfe9c89",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cbd46f8-09cb-40c4-8a64-3802eaf0cc81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "88b17e7e-be45-424c-a626-61d57964464f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>log_like</th>\n",
       "      <th>chi_stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>-6132.979553</td>\n",
       "      <td>5493.727061</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.963773</td>\n",
       "      <td>0.860876</td>\n",
       "      <td>0.979829</td>\n",
       "      <td>0.941195</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Rv2752c_p.Asn30Ser</td>\n",
       "      <td>-3399.923547</td>\n",
       "      <td>27.615050</td>\n",
       "      <td>1.480226e-07</td>\n",
       "      <td>0.985922</td>\n",
       "      <td>0.950711</td>\n",
       "      <td>0.981884</td>\n",
       "      <td>0.971760</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.000534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>-3398.741570</td>\n",
       "      <td>25.251095</td>\n",
       "      <td>5.033098e-07</td>\n",
       "      <td>0.985492</td>\n",
       "      <td>0.951108</td>\n",
       "      <td>0.981789</td>\n",
       "      <td>0.971824</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.001816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>rpoC_p.Ile491Thr</td>\n",
       "      <td>-3394.700163</td>\n",
       "      <td>17.168281</td>\n",
       "      <td>3.421011e-05</td>\n",
       "      <td>0.986078</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.005144</td>\n",
       "      <td>0.123464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>rpoC_p.Asn698Ser</td>\n",
       "      <td>-3394.005705</td>\n",
       "      <td>15.779366</td>\n",
       "      <td>7.117452e-05</td>\n",
       "      <td>0.986091</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.009514</td>\n",
       "      <td>0.256869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "      <td>-3393.225162</td>\n",
       "      <td>14.218279</td>\n",
       "      <td>1.627815e-04</td>\n",
       "      <td>0.986078</td>\n",
       "      <td>0.950909</td>\n",
       "      <td>0.981932</td>\n",
       "      <td>0.971856</td>\n",
       "      <td>0.017802</td>\n",
       "      <td>0.587479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              mutation     log_like     chi_stat          pval       AUC  \\\n",
       "0     rpoB_p.Ser450Leu -6132.979553  5493.727061  0.000000e+00  0.963773   \n",
       "26  Rv2752c_p.Asn30Ser -3399.923547    27.615050  1.480226e-07  0.985922   \n",
       "30   rpoC_p.Glu1092Asp -3398.741570    25.251095  5.033098e-07  0.985492   \n",
       "38    rpoC_p.Ile491Thr -3394.700163    17.168281  3.421011e-05  0.986078   \n",
       "41    rpoC_p.Asn698Ser -3394.005705    15.779366  7.117452e-05  0.986091   \n",
       "47   rpoC_p.Pro1040Arg -3393.225162    14.218279  1.627815e-04  0.986078   \n",
       "\n",
       "        Sens      Spec  accuracy   BH_pval  Bonferroni_pval  \n",
       "0   0.860876  0.979829  0.941195  0.000000         0.000000  \n",
       "26  0.950711  0.981884  0.971760  0.000045         0.000534  \n",
       "30  0.951108  0.981789  0.971824  0.000114         0.001816  \n",
       "38  0.950909  0.981932  0.971856  0.005144         0.123464  \n",
       "41  0.950909  0.981932  0.971856  0.009514         0.256869  \n",
       "47  0.950909  0.981932  0.971856  0.017802         0.587479  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRT_res.query(\"mutation in @LRT_bootstrap_StatsDiff.mutation.unique()\").query(\"BH_pval < 0.05\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51ff9fee-e0ea-4908-ad04-750f663165f2",
   "metadata": {},
   "source": [
    "# Model with Interaction Terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d6cf636e-5869-4426-88b5-f5ff038240ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "interact_coef = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_LRT05_sig_coef.csv\"))\n",
    "interact_coef_bs = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_LRT05_sig_BScoef.csv\"))\n",
    "\n",
    "# interact_res = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_LRT05_sig_BSsummaryStats.csv\"))\n",
    "\n",
    "interact_matrix = pd.read_pickle(os.path.join(analysis_dir, drug, f\"BINARY/interaction/model_matrix_{phenos_name}_LRT05_sig.pkl\"))\n",
    "num_samples = len(interact_matrix)\n",
    "\n",
    "# del interact_matrix\n",
    "interact_coef = get_pvalues_add_ci(interact_coef, interact_coef_bs, \"mutation\", num_samples, alpha=0.05)\n",
    "#interact_coef = add_pval_corrections()\n",
    "\n",
    "interact_coef = add_pval_corrections(interact_coef)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "48c1b7bd-2012-4dda-9bb2-b4e7fc0b59af",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_mut = interact_coef[\"mutation\"].str.split(\"+\", n=1, expand=True)\n",
    "split_mut.columns = [\"Mut1\", \"Mut2\"]\n",
    "\n",
    "interact_coef = pd.concat([interact_coef, split_mut], axis=1)\n",
    "\n",
    "mut_conf_dict = dict(zip(who_variants_combined.query(\"drug=='RIF'\")[\"mutation\"], who_variants_combined.query(\"drug=='RIF'\")[\"confidence\"]))\n",
    "\n",
    "interact_coef[\"conf1\"] = interact_coef[\"Mut1\"].map(mut_conf_dict)\n",
    "interact_coef[\"conf2\"] = interact_coef[\"Mut2\"].map(mut_conf_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b963c78c-4d6f-4399-8708-2ebcf0fbe6a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>conf1</th>\n",
       "      <th>conf2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.219059</td>\n",
       "      <td>1.307759e-01</td>\n",
       "      <td>0.276704</td>\n",
       "      <td>1.474453e-10</td>\n",
       "      <td>5.291179e-09</td>\n",
       "      <td>5.555738e-07</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>None</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>rpoB_p.Glu761Asp+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.045750</td>\n",
       "      <td>3.516931e-02</td>\n",
       "      <td>0.055883</td>\n",
       "      <td>6.531201e-17</td>\n",
       "      <td>4.825405e-15</td>\n",
       "      <td>2.460957e-13</td>\n",
       "      <td>rpoB_p.Glu761Asp</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>rpoB_p.Asp435Gly+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.038459</td>\n",
       "      <td>1.972758e-02</td>\n",
       "      <td>0.063935</td>\n",
       "      <td>4.819887e-04</td>\n",
       "      <td>4.597806e-03</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Asp435Gly</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>rpoB_p.Ile488Val+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.033812</td>\n",
       "      <td>2.370705e-02</td>\n",
       "      <td>0.043639</td>\n",
       "      <td>2.951732e-11</td>\n",
       "      <td>1.170750e-09</td>\n",
       "      <td>1.112212e-07</td>\n",
       "      <td>rpoB_p.Ile488Val</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>rpoB_p.Ser450Trp+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.026928</td>\n",
       "      <td>1.726181e-02</td>\n",
       "      <td>0.037756</td>\n",
       "      <td>4.941578e-07</td>\n",
       "      <td>1.023070e-05</td>\n",
       "      <td>1.861987e-03</td>\n",
       "      <td>rpoB_p.Ser450Trp</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>rpoB_p.Asp435Asn+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.026639</td>\n",
       "      <td>1.951951e-17</td>\n",
       "      <td>0.042539</td>\n",
       "      <td>3.542234e-03</td>\n",
       "      <td>2.269921e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Asp435Asn</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>rpoB_p.Met434Val+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.025688</td>\n",
       "      <td>6.916843e-03</td>\n",
       "      <td>0.045850</td>\n",
       "      <td>8.438221e-03</td>\n",
       "      <td>4.478200e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Met434Val</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>rpoB_p.His445Leu+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.025392</td>\n",
       "      <td>1.397709e-02</td>\n",
       "      <td>0.052171</td>\n",
       "      <td>7.719784e-03</td>\n",
       "      <td>4.209573e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.His445Leu</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>rpoB_p.Ser441Gln+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.022871</td>\n",
       "      <td>9.898698e-03</td>\n",
       "      <td>0.036071</td>\n",
       "      <td>9.790105e-04</td>\n",
       "      <td>7.950241e-03</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Ser441Gln</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>rpoB_p.Ile480Val+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.022659</td>\n",
       "      <td>9.347350e-03</td>\n",
       "      <td>0.030784</td>\n",
       "      <td>2.985122e-05</td>\n",
       "      <td>4.244505e-04</td>\n",
       "      <td>1.124794e-01</td>\n",
       "      <td>rpoB_p.Ile480Val</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>rpoB_p.His445Asp+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.020935</td>\n",
       "      <td>1.157580e-02</td>\n",
       "      <td>0.028511</td>\n",
       "      <td>5.809340e-06</td>\n",
       "      <td>9.600699e-05</td>\n",
       "      <td>2.188959e-02</td>\n",
       "      <td>rpoB_p.His445Asp</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>rpoB_p.Arg827Leu+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.016304</td>\n",
       "      <td>8.307722e-03</td>\n",
       "      <td>0.022358</td>\n",
       "      <td>2.931413e-06</td>\n",
       "      <td>5.284959e-05</td>\n",
       "      <td>1.104556e-02</td>\n",
       "      <td>rpoB_p.Arg827Leu</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>rpoB_p.Arg552His+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.012905</td>\n",
       "      <td>6.779440e-03</td>\n",
       "      <td>0.020109</td>\n",
       "      <td>7.100014e-05</td>\n",
       "      <td>8.947442e-04</td>\n",
       "      <td>2.675285e-01</td>\n",
       "      <td>rpoB_p.Arg552His</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>rpoB_p.Leu554Pro+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.010215</td>\n",
       "      <td>2.031653e-17</td>\n",
       "      <td>0.014629</td>\n",
       "      <td>2.129588e-03</td>\n",
       "      <td>1.508325e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Leu554Pro</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662</th>\n",
       "      <td>rpoB_p.His1028Tyr+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.008887</td>\n",
       "      <td>3.161060e-03</td>\n",
       "      <td>0.015932</td>\n",
       "      <td>6.243227e-03</td>\n",
       "      <td>3.619151e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.His1028Tyr</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>rpoB_p.Arg827Cys+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.008819</td>\n",
       "      <td>2.968043e-17</td>\n",
       "      <td>0.014148</td>\n",
       "      <td>5.773310e-03</td>\n",
       "      <td>3.372687e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Arg827Cys</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>rpoB_p.Arg827His+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.007739</td>\n",
       "      <td>2.739143e-03</td>\n",
       "      <td>0.013618</td>\n",
       "      <td>2.132171e-03</td>\n",
       "      <td>1.507321e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Arg827His</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>rpoB_p.Thr400Ala+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.006134</td>\n",
       "      <td>2.296825e-03</td>\n",
       "      <td>0.011786</td>\n",
       "      <td>7.393401e-03</td>\n",
       "      <td>4.090798e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Thr400Ala</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>697</th>\n",
       "      <td>rpoB_p.Ala286Val+rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.005465</td>\n",
       "      <td>2.082876e-03</td>\n",
       "      <td>0.009805</td>\n",
       "      <td>7.418594e-03</td>\n",
       "      <td>4.098719e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Ala286Val</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>rpoB_p.Leu430Pro+rpoC_p.Glu1092Asp</td>\n",
       "      <td>-0.060411</td>\n",
       "      <td>-1.071780e-01</td>\n",
       "      <td>-0.016505</td>\n",
       "      <td>6.979631e-03</td>\n",
       "      <td>3.925261e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>rpoB_p.Leu430Pro</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                mutation      coef       coef_LB   coef_UB  \\\n",
       "107                    rpoC_p.Glu1092Asp  0.219059  1.307759e-01  0.276704   \n",
       "51    rpoB_p.Glu761Asp+rpoC_p.Glu1092Asp  0.045750  3.516931e-02  0.055883   \n",
       "400   rpoB_p.Asp435Gly+rpoC_p.Glu1092Asp  0.038459  1.972758e-02  0.063935   \n",
       "96    rpoB_p.Ile488Val+rpoC_p.Glu1092Asp  0.033812  2.370705e-02  0.043639   \n",
       "184   rpoB_p.Ser450Trp+rpoC_p.Glu1092Asp  0.026928  1.726181e-02  0.037756   \n",
       "596   rpoB_p.Asp435Asn+rpoC_p.Glu1092Asp  0.026639  1.951951e-17  0.042539   \n",
       "726   rpoB_p.Met434Val+rpoC_p.Glu1092Asp  0.025688  6.916843e-03  0.045850   \n",
       "706   rpoB_p.His445Leu+rpoC_p.Glu1092Asp  0.025392  1.397709e-02  0.052171   \n",
       "469   rpoB_p.Ser441Gln+rpoC_p.Glu1092Asp  0.022871  9.898698e-03  0.036071   \n",
       "268   rpoB_p.Ile480Val+rpoC_p.Glu1092Asp  0.022659  9.347350e-03  0.030784   \n",
       "231   rpoB_p.His445Asp+rpoC_p.Glu1092Asp  0.020935  1.157580e-02  0.028511   \n",
       "211   rpoB_p.Arg827Leu+rpoC_p.Glu1092Asp  0.016304  8.307722e-03  0.022358   \n",
       "302   rpoB_p.Arg552His+rpoC_p.Glu1092Asp  0.012905  6.779440e-03  0.020109   \n",
       "538   rpoB_p.Leu554Pro+rpoC_p.Glu1092Asp  0.010215  2.031653e-17  0.014629   \n",
       "662  rpoB_p.His1028Tyr+rpoC_p.Glu1092Asp  0.008887  3.161060e-03  0.015932   \n",
       "657   rpoB_p.Arg827Cys+rpoC_p.Glu1092Asp  0.008819  2.968043e-17  0.014148   \n",
       "540   rpoB_p.Arg827His+rpoC_p.Glu1092Asp  0.007739  2.739143e-03  0.013618   \n",
       "696   rpoB_p.Thr400Ala+rpoC_p.Glu1092Asp  0.006134  2.296825e-03  0.011786   \n",
       "697   rpoB_p.Ala286Val+rpoC_p.Glu1092Asp  0.005465  2.082876e-03  0.009805   \n",
       "685   rpoB_p.Leu430Pro+rpoC_p.Glu1092Asp -0.060411 -1.071780e-01 -0.016505   \n",
       "\n",
       "             pval       BH_pval  Bonferroni_pval               Mut1  \\\n",
       "107  1.474453e-10  5.291179e-09     5.555738e-07  rpoC_p.Glu1092Asp   \n",
       "51   6.531201e-17  4.825405e-15     2.460957e-13   rpoB_p.Glu761Asp   \n",
       "400  4.819887e-04  4.597806e-03     1.000000e+00   rpoB_p.Asp435Gly   \n",
       "96   2.951732e-11  1.170750e-09     1.112212e-07   rpoB_p.Ile488Val   \n",
       "184  4.941578e-07  1.023070e-05     1.861987e-03   rpoB_p.Ser450Trp   \n",
       "596  3.542234e-03  2.269921e-02     1.000000e+00   rpoB_p.Asp435Asn   \n",
       "726  8.438221e-03  4.478200e-02     1.000000e+00   rpoB_p.Met434Val   \n",
       "706  7.719784e-03  4.209573e-02     1.000000e+00   rpoB_p.His445Leu   \n",
       "469  9.790105e-04  7.950241e-03     1.000000e+00   rpoB_p.Ser441Gln   \n",
       "268  2.985122e-05  4.244505e-04     1.124794e-01   rpoB_p.Ile480Val   \n",
       "231  5.809340e-06  9.600699e-05     2.188959e-02   rpoB_p.His445Asp   \n",
       "211  2.931413e-06  5.284959e-05     1.104556e-02   rpoB_p.Arg827Leu   \n",
       "302  7.100014e-05  8.947442e-04     2.675285e-01   rpoB_p.Arg552His   \n",
       "538  2.129588e-03  1.508325e-02     1.000000e+00   rpoB_p.Leu554Pro   \n",
       "662  6.243227e-03  3.619151e-02     1.000000e+00  rpoB_p.His1028Tyr   \n",
       "657  5.773310e-03  3.372687e-02     1.000000e+00   rpoB_p.Arg827Cys   \n",
       "540  2.132171e-03  1.507321e-02     1.000000e+00   rpoB_p.Arg827His   \n",
       "696  7.393401e-03  4.090798e-02     1.000000e+00   rpoB_p.Thr400Ala   \n",
       "697  7.418594e-03  4.098719e-02     1.000000e+00   rpoB_p.Ala286Val   \n",
       "685  6.979631e-03  3.925261e-02     1.000000e+00   rpoB_p.Leu430Pro   \n",
       "\n",
       "                  Mut2                      conf1             conf2  \n",
       "107               None           5) Not assoc w R               NaN  \n",
       "51   rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "400  rpoC_p.Glu1092Asp     2) Assoc w R - Interim  5) Not assoc w R  \n",
       "96   rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "184  rpoC_p.Glu1092Asp               1) Assoc w R  5) Not assoc w R  \n",
       "596  rpoC_p.Glu1092Asp                        NaN  5) Not assoc w R  \n",
       "726  rpoC_p.Glu1092Asp     2) Assoc w R - Interim  5) Not assoc w R  \n",
       "706  rpoC_p.Glu1092Asp               1) Assoc w R  5) Not assoc w R  \n",
       "469  rpoC_p.Glu1092Asp               1) Assoc w R  5) Not assoc w R  \n",
       "268  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "231  rpoC_p.Glu1092Asp               1) Assoc w R  5) Not assoc w R  \n",
       "211  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "302  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "538  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "662  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "657  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "540  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "696  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "697  rpoC_p.Glu1092Asp  3) Uncertain significance  5) Not assoc w R  \n",
       "685  rpoC_p.Glu1092Asp               1) Assoc w R  5) Not assoc w R  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_coef.query(\"BH_pval < 0.05\").loc[interact_coef[\"mutation\"].str.contains(\"rpoC_p.Glu1092Asp\")].sort_values(\"coef\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5943655d-ec4c-44aa-9a7b-aec66cadacc5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fae26d9c-f81b-4e79-b9c8-fda4e1bf82e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_for_interact = ['Rv2752c_p.Asn30Ser', 'rpoC_p.Glu1092Asp', 'rpoC_p.Ile491Thr',\n",
    "       'rpoC_p.Asn698Ser', 'rpoC_p.Pro1040Arg', 'rpoC_p.Ile491Val',\n",
    "       'rpoC_p.Glu750Asp', 'rpoA_p.Val183Gly', 'rpoC_p.Asn416Ser']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3b93b5f-93da-4454-9a06-3ce13afbe12a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a7a80b4a-32a7-442e-9750-f7eacca36c15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rv2752c_p.Asn30Ser 4\n",
      "rpoC_p.Glu1092Asp 125\n",
      "rpoC_p.Ile491Thr 3\n",
      "rpoC_p.Asn698Ser 4\n",
      "rpoC_p.Pro1040Arg 6\n",
      "rpoC_p.Ile491Val 8\n",
      "rpoC_p.Glu750Asp 4\n",
      "rpoA_p.Val183Gly 6\n",
      "rpoC_p.Asn416Ser 3\n"
     ]
    }
   ],
   "source": [
    "for mut in mutations_for_interact:\n",
    "    \n",
    "    print(mut, interact_model_matrix[interact_model_matrix.columns[interact_model_matrix.columns.str.contains(mut)]].shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f340f1e3-2301-4a45-85bb-1205ba8d5708",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c18df3ec-8542-43ff-bc34-b9bfb4b7b2a8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
