{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "86d79dbc-86b7-4a6d-b26c-3052aaeb12a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 150\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "\n",
    "import glob, os, yaml, subprocess, itertools, sparse, vcf\n",
    "from sklearn.model_selection import train_test_split, KFold, GridSearchCV\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "import sklearn.metrics\n",
    "from sklearn.decomposition import PCA\n",
    "import timeit\n",
    "import scipy.stats as st\n",
    "import statsmodels.api as sm\n",
    "import pickle, yaml, tracemalloc\n",
    "from scipy.stats import binomtest\n",
    "\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue'\n",
    "input_data_dir = '/n/data1/hms/dbmi/farhat/ye12/who'\n",
    "who_variants_combined = pd.read_csv(\"who_confidence_2021.csv\")\n",
    "\n",
    "import itertools\n",
    "from stats_utils import *\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "973d52eb-a53f-4ea0-b8de-35c1eecdd397",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path = \"/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Rifampicin/BINARY/tiers=1/phenos=WHO/dropAF_noSyn_unpooled\"\n",
    "model_matrix = pd.read_pickle(os.path.join(model_path, \"model_matrix.pkl\"))\n",
    "model_analysis = pd.read_csv(os.path.join(model_path, \"model_analysis.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98643c3d-d48c-4b99-83f4-eb5257046ce4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [mutation, coef, pval, BH_pval, Bonferroni_pval, Odds_Ratio, confidence]\n",
       "Index: []"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_analysis.query(\"coef==0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b04240ec-7618-4a51-be4a-b58f359d7416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(820, 7)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_analysis.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "da050829-e4d1-434d-a29b-f440fbc27c8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [mutation, coef, pval, BH_pval, Bonferroni_pval, Odds_Ratio, confidence]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_analysis.query(\"mutation=='rpoB_p.Asn437Ile'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1ef5c843-3a61-4a49-bd33-76355a6a82b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35362, 843)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "de7663b4-912c-4e2a-a2e2-55c5b472cf1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "853"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "825 + 28"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "83f77fb5-bf4f-4f11-ab85-5f908ddc95dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>mutation</th>\n",
       "      <th>rpoB_p.Asn437Ile</th>\n",
       "      <th>rpoB_p.Asn437_Asn438del</th>\n",
       "      <th>rpoB_p.Asp435Leu</th>\n",
       "      <th>rpoB_p.Gln436_Asn437del</th>\n",
       "      <th>rpoB_p.Glu207Gly</th>\n",
       "      <th>rpoB_p.Glu291Ala</th>\n",
       "      <th>rpoB_p.Glu297Asp</th>\n",
       "      <th>rpoB_p.Glu423Gly</th>\n",
       "      <th>rpoB_p.His445Phe</th>\n",
       "      <th>rpoB_p.His445_Lys446del</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB_p.Pro439Ala</th>\n",
       "      <th>rpoB_p.Pro439Leu</th>\n",
       "      <th>rpoB_p.Ser431_Met434del</th>\n",
       "      <th>rpoB_p.Ser431dup</th>\n",
       "      <th>rpoB_p.Ser450Ala</th>\n",
       "      <th>rpoB_p.Ser450Asn</th>\n",
       "      <th>rpoB_p.Ser450_Leu452del</th>\n",
       "      <th>rpoB_p.Thr427Gly</th>\n",
       "      <th>rpoB_p.Thr427Pro</th>\n",
       "      <th>rpoB_p.Thr427_Ser428del</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>933876</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946364</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946365</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946366</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946367</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35362 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "mutation   rpoB_p.Asn437Ile  rpoB_p.Asn437_Asn438del  rpoB_p.Asp435Leu  \\\n",
       "sample_id                                                                \n",
       "809                     0.0                      0.0               0.0   \n",
       "812                     0.0                      0.0               0.0   \n",
       "815                     0.0                      0.0               0.0   \n",
       "816                     0.0                      0.0               0.0   \n",
       "817                     0.0                      0.0               0.0   \n",
       "...                     ...                      ...               ...   \n",
       "933876                  0.0                      0.0               0.0   \n",
       "946364                  0.0                      0.0               0.0   \n",
       "946365                  0.0                      0.0               0.0   \n",
       "946366                  0.0                      0.0               0.0   \n",
       "946367                  0.0                      0.0               0.0   \n",
       "\n",
       "mutation   rpoB_p.Gln436_Asn437del  rpoB_p.Glu207Gly  rpoB_p.Glu291Ala  \\\n",
       "sample_id                                                                \n",
       "809                            0.0               0.0               0.0   \n",
       "812                            0.0               0.0               0.0   \n",
       "815                            0.0               0.0               0.0   \n",
       "816                            0.0               0.0               0.0   \n",
       "817                            0.0               0.0               0.0   \n",
       "...                            ...               ...               ...   \n",
       "933876                         0.0               0.0               0.0   \n",
       "946364                         0.0               0.0               0.0   \n",
       "946365                         0.0               0.0               0.0   \n",
       "946366                         0.0               0.0               0.0   \n",
       "946367                         0.0               0.0               0.0   \n",
       "\n",
       "mutation   rpoB_p.Glu297Asp  rpoB_p.Glu423Gly  rpoB_p.His445Phe  \\\n",
       "sample_id                                                         \n",
       "809                     0.0               0.0               0.0   \n",
       "812                     0.0               0.0               0.0   \n",
       "815                     0.0               0.0               0.0   \n",
       "816                     0.0               0.0               0.0   \n",
       "817                     0.0               0.0               0.0   \n",
       "...                     ...               ...               ...   \n",
       "933876                  0.0               0.0               0.0   \n",
       "946364                  0.0               0.0               0.0   \n",
       "946365                  0.0               0.0               0.0   \n",
       "946366                  0.0               0.0               0.0   \n",
       "946367                  0.0               0.0               0.0   \n",
       "\n",
       "mutation   rpoB_p.His445_Lys446del  ...  rpoB_p.Pro439Ala  rpoB_p.Pro439Leu  \\\n",
       "sample_id                           ...                                       \n",
       "809                            0.0  ...               0.0               0.0   \n",
       "812                            0.0  ...               0.0               0.0   \n",
       "815                            0.0  ...               0.0               0.0   \n",
       "816                            0.0  ...               0.0               0.0   \n",
       "817                            0.0  ...               0.0               0.0   \n",
       "...                            ...  ...               ...               ...   \n",
       "933876                         0.0  ...               0.0               0.0   \n",
       "946364                         0.0  ...               0.0               0.0   \n",
       "946365                         0.0  ...               0.0               0.0   \n",
       "946366                         0.0  ...               0.0               0.0   \n",
       "946367                         0.0  ...               0.0               0.0   \n",
       "\n",
       "mutation   rpoB_p.Ser431_Met434del  rpoB_p.Ser431dup  rpoB_p.Ser450Ala  \\\n",
       "sample_id                                                                \n",
       "809                            0.0               0.0               0.0   \n",
       "812                            0.0               0.0               0.0   \n",
       "815                            0.0               0.0               0.0   \n",
       "816                            0.0               0.0               0.0   \n",
       "817                            0.0               0.0               0.0   \n",
       "...                            ...               ...               ...   \n",
       "933876                         0.0               0.0               0.0   \n",
       "946364                         0.0               0.0               0.0   \n",
       "946365                         0.0               0.0               0.0   \n",
       "946366                         0.0               0.0               0.0   \n",
       "946367                         0.0               0.0               0.0   \n",
       "\n",
       "mutation   rpoB_p.Ser450Asn  rpoB_p.Ser450_Leu452del  rpoB_p.Thr427Gly  \\\n",
       "sample_id                                                                \n",
       "809                     0.0                      0.0               0.0   \n",
       "812                     0.0                      0.0               0.0   \n",
       "815                     0.0                      0.0               0.0   \n",
       "816                     0.0                      0.0               0.0   \n",
       "817                     0.0                      0.0               0.0   \n",
       "...                     ...                      ...               ...   \n",
       "933876                  0.0                      0.0               0.0   \n",
       "946364                  0.0                      0.0               0.0   \n",
       "946365                  0.0                      0.0               0.0   \n",
       "946366                  0.0                      0.0               0.0   \n",
       "946367                  0.0                      0.0               0.0   \n",
       "\n",
       "mutation   rpoB_p.Thr427Pro  rpoB_p.Thr427_Ser428del  \n",
       "sample_id                                             \n",
       "809                     0.0                      0.0  \n",
       "812                     0.0                      0.0  \n",
       "815                     0.0                      0.0  \n",
       "816                     0.0                      0.0  \n",
       "817                     0.0                      0.0  \n",
       "...                     ...                      ...  \n",
       "933876                  0.0                      0.0  \n",
       "946364                  0.0                      0.0  \n",
       "946365                  0.0                      0.0  \n",
       "946366                  0.0                      0.0  \n",
       "946367                  0.0                      0.0  \n",
       "\n",
       "[35362 rows x 28 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_matrix[model_matrix.columns[((model_matrix == 0).all())]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "73746e03-3f9b-42ee-ac68-f0f2e82b83b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = {'fit_time': np.array([0.46307302, 0.46552658, 0.46767092, 0.45148492, 0.46610522]), \n",
    "       'score_time': np.array([0.01107359, 0.00999832, 0.01002049, 0.00967932, 0.01317215]), \n",
    "       'test_roc_auc': np.array([0.97973397, 0.98943888, 0.95101111, 0.95274882, 0.96620066]), \n",
    "       'test_recall': np.array([0.95521739, 0.95523685, 0.89217391, 0.90434783, 0.93217391]), \n",
    "       'test_accuracy': np.array([0.97737876, 0.97879259, 0.94937783, 0.95220588, 0.95616516]), \n",
    "       'test_balanced_accuracy': np.array([0.97163761, 0.97269387, 0.93456139, 0.93981012, 0.94995116])\n",
    "      }\n",
    "\n",
    "res['test_spec'] = 2 * res['test_balanced_accuracy'] - res['test_recall']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "33e78587-ee3d-4fd2-86b5-9ec933b0ca91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.967826688"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res['test_roc_auc'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "be90a6f5-3a3e-442a-bb5c-f4a349da6a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9278299779999999"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res['test_recall'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f8c0ef68-994f-446b-8de7-6fead8a70b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9796316820000002"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res['test_spec'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "02aba91f-d66b-4cf1-bb31-9dd91fdeb281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.962784044"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res['test_accuracy'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "999e9cc0-b4ec-4cf5-b9e4-62587ead0395",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9537308300000001"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res['test_balanced_accuracy'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7389b31-18d7-4d55-ac38-ec7ec6424c54",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a1777ff8-c722-4558-804c-fae56eab6141",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model for 35401 WHO phenotypes on 231 Tier 1+2 mutations!\n"
     ]
    }
   ],
   "source": [
    "drug = \"Rifampicin\"\n",
    "config_file = \"../config_files/binary_07.yaml\"\n",
    "kwargs = yaml.safe_load(open(config_file))\n",
    "\n",
    "analysis_dir = kwargs[\"output_dir\"]\n",
    "num_PCs = kwargs[\"num_PCs\"]\n",
    "tiers_lst = kwargs[\"tiers_lst\"]\n",
    "pheno_category_lst = kwargs[\"pheno_category_lst\"]\n",
    "\n",
    "# make sure that both phenotypes are included\n",
    "if \"ALL\" in pheno_category_lst:\n",
    "    phenos_name = \"ALL\"\n",
    "    pheno_category_lst = [\"ALL\", \"WHO\"]\n",
    "else:\n",
    "    phenos_name = \"WHO\"\n",
    "    \n",
    "mutations = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY\", \"significant_tiers=1+2_variants.csv\"))\n",
    "mutations[\"Tier\"] = mutations[\"Tier\"].astype(str)\n",
    "mutations = mutations.query(\"Tier in @tiers_lst\").mutation.values\n",
    "\n",
    "    \n",
    "############# STEP 1: READ IN THE PREVIOUSLY GENERATED MATRICES #############\n",
    "    \n",
    "    \n",
    "df_phenos = pd.read_csv(os.path.join(analysis_dir, drug, \"phenos_binary.csv\")).query(\"phenotypic_category in @pheno_category_lst\").set_index(\"sample_id\")\n",
    "print(f\"Fitting model for {len(df_phenos)} {phenos_name} phenotypes on {len(mutations)} Tier {'+'.join(tiers_lst)} mutations!\")\n",
    "\n",
    "# use the model matrix from the unpooled model\n",
    "matrix = pd.read_pickle(os.path.join(analysis_dir, drug, f\"BINARY/tiers={'+'.join(tiers_lst)}/phenos={phenos_name}/dropAF_noSyn_unpooled/model_matrix.pkl\"))\n",
    "\n",
    "if len(set(mutations) - set(matrix.columns)) > 0:\n",
    "    #raise ValueError(f\"Mutations {set(mutations) - set(matrix.columns)} are not in the model matrix\")\n",
    "    mutations = list(set(mutations).intersection(matrix.columns))\n",
    "    \n",
    "matrix = matrix[mutations]\n",
    "\n",
    "# Read in the PC coordinates dataframe, then keep only the desired number of principal components\n",
    "eigenvec_df = pd.read_csv(\"../data/eigenvec_10PC.csv\", index_col=[0]).iloc[:, :num_PCs]\n",
    "matrix = matrix.merge(eigenvec_df, left_index=True, right_index=True, how=\"left\")\n",
    "\n",
    "df_phenos = df_phenos.loc[matrix.index]\n",
    "assert sum(matrix.index != df_phenos.index.values) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e2bf22ce-e0ad-4b87-8085-89e173f72a98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rpoB_p.His835Arg</th>\n",
       "      <th>rpoB_p.Leu378Arg</th>\n",
       "      <th>rpoB_p.Leu430_Ser431insArg</th>\n",
       "      <th>rpoC_p.Leu507Val</th>\n",
       "      <th>rpoB_p.Ser450Met</th>\n",
       "      <th>Rv2752c_p.Ser189*</th>\n",
       "      <th>rpoC_p.Pro1040Thr</th>\n",
       "      <th>rpoB_p.Gln432Glu</th>\n",
       "      <th>rpoB_p.Leu443Trp</th>\n",
       "      <th>rpoB_p.Asp435Tyr</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoC_p.Glu750Asp</th>\n",
       "      <th>rpoC_p.His525Asn</th>\n",
       "      <th>rpoA_p.Gly31Ala</th>\n",
       "      <th>rpoB_p.Pro700Ser</th>\n",
       "      <th>rpoC_p.Thr812Ile</th>\n",
       "      <th>PC0</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>0.007680</td>\n",
       "      <td>-0.007175</td>\n",
       "      <td>-0.000713</td>\n",
       "      <td>-0.000181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.007675</td>\n",
       "      <td>-0.007189</td>\n",
       "      <td>-0.000707</td>\n",
       "      <td>-0.000202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>0.007654</td>\n",
       "      <td>-0.007125</td>\n",
       "      <td>-0.000733</td>\n",
       "      <td>-0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.007694</td>\n",
       "      <td>-0.007159</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>-0.000184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>0.007686</td>\n",
       "      <td>-0.007170</td>\n",
       "      <td>-0.000709</td>\n",
       "      <td>-0.000187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>933876</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004837</td>\n",
       "      <td>-0.004983</td>\n",
       "      <td>-0.001063</td>\n",
       "      <td>-0.000114</td>\n",
       "      <td>0.000083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946364</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004849</td>\n",
       "      <td>-0.004961</td>\n",
       "      <td>-0.001047</td>\n",
       "      <td>-0.000157</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946365</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.004965</td>\n",
       "      <td>-0.002993</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.006936</td>\n",
       "      <td>-0.000773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946366</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001759</td>\n",
       "      <td>0.003259</td>\n",
       "      <td>0.010861</td>\n",
       "      <td>-0.000595</td>\n",
       "      <td>-0.000034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946367</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004850</td>\n",
       "      <td>-0.004960</td>\n",
       "      <td>-0.001035</td>\n",
       "      <td>-0.000162</td>\n",
       "      <td>0.000058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30984 rows × 233 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rpoB_p.His835Arg  rpoB_p.Leu378Arg  rpoB_p.Leu430_Ser431insArg  \\\n",
       "sample_id                                                                   \n",
       "809                     0.0               0.0                         0.0   \n",
       "812                     0.0               0.0                         0.0   \n",
       "815                     0.0               0.0                         0.0   \n",
       "816                     0.0               0.0                         0.0   \n",
       "817                     0.0               0.0                         0.0   \n",
       "...                     ...               ...                         ...   \n",
       "933876                  0.0               0.0                         0.0   \n",
       "946364                  0.0               0.0                         0.0   \n",
       "946365                  0.0               0.0                         0.0   \n",
       "946366                  0.0               0.0                         0.0   \n",
       "946367                  0.0               0.0                         0.0   \n",
       "\n",
       "           rpoC_p.Leu507Val  rpoB_p.Ser450Met  Rv2752c_p.Ser189*  \\\n",
       "sample_id                                                          \n",
       "809                     0.0               0.0                0.0   \n",
       "812                     0.0               0.0                0.0   \n",
       "815                     0.0               0.0                0.0   \n",
       "816                     0.0               0.0                0.0   \n",
       "817                     0.0               0.0                0.0   \n",
       "...                     ...               ...                ...   \n",
       "933876                  0.0               0.0                0.0   \n",
       "946364                  0.0               0.0                0.0   \n",
       "946365                  0.0               0.0                0.0   \n",
       "946366                  0.0               0.0                0.0   \n",
       "946367                  0.0               0.0                0.0   \n",
       "\n",
       "           rpoC_p.Pro1040Thr  rpoB_p.Gln432Glu  rpoB_p.Leu443Trp  \\\n",
       "sample_id                                                          \n",
       "809                      0.0               0.0               0.0   \n",
       "812                      0.0               0.0               0.0   \n",
       "815                      0.0               0.0               0.0   \n",
       "816                      0.0               0.0               0.0   \n",
       "817                      0.0               0.0               0.0   \n",
       "...                      ...               ...               ...   \n",
       "933876                   0.0               0.0               0.0   \n",
       "946364                   0.0               0.0               0.0   \n",
       "946365                   0.0               0.0               0.0   \n",
       "946366                   0.0               0.0               0.0   \n",
       "946367                   0.0               0.0               0.0   \n",
       "\n",
       "           rpoB_p.Asp435Tyr  ...  rpoC_p.Glu750Asp  rpoC_p.His525Asn  \\\n",
       "sample_id                    ...                                       \n",
       "809                     0.0  ...               0.0               0.0   \n",
       "812                     0.0  ...               0.0               0.0   \n",
       "815                     0.0  ...               0.0               0.0   \n",
       "816                     0.0  ...               0.0               0.0   \n",
       "817                     0.0  ...               0.0               0.0   \n",
       "...                     ...  ...               ...               ...   \n",
       "933876                  0.0  ...               0.0               0.0   \n",
       "946364                  0.0  ...               0.0               0.0   \n",
       "946365                  0.0  ...               0.0               0.0   \n",
       "946366                  0.0  ...               0.0               0.0   \n",
       "946367                  0.0  ...               0.0               0.0   \n",
       "\n",
       "           rpoA_p.Gly31Ala  rpoB_p.Pro700Ser  rpoC_p.Thr812Ile       PC0  \\\n",
       "sample_id                                                                  \n",
       "809                    0.0               0.0               0.0  0.000251   \n",
       "812                    0.0               0.0               0.0  0.000240   \n",
       "815                    0.0               0.0               0.0  0.000184   \n",
       "816                    0.0               0.0               0.0  0.000209   \n",
       "817                    0.0               0.0               0.0  0.000235   \n",
       "...                    ...               ...               ...       ...   \n",
       "933876                 0.0               0.0               0.0  0.004837   \n",
       "946364                 0.0               0.0               0.0  0.004849   \n",
       "946365                 0.0               0.0               0.0 -0.004965   \n",
       "946366                 0.0               0.0               0.0  0.001759   \n",
       "946367                 0.0               0.0               0.0  0.004850   \n",
       "\n",
       "                PC1       PC2       PC3       PC4  \n",
       "sample_id                                          \n",
       "809        0.007680 -0.007175 -0.000713 -0.000181  \n",
       "812        0.007675 -0.007189 -0.000707 -0.000202  \n",
       "815        0.007654 -0.007125 -0.000733 -0.000033  \n",
       "816        0.007694 -0.007159 -0.000706 -0.000184  \n",
       "817        0.007686 -0.007170 -0.000709 -0.000187  \n",
       "...             ...       ...       ...       ...  \n",
       "933876    -0.004983 -0.001063 -0.000114  0.000083  \n",
       "946364    -0.004961 -0.001047 -0.000157  0.000073  \n",
       "946365    -0.002993  0.000277  0.006936 -0.000773  \n",
       "946366     0.003259  0.010861 -0.000595 -0.000034  \n",
       "946367    -0.004960 -0.001035 -0.000162  0.000058  \n",
       "\n",
       "[30984 rows x 233 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc8ef35-e6a4-457e-8e23-30fb66d2c98e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "07de7321-bf6a-40ca-9cd5-51a5df1b16c0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# RIFAMPICIN TEST CASE\n",
    "\n",
    "# 0. Original Analysis: 137 new significant resistance-associated variants\n",
    "\n",
    "## But, there are 152 with OR > 1 and BH p-val < 0.01 that are in tier 2 genes (rpoB is the only tier 1)\n",
    "## These 152 mutations were studied in the next two analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d46b8cbd-3e49-4b4d-94fc-fb3f944f4ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = \"BINARY\"\n",
    "phenos_name = \"WHO\"\n",
    "drug = \"Rifampicin\"\n",
    "drug_WHO_abbr = \"RIF\"\n",
    "\n",
    "# get the number of high confidence resistance-associated \n",
    "tiers1_stats = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/tiers=1/phenos=WHO/dropAF_noSyn_unpooled\", \"model_analysis_with_stats.csv\"))\n",
    "tiers12_stats = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/tiers=1+2/phenos=WHO/dropAF_noSyn_unpooled\", \"model_analysis_with_stats.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4ef503fe-42a3-4981-842d-59e4a3a05a65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>confidence</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>NPV_LB</th>\n",
       "      <th>NPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>rpoB_p.Asn24Lys</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>759878</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.057381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.669910</td>\n",
       "      <td>0.679703</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000762</td>\n",
       "      <td>0.999845</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.999444</td>\n",
       "      <td>1.000034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>rpoB_p.Phe503Val</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761313</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.019053</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.669872</td>\n",
       "      <td>0.679665</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000484</td>\n",
       "      <td>0.999845</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.999743</td>\n",
       "      <td>1.000083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>rpoB_p.Asn501Lys</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761309</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.019053</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.669872</td>\n",
       "      <td>0.679665</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000484</td>\n",
       "      <td>0.999845</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.999743</td>\n",
       "      <td>1.000083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation  predicted_effect position                 confidence  \\\n",
       "58    rpoB_p.Asn24Lys  missense_variant   759878  3) Uncertain significance   \n",
       "196  rpoB_p.Phe503Val  missense_variant   761313  3) Uncertain significance   \n",
       "197  rpoB_p.Asn501Lys  missense_variant   761309  3) Uncertain significance   \n",
       "\n",
       "     Odds_Ratio  pval  BH_pval  Bonferroni_pval  Significant  Num_Isolates  \\\n",
       "58     1.057381   0.0      0.0              0.0            1             3   \n",
       "196    1.019053   0.0      0.0              0.0            1             1   \n",
       "197    1.019053   0.0      0.0              0.0            1             1   \n",
       "\n",
       "     ...    NPV_LB    NPV_UB   Sens_LB   Sens_UB   Spec_LB  Spec_UB  LR+_LB  \\\n",
       "58   ...  0.669910  0.679703  0.000054  0.000762  0.999845      1.0     NaN   \n",
       "196  ...  0.669872  0.679665  0.000002  0.000484  0.999845      1.0     NaN   \n",
       "197  ...  0.669872  0.679665  0.000002  0.000484  0.999845      1.0     NaN   \n",
       "\n",
       "     LR+_UB    LR-_LB    LR-_UB  \n",
       "58      inf  0.999444  1.000034  \n",
       "196     inf  0.999743  1.000083  \n",
       "197     inf  0.999743  1.000083  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiers1_stats.query(\"mutation in ['rpoB_p.Phe503Val', 'rpoB_p.Asn24Lys', 'rpoB_p.Asn501Lys']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "09e35f41-2ce5-4db8-9c0b-a749ee59c09f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_significant_genes_by_tier(tier1_model, tier2_model, LRT_res):\n",
    "\n",
    "#     print(f\"Total numbers of mutations: {tier1_model.shape}, {tier2_model.shape}\")\n",
    "\n",
    "#     tier1_significant = tier1_model.query(\"Odds_Ratio > 1 & BH_pval < 0.05\")\n",
    "#     tier2_significant = tier2_model.query(\"Odds_Ratio > 1 & BH_pval < 0.01\")\n",
    "\n",
    "#     print(f\"Numbers of significant mutations: {len(set(tier1_significant.mutation))}, {len(set(tier2_significant.mutation) - set(tier1_significant.mutation))}\")\n",
    "    \n",
    "#     combined_df =pd.concat([tier1_significant, tier2_significant], axis=0)\n",
    "#     combined_df[\"Tier\"] = [1]*len(tier1_significant) + [2]*len(tier2_significant)\n",
    "#     return combined_df.drop_duplicates(subset=\"mutation\", keep=\"first\").merge(LRT_res[[\"mutation\", 'LRT_pval', 'LRT_BH_pval', 'LRT_Bonferroni_pval']], on=\"mutation\", how=\"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5ac0441-6a40-4c56-98fd-a46ae44e6e8e",
   "metadata": {},
   "source": [
    "# 1. Excluded Samples Analysis: 54 new significant resistance-associated variants\n",
    "\n",
    "Removed ~5,000 samples that contain both high confidence rpoB mutations (Category 1 or 2 in 2021 catalogue) and any of the ~150 significant tier2 mutations. \n",
    "\n",
    "An L2-penalized regression was fit to determine the effect sizes of tier 2 mutations, independent of resistance-associated mutations they may occur with. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ee479d3c-82e4-4390-8956-620fefc7b84e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_df = pd.read_csv(os.path.join(analysis_dir, f\"{drug}/{folder}/exclude_comutation/WHOphenos_univariate_stats.csv\"))\n",
    "# exclude_df = exclude_df.loc[~exclude_df[\"mutation\"].str.contains(\"PC\")]\n",
    "# exclude_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']] = exclude_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']].astype(int)\n",
    "\n",
    "# exclude_df.query(\"Odds_Ratio > 1 & BH_pval < 0.01 & TP >= 5 & PPV_LB >= 0.25 & ~mutation.str.contains('rpoB')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1ce0aaa4-a55c-4658-ac8c-3836f6a67f53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_df.query(\"mutation=='rpoC_p.Glu1092Asp'\")[['mutation', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval',\n",
    "#        'Bonferroni_pval', 'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN',\n",
    "#        'FN', 'PPV_LB', 'PPV', 'PPV_UB', 'NPV', 'Sens', 'Spec']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b8b21b1-b6a4-4a9d-a822-5806d1916379",
   "metadata": {},
   "source": [
    "# 2. Likelihood Ratio Test: 5/152 of the Above Tier 2 Mutations are Significant\n",
    "\n",
    "152 tier 2 variants were found significant in the first analysis. We then performed a likelihood ratio test:\n",
    "\n",
    "A mutation is removed from the original model input matrix, and then a new model is fit. \n",
    "\n",
    "We then compare the log-likelihoods of the original model and the model with 1 less mutation. If the p-value is significant, it means that removing the mutation significantly changes the fit of the model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "ee7cd7fb-412b-4f34-9d91-23d88530bb1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Rifampicin/BINARY/tiers=1/phenos=WHO/dropAF_noSyn_unpooled'"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.join(analysis_dir, drug, f\"BINARY/tiers=1/phenos={phenos_name}/dropAF_noSyn_unpooled\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "84632d9d-fd85-4532-a823-15b8b33dde51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "815 2845\n"
     ]
    }
   ],
   "source": [
    "phenos_name = \"WHO\"\n",
    "out_dir = os.path.join(analysis_dir, drug, f\"BINARY/LRT/phenos={phenos_name}\")\n",
    "\n",
    "LRTresults_tier1 = pd.read_csv(os.path.join(out_dir, f\"results_tier={'1'}.csv\")).rename(columns={\"Unnamed: 0\": \"mutation\"})\n",
    "LRTresults_tier2 = pd.read_csv(os.path.join(out_dir, f\"results_tier={'2'}.csv\")).rename(columns={\"Unnamed: 0\": \"mutation\"})\n",
    "\n",
    "# remove the FULL model row, which is the first row\n",
    "LRTresults_tier1 = add_pval_corrections(LRTresults_tier1.iloc[1:, ])\n",
    "LRTresults_tier2 = add_pval_corrections(LRTresults_tier2.iloc[1:, ])\n",
    "\n",
    "tier1_model_permute = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/tiers=1/phenos={phenos_name}/dropAF_noSyn_unpooled/model_analysis.csv\")).query(\"~mutation.str.contains('PC')\")\n",
    "tier2_model_permute = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/tiers=1+2/phenos={phenos_name}/dropAF_noSyn_unpooled/model_analysis.csv\")).query(\"~mutation.str.contains('PC')\")\n",
    "\n",
    "# remove the tier 1 genes for the purposes of this analysis\n",
    "tier2_model_permute = tier2_model_permute.query(\"mutation not in @tier1_model_permute.mutation\")\n",
    "\n",
    "# check that all mutations in the permutation dataframe are in the LRT dataframe. The only difference should be the FULL model row\n",
    "# print(set(tier1_model_permute.mutation).symmetric_difference(LRTresults_tier1.index))\n",
    "# print(set(tier2_model_permute.mutation).symmetric_difference(LRTresults_tier2.index))\n",
    "assert len(set(tier1_model_permute.mutation).symmetric_difference(LRTresults_tier1.mutation)) == 0\n",
    "assert len(set(tier2_model_permute.mutation).symmetric_difference(LRTresults_tier2.mutation)) == 0\n",
    "\n",
    "# combine results into a single dataframe for easy searching\n",
    "tier1_model = tier1_model_permute.merge(LRTresults_tier1[[\"mutation\", \"pval\", \"BH_pval\", \"Bonferroni_pval\"]].rename(columns={\"pval\": \"LRT_pval\", \n",
    "                                                                                                               \"BH_pval\": \"LRT_BH_pval\", \n",
    "                                                                                                               \"Bonferroni_pval\": \"LRT_Bonferroni_pval\"\n",
    "                                                                                                              }), on=\"mutation\", how=\"left\")\n",
    "\n",
    "tier2_model = tier2_model_permute.merge(LRTresults_tier2[[\"mutation\", \"pval\", \"BH_pval\", \"Bonferroni_pval\"]].rename(columns={\"pval\": \"LRT_pval\", \n",
    "                                                                                                               \"BH_pval\": \"LRT_BH_pval\", \n",
    "                                                                                                               \"Bonferroni_pval\": \"LRT_Bonferroni_pval\"\n",
    "                                                                                                              }), on=\"mutation\", how=\"left\")\n",
    "\n",
    "print(len(tier1_model), len(tier2_model))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6c4aefb8-9837-49ef-aed9-28f732defb3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3) Uncertain significance    61\n",
       "NaN                          19\n",
       "2) Assoc w R - Interim       15\n",
       "5) Not assoc w R              5\n",
       "Name: confidence, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of Tier 1 mutations determined significant by original analysis, but not in LRT\n",
    "thresh = 0.05\n",
    "tier1_model.query(\"BH_pval < @thresh & LRT_BH_pval > @thresh\").confidence.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bb824e0e-3baf-48ea-aa9a-8a24cd23c58d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78 significant coef, but not LRT\n",
      "2 significant LRT, but not coef\n",
      "51 significant in both\n"
     ]
    }
   ],
   "source": [
    "# mutations picked up by the LRT, but not the coefficient association. \n",
    "# They are not significant in the tiers 1+2 model either \n",
    "thresh = 0.05\n",
    "\n",
    "print(f\"{len(tier1_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval > @thresh'))} significant coef, but not LRT\")\n",
    "print(f\"{len(tier1_model.query('Odds_Ratio > 1 & BH_pval > @thresh & LRT_BH_pval < @thresh'))} significant LRT, but not coef\")\n",
    "print(f\"{len(tier1_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval < @thresh'))} significant in both\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ac0dda9c-bbfb-4a08-973d-3923e83386ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.05\n",
    "len(tier1_model.query('Odds_Ratio > 1 & (BH_pval < @thresh & LRT_BH_pval < @thresh)'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4188cc4c-8f6c-45bc-81c2-7c235dadde4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "len(tier2_model.query('Odds_Ratio > 1 & (BH_pval < @thresh & LRT_BH_pval < @thresh)'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6751efcd-707c-44f3-b0a2-7088fcc1e4eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.195840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.216332</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>5.033098e-07</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.001432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Rv2752c_p.Asn30Ser</td>\n",
       "      <td>0.099958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.105124</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.480226e-07</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.000421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              mutation      coef  pval  BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "2    rpoC_p.Glu1092Asp  0.195840   0.0      0.0              0.0    1.216332   \n",
       "19  Rv2752c_p.Asn30Ser  0.099958   0.0      0.0              0.0    1.105124   \n",
       "\n",
       "                   confidence      LRT_pval  LRT_BH_pval  LRT_Bonferroni_pval  \n",
       "2            5) Not assoc w R  5.033098e-07     0.000716             0.001432  \n",
       "19  3) Uncertain significance  1.480226e-07     0.000421             0.000421  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tier2_model.query('Odds_Ratio > 1 & (BH_pval < @thresh & LRT_BH_pval < @thresh)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f388f24e-4500-4f2a-81f2-f4953b70a34a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>glpK_p.Val192fs</td>\n",
       "      <td>0.026607</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.262114</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.026964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mutation      coef   pval   BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "203  glpK_p.Val192fs  0.026607  0.033  0.262114              1.0    1.026964   \n",
       "\n",
       "    confidence  LRT_pval  LRT_BH_pval  LRT_Bonferroni_pval  \n",
       "203        NaN       1.0          1.0                  1.0  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this was the other mutation they drew attention to\n",
    "tier2_model.query(\"mutation=='glpK_p.Val192fs'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "68e95ff2-7e0a-4ca7-a166-b99c61dc9b15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "98 significant coef, but not LRT\n",
      "0 significant LRT, but not coef\n",
      "2 significant in both\n"
     ]
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "\n",
    "print(f\"{len(tier2_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval > @thresh'))} significant coef, but not LRT\")\n",
    "print(f\"{len(tier2_model.query('Odds_Ratio > 1 & BH_pval > @thresh & LRT_BH_pval < @thresh'))} significant LRT, but not coef\")\n",
    "print(f\"{len(tier2_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval < @thresh'))} significant in both\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0e2d22d8-713c-4927-9596-a90281a1088f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>confidence</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>Total_Isolates</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>TN</th>\n",
       "      <th>FN</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR-</th>\n",
       "      <th>PPV_LB</th>\n",
       "      <th>PPV_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>1.216332</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2697</td>\n",
       "      <td>30984</td>\n",
       "      <td>1711</td>\n",
       "      <td>986</td>\n",
       "      <td>19935</td>\n",
       "      <td>8352</td>\n",
       "      <td>0.634409</td>\n",
       "      <td>0.704741</td>\n",
       "      <td>0.170029</td>\n",
       "      <td>0.95287</td>\n",
       "      <td>3.60768</td>\n",
       "      <td>0.871022</td>\n",
       "      <td>0.615909</td>\n",
       "      <td>0.652616</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation  predicted_effect        confidence  Odds_Ratio  \\\n",
       "16  rpoC_p.Glu1092Asp  missense_variant  5) Not assoc w R    1.216332   \n",
       "\n",
       "    BH_pval  Num_Isolates  Total_Isolates    TP   FP     TN    FN       PPV  \\\n",
       "16      0.0          2697           30984  1711  986  19935  8352  0.634409   \n",
       "\n",
       "         NPV      Sens     Spec      LR+       LR-    PPV_LB    PPV_UB  \n",
       "16  0.704741  0.170029  0.95287  3.60768  0.871022  0.615909  0.652616  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiers12_stats.query(\"mutation=='rpoC_p.Glu1092Asp'\")[['mutation', 'predicted_effect', 'confidence', 'Odds_Ratio',\n",
    "       'BH_pval', 'Num_Isolates',\n",
    "       'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV', 'NPV', 'Sens', 'Spec',\n",
    "       'LR+', 'LR-', 'PPV_LB', 'PPV_UB']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0399c22c-5542-402a-9bf4-82c1d86c1646",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rpoC_p.Glu1092Asp', 'Rv2752c_p.Asn30Ser'], dtype=object)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "tier2_model.query('BH_pval < @thresh & LRT_BH_pval < @thresh')[\"mutation\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "3092d178-f2a7-4456-9ca6-bba7d2694307",
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh = 0.05\n",
    "tier1_sig_variants = tier1_model.query('Odds_Ratio > 1 & (BH_pval < @thresh | LRT_BH_pval < @thresh)')[\"mutation\"].values\n",
    "\n",
    "thresh = 0.01\n",
    "tier2_sig_variants = tier2_model.query('Odds_Ratio > 1 & (BH_pval < @thresh | LRT_BH_pval < @thresh)')[\"mutation\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b938be6e-4867-4039-ad34-91c9e7a2f53b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(131, 100)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tier1_sig_variants), len(tier2_sig_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "480d217d-0b8d-4fda-9514-54002dca8e4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "231\n"
     ]
    }
   ],
   "source": [
    "combined_variants = np.concatenate([tier1_sig_variants, tier2_sig_variants])\n",
    "print(len(np.unique(combined_variants)))\n",
    "        \n",
    "combined_variants_df = pd.DataFrame({\"mutation\": combined_variants, \"Tier\": [\"1\"]*len(tier1_sig_variants) + [\"2\"]*len(tier2_sig_variants)})\n",
    "\n",
    "combined_variants_df.to_csv(os.path.join(analysis_dir, drug, \"BINARY\", \"significant_tiers=1+2_variants.csv\"), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "cfa61ac4-3456-4324-a9c7-603c05f8b431",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>Tier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpoB_p.Asp435Val</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rpoB_p.His445Tyr</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rpoB_p.His445Asp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rpoB_p.Ser450Trp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>glpK_c.-7A&gt;G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Rv2477c_p.Phe429Leu</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>glpK_p.Asp215Gly</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>Rv2752c_p.Ser189*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>rpoC_p.Gln262His</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                mutation Tier\n",
       "0       rpoB_p.Ser450Leu    1\n",
       "1       rpoB_p.Asp435Val    1\n",
       "2       rpoB_p.His445Tyr    1\n",
       "3       rpoB_p.His445Asp    1\n",
       "4       rpoB_p.Ser450Trp    1\n",
       "..                   ...  ...\n",
       "226         glpK_c.-7A>G    2\n",
       "227  Rv2477c_p.Phe429Leu    2\n",
       "228     glpK_p.Asp215Gly    2\n",
       "229    Rv2752c_p.Ser189*    2\n",
       "230     rpoC_p.Gln262His    2\n",
       "\n",
       "[231 rows x 2 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a65ddcec-4625-4a8e-9ca4-74dd7d70af8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "71605d31-6ea9-4607-93d1-50075f50b815",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Rifampicin/BINARY/significant_tiers=1+2_variants.csv'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.join(analysis_dir, drug, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bee0539-739c-44bf-9a4c-88ed59f6e995",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis/prediction_model.py config_files/binary_03.yaml Rifampicin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5ebc9888-8189-4bb3-8b9f-1a0f83d53b79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model for 35401 WHO phenotypes on 131 mutations!\n"
     ]
    }
   ],
   "source": [
    "config_file = \"../config_files/binary_03.yaml\"\n",
    "kwargs = yaml.safe_load(open(config_file))\n",
    "\n",
    "significant_mut_file = os.path.join(analysis_dir, drug, \"BINARY\", \"significant_tiers=1+2_variants.csv\")\n",
    "\n",
    "analysis_dir = kwargs[\"output_dir\"]\n",
    "num_PCs = kwargs[\"num_PCs\"]\n",
    "tiers_lst = kwargs[\"tiers_lst\"]\n",
    "pheno_category_lst = kwargs[\"pheno_category_lst\"]\n",
    "\n",
    "# make sure that both phenotypes are included\n",
    "if \"ALL\" in pheno_category_lst:\n",
    "    phenos_name = \"ALL\"\n",
    "    pheno_category_lst = [\"ALL\", \"WHO\"]\n",
    "else:\n",
    "    phenos_name = \"WHO\"\n",
    "\n",
    "    \n",
    "df_phenos = pd.read_csv(os.path.join(analysis_dir, drug, \"phenos_binary.csv\")).query(\"phenotypic_category in @pheno_category_lst\").set_index(\"sample_id\")\n",
    "mutations = pd.read_csv(significant_mut_file)\n",
    "mutations[\"Tier\"] = mutations[\"Tier\"].astype(str)\n",
    "mutations = mutations.query(\"Tier in @tiers_lst\")\n",
    "print(f\"Fitting model for {len(df_phenos)} {phenos_name} phenotypes on {len(mutations)} mutations!\")\n",
    "\n",
    "# read in only the genotypes files for the tiers for this model\n",
    "df_model = pd.concat([pd.read_csv(os.path.join(analysis_dir, drug, f\"genos_{num}.csv.gz\"), compression=\"gzip\", low_memory=False) for num in tiers_lst])\n",
    "\n",
    "# then keep only samples of the desired phenotype\n",
    "df_model = df_model.loc[df_model[\"sample_id\"].isin(df_phenos.index)]\n",
    "\n",
    "# keep only mutations for the small model\n",
    "df_model[\"mutation\"] = df_model[\"resolved_symbol\"] + \"_\" + df_model[\"variant_category\"]\n",
    "df_model = df_model.query(\"mutation in @mutations.mutation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "bf3db64d-8d49-42a6-9d65-70fbdf9b4f7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35401, 131)\n",
      "(35376, 131)\n"
     ]
    }
   ],
   "source": [
    "matrix = df_model.pivot(index=\"sample_id\", columns=\"mutation\", values=\"variant_binary_status\")\n",
    "del df_model\n",
    "print(matrix.shape)\n",
    "\n",
    "matrix = matrix.dropna(axis=0)\n",
    "print(matrix.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cf9e189b-eb7d-4207-9915-7a27a6f65335",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35376, 131)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix[matrix.columns[~((matrix == 0).all())]].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a5dc4845-0a0c-4f74-850d-c86ee06fa990",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the PC coordinates dataframe, then keep only the desired number of principal components\n",
    "eigenvec_df = pd.read_csv(\"../data/eigenvec_10PC.csv\", index_col=[0]).iloc[:, :num_PCs]\n",
    "matrix = matrix.merge(eigenvec_df, left_index=True, right_index=True, how=\"left\")\n",
    "\n",
    "df_phenos = df_phenos.loc[matrix.index]\n",
    "assert sum(matrix.index != df_phenos.index.values) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "59c2aeb5-ce1f-4954-a58f-4b4b90f0e117",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35376, 136)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c3b27aa1-3164-4000-abc4-d6597396a741",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rpoB_c.-142C&gt;T</th>\n",
       "      <th>rpoB_c.-165C&gt;T</th>\n",
       "      <th>rpoB_c.-179C&gt;T</th>\n",
       "      <th>rpoB_c.-180A&gt;C</th>\n",
       "      <th>rpoB_c.-209C&gt;G</th>\n",
       "      <th>rpoB_c.-21G&gt;C</th>\n",
       "      <th>rpoB_p.Ala286Val</th>\n",
       "      <th>rpoB_p.Ala2Thr</th>\n",
       "      <th>rpoB_p.Ala584Asp</th>\n",
       "      <th>rpoB_p.Ala584Gly</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB_p.Val26Ala</th>\n",
       "      <th>rpoB_p.Val496Ala</th>\n",
       "      <th>rpoB_p.Val496Met</th>\n",
       "      <th>rpoB_p.Val534Met</th>\n",
       "      <th>rpoB_p.Val562Ala</th>\n",
       "      <th>PC0</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>0.007680</td>\n",
       "      <td>-0.007175</td>\n",
       "      <td>-0.000713</td>\n",
       "      <td>-0.000181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.007675</td>\n",
       "      <td>-0.007189</td>\n",
       "      <td>-0.000707</td>\n",
       "      <td>-0.000202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>0.007654</td>\n",
       "      <td>-0.007125</td>\n",
       "      <td>-0.000733</td>\n",
       "      <td>-0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.007694</td>\n",
       "      <td>-0.007159</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>-0.000184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>0.007686</td>\n",
       "      <td>-0.007170</td>\n",
       "      <td>-0.000709</td>\n",
       "      <td>-0.000187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>933876</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004837</td>\n",
       "      <td>-0.004983</td>\n",
       "      <td>-0.001063</td>\n",
       "      <td>-0.000114</td>\n",
       "      <td>0.000083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946364</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004849</td>\n",
       "      <td>-0.004961</td>\n",
       "      <td>-0.001047</td>\n",
       "      <td>-0.000157</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946365</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.004965</td>\n",
       "      <td>-0.002993</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.006936</td>\n",
       "      <td>-0.000773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946366</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001759</td>\n",
       "      <td>0.003259</td>\n",
       "      <td>0.010861</td>\n",
       "      <td>-0.000595</td>\n",
       "      <td>-0.000034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946367</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004850</td>\n",
       "      <td>-0.004960</td>\n",
       "      <td>-0.001035</td>\n",
       "      <td>-0.000162</td>\n",
       "      <td>0.000058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35376 rows × 136 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rpoB_c.-142C>T  rpoB_c.-165C>T  rpoB_c.-179C>T  rpoB_c.-180A>C  \\\n",
       "sample_id                                                                   \n",
       "809                   0.0             0.0             0.0             0.0   \n",
       "812                   0.0             0.0             0.0             0.0   \n",
       "815                   0.0             0.0             0.0             0.0   \n",
       "816                   0.0             0.0             0.0             0.0   \n",
       "817                   0.0             0.0             0.0             0.0   \n",
       "...                   ...             ...             ...             ...   \n",
       "933876                0.0             0.0             0.0             0.0   \n",
       "946364                0.0             0.0             0.0             0.0   \n",
       "946365                0.0             0.0             0.0             0.0   \n",
       "946366                0.0             0.0             0.0             0.0   \n",
       "946367                0.0             0.0             0.0             0.0   \n",
       "\n",
       "           rpoB_c.-209C>G  rpoB_c.-21G>C  rpoB_p.Ala286Val  rpoB_p.Ala2Thr  \\\n",
       "sample_id                                                                    \n",
       "809                   0.0            0.0               0.0             0.0   \n",
       "812                   0.0            0.0               0.0             0.0   \n",
       "815                   0.0            0.0               0.0             0.0   \n",
       "816                   0.0            0.0               0.0             0.0   \n",
       "817                   0.0            0.0               0.0             0.0   \n",
       "...                   ...            ...               ...             ...   \n",
       "933876                0.0            0.0               0.0             0.0   \n",
       "946364                0.0            0.0               0.0             0.0   \n",
       "946365                0.0            0.0               0.0             0.0   \n",
       "946366                0.0            0.0               0.0             0.0   \n",
       "946367                0.0            0.0               0.0             0.0   \n",
       "\n",
       "           rpoB_p.Ala584Asp  rpoB_p.Ala584Gly  ...  rpoB_p.Val26Ala  \\\n",
       "sample_id                                      ...                    \n",
       "809                     0.0               0.0  ...              0.0   \n",
       "812                     0.0               0.0  ...              0.0   \n",
       "815                     0.0               0.0  ...              0.0   \n",
       "816                     0.0               0.0  ...              0.0   \n",
       "817                     0.0               0.0  ...              0.0   \n",
       "...                     ...               ...  ...              ...   \n",
       "933876                  0.0               0.0  ...              0.0   \n",
       "946364                  0.0               0.0  ...              0.0   \n",
       "946365                  0.0               0.0  ...              0.0   \n",
       "946366                  0.0               0.0  ...              0.0   \n",
       "946367                  0.0               0.0  ...              0.0   \n",
       "\n",
       "           rpoB_p.Val496Ala  rpoB_p.Val496Met  rpoB_p.Val534Met  \\\n",
       "sample_id                                                         \n",
       "809                     0.0               0.0               0.0   \n",
       "812                     0.0               0.0               0.0   \n",
       "815                     0.0               0.0               0.0   \n",
       "816                     0.0               0.0               0.0   \n",
       "817                     0.0               0.0               0.0   \n",
       "...                     ...               ...               ...   \n",
       "933876                  0.0               0.0               0.0   \n",
       "946364                  0.0               0.0               0.0   \n",
       "946365                  0.0               0.0               0.0   \n",
       "946366                  0.0               0.0               0.0   \n",
       "946367                  0.0               0.0               0.0   \n",
       "\n",
       "           rpoB_p.Val562Ala       PC0       PC1       PC2       PC3       PC4  \n",
       "sample_id                                                                      \n",
       "809                     0.0  0.000251  0.007680 -0.007175 -0.000713 -0.000181  \n",
       "812                     0.0  0.000240  0.007675 -0.007189 -0.000707 -0.000202  \n",
       "815                     0.0  0.000184  0.007654 -0.007125 -0.000733 -0.000033  \n",
       "816                     0.0  0.000209  0.007694 -0.007159 -0.000706 -0.000184  \n",
       "817                     0.0  0.000235  0.007686 -0.007170 -0.000709 -0.000187  \n",
       "...                     ...       ...       ...       ...       ...       ...  \n",
       "933876                  0.0  0.004837 -0.004983 -0.001063 -0.000114  0.000083  \n",
       "946364                  0.0  0.004849 -0.004961 -0.001047 -0.000157  0.000073  \n",
       "946365                  0.0 -0.004965 -0.002993  0.000277  0.006936 -0.000773  \n",
       "946366                  0.0  0.001759  0.003259  0.010861 -0.000595 -0.000034  \n",
       "946367                  0.0  0.004850 -0.004960 -0.001035 -0.000162  0.000058  \n",
       "\n",
       "[35376 rows x 136 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "73ebfcf9-88e1-41bd-9940-4842324b14b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotypic_category</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>933876</th>\n",
       "      <td>WHO</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946364</th>\n",
       "      <td>WHO</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946365</th>\n",
       "      <td>WHO</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946366</th>\n",
       "      <td>WHO</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946367</th>\n",
       "      <td>WHO</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35376 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          phenotypic_category  phenotype\n",
       "sample_id                               \n",
       "809                       WHO          0\n",
       "812                       WHO          0\n",
       "815                       WHO          0\n",
       "816                       WHO          0\n",
       "817                       WHO          0\n",
       "...                       ...        ...\n",
       "933876                    WHO          1\n",
       "946364                    WHO          1\n",
       "946365                    WHO          1\n",
       "946366                    WHO          1\n",
       "946367                    WHO          1\n",
       "\n",
       "[35376 rows x 2 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6bc0e53-cb85-4a70-8d7f-abf3b47ce6a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0548d40-6492-4fc0-9aaf-9cdfae183a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_model[\"mutation\"] = df_model[\"resolved_symbol\"] + \"_\" + df_model[\"variant_category\"]\n",
    "df_model = df_model.query(\"mutation in @mutations.mutation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03339f29-60f9-4b4b-999f-a5ba84b2d9bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ed83c0d-3096-4273-b4a4-4c570839dc18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eca2d07-5f1e-430c-b621-30fea69cd230",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "0ff3b76f-2bbe-431f-89c9-7769e62cde44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Rifampicin/BINARY'"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config_file = \"config_files/\"\n",
    "kwargs = yaml.safe_load(open(config_file))\n",
    "\n",
    "analysis_dir = kwargs[\"output_dir\"]\n",
    "num_PCs = kwargs[\"num_PCs\"]\n",
    "\n",
    "pheno_category_lst = kwargs[\"pheno_category_lst\"]\n",
    "# make sure that both phenotypes are included\n",
    "if \"ALL\" in pheno_category_lst:\n",
    "    phenos_name = \"ALL\"\n",
    "    pheno_category_lst = [\"ALL\", \"WHO\"]\n",
    "else:\n",
    "    phenos_name = \"WHO\"\n",
    "\n",
    "    \n",
    "mutations_lst = pd.read_csv(significat_mut_path, sep=\"\\t\", header=None)[0].values\n",
    "print(f\"Fitting model for {phenos_name} phenotypes on {len(mutations_lst)} mutations!\")\n",
    "os.path.join(analysis_dir, drug, \"BINARY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce02d13-9ceb-4dea-a09b-c87c5498d739",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "651dd7e7-b428-4b87-bc0f-5946578a8e19",
   "metadata": {},
   "source": [
    "# Perform bootstrapping for all 152 mutations to see how much AUC, Sens, Spec, and Accuracy Change\n",
    "\n",
    "If the confidence interval lies above 0, it means that removing the feature SIGNIFICANTLY DECREASES THE PREDICTIVE POWER OF THE MODEL\n",
    "\n",
    "CI > 0 = FEATURE IS IMPORTANT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b9d72b2-6e0e-4bc7-b8ad-39cf938671da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # bootstrapped the difference: full model stat - (model - 1) stat\n",
    "# LRT_bootstrap_StatsDiff = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_bootstrap_StatsDiff.csv.gz\"), compression=\"gzip\")\n",
    "# print(len(LRT_bootstrap_StatsDiff.mutation.unique()))\n",
    "# LRT_bootstrap_StatsDiff2 = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_bootstrap_StatsDiff_2.csv.gz\"), compression=\"gzip\")\n",
    "# print(len(LRT_bootstrap_StatsDiff2.mutation.unique()))\n",
    "\n",
    "# LRT_bootstrap_StatsDiff = pd.concat([LRT_bootstrap_StatsDiff, LRT_bootstrap_StatsDiff2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "0924c0cd-55e4-495c-8686-d492251baff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SpearmanrResult(correlation=0.9338806280906226, pvalue=2.6137422278744596e-05)\n"
     ]
    }
   ],
   "source": [
    "# for metric in ['AUC']:\n",
    "\n",
    "#     print(st.spearmanr(LRT_res.query(\"mutation in @LRT_bootstrap_StatsDiff.mutation.unique()\")[\"BH_pval\"], \n",
    "#                  pval_df[metric]\n",
    "#                 ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "f4451bac-538f-4944-8807-7a2823193781",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pval_df = pd.DataFrame(columns=['AUC', 'Sens', 'Spec', 'accuracy'])\n",
    "\n",
    "# for mut in LRT_bootstrap_StatsDiff.mutation.unique():\n",
    "    \n",
    "#     print(\"\\n\" + mut)\n",
    "    \n",
    "#     for metric in ['AUC']:\n",
    "    \n",
    "# #         lower, upper = np.percentile(LRT_bootstrap_StatsDiff.query(\"mutation==@mut\")[metric], q=[2.5, 97.5])\n",
    "    \n",
    "# #         if lower > 0 and upper > 0:\n",
    "# #             print(f\"    {metric}: {round(lower, 4)}, {round(upper, 4)}\")\n",
    "\n",
    "#         pval = 1 - np.mean(LRT_bootstrap_StatsDiff.query('mutation==@mut')[metric].values > 0)\n",
    "        \n",
    "#         if pval < 0.05:\n",
    "#             print(f\"{metric}: {round(pval, 2)}\")\n",
    "            \n",
    "#         pval_df.loc[mut, metric] = pval"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51ff9fee-e0ea-4908-ad04-750f663165f2",
   "metadata": {},
   "source": [
    "# Model with Interaction Terms\n",
    "\n",
    "Chose 2 variants because they were the only Tier 2 genes that were significantly associated in the Ridge regression and were significant in the LRT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c88b8b80-1690-4a6c-8b5f-1dddf193a3f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.195840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.216332</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>5.033098e-07</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.001432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Rv2752c_p.Asn30Ser</td>\n",
       "      <td>0.099958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.105124</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.480226e-07</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.000421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              mutation      coef  pval  BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "2    rpoC_p.Glu1092Asp  0.195840   0.0      0.0              0.0    1.216332   \n",
       "19  Rv2752c_p.Asn30Ser  0.099958   0.0      0.0              0.0    1.105124   \n",
       "\n",
       "                   confidence      LRT_pval  LRT_BH_pval  LRT_Bonferroni_pval  \n",
       "2            5) Not assoc w R  5.033098e-07     0.000716             0.001432  \n",
       "19  3) Uncertain significance  1.480226e-07     0.000421             0.000421  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "tier2_model.query('BH_pval < @thresh & LRT_BH_pval < @thresh')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "0a490028-8614-4b27-86bf-51612d571e1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30984, 3737)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "interact_mut = tier2_model.query('BH_pval < @thresh & LRT_BH_pval < @thresh')[\"mutation\"].values\n",
    "df_phenos = pd.read_csv(os.path.join(analysis_dir, drug, \"phenos_binary.csv\")).set_index(\"sample_id\")\n",
    "#interact_mut = ['rpoC_p.Glu1092Asp', 'Rv2752c_p.Asn30Ser']\n",
    "\n",
    "# model_matrix = pd.read_pickle(os.path.join(analysis_dir, drug, \"BINARY/tiers=1+2/phenos=WHO/dropAF_noSyn_unpooled/model_matrix.pkl\"))\n",
    "# interact_matrix = model_matrix.copy()\n",
    "\n",
    "# # get a list of all pairs of mutations to interact\n",
    "# interactions = list(itertools.product(*[interact_mut, interact_matrix.columns[interact_matrix.columns.str.contains(\"rpoB\")]]))\n",
    "\n",
    "# # add interactions\n",
    "# for (mut1, mut2) in interactions:\n",
    "#     interact_matrix[f\"{mut1}+{mut2}\"] = interact_matrix[mut1] * interact_matrix[mut2]\n",
    "    \n",
    "# # drop any columns with no signal\n",
    "# interact_matrix = interact_matrix[interact_matrix.columns[~((interact_matrix == 0).all())]]\n",
    "\n",
    "# assert len(set(model_matrix.columns) - set(interact_matrix.columns)) == 0\n",
    "# print(len(interact_matrix), len(model_matrix))\n",
    "\n",
    "# interact_matrix.to_pickle(os.path.join(analysis_dir, drug, f\"BINARY/interaction/model_matrix_{phenos_name}.pkl\"))\n",
    "\n",
    "interact_matrix = pd.read_pickle(os.path.join(analysis_dir, drug, f\"BINARY/interaction/model_matrix_{phenos_name}.pkl\"))\n",
    "\n",
    "interact_matrix = interact_matrix.merge(df_phenos, left_index=True, right_index=True, how=\"left\")\n",
    "del interact_matrix[\"phenotypic_category\"]\n",
    "\n",
    "interact_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "61501b9b-818c-474a-8f38-82074d1e4b53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rpoC_p.Glu1092Asp 125\n",
      "Rv2752c_p.Asn30Ser 4\n"
     ]
    }
   ],
   "source": [
    "# so much co-occurrence of rpoC_p.Glu1092Asp with rpoB mutations. Can also check proportion that are high confidence rpoB mutations\n",
    "for mut in interact_mut:\n",
    "    print(mut, interact_matrix[interact_matrix.columns[interact_matrix.columns.str.contains(mut)]].shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "97e9b440-e22c-4b3a-8bb0-490cb0fc96e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenos_name = \"WHO\"\n",
    "interact_res = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_coef.csv\"))\n",
    "interact_permute = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_coef_permutation.csv\"))\n",
    "\n",
    "# assess significance using the results of the permutation test\n",
    "for i, row in interact_res.iterrows():\n",
    "    # p-value is the proportion of permutation coefficients that are AT LEAST AS EXTREME as the test statistic\n",
    "    if row[\"coef\"] > 0:\n",
    "        interact_res.loc[i, \"pval\"] = np.mean(interact_permute[row[\"mutation\"]] >= row[\"coef\"])\n",
    "    else:\n",
    "        interact_res.loc[i, \"pval\"] = np.mean(interact_permute[row[\"mutation\"]] <= row[\"coef\"])\n",
    "        \n",
    "interact_res = add_pval_corrections(interact_res)\n",
    "interact_res[\"Odds_Ratio\"] = np.exp(interact_res[\"coef\"])\n",
    "\n",
    "interact_res[[\"Mut1\", \"Mut2\"]] = interact_res[\"mutation\"].str.split(\"+\", expand=True)\n",
    "interact_res = interact_res.merge(who_variants_combined.query(\"drug=='RIF'\"), left_on=\"Mut2\", right_on=\"mutation\", how=\"left\")\n",
    "del interact_res[\"mutation_y\"]\n",
    "del interact_res[\"drug\"]\n",
    "interact_res.rename(columns={\"mutation_x\": \"mutation\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "42d4f33d-4f57-4cc7-92b8-cf3ecb1cc77a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3206</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.219309</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.245216</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3676</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Ile491Tyr</td>\n",
       "      <td>0.026609</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.026966</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Ile491Tyr</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3693</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Met434_Asn437delinsIle</td>\n",
       "      <td>0.018817</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.018995</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Met434_Asn437delinsIle</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3623</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp</td>\n",
       "      <td>-0.027974</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972414</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3725</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Val170Phe</td>\n",
       "      <td>-0.057892</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.038700</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.943752</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Val170Phe</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3708</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu</td>\n",
       "      <td>0.057540</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.048584</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.059228</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             mutation      coef   pval  \\\n",
       "3206                                rpoC_p.Glu1092Asp  0.219309  0.000   \n",
       "3676               rpoC_p.Glu1092Asp+rpoB_p.Ile491Tyr  0.026609  0.000   \n",
       "3693  rpoC_p.Glu1092Asp+rpoB_p.Met434_Asn437delinsIle  0.018817  0.000   \n",
       "3623               rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp -0.027974  0.000   \n",
       "3725               rpoC_p.Glu1092Asp+rpoB_p.Val170Phe -0.057892  0.003   \n",
       "3708               rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu  0.057540  0.004   \n",
       "\n",
       "       BH_pval  Bonferroni_pval  Odds_Ratio               Mut1  \\\n",
       "3206  0.000000              0.0    1.245216  rpoC_p.Glu1092Asp   \n",
       "3676  0.000000              0.0    1.026966  rpoC_p.Glu1092Asp   \n",
       "3693  0.000000              0.0    1.018995  rpoC_p.Glu1092Asp   \n",
       "3623  0.000000              0.0    0.972414  rpoC_p.Glu1092Asp   \n",
       "3725  0.038700              1.0    0.943752  rpoC_p.Glu1092Asp   \n",
       "3708  0.048584              1.0    1.059228  rpoC_p.Glu1092Asp   \n",
       "\n",
       "                               Mut2                 confidence  \n",
       "3206                           None                        NaN  \n",
       "3676               rpoB_p.Ile491Tyr  3) Uncertain significance  \n",
       "3693  rpoB_p.Met434_Asn437delinsIle     2) Assoc w R - Interim  \n",
       "3623               rpoB_p.Asn260Asp                        NaN  \n",
       "3725               rpoB_p.Val170Phe               1) Assoc w R  \n",
       "3708               rpoB_p.Ser450Leu               1) Assoc w R  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation.str.contains('rpoC_p.Glu1092Asp') & BH_pval < 0.05\").sort_values([\"BH_pval\", \"Odds_Ratio\"], \n",
    "                                                                                           ascending=[True, False]\n",
    "                                                                                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "50e05e3c-ae52-41db-aa52-c197e98d458d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2936</th>\n",
       "      <td>rpoB_p.Val170Phe</td>\n",
       "      <td>0.191617</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.211206</td>\n",
       "      <td>rpoB_p.Val170Phe</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3725</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Val170Phe</td>\n",
       "      <td>-0.057892</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0387</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.943752</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Val170Phe</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                mutation      coef   pval  BH_pval  \\\n",
       "2936                    rpoB_p.Val170Phe  0.191617  0.000   0.0000   \n",
       "3725  rpoC_p.Glu1092Asp+rpoB_p.Val170Phe -0.057892  0.003   0.0387   \n",
       "\n",
       "      Bonferroni_pval  Odds_Ratio               Mut1              Mut2  \\\n",
       "2936              0.0    1.211206   rpoB_p.Val170Phe              None   \n",
       "3725              1.0    0.943752  rpoC_p.Glu1092Asp  rpoB_p.Val170Phe   \n",
       "\n",
       "        confidence  \n",
       "2936           NaN  \n",
       "3725  1) Assoc w R  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation.str.contains('rpoB_p.Val170Phe')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "9344d6c2-420a-4320-a06e-1f30196b8ee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>-0.027974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972414</td>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3623</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp</td>\n",
       "      <td>-0.027974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972414</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                mutation      coef  pval  BH_pval  \\\n",
       "2395                    rpoB_p.Asn260Asp -0.027974   0.0      0.0   \n",
       "3623  rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp -0.027974   0.0      0.0   \n",
       "\n",
       "      Bonferroni_pval  Odds_Ratio               Mut1              Mut2  \\\n",
       "2395              0.0    0.972414   rpoB_p.Asn260Asp              None   \n",
       "3623              0.0    0.972414  rpoC_p.Glu1092Asp  rpoB_p.Asn260Asp   \n",
       "\n",
       "     confidence  \n",
       "2395        NaN  \n",
       "3623        NaN  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation.str.contains('rpoB_p.Asn260Asp')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "9710b1d6-fad5-4b8b-ba48-d96227834be3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phenotype  rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu\n",
       "0          0.0                                   20907\n",
       "           1.0                                      14\n",
       "1          0.0                                    8682\n",
       "           1.0                                    1381\n",
       "Name: rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu, dtype: int64"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_matrix.groupby(\"phenotype\")['rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "72d42bbb-ce7e-4e64-9af7-d31b0acd9ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.045023237800154915"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_matrix['rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "d8a9eb00-3c8e-445c-94c5-07a3c6e0f7d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9971119133574007"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1381 / 1385"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "bb4d7011-21b7-4c87-b754-4c9adbf2d6a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.08704492641363284"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_matrix['rpoC_p.Glu1092Asp'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "4adc8ca7-e960-47f8-86e6-ac6f4fc81f85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.21449780531887425"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_matrix['rpoB_p.Ser450Leu'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cea1cf2d-e073-428b-8dd6-0acadb10438f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Rifampicin/BINARY/tiers=1/phenos=WHO/dropAF_noSyn_unpooled'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "947aaa3a-da07-4ffd-9d68-9e6e361ed275",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9092692c-9ff2-41de-8481-7a8e36f5bfa4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d96600d-2456-4d00-a536-47d985a2117f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
