{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "86d79dbc-86b7-4a6d-b26c-3052aaeb12a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 150\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "\n",
    "import glob, os, yaml, subprocess, itertools, sparse, vcf\n",
    "from sklearn.model_selection import train_test_split, KFold, GridSearchCV\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "import sklearn.metrics\n",
    "from sklearn.decomposition import PCA\n",
    "import timeit\n",
    "import scipy.stats as st\n",
    "import statsmodels.api as sm\n",
    "import pickle, yaml, tracemalloc\n",
    "from scipy.stats import binomtest\n",
    "\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue'\n",
    "input_data_dir = '/n/data1/hms/dbmi/farhat/ye12/who'\n",
    "who_variants_combined = pd.read_csv(\"who_confidence_2021.csv\")\n",
    "\n",
    "import itertools\n",
    "from stats_utils import *\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07de7321-bf6a-40ca-9cd5-51a5df1b16c0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# RIFAMPICIN TEST CASE\n",
    "\n",
    "# 0. Original Analysis: 137 new significant resistance-associated variants\n",
    "\n",
    "## But, there are 152 with OR > 1 and BH p-val < 0.01 that are in tier 2 genes (rpoB is the only tier 1)\n",
    "## These 152 mutations were studied in the next two analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d46b8cbd-3e49-4b4d-94fc-fb3f944f4ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = \"BINARY\"\n",
    "phenos_name = \"WHO\"\n",
    "drug = \"Rifampicin\"\n",
    "drug_WHO_abbr = \"RIF\"\n",
    "\n",
    "# get the number of high confidence resistance-associated \n",
    "high_Conf_variants = who_variants_combined.loc[(who_variants_combined[\"drug\"]==drug_WHO_abbr) & \n",
    "                                               (who_variants_combined[\"confidence\"].str.contains(\"|\".join([\"1\", \"2\"])))\n",
    "                                              ].mutation.unique()\n",
    "\n",
    "# these have Odds_Ratio > 1, BH pval < 0.01, and are not in the Tier 1 analysis (for RIF, it means they are not in rpoB)\n",
    "# DIDN'T THRESHOLD USING THE UNIVARIATE STATS BECAUSE THOSE WILL BE COMPUTED AT THE VERY END AND SHOULDN'T BE INVOLVED IN MAKING DECISIONS ABOUT THE REGRESSION ANALYSIS\n",
    "og_models = pd.read_excel(\"../results/BINARY/Rifampicin.xlsx\", sheet_name=[\"Model_3\", \"Model_7\", \"Model_11\", \"Model_15\"])\n",
    "#print(og_who_model.query(\"Odds_Ratio > 1 & BH_pval < 0.01 & PPV_LB > 0.25 & TP >= 5 & ~mutation.str.contains('rpoB')\").shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "09e35f41-2ce5-4db8-9c0b-a749ee59c09f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_significant_genes_by_tier(tier1_model, tier2_model, LRT_res):\n",
    "\n",
    "    print(f\"Total numbers of mutations: {tier1_model.shape}, {tier2_model.shape}\")\n",
    "\n",
    "    tier1_significant = tier1_model.query(\"Odds_Ratio > 1 & BH_pval < 0.05\")\n",
    "    tier2_significant = tier2_model.query(\"Odds_Ratio > 1 & BH_pval < 0.01\")\n",
    "\n",
    "    print(f\"Numbers of significant mutations: {len(set(tier1_significant.mutation))}, {len(set(tier2_significant.mutation) - set(tier1_significant.mutation))}\")\n",
    "    \n",
    "    combined_df =pd.concat([tier1_significant, tier2_significant], axis=0)\n",
    "    combined_df[\"Tier\"] = [1]*len(tier1_significant) + [2]*len(tier2_significant)\n",
    "    return combined_df.drop_duplicates(subset=\"mutation\", keep=\"first\").merge(LRT_res[[\"mutation\", 'LRT_pval', 'LRT_BH_pval', 'LRT_Bonferroni_pval']], on=\"mutation\", how=\"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5ac0441-6a40-4c56-98fd-a46ae44e6e8e",
   "metadata": {},
   "source": [
    "# 1. Excluded Samples Analysis: 54 new significant resistance-associated variants\n",
    "\n",
    "Removed ~5,000 samples that contain both high confidence rpoB mutations (Category 1 or 2 in 2021 catalogue) and any of the ~150 significant tier2 mutations. \n",
    "\n",
    "An L2-penalized regression was fit to determine the effect sizes of tier 2 mutations, independent of resistance-associated mutations they may occur with. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ee479d3c-82e4-4390-8956-620fefc7b84e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_df = pd.read_csv(os.path.join(analysis_dir, f\"{drug}/{folder}/exclude_comutation/WHOphenos_univariate_stats.csv\"))\n",
    "# exclude_df = exclude_df.loc[~exclude_df[\"mutation\"].str.contains(\"PC\")]\n",
    "# exclude_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']] = exclude_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']].astype(int)\n",
    "\n",
    "# exclude_df.query(\"Odds_Ratio > 1 & BH_pval < 0.01 & TP >= 5 & PPV_LB >= 0.25 & ~mutation.str.contains('rpoB')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1ce0aaa4-a55c-4658-ac8c-3836f6a67f53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_df.query(\"mutation=='rpoC_p.Glu1092Asp'\")[['mutation', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval',\n",
    "#        'Bonferroni_pval', 'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN',\n",
    "#        'FN', 'PPV_LB', 'PPV', 'PPV_UB', 'NPV', 'Sens', 'Spec']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b8b21b1-b6a4-4a9d-a822-5806d1916379",
   "metadata": {},
   "source": [
    "# 2. Likelihood Ratio Test: 5/152 of the Above Tier 2 Mutations are Significant\n",
    "\n",
    "152 tier 2 variants were found significant in the first analysis. We then performed a likelihood ratio test:\n",
    "\n",
    "A mutation is removed from the original model input matrix, and then a new model is fit. \n",
    "\n",
    "We then compare the log-likelihoods of the original model and the model with 1 less mutation. If the p-value is significant, it means that removing the mutation significantly changes the fit of the model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "38d6d423-75b3-475b-a057-5bb9f6f80c59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log_like</th>\n",
       "      <th>chi_stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>AUC</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-6060.916834</td>\n",
       "      <td>2301.327422</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.964044</td>\n",
       "      <td>0.904443</td>\n",
       "      <td>0.978961</td>\n",
       "      <td>0.954725</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-5277.604669</td>\n",
       "      <td>734.703093</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.969290</td>\n",
       "      <td>0.923311</td>\n",
       "      <td>0.981392</td>\n",
       "      <td>0.962502</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-4998.790841</td>\n",
       "      <td>177.075436</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.973489</td>\n",
       "      <td>0.932702</td>\n",
       "      <td>0.979003</td>\n",
       "      <td>0.963944</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-4959.474858</td>\n",
       "      <td>98.443469</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.973846</td>\n",
       "      <td>0.934093</td>\n",
       "      <td>0.979129</td>\n",
       "      <td>0.964482</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-5037.650604</td>\n",
       "      <td>254.794962</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972094</td>\n",
       "      <td>0.929484</td>\n",
       "      <td>0.981769</td>\n",
       "      <td>0.964764</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>-4910.148492</td>\n",
       "      <td>-0.209262</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.974210</td>\n",
       "      <td>0.934788</td>\n",
       "      <td>0.978961</td>\n",
       "      <td>0.964595</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>-4910.085364</td>\n",
       "      <td>-0.335519</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.974170</td>\n",
       "      <td>0.932441</td>\n",
       "      <td>0.981141</td>\n",
       "      <td>0.965302</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>-4910.203516</td>\n",
       "      <td>-0.099214</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.974200</td>\n",
       "      <td>0.934788</td>\n",
       "      <td>0.978961</td>\n",
       "      <td>0.964595</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813</th>\n",
       "      <td>-4909.935748</td>\n",
       "      <td>-0.634751</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.974187</td>\n",
       "      <td>0.934788</td>\n",
       "      <td>0.978961</td>\n",
       "      <td>0.964595</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>814</th>\n",
       "      <td>-4910.218360</td>\n",
       "      <td>-0.069527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.974185</td>\n",
       "      <td>0.934788</td>\n",
       "      <td>0.978961</td>\n",
       "      <td>0.964595</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>815 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        log_like     chi_stat  pval       AUC      Sens      Spec  accuracy  \\\n",
       "0   -6060.916834  2301.327422   0.0  0.964044  0.904443  0.978961  0.954725   \n",
       "1   -5277.604669   734.703093   0.0  0.969290  0.923311  0.981392  0.962502   \n",
       "2   -4998.790841   177.075436   0.0  0.973489  0.932702  0.979003  0.963944   \n",
       "3   -4959.474858    98.443469   0.0  0.973846  0.934093  0.979129  0.964482   \n",
       "4   -5037.650604   254.794962   0.0  0.972094  0.929484  0.981769  0.964764   \n",
       "..           ...          ...   ...       ...       ...       ...       ...   \n",
       "810 -4910.148492    -0.209262   1.0  0.974210  0.934788  0.978961  0.964595   \n",
       "811 -4910.085364    -0.335519   1.0  0.974170  0.932441  0.981141  0.965302   \n",
       "812 -4910.203516    -0.099214   1.0  0.974200  0.934788  0.978961  0.964595   \n",
       "813 -4909.935748    -0.634751   1.0  0.974187  0.934788  0.978961  0.964595   \n",
       "814 -4910.218360    -0.069527   1.0  0.974185  0.934788  0.978961  0.964595   \n",
       "\n",
       "     BH_pval  Bonferroni_pval  \n",
       "0        0.0              0.0  \n",
       "1        0.0              0.0  \n",
       "2        0.0              0.0  \n",
       "3        0.0              0.0  \n",
       "4        0.0              0.0  \n",
       "..       ...              ...  \n",
       "810      1.0              1.0  \n",
       "811      1.0              1.0  \n",
       "812      1.0              1.0  \n",
       "813      1.0              1.0  \n",
       "814      1.0              1.0  \n",
       "\n",
       "[815 rows x 9 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LRTresults_tier1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "81d4bcd6-8487-4e4e-8fa2-da56af948fcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Rifampicin/BINARY/tiers=1/phenos=WHO/dropAF_noSyn_unpooled'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.join(analysis_dir, drug, f\"BINARY/tiers=1/phenos={phenos_name}/dropAF_noSyn_unpooled\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "84632d9d-fd85-4532-a823-15b8b33dde51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "815 2845\n"
     ]
    }
   ],
   "source": [
    "phenos_name = \"WHO\"\n",
    "out_dir = os.path.join(analysis_dir, drug, f\"BINARY/LRT/phenos={phenos_name}\")\n",
    "\n",
    "LRTresults_tier1 = pd.read_csv(os.path.join(out_dir, f\"results_tier={'1'}.csv\")).rename(columns={\"Unnamed: 0\": \"mutation\"})\n",
    "LRTresults_tier2 = pd.read_csv(os.path.join(out_dir, f\"results_tier={'2'}.csv\")).rename(columns={\"Unnamed: 0\": \"mutation\"})\n",
    "\n",
    "# remove the FULL model row, which is the first row\n",
    "LRTresults_tier1 = add_pval_corrections(LRTresults_tier1.iloc[1:, ])\n",
    "LRTresults_tier2 = add_pval_corrections(LRTresults_tier2.iloc[1:, ])\n",
    "\n",
    "tier1_model_permute = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/tiers=1/phenos={phenos_name}/dropAF_noSyn_unpooled/model_analysis.csv\")).query(\"~mutation.str.contains('PC')\")\n",
    "tier2_model_permute = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/tiers=1+2/phenos={phenos_name}/dropAF_noSyn_unpooled/model_analysis.csv\")).query(\"~mutation.str.contains('PC')\")\n",
    "\n",
    "# remove the tier 1 genes for the purposes of this analysis\n",
    "tier2_model_permute = tier2_model_permute.query(\"mutation not in @tier1_model_permute.mutation\")\n",
    "\n",
    "# check that all mutations in the permutation dataframe are in the LRT dataframe. The only difference should be the FULL model row\n",
    "# print(set(tier1_model_permute.mutation).symmetric_difference(LRTresults_tier1.index))\n",
    "# print(set(tier2_model_permute.mutation).symmetric_difference(LRTresults_tier2.index))\n",
    "assert len(set(tier1_model_permute.mutation).symmetric_difference(LRTresults_tier1.mutation)) == 0\n",
    "assert len(set(tier2_model_permute.mutation).symmetric_difference(LRTresults_tier2.mutation)) == 0\n",
    "\n",
    "# combine results into a single dataframe for easy searching\n",
    "tier1_model = tier1_model_permute.merge(LRTresults_tier1[[\"mutation\", \"pval\", \"BH_pval\", \"Bonferroni_pval\"]].rename(columns={\"pval\": \"LRT_pval\", \n",
    "                                                                                                               \"BH_pval\": \"LRT_BH_pval\", \n",
    "                                                                                                               \"Bonferroni_pval\": \"LRT_Bonferroni_pval\"\n",
    "                                                                                                              }), on=\"mutation\", how=\"left\")\n",
    "\n",
    "tier2_model = tier2_model_permute.merge(LRTresults_tier2[[\"mutation\", \"pval\", \"BH_pval\", \"Bonferroni_pval\"]].rename(columns={\"pval\": \"LRT_pval\", \n",
    "                                                                                                               \"BH_pval\": \"LRT_BH_pval\", \n",
    "                                                                                                               \"Bonferroni_pval\": \"LRT_Bonferroni_pval\"\n",
    "                                                                                                              }), on=\"mutation\", how=\"left\")\n",
    "\n",
    "print(len(tier1_model), len(tier2_model))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6c4aefb8-9837-49ef-aed9-28f732defb3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3) Uncertain significance    34\n",
       "NaN                          16\n",
       "2) Assoc w R - Interim       14\n",
       "5) Not assoc w R              4\n",
       "Name: confidence, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of Tier 1 mutations determined significant by original analysis, but not in LRT\n",
    "thresh = 0.05\n",
    "tier1_model.query(\"BH_pval < @thresh & LRT_BH_pval > @thresh\").confidence.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bb824e0e-3baf-48ea-aa9a-8a24cd23c58d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54 significant coef, but not LRT\n",
      "4 significant LRT, but not coef\n",
      "47 significant in both\n"
     ]
    }
   ],
   "source": [
    "# mutations picked up by the LRT, but not the coefficient association. \n",
    "# They are not significant in the tiers 1+2 model either \n",
    "thresh = 0.05\n",
    "\n",
    "print(f\"{len(tier1_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval > @thresh'))} significant coef, but not LRT\")\n",
    "print(f\"{len(tier1_model.query('Odds_Ratio > 1 & BH_pval > @thresh & LRT_BH_pval < @thresh'))} significant LRT, but not coef\")\n",
    "print(f\"{len(tier1_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval < @thresh'))} significant in both\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3f24971a-451c-4499-9646-c2dd9b03832c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>rpoB_p.Asn438del</td>\n",
       "      <td>0.071560</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.410000</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.074182</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.646256e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>rpoB_p.Leu430_Ser431insArg</td>\n",
       "      <td>0.050143</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.615000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.051422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.782667e-05</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.022679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>rpoB_p.Gln436del</td>\n",
       "      <td>0.048461</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.751667</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.049655</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.692521e-05</td>\n",
       "      <td>0.000811</td>\n",
       "      <td>0.013794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>rpoB_p.Asp435_Gln436delinsGlu</td>\n",
       "      <td>0.035628</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.961379</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.036270</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.195996e-03</td>\n",
       "      <td>0.034812</td>\n",
       "      <td>0.974737</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         mutation      coef   pval   BH_pval  Bonferroni_pval  \\\n",
       "45               rpoB_p.Asn438del  0.071560  0.001  0.410000             0.82   \n",
       "65     rpoB_p.Leu430_Ser431insArg  0.050143  0.003  0.615000             1.00   \n",
       "70               rpoB_p.Gln436del  0.048461  0.011  0.751667             1.00   \n",
       "97  rpoB_p.Asp435_Gln436delinsGlu  0.035628  0.034  0.961379             1.00   \n",
       "\n",
       "    Odds_Ratio              confidence      LRT_pval  LRT_BH_pval  \\\n",
       "45    1.074182  2) Assoc w R - Interim  1.646256e-08     0.000002   \n",
       "65    1.051422                     NaN  2.782667e-05     0.001194   \n",
       "70    1.049655  2) Assoc w R - Interim  1.692521e-05     0.000811   \n",
       "97    1.036270  2) Assoc w R - Interim  1.195996e-03     0.034812   \n",
       "\n",
       "    LRT_Bonferroni_pval  \n",
       "45             0.000013  \n",
       "65             0.022679  \n",
       "70             0.013794  \n",
       "97             0.974737  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mutations picked up by the LRT, but not the coefficient association. \n",
    "# They are not significant in the tiers 1+2 model either \n",
    "thresh = 0.05\n",
    "tier1_model.query(\"Odds_Ratio > 1 & BH_pval > @thresh & LRT_BH_pval < @thresh\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4da438c6-1634-41db-9689-dfb4557008cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NaN                          111\n",
       "3) Uncertain significance     20\n",
       "5) Not assoc w R               5\n",
       "Name: confidence, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "tier2_model.query(\"BH_pval < @thresh & LRT_BH_pval > @thresh\").confidence.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f388f24e-4500-4f2a-81f2-f4953b70a34a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>glpK_p.Val192fs</td>\n",
       "      <td>0.026607</td>\n",
       "      <td>0.033</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.026964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mutation      coef   pval  BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "203  glpK_p.Val192fs  0.026607  0.033      1.0              1.0    1.026964   \n",
       "\n",
       "    confidence  LRT_pval  LRT_BH_pval  LRT_Bonferroni_pval  \n",
       "203        NaN       1.0          1.0                  1.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tier2_model.query(\"mutation=='glpK_p.Val192fs'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "68e95ff2-7e0a-4ca7-a166-b99c61dc9b15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "98 significant coef, but not LRT\n",
      "0 significant LRT, but not coef\n",
      "2 significant in both\n"
     ]
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "\n",
    "print(f\"{len(tier2_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval > @thresh'))} significant coef, but not LRT\")\n",
    "print(f\"{len(tier2_model.query('Odds_Ratio > 1 & BH_pval > @thresh & LRT_BH_pval < @thresh'))} significant LRT, but not coef\")\n",
    "print(f\"{len(tier2_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval < @thresh'))} significant in both\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "26826123-bc4a-416f-97c0-cda738d49334",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoC_p.Val483Gly</td>\n",
       "      <td>0.300945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.351136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006608</td>\n",
       "      <td>0.458527</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpoC_p.Ile491Thr</td>\n",
       "      <td>0.199407</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.220678</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.024332</td>\n",
       "      <td>0.097328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rpoC_p.Ile491Val</td>\n",
       "      <td>0.194982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.215289</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.000808</td>\n",
       "      <td>0.176888</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rpoC_p.Asn698Ser</td>\n",
       "      <td>0.190240</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.209540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.033749</td>\n",
       "      <td>0.202492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rpoC_p.Pro1040Arg</td>\n",
       "      <td>0.179407</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.196508</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.051457</td>\n",
       "      <td>0.463113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>glpK_c.-7A&gt;G</td>\n",
       "      <td>0.036804</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.037489</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010680</td>\n",
       "      <td>0.523861</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>Rv2477c_p.Phe429Leu</td>\n",
       "      <td>0.030412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.030879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.042472</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>glpK_p.Asp215Gly</td>\n",
       "      <td>0.023232</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.023504</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.827897</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Rv2752c_p.Ser189*</td>\n",
       "      <td>0.022385</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.022638</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.825366</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>rpoC_p.Gln262His</td>\n",
       "      <td>0.020150</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.020354</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.822674</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                mutation      coef  pval  BH_pval  Bonferroni_pval  \\\n",
       "0       rpoC_p.Val483Gly  0.300945   0.0      0.0              0.0   \n",
       "1       rpoC_p.Ile491Thr  0.199407   0.0      0.0              0.0   \n",
       "3       rpoC_p.Ile491Val  0.194982   0.0      0.0              0.0   \n",
       "4       rpoC_p.Asn698Ser  0.190240   0.0      0.0              0.0   \n",
       "5      rpoC_p.Pro1040Arg  0.179407   0.0      0.0              0.0   \n",
       "..                   ...       ...   ...      ...              ...   \n",
       "138         glpK_c.-7A>G  0.036804   0.0      0.0              0.0   \n",
       "175  Rv2477c_p.Phe429Leu  0.030412   0.0      0.0              0.0   \n",
       "234     glpK_p.Asp215Gly  0.023232   0.0      0.0              0.0   \n",
       "245    Rv2752c_p.Ser189*  0.022385   0.0      0.0              0.0   \n",
       "283     rpoC_p.Gln262His  0.020150   0.0      0.0              0.0   \n",
       "\n",
       "     Odds_Ratio                 confidence  LRT_pval  LRT_BH_pval  \\\n",
       "0      1.351136                        NaN  0.006608     0.458527   \n",
       "1      1.220678                        NaN  0.000034     0.024332   \n",
       "3      1.215289  3) Uncertain significance  0.000808     0.176888   \n",
       "4      1.209540                        NaN  0.000071     0.033749   \n",
       "5      1.196508  3) Uncertain significance  0.000163     0.051457   \n",
       "..          ...                        ...       ...          ...   \n",
       "138    1.037489                        NaN  0.010680     0.523861   \n",
       "175    1.030879                        NaN  0.042472     1.000000   \n",
       "234    1.023504                        NaN  0.827897     1.000000   \n",
       "245    1.022638                        NaN  0.825366     1.000000   \n",
       "283    1.020354                        NaN  0.822674     1.000000   \n",
       "\n",
       "     LRT_Bonferroni_pval  \n",
       "0               1.000000  \n",
       "1               0.097328  \n",
       "3               1.000000  \n",
       "4               0.202492  \n",
       "5               0.463113  \n",
       "..                   ...  \n",
       "138             1.000000  \n",
       "175             1.000000  \n",
       "234             1.000000  \n",
       "245             1.000000  \n",
       "283             1.000000  \n",
       "\n",
       "[98 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tier2_model.query('Odds_Ratio > 1 & BH_pval < @thresh & LRT_BH_pval > @thresh')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1c0fd3a0-91d4-4061-baf7-25365d275efb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>confidence</th>\n",
       "      <th>LRT_pval</th>\n",
       "      <th>LRT_BH_pval</th>\n",
       "      <th>LRT_Bonferroni_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoC_p.Val483Gly</td>\n",
       "      <td>0.300945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.351136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.607943e-03</td>\n",
       "      <td>0.458527</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.195840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.216332</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>5.033098e-07</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.001432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rpoC_p.Val483Ala</td>\n",
       "      <td>0.159044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.172390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.559949e-02</td>\n",
       "      <td>0.625078</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mutation      coef  pval  BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "0   rpoC_p.Val483Gly  0.300945   0.0      0.0              0.0    1.351136   \n",
       "2  rpoC_p.Glu1092Asp  0.195840   0.0      0.0              0.0    1.216332   \n",
       "6   rpoC_p.Val483Ala  0.159044   0.0      0.0              0.0    1.172390   \n",
       "\n",
       "         confidence      LRT_pval  LRT_BH_pval  LRT_Bonferroni_pval  \n",
       "0               NaN  6.607943e-03     0.458527             1.000000  \n",
       "2  5) Not assoc w R  5.033098e-07     0.000716             0.001432  \n",
       "6               NaN  1.559949e-02     0.625078             1.000000  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tier2_model.query(\"mutation in ['rpoC_p.Val483Ala', 'rpoC_p.Val483Gly', 'rpoC_p.Glu1092Asp']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327d03d6-f2b0-4dda-a2fd-26fca039de90",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e2d22d8-713c-4927-9596-a90281a1088f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0399c22c-5542-402a-9bf4-82c1d86c1646",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rpoC_p.Glu1092Asp', 'Rv2752c_p.Asn30Ser'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "tier2_model.query('BH_pval < @thresh & LRT_BH_pval < @thresh')[\"mutation\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "63f2e03f-c90e-46c2-9de5-f35d47c9dd30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3) Uncertain significance    1515\n",
       "2) Assoc w R - Interim         93\n",
       "5) Not assoc w R               28\n",
       "1) Assoc w R                   24\n",
       "Name: confidence, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who_variants_combined.query(\"drug=='RIF'\").confidence.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "651dd7e7-b428-4b87-bc0f-5946578a8e19",
   "metadata": {},
   "source": [
    "# Perform bootstrapping for all 152 mutations to see how much AUC, Sens, Spec, and Accuracy Change\n",
    "\n",
    "If the confidence interval lies above 0, it means that removing the feature SIGNIFICANTLY DECREASES THE PREDICTIVE POWER OF THE MODEL\n",
    "\n",
    "CI > 0 = FEATURE IS IMPORTANT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b9d72b2-6e0e-4bc7-b8ad-39cf938671da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # bootstrapped the difference: full model stat - (model - 1) stat\n",
    "# LRT_bootstrap_StatsDiff = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_bootstrap_StatsDiff.csv.gz\"), compression=\"gzip\")\n",
    "# print(len(LRT_bootstrap_StatsDiff.mutation.unique()))\n",
    "# LRT_bootstrap_StatsDiff2 = pd.read_csv(os.path.join(analysis_dir, drug, \"BINARY/LRT\", f\"{phenos_name}phenos_bootstrap_StatsDiff_2.csv.gz\"), compression=\"gzip\")\n",
    "# print(len(LRT_bootstrap_StatsDiff2.mutation.unique()))\n",
    "\n",
    "# LRT_bootstrap_StatsDiff = pd.concat([LRT_bootstrap_StatsDiff, LRT_bootstrap_StatsDiff2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "0924c0cd-55e4-495c-8686-d492251baff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SpearmanrResult(correlation=0.9338806280906226, pvalue=2.6137422278744596e-05)\n"
     ]
    }
   ],
   "source": [
    "for metric in ['AUC']:\n",
    "\n",
    "    print(st.spearmanr(LRT_res.query(\"mutation in @LRT_bootstrap_StatsDiff.mutation.unique()\")[\"BH_pval\"], \n",
    "                 pval_df[metric]\n",
    "                ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "f4451bac-538f-4944-8807-7a2823193781",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "rpoB_p.Ser450Leu\n",
      "AUC: 0.0\n",
      "\n",
      "Rv2752c_p.Asn30Ser\n",
      "AUC: 0.01\n",
      "\n",
      "rpoC_p.Glu1092Asp\n",
      "AUC: 0.01\n",
      "\n",
      "rpoC_p.Ile491Thr\n",
      "\n",
      "rpoC_p.Asn698Ser\n",
      "AUC: 0.02\n",
      "\n",
      "rpoC_p.Pro1040Arg\n",
      "\n",
      "glpK_p.Leu152Arg\n",
      "\n",
      "Rv2752c_p.Val396Gly\n",
      "\n",
      "lpqB_p.Asp370Glu\n",
      "\n",
      "rpoA_c.-316G>A\n",
      "\n",
      "Rv1129c_p.Ser362Thr\n"
     ]
    }
   ],
   "source": [
    "pval_df = pd.DataFrame(columns=['AUC', 'Sens', 'Spec', 'accuracy'])\n",
    "\n",
    "for mut in LRT_bootstrap_StatsDiff.mutation.unique():\n",
    "    \n",
    "    print(\"\\n\" + mut)\n",
    "    \n",
    "    for metric in ['AUC']:\n",
    "    \n",
    "#         lower, upper = np.percentile(LRT_bootstrap_StatsDiff.query(\"mutation==@mut\")[metric], q=[2.5, 97.5])\n",
    "    \n",
    "#         if lower > 0 and upper > 0:\n",
    "#             print(f\"    {metric}: {round(lower, 4)}, {round(upper, 4)}\")\n",
    "\n",
    "        pval = 1 - np.mean(LRT_bootstrap_StatsDiff.query('mutation==@mut')[metric].values > 0)\n",
    "        \n",
    "        if pval < 0.05:\n",
    "            print(f\"{metric}: {round(pval, 2)}\")\n",
    "            \n",
    "        pval_df.loc[mut, metric] = pval"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51ff9fee-e0ea-4908-ad04-750f663165f2",
   "metadata": {},
   "source": [
    "# Model with Interaction Terms\n",
    "\n",
    "Chose 2 variants because they were the only Tier 2 genes that were significantly associated in the Ridge regression and were significant in the LRT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a490028-8614-4b27-86bf-51612d571e1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30984, 3736)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 0.01\n",
    "interact_mut = tier2_model.query('BH_pval < @thresh & LRT_BH_pval < @thresh')[\"mutation\"].values\n",
    "#interact_mut = ['rpoC_p.Glu1092Asp', 'Rv2752c_p.Asn30Ser']\n",
    "\n",
    "# model_matrix = pd.read_pickle(os.path.join(analysis_dir, drug, \"BINARY/tiers=1+2/phenos=WHO/dropAF_noSyn_unpooled/model_matrix.pkl\"))\n",
    "# interact_matrix = model_matrix.copy()\n",
    "\n",
    "# # get a list of all pairs of mutations to interact\n",
    "# interactions = list(itertools.product(*[interact_mut, interact_matrix.columns[interact_matrix.columns.str.contains(\"rpoB\")]]))\n",
    "\n",
    "# # add interactions\n",
    "# for (mut1, mut2) in interactions:\n",
    "#     interact_matrix[f\"{mut1}+{mut2}\"] = interact_matrix[mut1] * interact_matrix[mut2]\n",
    "    \n",
    "# # drop any columns with no signal\n",
    "# interact_matrix = interact_matrix[interact_matrix.columns[~((interact_matrix == 0).all())]]\n",
    "\n",
    "# assert len(set(model_matrix.columns) - set(interact_matrix.columns)) == 0\n",
    "# print(len(interact_matrix), len(model_matrix))\n",
    "\n",
    "# interact_matrix.to_pickle(os.path.join(analysis_dir, drug, f\"BINARY/interaction/model_matrix_{phenos_name}.pkl\"))\n",
    "\n",
    "interact_matrix = pd.read_pickle(os.path.join(analysis_dir, drug, f\"BINARY/interaction/model_matrix_{phenos_name}.pkl\"))\n",
    "interact_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "61501b9b-818c-474a-8f38-82074d1e4b53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rpoC_p.Glu1092Asp 125\n",
      "Rv2752c_p.Asn30Ser 4\n"
     ]
    }
   ],
   "source": [
    "# so much co-occurrence of rpoC_p.Glu1092Asp with rpoB mutations. Can also check proportion that are high confidence rpoB mutations\n",
    "for mut in interact_mut:\n",
    "    print(mut, interact_matrix[interact_matrix.columns[interact_matrix.columns.str.contains(mut)]].shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "97e9b440-e22c-4b3a-8bb0-490cb0fc96e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenos_name = \"WHO\"\n",
    "interact_res = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_coef.csv\"))\n",
    "interact_permute = pd.read_csv(os.path.join(analysis_dir, drug, f\"BINARY/interaction/{phenos_name}_coef_permutation.csv\"))\n",
    "\n",
    "# assess significance using the results of the permutation test\n",
    "for i, row in interact_res.iterrows():\n",
    "    # p-value is the proportion of permutation coefficients that are AT LEAST AS EXTREME as the test statistic\n",
    "    if row[\"coef\"] > 0:\n",
    "        interact_res.loc[i, \"pval\"] = np.mean(interact_permute[row[\"mutation\"]] >= row[\"coef\"])\n",
    "    else:\n",
    "        interact_res.loc[i, \"pval\"] = np.mean(interact_permute[row[\"mutation\"]] <= row[\"coef\"])\n",
    "        \n",
    "interact_res = add_pval_corrections(interact_res)\n",
    "interact_res[\"Odds_Ratio\"] = np.exp(interact_res[\"coef\"])\n",
    "\n",
    "interact_res[[\"Mut1\", \"Mut2\"]] = interact_res[\"mutation\"].str.split(\"+\", expand=True)\n",
    "interact_res = interact_res.merge(who_variants_combined.query(\"drug=='RIF'\"), left_on=\"Mut2\", right_on=\"mutation\", how=\"left\")\n",
    "del interact_res[\"mutation_y\"]\n",
    "interact_res.rename(columns={\"mutation_x\": \"mutation\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "42d4f33d-4f57-4cc7-92b8-cf3ecb1cc77a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>drug</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.219309</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.245216</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Ile491Tyr</td>\n",
       "      <td>0.026609</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.026966</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Ile491Tyr</td>\n",
       "      <td>RIF</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Met434_Asn437delinsIle</td>\n",
       "      <td>0.018817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.018995</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Met434_Asn437delinsIle</td>\n",
       "      <td>RIF</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp</td>\n",
       "      <td>-0.027974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972414</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            mutation      coef  pval  BH_pval  \\\n",
       "219                                rpoC_p.Glu1092Asp  0.219309   0.0      0.0   \n",
       "170               rpoC_p.Glu1092Asp+rpoB_p.Ile491Tyr  0.026609   0.0      0.0   \n",
       "169  rpoC_p.Glu1092Asp+rpoB_p.Met434_Asn437delinsIle  0.018817   0.0      0.0   \n",
       "171               rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp -0.027974   0.0      0.0   \n",
       "\n",
       "     Bonferroni_pval  Odds_Ratio               Mut1  \\\n",
       "219              0.0    1.245216  rpoC_p.Glu1092Asp   \n",
       "170              0.0    1.026966  rpoC_p.Glu1092Asp   \n",
       "169              0.0    1.018995  rpoC_p.Glu1092Asp   \n",
       "171              0.0    0.972414  rpoC_p.Glu1092Asp   \n",
       "\n",
       "                              Mut2 drug                 confidence  \n",
       "219                           None  NaN                        NaN  \n",
       "170               rpoB_p.Ile491Tyr  RIF  3) Uncertain significance  \n",
       "169  rpoB_p.Met434_Asn437delinsIle  RIF     2) Assoc w R - Interim  \n",
       "171               rpoB_p.Asn260Asp  NaN                        NaN  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation.str.contains('rpoC_p.Glu1092Asp') & BH_pval < 0.05\").sort_values([\"BH_pval\", \"Odds_Ratio\"], \n",
    "                                                                                           ascending=[True, False]\n",
    "                                                                                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "50e05e3c-ae52-41db-aa52-c197e98d458d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>drug</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>rpoC_p.Val483Ala</td>\n",
       "      <td>0.16795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.182877</td>\n",
       "      <td>rpoC_p.Val483Ala</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation     coef  pval  BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "175  rpoC_p.Val483Ala  0.16795   0.0      0.0              0.0    1.182877   \n",
       "\n",
       "                 Mut1  Mut2 drug confidence  \n",
       "175  rpoC_p.Val483Ala  None  NaN        NaN  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation=='rpoC_p.Val483Ala'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "9344d6c2-420a-4320-a06e-1f30196b8ee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>drug</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>-0.027974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972414</td>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mutation      coef  pval  BH_pval  Bonferroni_pval  Odds_Ratio  \\\n",
       "91  rpoB_p.Asn260Asp -0.027974   0.0      0.0              0.0    0.972414   \n",
       "\n",
       "                Mut1  Mut2 drug confidence  \n",
       "91  rpoB_p.Asn260Asp  None  NaN        NaN  "
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation=='rpoB_p.Asn260Asp'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "a8c0efc9-08e2-4c65-825c-c6f7cf3ffecb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30984, 3736)"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "fb3004f7-7d49-4a10-b8ed-d175c491987c",
   "metadata": {},
   "outputs": [],
   "source": [
    "interact_matrix = interact_matrix.merge(df_phenos, left_index=True, right_index=True)\n",
    "del interact_matrix[\"phenotypic_category\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b69ff2be-3411-42a3-9139-9fc5f0173dde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>Mut1</th>\n",
       "      <th>Mut2</th>\n",
       "      <th>drug</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>0.219309</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.245216</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Ile491Tyr</td>\n",
       "      <td>0.026609</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.026966</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Ile491Tyr</td>\n",
       "      <td>RIF</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Met434_Asn437delinsIle</td>\n",
       "      <td>0.018817</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.018995</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Met434_Asn437delinsIle</td>\n",
       "      <td>RIF</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp</td>\n",
       "      <td>-0.027974</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.972414</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Asn260Asp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu</td>\n",
       "      <td>0.057540</td>\n",
       "      <td>0.004</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.059228</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>RIF</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Asp435Gly</td>\n",
       "      <td>0.038386</td>\n",
       "      <td>0.027</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.039132</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Asp435Gly</td>\n",
       "      <td>RIF</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Val168Ala</td>\n",
       "      <td>0.033776</td>\n",
       "      <td>0.006</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.034352</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Val168Ala</td>\n",
       "      <td>RIF</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Ser450Trp</td>\n",
       "      <td>0.026845</td>\n",
       "      <td>0.019</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.027209</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Ser450Trp</td>\n",
       "      <td>RIF</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Asp435Asn</td>\n",
       "      <td>0.026641</td>\n",
       "      <td>0.029</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.026999</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Asp435Asn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.His445Leu</td>\n",
       "      <td>0.025338</td>\n",
       "      <td>0.043</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.025661</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.His445Leu</td>\n",
       "      <td>RIF</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Leu554Pro</td>\n",
       "      <td>0.010183</td>\n",
       "      <td>0.037</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.010235</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Leu554Pro</td>\n",
       "      <td>RIF</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Leu452Pro</td>\n",
       "      <td>-0.032888</td>\n",
       "      <td>0.026</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.967646</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Leu452Pro</td>\n",
       "      <td>RIF</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Val170Phe</td>\n",
       "      <td>-0.057892</td>\n",
       "      <td>0.003</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.943752</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Val170Phe</td>\n",
       "      <td>RIF</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>rpoC_p.Glu1092Asp+rpoB_p.Leu430Pro</td>\n",
       "      <td>-0.060548</td>\n",
       "      <td>0.010</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.941249</td>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>rpoB_p.Leu430Pro</td>\n",
       "      <td>RIF</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            mutation      coef   pval  \\\n",
       "219                                rpoC_p.Glu1092Asp  0.219309  0.000   \n",
       "170               rpoC_p.Glu1092Asp+rpoB_p.Ile491Tyr  0.026609  0.000   \n",
       "169  rpoC_p.Glu1092Asp+rpoB_p.Met434_Asn437delinsIle  0.018817  0.000   \n",
       "171               rpoC_p.Glu1092Asp+rpoB_p.Asn260Asp -0.027974  0.000   \n",
       "300               rpoC_p.Glu1092Asp+rpoB_p.Ser450Leu  0.057540  0.004   \n",
       "429               rpoC_p.Glu1092Asp+rpoB_p.Asp435Gly  0.038386  0.027   \n",
       "330               rpoC_p.Glu1092Asp+rpoB_p.Val168Ala  0.033776  0.006   \n",
       "405               rpoC_p.Glu1092Asp+rpoB_p.Ser450Trp  0.026845  0.019   \n",
       "436               rpoC_p.Glu1092Asp+rpoB_p.Asp435Asn  0.026641  0.029   \n",
       "488               rpoC_p.Glu1092Asp+rpoB_p.His445Leu  0.025338  0.043   \n",
       "468               rpoC_p.Glu1092Asp+rpoB_p.Leu554Pro  0.010183  0.037   \n",
       "425               rpoC_p.Glu1092Asp+rpoB_p.Leu452Pro -0.032888  0.026   \n",
       "287               rpoC_p.Glu1092Asp+rpoB_p.Val170Phe -0.057892  0.003   \n",
       "351               rpoC_p.Glu1092Asp+rpoB_p.Leu430Pro -0.060548  0.010   \n",
       "\n",
       "     BH_pval  Bonferroni_pval  Odds_Ratio               Mut1  \\\n",
       "219      0.0              0.0    1.245216  rpoC_p.Glu1092Asp   \n",
       "170      0.0              0.0    1.026966  rpoC_p.Glu1092Asp   \n",
       "169      0.0              0.0    1.018995  rpoC_p.Glu1092Asp   \n",
       "171      0.0              0.0    0.972414  rpoC_p.Glu1092Asp   \n",
       "300      1.0              1.0    1.059228  rpoC_p.Glu1092Asp   \n",
       "429      1.0              1.0    1.039132  rpoC_p.Glu1092Asp   \n",
       "330      1.0              1.0    1.034352  rpoC_p.Glu1092Asp   \n",
       "405      1.0              1.0    1.027209  rpoC_p.Glu1092Asp   \n",
       "436      1.0              1.0    1.026999  rpoC_p.Glu1092Asp   \n",
       "488      1.0              1.0    1.025661  rpoC_p.Glu1092Asp   \n",
       "468      1.0              1.0    1.010235  rpoC_p.Glu1092Asp   \n",
       "425      1.0              1.0    0.967646  rpoC_p.Glu1092Asp   \n",
       "287      1.0              1.0    0.943752  rpoC_p.Glu1092Asp   \n",
       "351      1.0              1.0    0.941249  rpoC_p.Glu1092Asp   \n",
       "\n",
       "                              Mut2 drug                 confidence  \n",
       "219                           None  NaN                        NaN  \n",
       "170               rpoB_p.Ile491Tyr  RIF  3) Uncertain significance  \n",
       "169  rpoB_p.Met434_Asn437delinsIle  RIF     2) Assoc w R - Interim  \n",
       "171               rpoB_p.Asn260Asp  NaN                        NaN  \n",
       "300               rpoB_p.Ser450Leu  RIF               1) Assoc w R  \n",
       "429               rpoB_p.Asp435Gly  RIF     2) Assoc w R - Interim  \n",
       "330               rpoB_p.Val168Ala  RIF  3) Uncertain significance  \n",
       "405               rpoB_p.Ser450Trp  RIF               1) Assoc w R  \n",
       "436               rpoB_p.Asp435Asn  NaN                        NaN  \n",
       "488               rpoB_p.His445Leu  RIF               1) Assoc w R  \n",
       "468               rpoB_p.Leu554Pro  RIF  3) Uncertain significance  \n",
       "425               rpoB_p.Leu452Pro  RIF               1) Assoc w R  \n",
       "287               rpoB_p.Val170Phe  RIF               1) Assoc w R  \n",
       "351               rpoB_p.Leu430Pro  RIF               1) Assoc w R  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_res.query(\"mutation.str.contains('rpoC_p.Glu1092Asp') & pval < 0.05\").sort_values([\"BH_pval\", \"Odds_Ratio\"], \n",
    "                                                                                           ascending=[True, False]\n",
    "                                                                                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "271a6915-6fa4-46a0-bc6c-87c341ec5f17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>confidence</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4680</th>\n",
       "      <td>RIF</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>rpoB_p.Ile491Tyr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     drug                 confidence          mutation\n",
       "4680  RIF  3) Uncertain significance  rpoB_p.Ile491Tyr"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who_variants_combined.query(\"mutation=='rpoB_p.Ile491Tyr'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "947aaa3a-da07-4ffd-9d68-9e6e361ed275",
   "metadata": {},
   "outputs": [],
   "source": [
    "tiers1_stats = pd.read_csv(os.path.join(analysis_paths[0], \"model_analysis_with_stats.csv\"))\n",
    "tiers12_stats = pd.read_csv(os.path.join(analysis_paths[1], \"model_analysis_with_stats.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "7ee996f7-1941-435f-8aef-474a7968aac6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>confidence</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>NPV_LB</th>\n",
       "      <th>NPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761154;761155</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>13.996830</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>7536</td>\n",
       "      <td>...</td>\n",
       "      <td>0.849854</td>\n",
       "      <td>0.858186</td>\n",
       "      <td>0.638086</td>\n",
       "      <td>0.655640</td>\n",
       "      <td>0.995089</td>\n",
       "      <td>0.996740</td>\n",
       "      <td>131.630401</td>\n",
       "      <td>196.405333</td>\n",
       "      <td>0.345859</td>\n",
       "      <td>0.363410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpoB_p.Asp435Val</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761110</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>2.453635</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>695</td>\n",
       "      <td>...</td>\n",
       "      <td>0.683125</td>\n",
       "      <td>0.692907</td>\n",
       "      <td>0.055387</td>\n",
       "      <td>0.064131</td>\n",
       "      <td>0.999284</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>81.961356</td>\n",
       "      <td>305.112605</td>\n",
       "      <td>0.936382</td>\n",
       "      <td>0.945054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rpoB_p.His445Tyr</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761138;761139</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>2.028256</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>458</td>\n",
       "      <td>...</td>\n",
       "      <td>0.678422</td>\n",
       "      <td>0.688210</td>\n",
       "      <td>0.035491</td>\n",
       "      <td>0.042651</td>\n",
       "      <td>0.999229</td>\n",
       "      <td>0.999799</td>\n",
       "      <td>49.686851</td>\n",
       "      <td>173.868388</td>\n",
       "      <td>0.957910</td>\n",
       "      <td>0.965003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rpoB_p.His445Asp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761139</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.953321</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>385</td>\n",
       "      <td>...</td>\n",
       "      <td>0.677139</td>\n",
       "      <td>0.686928</td>\n",
       "      <td>0.029849</td>\n",
       "      <td>0.036471</td>\n",
       "      <td>0.999511</td>\n",
       "      <td>0.999932</td>\n",
       "      <td>65.271465</td>\n",
       "      <td>380.899057</td>\n",
       "      <td>0.963895</td>\n",
       "      <td>0.970440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rpoB_p.Ser450Trp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761155</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.595500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>198</td>\n",
       "      <td>...</td>\n",
       "      <td>0.673539</td>\n",
       "      <td>0.683331</td>\n",
       "      <td>0.014594</td>\n",
       "      <td>0.019391</td>\n",
       "      <td>0.999571</td>\n",
       "      <td>0.999954</td>\n",
       "      <td>37.397049</td>\n",
       "      <td>270.739868</td>\n",
       "      <td>0.980940</td>\n",
       "      <td>0.985659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>rpoB_p.Asp853_Glu854dup</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>762368;762365;762352</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.944660</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>0.320645</td>\n",
       "      <td>0.330445</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>0.001893</td>\n",
       "      <td>0.999679</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.998195</td>\n",
       "      <td>0.999123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>rpoB_p.Ile925Val</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>762579</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.938021</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>0.320866</td>\n",
       "      <td>0.330675</td>\n",
       "      <td>0.002399</td>\n",
       "      <td>0.003845</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>0.999700</td>\n",
       "      <td>2.120413</td>\n",
       "      <td>9.123034</td>\n",
       "      <td>0.996785</td>\n",
       "      <td>0.998485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>rpoB_c.-83A&gt;G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>759724</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.936417</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.546667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0.320480</td>\n",
       "      <td>0.330277</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.001089</td>\n",
       "      <td>0.999516</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1.022873</td>\n",
       "      <td>58.144978</td>\n",
       "      <td>0.999046</td>\n",
       "      <td>0.999786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813</th>\n",
       "      <td>rpoB_p.Asn381His</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>760947</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.898769</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>76</td>\n",
       "      <td>...</td>\n",
       "      <td>0.321046</td>\n",
       "      <td>0.330856</td>\n",
       "      <td>0.002510</td>\n",
       "      <td>0.003985</td>\n",
       "      <td>0.999679</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.996100</td>\n",
       "      <td>0.997530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>814</th>\n",
       "      <td>rpoB_c.-61C&gt;T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>759746</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.826835</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.341216</td>\n",
       "      <td>0.351782</td>\n",
       "      <td>0.136921</td>\n",
       "      <td>0.145803</td>\n",
       "      <td>0.940186</td>\n",
       "      <td>0.948640</td>\n",
       "      <td>2.347780</td>\n",
       "      <td>2.764205</td>\n",
       "      <td>0.902961</td>\n",
       "      <td>0.915308</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>815 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mutation       predicted_effect              position  \\\n",
       "0           rpoB_p.Ser450Leu       missense_variant         761154;761155   \n",
       "1           rpoB_p.Asp435Val       missense_variant                761110   \n",
       "2           rpoB_p.His445Tyr       missense_variant         761138;761139   \n",
       "3           rpoB_p.His445Asp       missense_variant                761139   \n",
       "4           rpoB_p.Ser450Trp       missense_variant                761155   \n",
       "..                       ...                    ...                   ...   \n",
       "810  rpoB_p.Asp853_Glu854dup      inframe_insertion  762368;762365;762352   \n",
       "811         rpoB_p.Ile925Val       missense_variant                762579   \n",
       "812            rpoB_c.-83A>G  upstream_gene_variant                759724   \n",
       "813         rpoB_p.Asn381His       missense_variant                760947   \n",
       "814            rpoB_c.-61C>T  upstream_gene_variant                759746   \n",
       "\n",
       "                    confidence  Odds_Ratio   pval   BH_pval  Bonferroni_pval  \\\n",
       "0                 1) Assoc w R   13.996830  0.000  0.000000              0.0   \n",
       "1                 1) Assoc w R    2.453635  0.000  0.000000              0.0   \n",
       "2                 1) Assoc w R    2.028256  0.000  0.000000              0.0   \n",
       "3                 1) Assoc w R    1.953321  0.000  0.000000              0.0   \n",
       "4                 1) Assoc w R    1.595500  0.000  0.000000              0.0   \n",
       "..                         ...         ...    ...       ...              ...   \n",
       "810           5) Not assoc w R    0.944660  0.000  0.000000              0.0   \n",
       "811           5) Not assoc w R    0.938021  0.000  0.000000              0.0   \n",
       "812  3) Uncertain significance    0.936417  0.002  0.546667              1.0   \n",
       "813           5) Not assoc w R    0.898769  0.000  0.000000              0.0   \n",
       "814           5) Not assoc w R    0.826835  0.000  0.000000              0.0   \n",
       "\n",
       "     Significant  Num_Isolates  ...    NPV_LB    NPV_UB   Sens_LB   Sens_UB  \\\n",
       "0              1          7536  ...  0.849854  0.858186  0.638086  0.655640   \n",
       "1              1           695  ...  0.683125  0.692907  0.055387  0.064131   \n",
       "2              1           458  ...  0.678422  0.688210  0.035491  0.042651   \n",
       "3              1           385  ...  0.677139  0.686928  0.029849  0.036471   \n",
       "4              1           198  ...  0.673539  0.683331  0.014594  0.019391   \n",
       "..           ...           ...  ...       ...       ...       ...       ...   \n",
       "810            1            32  ...  0.320645  0.330445  0.000917  0.001893   \n",
       "811            1            81  ...  0.320866  0.330675  0.002399  0.003845   \n",
       "812            0            17  ...  0.320480  0.330277  0.000383  0.001089   \n",
       "813            1            76  ...  0.321046  0.330856  0.002510  0.003985   \n",
       "814            1          4010  ...  0.341216  0.351782  0.136921  0.145803   \n",
       "\n",
       "      Spec_LB   Spec_UB      LR+_LB      LR+_UB    LR-_LB    LR-_UB  \n",
       "0    0.995089  0.996740  131.630401  196.405333  0.345859  0.363410  \n",
       "1    0.999284  0.999828   81.961356  305.112605  0.936382  0.945054  \n",
       "2    0.999229  0.999799   49.686851  173.868388  0.957910  0.965003  \n",
       "3    0.999511  0.999932   65.271465  380.899057  0.963895  0.970440  \n",
       "4    0.999571  0.999954   37.397049  270.739868  0.980940  0.985659  \n",
       "..        ...       ...         ...         ...       ...       ...  \n",
       "810  0.999679  1.000000         NaN         inf  0.998195  0.999123  \n",
       "811  0.998630  0.999700    2.120413    9.123034  0.996785  0.998485  \n",
       "812  0.999516  0.999998    1.022873   58.144978  0.999046  0.999786  \n",
       "813  0.999679  1.000000         NaN         inf  0.996100  0.997530  \n",
       "814  0.940186  0.948640    2.347780    2.764205  0.902961  0.915308  \n",
       "\n",
       "[815 rows x 33 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiers1_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "6ed656ea-d0e3-45c1-a76e-020b18d3ae50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>confidence</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>TN</th>\n",
       "      <th>FN</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpoB_p.Ser450Leu</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761154;761155</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>13.996830</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>7536</td>\n",
       "      <td>...</td>\n",
       "      <td>7440</td>\n",
       "      <td>96</td>\n",
       "      <td>23765</td>\n",
       "      <td>4061</td>\n",
       "      <td>0.987261</td>\n",
       "      <td>0.854057</td>\n",
       "      <td>0.646900</td>\n",
       "      <td>0.995977</td>\n",
       "      <td>160.788410</td>\n",
       "      <td>0.354526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpoB_p.Asp435Val</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761110</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>2.453635</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>695</td>\n",
       "      <td>...</td>\n",
       "      <td>686</td>\n",
       "      <td>9</td>\n",
       "      <td>23852</td>\n",
       "      <td>10815</td>\n",
       "      <td>0.987050</td>\n",
       "      <td>0.688032</td>\n",
       "      <td>0.059647</td>\n",
       "      <td>0.999623</td>\n",
       "      <td>158.137418</td>\n",
       "      <td>0.940708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rpoB_p.His445Tyr</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761138;761139</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>2.028256</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>458</td>\n",
       "      <td>...</td>\n",
       "      <td>448</td>\n",
       "      <td>10</td>\n",
       "      <td>23851</td>\n",
       "      <td>11053</td>\n",
       "      <td>0.978166</td>\n",
       "      <td>0.683331</td>\n",
       "      <td>0.038953</td>\n",
       "      <td>0.999581</td>\n",
       "      <td>92.946074</td>\n",
       "      <td>0.961450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rpoB_p.His445Asp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761139</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.953321</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>385</td>\n",
       "      <td>...</td>\n",
       "      <td>380</td>\n",
       "      <td>5</td>\n",
       "      <td>23856</td>\n",
       "      <td>11121</td>\n",
       "      <td>0.987013</td>\n",
       "      <td>0.682048</td>\n",
       "      <td>0.033041</td>\n",
       "      <td>0.999790</td>\n",
       "      <td>157.676376</td>\n",
       "      <td>0.967162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rpoB_p.Ser450Trp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>761155</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.595500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>198</td>\n",
       "      <td>...</td>\n",
       "      <td>194</td>\n",
       "      <td>4</td>\n",
       "      <td>23857</td>\n",
       "      <td>11307</td>\n",
       "      <td>0.979798</td>\n",
       "      <td>0.678450</td>\n",
       "      <td>0.016868</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>100.622424</td>\n",
       "      <td>0.983297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>rpoB_p.Asp853_Glu854dup</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>762368;762365;762352</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.944660</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>11501</td>\n",
       "      <td>23829</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.325531</td>\n",
       "      <td>0.001341</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.998659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>rpoB_p.Ile925Val</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>762579</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.938021</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>73</td>\n",
       "      <td>8</td>\n",
       "      <td>11493</td>\n",
       "      <td>23788</td>\n",
       "      <td>0.901235</td>\n",
       "      <td>0.325756</td>\n",
       "      <td>0.003059</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>4.398249</td>\n",
       "      <td>0.997635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>rpoB_c.-83A&gt;G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>759724</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.936417</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.546667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>11500</td>\n",
       "      <td>23845</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.000671</td>\n",
       "      <td>0.999913</td>\n",
       "      <td>7.711999</td>\n",
       "      <td>0.999416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813</th>\n",
       "      <td>rpoB_p.Asn381His</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>760947</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.898769</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>76</td>\n",
       "      <td>...</td>\n",
       "      <td>76</td>\n",
       "      <td>0</td>\n",
       "      <td>11501</td>\n",
       "      <td>23785</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.325937</td>\n",
       "      <td>0.003185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.996815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>814</th>\n",
       "      <td>rpoB_c.-61C&gt;T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>759746</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.826835</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4010</td>\n",
       "      <td>...</td>\n",
       "      <td>3372</td>\n",
       "      <td>638</td>\n",
       "      <td>10863</td>\n",
       "      <td>20489</td>\n",
       "      <td>0.840898</td>\n",
       "      <td>0.346485</td>\n",
       "      <td>0.141318</td>\n",
       "      <td>0.944527</td>\n",
       "      <td>2.547498</td>\n",
       "      <td>0.909113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>815 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mutation       predicted_effect              position  \\\n",
       "0           rpoB_p.Ser450Leu       missense_variant         761154;761155   \n",
       "1           rpoB_p.Asp435Val       missense_variant                761110   \n",
       "2           rpoB_p.His445Tyr       missense_variant         761138;761139   \n",
       "3           rpoB_p.His445Asp       missense_variant                761139   \n",
       "4           rpoB_p.Ser450Trp       missense_variant                761155   \n",
       "..                       ...                    ...                   ...   \n",
       "810  rpoB_p.Asp853_Glu854dup      inframe_insertion  762368;762365;762352   \n",
       "811         rpoB_p.Ile925Val       missense_variant                762579   \n",
       "812            rpoB_c.-83A>G  upstream_gene_variant                759724   \n",
       "813         rpoB_p.Asn381His       missense_variant                760947   \n",
       "814            rpoB_c.-61C>T  upstream_gene_variant                759746   \n",
       "\n",
       "                    confidence  Odds_Ratio   pval   BH_pval  Bonferroni_pval  \\\n",
       "0                 1) Assoc w R   13.996830  0.000  0.000000              0.0   \n",
       "1                 1) Assoc w R    2.453635  0.000  0.000000              0.0   \n",
       "2                 1) Assoc w R    2.028256  0.000  0.000000              0.0   \n",
       "3                 1) Assoc w R    1.953321  0.000  0.000000              0.0   \n",
       "4                 1) Assoc w R    1.595500  0.000  0.000000              0.0   \n",
       "..                         ...         ...    ...       ...              ...   \n",
       "810           5) Not assoc w R    0.944660  0.000  0.000000              0.0   \n",
       "811           5) Not assoc w R    0.938021  0.000  0.000000              0.0   \n",
       "812  3) Uncertain significance    0.936417  0.002  0.546667              1.0   \n",
       "813           5) Not assoc w R    0.898769  0.000  0.000000              0.0   \n",
       "814           5) Not assoc w R    0.826835  0.000  0.000000              0.0   \n",
       "\n",
       "     Significant  Num_Isolates  ...    TP   FP     TN     FN       PPV  \\\n",
       "0              1          7536  ...  7440   96  23765   4061  0.987261   \n",
       "1              1           695  ...   686    9  23852  10815  0.987050   \n",
       "2              1           458  ...   448   10  23851  11053  0.978166   \n",
       "3              1           385  ...   380    5  23856  11121  0.987013   \n",
       "4              1           198  ...   194    4  23857  11307  0.979798   \n",
       "..           ...           ...  ...   ...  ...    ...    ...       ...   \n",
       "810            1            32  ...    32    0  11501  23829  1.000000   \n",
       "811            1            81  ...    73    8  11493  23788  0.901235   \n",
       "812            0            17  ...    16    1  11500  23845  0.941176   \n",
       "813            1            76  ...    76    0  11501  23785  1.000000   \n",
       "814            1          4010  ...  3372  638  10863  20489  0.840898   \n",
       "\n",
       "          NPV      Sens      Spec         LR+       LR-  \n",
       "0    0.854057  0.646900  0.995977  160.788410  0.354526  \n",
       "1    0.688032  0.059647  0.999623  158.137418  0.940708  \n",
       "2    0.683331  0.038953  0.999581   92.946074  0.961450  \n",
       "3    0.682048  0.033041  0.999790  157.676376  0.967162  \n",
       "4    0.678450  0.016868  0.999832  100.622424  0.983297  \n",
       "..        ...       ...       ...         ...       ...  \n",
       "810  0.325531  0.001341  1.000000         inf  0.998659  \n",
       "811  0.325756  0.003059  0.999304    4.398249  0.997635  \n",
       "812  0.325364  0.000671  0.999913    7.711999  0.999416  \n",
       "813  0.325937  0.003185  1.000000         inf  0.996815  \n",
       "814  0.346485  0.141318  0.944527    2.547498  0.909113  \n",
       "\n",
       "[815 rows x 21 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiers1_stats[tiers1_stats.columns[~tiers1_stats.columns.str.contains(\"|\".join([\"LB\", \"UB\"]))]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "42f8a24b-9112-4802-ad3a-c5815e407c4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>confidence</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>TN</th>\n",
       "      <th>FN</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rpoC_p.Val483Gly</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>764817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.351136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>900</td>\n",
       "      <td>...</td>\n",
       "      <td>883</td>\n",
       "      <td>17</td>\n",
       "      <td>20904</td>\n",
       "      <td>9180</td>\n",
       "      <td>0.981111</td>\n",
       "      <td>0.694854</td>\n",
       "      <td>0.087747</td>\n",
       "      <td>0.999187</td>\n",
       "      <td>107.985825</td>\n",
       "      <td>0.912995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rpoC_p.Glu1092Asp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>766645</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>1.216332</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2697</td>\n",
       "      <td>...</td>\n",
       "      <td>1711</td>\n",
       "      <td>986</td>\n",
       "      <td>19935</td>\n",
       "      <td>8352</td>\n",
       "      <td>0.634409</td>\n",
       "      <td>0.704741</td>\n",
       "      <td>0.170029</td>\n",
       "      <td>0.952870</td>\n",
       "      <td>3.607680</td>\n",
       "      <td>0.871022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rpoC_p.Val483Ala</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>764817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.172390</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>236</td>\n",
       "      <td>...</td>\n",
       "      <td>233</td>\n",
       "      <td>3</td>\n",
       "      <td>20918</td>\n",
       "      <td>9830</td>\n",
       "      <td>0.987288</td>\n",
       "      <td>0.680304</td>\n",
       "      <td>0.023154</td>\n",
       "      <td>0.999857</td>\n",
       "      <td>161.469178</td>\n",
       "      <td>0.976986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation  predicted_effect position        confidence  \\\n",
       "9    rpoC_p.Val483Gly  missense_variant   764817               NaN   \n",
       "16  rpoC_p.Glu1092Asp  missense_variant   766645  5) Not assoc w R   \n",
       "27   rpoC_p.Val483Ala  missense_variant   764817               NaN   \n",
       "\n",
       "    Odds_Ratio  pval  BH_pval  Bonferroni_pval  Significant  Num_Isolates  \\\n",
       "9     1.351136   0.0      0.0              0.0            1           900   \n",
       "16    1.216332   0.0      0.0              0.0            1          2697   \n",
       "27    1.172390   0.0      0.0              0.0            1           236   \n",
       "\n",
       "    ...    TP   FP     TN    FN       PPV       NPV      Sens      Spec  \\\n",
       "9   ...   883   17  20904  9180  0.981111  0.694854  0.087747  0.999187   \n",
       "16  ...  1711  986  19935  8352  0.634409  0.704741  0.170029  0.952870   \n",
       "27  ...   233    3  20918  9830  0.987288  0.680304  0.023154  0.999857   \n",
       "\n",
       "           LR+       LR-  \n",
       "9   107.985825  0.912995  \n",
       "16    3.607680  0.871022  \n",
       "27  161.469178  0.976986  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiers12_stats[tiers12_stats.columns[~tiers12_stats.columns.str.contains(\"|\".join([\"LB\", \"UB\"]))]].query(\"mutation in ['rpoC_p.Glu1092Asp', 'rpoC_p.Val483Ala', 'rpoC_p.Val483Gly']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "39c1efde-d4d7-481b-84ac-ea3ba0172502",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PPV_LB</th>\n",
       "      <th>PPV_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.969929</td>\n",
       "      <td>0.988959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.615909</td>\n",
       "      <td>0.652616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.963302</td>\n",
       "      <td>0.997371</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      PPV_LB    PPV_UB\n",
       "9   0.969929  0.988959\n",
       "16  0.615909  0.652616\n",
       "27  0.963302  0.997371"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiers12_stats.query(\"mutation in ['rpoC_p.Glu1092Asp', 'rpoC_p.Val483Ala', 'rpoC_p.Val483Gly']\")[[\"PPV_LB\", \"PPV_UB\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d96600d-2456-4d00-a536-47d985a2117f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
