{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import sklearn\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "import glob, os, yaml, subprocess, itertools, sparse, pickle\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)\n",
    "samples_summary = pd.read_csv(\"../data/samples_summary.csv\")\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2439e0a-1ad9-4471-9e91-ac841f82fba2",
   "metadata": {},
   "source": [
    "# Combined Analysis Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "00729d49-7aeb-457e-af65-74a6bf488b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def final_processing(drug):\n",
    "    '''\n",
    "    Functions for processing outputs before sending to everyone else.\n",
    "    \n",
    "    1. Remove principal components (will describe them separately)\n",
    "    2. Add LOF to the predicted_effect column for pooled LOF mutations\n",
    "    3. Remove genome_index column (should actually do that earlier, but will fix later)\n",
    "    4. Remove the logistic regression coefficient columns (they will prefer to work with odds ratios)\n",
    "    5. Any other column renaming or dropping for clarity\n",
    "    '''\n",
    "    \n",
    "    analysis_df = pd.read_csv(os.path.join(analysis_dir, drug, \"final_analysis.csv\"))\n",
    "    analysis_df.rename(columns={\"orig_variant\": \"mutation\"}, inplace=True)\n",
    "    \n",
    "    # remove logReg coefficients. Keep only odds ratios. Remove the other two columns, which were present mainly for me to see\n",
    "    # if we were picking up many mutations that were in the 2021 mutation catalog\n",
    "    del analysis_df[\"confidence_WHO_2021\"]\n",
    "    analysis_df = analysis_df[analysis_df.columns[~analysis_df.columns.str.contains(\"coef\")]]\n",
    "    \n",
    "    # remove significant principal components and replace the NaNs in the predicted effect column for the gene loss of functions\n",
    "    analysis_df = analysis_df.loc[~analysis_df[\"mutation\"].str.contains(\"PC\")]\n",
    "    \n",
    "    # predicted effect should not be NaN for anything. position is NaN only for the pooled LOF mutations\n",
    "    assert len(analysis_df.loc[pd.isnull(analysis_df[\"predicted_effect\"])]) == 0\n",
    "    \n",
    "    assert len(analysis_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']].dropna()) == len(analysis_df)\n",
    "    analysis_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']] = analysis_df[['Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN']].astype(int)\n",
    "    \n",
    "    # reorder columns\n",
    "    analysis_df = analysis_df[['mutation', 'predicted_effect', 'OR_LB', 'Odds_Ratio', 'OR_UB', 'pval', 'BH_pval', 'Bonferroni_pval',\n",
    "       'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV_LB', 'PPV', 'PPV_UB', 'Sens_LB', 'Sens', 'Sens_UB', 'Spec_LB', 'Spec', 'Spec_UB',\n",
    "       'LR+_LB', 'LR+', 'LR+_UB', 'LR-_LB', 'LR-', 'LR-_UB', 'Tier', 'Phenos', 'unpooled', 'synonymous']]\n",
    "            \n",
    "    return analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d9f2bfc4-14a1-4afb-9b97-79b58636f547",
   "metadata": {},
   "outputs": [],
   "source": [
    "# those that are actually done\n",
    "finished_drugs = np.sort(['Capreomycin', 'Amikacin', 'Kanamycin', 'Ethionamide', 'Levofloxacin', 'Clofazimine', 'Linezolid', 'Moxifloxacin', 'Bedaquiline', 'Delamanid'])\n",
    "\n",
    "drug_analyses = {}\n",
    "for drug in finished_drugs:\n",
    "    drug_analyses[drug] = final_processing(drug)\n",
    "    \n",
    "# write results to an Excel file, where each sheet is named for a drug\n",
    "with pd.ExcelWriter(\"Farhat_logReg_analysis.xlsx\") as file:\n",
    "   \n",
    "    for key, val in drug_analyses.items():\n",
    "        val.to_excel(file, sheet_name=key, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "08828f54-4462-43be-b69d-dd961bff7f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Delamanid\"\n",
    "res_df = pd.read_csv(os.path.join(analysis_dir, drug, \"final_analysis.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "21f8c3e9-4321-44f5-8fec-e57e5d5dfd08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Phenos</th>\n",
       "      <th>unpooled</th>\n",
       "      <th>synonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_lof</td>\n",
       "      <td>0.1808</td>\n",
       "      <td>0.1573</td>\n",
       "      <td>0.2015</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1982</td>\n",
       "      <td>1.1704</td>\n",
       "      <td>1.2233</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PC0</td>\n",
       "      <td>0.1553</td>\n",
       "      <td>0.0518</td>\n",
       "      <td>0.2549</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.9491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1680</td>\n",
       "      <td>1.0532</td>\n",
       "      <td>1.2903</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PC1</td>\n",
       "      <td>0.1460</td>\n",
       "      <td>0.0304</td>\n",
       "      <td>0.2356</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1571</td>\n",
       "      <td>1.0308</td>\n",
       "      <td>1.2657</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ddn_p.Leu49Pro</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>0.1530</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.1476</td>\n",
       "      <td>1.1241</td>\n",
       "      <td>1.1654</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PC4</td>\n",
       "      <td>0.1373</td>\n",
       "      <td>0.0372</td>\n",
       "      <td>0.2337</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0.0255</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1472</td>\n",
       "      <td>1.0379</td>\n",
       "      <td>1.2633</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>fgd1_c.774G&gt;A</td>\n",
       "      <td>-0.0641</td>\n",
       "      <td>-0.0745</td>\n",
       "      <td>-0.0505</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9379</td>\n",
       "      <td>0.9282</td>\n",
       "      <td>0.9507</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>fbiC_c.967C&gt;T</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>-0.0758</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9374</td>\n",
       "      <td>0.9270</td>\n",
       "      <td>0.9490</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>fbiB_p.Leu447Arg</td>\n",
       "      <td>-0.0649</td>\n",
       "      <td>-0.0788</td>\n",
       "      <td>-0.0504</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9371</td>\n",
       "      <td>0.9242</td>\n",
       "      <td>0.9509</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>fbiC_c.-11G&gt;A</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>-0.0857</td>\n",
       "      <td>-0.0481</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9342</td>\n",
       "      <td>0.9179</td>\n",
       "      <td>0.9530</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>fgd1_c.465C&gt;T</td>\n",
       "      <td>-0.0783</td>\n",
       "      <td>-0.0901</td>\n",
       "      <td>-0.0657</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9247</td>\n",
       "      <td>0.9139</td>\n",
       "      <td>0.9364</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>958 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0             ddn_lof  0.1808   0.1573   0.2015 0.0000   0.0000   \n",
       "1                 PC0  0.1553   0.0518   0.2549 0.0016   0.0123   \n",
       "2                 PC1  0.1460   0.0304   0.2356 0.0024   0.0157   \n",
       "3      ddn_p.Leu49Pro  0.1377   0.1170   0.1530 0.0000   0.0000   \n",
       "4                 PC4  0.1373   0.0372   0.2337 0.0045   0.0255   \n",
       "..                ...     ...      ...      ...    ...      ...   \n",
       "953     fgd1_c.774G>A -0.0641  -0.0745  -0.0505 0.0000   0.0000   \n",
       "954     fbiC_c.967C>T -0.0646  -0.0758  -0.0524 0.0000   0.0000   \n",
       "955  fbiB_p.Leu447Arg -0.0649  -0.0788  -0.0504 0.0000   0.0000   \n",
       "956     fbiC_c.-11G>A -0.0680  -0.0857  -0.0481 0.0000   0.0000   \n",
       "957     fgd1_c.465C>T -0.0783  -0.0901  -0.0657 0.0000   0.0000   \n",
       "\n",
       "     Bonferroni_pval        confidence_WHO_2021  Odds_Ratio  OR_LB  OR_UB  \\\n",
       "0             0.0000                        NaN      1.1982 1.1704 1.2233   \n",
       "1             0.9491                        NaN      1.1680 1.0532 1.2903   \n",
       "2             1.0000                        NaN      1.1571 1.0308 1.2657   \n",
       "3             0.0000     2) Assoc w R - Interim      1.1476 1.1241 1.1654   \n",
       "4             1.0000                        NaN      1.1472 1.0379 1.2633   \n",
       "..               ...                        ...         ...    ...    ...   \n",
       "953           0.0000                        NaN      0.9379 0.9282 0.9507   \n",
       "954           0.0000                        NaN      0.9374 0.9270 0.9490   \n",
       "955           0.0000  3) Uncertain significance      0.9371 0.9242 0.9509   \n",
       "956           0.0000  3) Uncertain significance      0.9342 0.9179 0.9530   \n",
       "957           0.0000                        NaN      0.9247 0.9139 0.9364   \n",
       "\n",
       "     Tier Phenos  unpooled  synonymous  \n",
       "0       2    ALL         0           0  \n",
       "1       2    ALL         1           0  \n",
       "2       2    ALL         1           0  \n",
       "3       2    ALL         1           0  \n",
       "4       2    ALL         1           0  \n",
       "..    ...    ...       ...         ...  \n",
       "953     2    ALL         0           1  \n",
       "954     2    ALL         0           1  \n",
       "955     2    ALL         1           0  \n",
       "956     2    ALL         1           0  \n",
       "957     2    ALL         0           1  \n",
       "\n",
       "[958 rows x 15 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed4a42c-86c9-4a43-9382-7f9723720ada",
   "metadata": {},
   "source": [
    "# Catalog-Based Method Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f78280-0749-4fe9-8cd9-079fa016922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_logReg_summary(out_dir):\n",
    "    \n",
    "    model_summary = pd.read_csv(os.path.join(out_dir, \"logReg_summary.csv\"))\n",
    "    model_analysis = pd.read_csv(os.path.join(out_dir, \"model_analysis.csv\"))\n",
    "    \n",
    "    summary_sens = model_summary.loc[0, \"Sens\"]\n",
    "    summary_spec = model_summary.loc[0, \"Spec\"]\n",
    "    \n",
    "    #max_idx = [np.argmax(model_analysis[\"Sens\"]), np.argmax(model_analysis[\"Spec\"])]\n",
    "    \n",
    "    better_variant = model_analysis.query(\"Sens >= @summary_sens & Spec >= @summary_spec\")\n",
    "    \n",
    "    if len(better_variant) == 0:\n",
    "        print(\"No variants have comparable sensitivity AND specificity\")\n",
    "    else:\n",
    "        print(better_variant)\n",
    "    return model_summary, model_analysis\n",
    "    \n",
    "    #return model_summary, model_analysis.iloc[max_idx, :][[\"orig_variant\", \"coef\", \"Sens\", \"Spec\", \"accuracy\", \"balanced_accuracy\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20dfc9fa-8f4b-4e2d-81b2-489754985444",
   "metadata": {},
   "outputs": [],
   "source": [
    "#out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Levofloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn\"\n",
    "\n",
    "summary, analysis = get_logReg_summary(out_dir)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ef135c-6253-4a64-9053-a08410585184",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.query(\"Sens > 0.9006\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c620b336-0167-4f00-b47b-e4bc031a045d",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.loc[analysis[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36145380-d850-475c-a6fb-e129a2824e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.query(\"Spec > 0.9575\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e714d4af-4376-4e03-874e-0d4566d278a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Moxifloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "\n",
    "get_logReg_summary(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec5a07e-ae8a-4de2-911f-d4a9ad4e6446",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pza = pd.read_csv(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn/phenos.csv\")\n",
    "\n",
    "# the missing ones might be M. cannettii, most similar to L6 based on the other lineage callers\n",
    "lineages = pd.read_pickle(\"../data/combined_lineage_sample_IDs.pkl\")\n",
    "lineages[\"Lineage\"] = lineages[\"Lineage\"].fillna(\"6\")\n",
    "lineages[\"Lineage_1\"] = lineages[\"Lineage_1\"].fillna(\"6\")\n",
    "\n",
    "lineages = lineages[[\"Sample Name\", \"Sample ID\", \"Lineage_1\"]]\n",
    "lineages[\"Lineage\"] = [str(val).split(\".\")[0] for val in lineages[\"Lineage_1\"].values]\n",
    "lineages.loc[lineages[\"Lineage\"].str.contains(\"BOV\"), \"Lineage\"] = \"M. bovis\"\n",
    "\n",
    "assert len(lineages.loc[pd.isnull(lineages[\"Lineage\"])]) == 0\n",
    "\n",
    "########## KEEP ONLY ISOLATES WITH ALL 3 PIECES OF DATA ##########\n",
    "\n",
    "# get only isolates with data for everyting: SNP matrix, in the model, and lineages\n",
    "combined = lineages.merge(df_pza, left_on=\"Sample ID\", right_on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56d7652e-ec58-4e0a-88ef-86e80fcc025a",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.groupby(\"Lineage\")[\"phenotype\"].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3334fbf-76ee-49e0-a0b9-0dccff1727c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.query(\"Lineage == 'M. bovis'\").Lineage_1.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c225d9c-faf1-4141-884c-034c55fa03de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
