{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import sklearn\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "import glob, os, yaml, subprocess, itertools, sparse, pickle\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)\n",
    "samples_summary = pd.read_csv(\"../data/samples_summary.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6a5a6a1-d7ff-4e89-8765-c19c0311edb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dlm = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1/phenos=WHO/dropAF_noSyn/model_analysis.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "feb87253-caba-4942-bfa8-06e337ac976a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_lof</td>\n",
       "      <td>0.2970</td>\n",
       "      <td>0.1632</td>\n",
       "      <td>0.3735</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3458</td>\n",
       "      <td>1.1773</td>\n",
       "      <td>1.4528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fgd1_lof</td>\n",
       "      <td>0.1569</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.2102</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0428</td>\n",
       "      <td>0.2570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1699</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.2339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ddn_p.Val127Phe</td>\n",
       "      <td>0.1557</td>\n",
       "      <td>-0.0246</td>\n",
       "      <td>0.2248</td>\n",
       "      <td>0.0023</td>\n",
       "      <td>0.0283</td>\n",
       "      <td>0.1414</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1685</td>\n",
       "      <td>0.9757</td>\n",
       "      <td>1.2520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fbiB_p.Lys448Arg</td>\n",
       "      <td>0.1363</td>\n",
       "      <td>-0.0291</td>\n",
       "      <td>0.2688</td>\n",
       "      <td>0.0445</td>\n",
       "      <td>0.1105</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3642877.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1460</td>\n",
       "      <td>0.9713</td>\n",
       "      <td>1.3085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fbiC_lof</td>\n",
       "      <td>0.1302</td>\n",
       "      <td>-0.0535</td>\n",
       "      <td>0.2157</td>\n",
       "      <td>0.0367</td>\n",
       "      <td>0.0989</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1390</td>\n",
       "      <td>0.9479</td>\n",
       "      <td>1.2408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>fbiC_p.Val41Met</td>\n",
       "      <td>-0.0466</td>\n",
       "      <td>-0.0827</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0245</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1303051.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9544</td>\n",
       "      <td>0.9206</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ddn_p.Glu83Asp</td>\n",
       "      <td>-0.0598</td>\n",
       "      <td>-0.0947</td>\n",
       "      <td>-0.0201</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.0198</td>\n",
       "      <td>0.0595</td>\n",
       "      <td>3987092.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9419</td>\n",
       "      <td>0.9096</td>\n",
       "      <td>0.9801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>fbiC_c.-32A&gt;G</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>-0.0976</td>\n",
       "      <td>-0.0191</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>0.0187</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>1302899.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9374</td>\n",
       "      <td>0.9071</td>\n",
       "      <td>0.9810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>fbiA_p.Arg304Gln</td>\n",
       "      <td>-0.0796</td>\n",
       "      <td>-0.1329</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0465</td>\n",
       "      <td>0.1109</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3641453.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.8755</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>PC0</td>\n",
       "      <td>-0.2255</td>\n",
       "      <td>-0.3787</td>\n",
       "      <td>-0.0700</td>\n",
       "      <td>0.0022</td>\n",
       "      <td>0.0346</td>\n",
       "      <td>0.1383</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7981</td>\n",
       "      <td>0.6848</td>\n",
       "      <td>0.9324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0            ddn_lof  0.2970   0.1632   0.3735 0.0000   0.0000   \n",
       "1           fgd1_lof  0.1569  -0.0000   0.2102 0.0041   0.0428   \n",
       "2    ddn_p.Val127Phe  0.1557  -0.0246   0.2248 0.0023   0.0283   \n",
       "3   fbiB_p.Lys448Arg  0.1363  -0.0291   0.2688 0.0445   0.1105   \n",
       "4           fbiC_lof  0.1302  -0.0535   0.2157 0.0367   0.0989   \n",
       "..               ...     ...      ...      ...    ...      ...   \n",
       "57   fbiC_p.Val41Met -0.0466  -0.0827   0.0000 0.0245   0.0843   \n",
       "58    ddn_p.Glu83Asp -0.0598  -0.0947  -0.0201 0.0010   0.0198   \n",
       "59     fbiC_c.-32A>G -0.0646  -0.0976  -0.0191 0.0006   0.0187   \n",
       "60  fbiA_p.Arg304Gln -0.0796  -0.1329   0.0000 0.0465   0.1109   \n",
       "61               PC0 -0.2255  -0.3787  -0.0700 0.0022   0.0346   \n",
       "\n",
       "    Bonferroni_pval  genome_index        confidence_WHO_2021  Odds_Ratio  \\\n",
       "0            0.0000           NaN                        NaN      1.3458   \n",
       "1            0.2570           NaN                        NaN      1.1699   \n",
       "2            0.1414           NaN                        NaN      1.1685   \n",
       "3            1.0000  3642877.0000  3) Uncertain significance      1.1460   \n",
       "4            1.0000           NaN                        NaN      1.1390   \n",
       "..              ...           ...                        ...         ...   \n",
       "57           1.0000  1303051.0000  3) Uncertain significance      0.9544   \n",
       "58           0.0595  3987092.0000  3) Uncertain significance      0.9419   \n",
       "59           0.0374  1302899.0000  3) Uncertain significance      0.9374   \n",
       "60           1.0000  3641453.0000  3) Uncertain significance      0.9235   \n",
       "61           0.1383           NaN                        NaN      0.7981   \n",
       "\n",
       "    OR_LB  OR_UB  \n",
       "0  1.1773 1.4528  \n",
       "1  1.0000 1.2339  \n",
       "2  0.9757 1.2520  \n",
       "3  0.9713 1.3085  \n",
       "4  0.9479 1.2408  \n",
       "..    ...    ...  \n",
       "57 0.9206 1.0000  \n",
       "58 0.9096 0.9801  \n",
       "59 0.9071 0.9810  \n",
       "60 0.8755 1.0000  \n",
       "61 0.6848 0.9324  \n",
       "\n",
       "[62 rows x 12 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dlm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ec21246b-2ad4-4fc4-93ad-fa1bf0e3a25d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['3) Uncertain significance', '2) Assoc w R - Interim'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who_variants.query(\"drug=='DLM'\").confidence.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f6d5a06a-13fc-430e-ae3f-de036704202a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>drug</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence</th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5248</th>\n",
       "      <td>3804</td>\n",
       "      <td>DLM</td>\n",
       "      <td>3986989</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>ddn</td>\n",
       "      <td>ddn_L49P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 drug genome_index              confidence gene   variant\n",
       "5248        3804  DLM      3986989  2) Assoc w R - Interim  ddn  ddn_L49P"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who_variants.query(\"drug=='DLM' & confidence=='2) Assoc w R - Interim'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a90c98-bf57-4f7c-9ac9-09fddea90075",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9893d9bc-a8ba-46b8-9288-b002ee5e1099",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bc974a8f-b5f8-4b53-b15a-16350d974bc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>Genos</th>\n",
       "      <th>Binary_Phenos</th>\n",
       "      <th>MICs</th>\n",
       "      <th>Lineages</th>\n",
       "      <th>SNP_Matrix</th>\n",
       "      <th>Tier1_LOF</th>\n",
       "      <th>Tier1_MultiInframe</th>\n",
       "      <th>Tier2_LOF</th>\n",
       "      <th>Tier2_MultiInframe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Delamanid</td>\n",
       "      <td>11803</td>\n",
       "      <td>11803</td>\n",
       "      <td>11353</td>\n",
       "      <td>11491</td>\n",
       "      <td>11803</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Levofloxacin</td>\n",
       "      <td>27576</td>\n",
       "      <td>27576</td>\n",
       "      <td>11447</td>\n",
       "      <td>26909</td>\n",
       "      <td>27576</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Linezolid</td>\n",
       "      <td>18010</td>\n",
       "      <td>18010</td>\n",
       "      <td>12626</td>\n",
       "      <td>17062</td>\n",
       "      <td>18010</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Moxifloxacin</td>\n",
       "      <td>22783</td>\n",
       "      <td>22783</td>\n",
       "      <td>12753</td>\n",
       "      <td>22116</td>\n",
       "      <td>22783</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Drug  Genos  Binary_Phenos   MICs  Lineages  SNP_Matrix  \\\n",
       "4      Delamanid  11803          11803  11353     11491       11803   \n",
       "9   Levofloxacin  27576          27576  11447     26909       27576   \n",
       "10     Linezolid  18010          18010  12626     17062       18010   \n",
       "11  Moxifloxacin  22783          22783  12753     22116       22783   \n",
       "\n",
       "    Tier1_LOF  Tier1_MultiInframe  Tier2_LOF  Tier2_MultiInframe  \n",
       "4           0                   0          0                   0  \n",
       "9           0                   0          6                   0  \n",
       "10          0                   0          0                   0  \n",
       "11          0                   0          8                   0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_summary.query(\"Tier1_LOF==0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f227f3bb-d702-4d9c-a8a5-b5bc84ba3042",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_path = \"/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1/phenos=WHO/dropAF_noSyn\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "17d2f574-a45a-41fc-81f5-c7cc33f3b72b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1/phenos=WHO/dropAF_noSyn'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "91369438-5bbc-4a86-a2f5-0b4ff33fc134",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[2 if \"+2\" in add_path else 1][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "eb92c414-3b8a-4c66-8758-3f6ffeec380b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WHO'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"ALL\" if \"ALL\" in add_path else \"WHO\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "20cfe1b0-1b71-4e33-8862-97b7e1c50a05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(\"unpooled\" in add_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "49dd5fff-0cc3-48f9-872d-534963c93a67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(\"withSyn\" in add_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5e9e6951-0937-44c6-a969-b4ec9d0d186b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'drop'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"AF\" if \"encodeAF\" in add_path else \"drop\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcd21f26-5d64-444c-92a1-a35038fb03b5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a2439e0a-1ad9-4471-9e91-ac841f82fba2",
   "metadata": {},
   "source": [
    "# Combined Analysis Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "00729d49-7aeb-457e-af65-74a6bf488b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def final_processing(drug):\n",
    "    '''\n",
    "    Functions for processing outputs before sending to everyone else.\n",
    "    \n",
    "    1. Remove principal components (will describe them separately)\n",
    "    2. Add LOF to the predicted_effect column for pooled LOF mutations\n",
    "    3. Remove genome_index column (should actually do that earlier, but will fix later)\n",
    "    4. Remove the logistic regression coefficient columns (they will prefer to work with odds ratios)\n",
    "    5. Any other column renaming or dropping for clarity\n",
    "    '''\n",
    "    \n",
    "    analysis_df = pd.read_csv(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/{drug}/final_analysis.csv\")\n",
    "    analysis_df.rename(columns={\"orig_variant\": \"mutation\", \"Tier1_only\": \"Tier\", \"WHO_phenos\": \"Phenos\"}, inplace=True)\n",
    "    \n",
    "    # remove logReg coefficients. Keep only odds ratios. Remove the other two columns, which were present mainly for me to see\n",
    "    # if we were picking up many mutations that were in the 2021 mutation catalog\n",
    "    del analysis_df[\"genome_index\"]\n",
    "    del analysis_df[\"confidence_WHO_2021\"]\n",
    "    analysis_df = analysis_df[analysis_df.columns[~analysis_df.columns.str.contains(\"coef\")]]\n",
    "    \n",
    "    # remove significant principal components and replace the NaNs in the predicted effect column for the gene loss of functions\n",
    "    analysis_df = analysis_df.loc[~analysis_df[\"mutation\"].str.contains(\"PC\")]\n",
    "    analysis_df.loc[analysis_df[\"mutation\"].str.contains(\"lof\"), \"predicted_effect\"] = \"LOF\"\n",
    "    \n",
    "    # predicted effect should not be NaN for anything. position is NaN only for the pooled LOF mutations\n",
    "    assert len(analysis_df.loc[pd.isnull(analysis_df[\"predicted_effect\"])]) == 0\n",
    "    assert len(analysis_df.loc[(~analysis_df[\"mutation\"].str.contains(\"lof\")) & (pd.isnull(analysis_df[\"position\"]))]) == 0\n",
    "    \n",
    "    analysis_df[\"Tier\"] = analysis_df[\"Tier\"].map({1: 1, 0: 2})\n",
    "    analysis_df[\"Phenos\"] = analysis_df[\"Phenos\"].map({1: \"WHO\", 0: \"ALL\"})\n",
    "    \n",
    "    # reorder columns\n",
    "    analysis_df = analysis_df[['mutation', 'predicted_effect', 'position', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval', 'Bonferroni_pval',\n",
    "       'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'Sens_LB', 'Sens', 'Sens_UB', 'Spec_LB', 'Spec', 'Spec_UB', 'PPV', 'PPV_LB', 'PPV_UB',\n",
    "       'LR+_LB', 'LR+', 'LR+_UB', 'LR-_LB', 'LR-', 'LR-_UB', 'Tier', 'Phenos', 'poolLOF', 'Syn']]\n",
    "    \n",
    "    analysis_df[['poolLOF', \"Syn\"]] = analysis_df[['poolLOF', \"Syn\"]].astype(int)\n",
    "        \n",
    "    return analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "26ac6d98-c680-47eb-8a77-0a12f7121517",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis\"\n",
    "finished_drugs = []\n",
    "\n",
    "for drug in os.listdir(analysis_dir):\n",
    "    \n",
    "    if os.path.isfile(os.path.join(analysis_dir, drug, \"final_analysis.csv\")):\n",
    "        finished_drugs.append(drug)\n",
    "        # drug_analyses.append(final_processing(drug))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "1f27f833-b810-4a22-9cd8-1ec10b70a75b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Levofloxacin', 'Pyrazinamide', 'Streptomycin', 'Amikacin', 'Clofazimine', 'Linezolid', 'Moxifloxacin', 'Kanamycin', 'Bedaquiline', 'Capreomycin', 'Delamanid', 'Ethionamide']\n"
     ]
    }
   ],
   "source": [
    "print(finished_drugs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "d9f2bfc4-14a1-4afb-9b97-79b58636f547",
   "metadata": {},
   "outputs": [],
   "source": [
    "# those that are actually done\n",
    "finished_drugs = ['Pyrazinamide', 'Amikacin', 'Clofazimine', 'Linezolid', 'Moxifloxacin', 'Kanamycin', 'Bedaquiline', 'Capreomycin', 'Delamanid']\n",
    "\n",
    "drug_analyses = {}\n",
    "for drug in finished_drugs:\n",
    "    drug_analyses[drug] = final_processing(drug)\n",
    "    \n",
    "# write results to an Excel file, where each sheet is named for a drug\n",
    "with pd.ExcelWriter(\"Farhat_logReg_analysis.xlsx\") as file:\n",
    "   \n",
    "    for key, val in drug_analyses.items():\n",
    "        val.to_excel(file, sheet_name=key, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "dd2ede15-4b01-4a16-8565-531b1925806c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17dbb6e3-d3d8-4427-963b-f2c88919552c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d471f30-5ad3-4545-aaae-53210b0d97a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "52ad210e-7627-48c7-84dd-e7d3e365d3fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>...</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Phenos</th>\n",
       "      <th>poolLOF</th>\n",
       "      <th>Syn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [mutation, predicted_effect, position, Odds_Ratio, OR_LB, OR_UB, pval, BH_pval, Bonferroni_pval, confidence_WHO_2021, Num_Isolates, Total_Isolates, TP, FP, TN, FN, Sens_LB, Sens, Sens_UB, Spec_LB, Spec, Spec_UB, PPV, PPV_LB, PPV_UB, LR+_LB, LR+, LR+_UB, LR-_LB, LR-, LR-_UB, Tier, Phenos, poolLOF, Syn]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 35 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_analyses[4].query(\"Tier==2&Phenos=='WHO'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "d0f1c384-b661-4f12-b071-ff944fd1ad2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>...</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Phenos</th>\n",
       "      <th>poolLOF</th>\n",
       "      <th>Syn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1591</th>\n",
       "      <td>panD_c.405G&gt;C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>4043877</td>\n",
       "      <td>0.9855</td>\n",
       "      <td>0.9794</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0368</td>\n",
       "      <td>0.0489</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665</th>\n",
       "      <td>clpC1_c.1863G&gt;A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>4038842</td>\n",
       "      <td>0.9852</td>\n",
       "      <td>0.9795</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0362</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1817</th>\n",
       "      <td>clpC1_c.1842G&gt;T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>4038863;4038860</td>\n",
       "      <td>0.9826</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0400</td>\n",
       "      <td>0.0499</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation    predicted_effect         position  Odds_Ratio  OR_LB  \\\n",
       "1591    panD_c.405G>C  synonymous_variant          4043877      0.9855 0.9794   \n",
       "1665  clpC1_c.1863G>A  synonymous_variant          4038842      0.9852 0.9795   \n",
       "1817  clpC1_c.1842G>T  synonymous_variant  4038863;4038860      0.9826 0.9726   \n",
       "\n",
       "      OR_UB   pval  BH_pval  Bonferroni_pval confidence_WHO_2021  ...  LR+_LB  \\\n",
       "1591 1.0000 0.0368   0.0489           1.0000                 NaN  ...     NaN   \n",
       "1665 1.0000 0.0362   0.0490           1.0000                 NaN  ...     NaN   \n",
       "1817 1.0000 0.0400   0.0499           1.0000                 NaN  ...     NaN   \n",
       "\n",
       "      LR+  LR+_UB  LR-_LB    LR-  LR-_UB  Tier  Phenos  poolLOF  Syn  \n",
       "1591  inf     NaN     NaN 1.0000     NaN     1     ALL        1    1  \n",
       "1665  inf     NaN     NaN 1.0000     NaN     1     ALL        1    1  \n",
       "1817  inf     NaN     NaN 1.0000     NaN     1     ALL        1    1  \n",
       "\n",
       "[3 rows x 35 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_analyses[0].query(\"Tier==1&Phenos=='ALL'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "66beb441-6ac2-4879-bf36-d39e8bc88007",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>Tier1</th>\n",
       "      <th>WHO_phenos</th>\n",
       "      <th>poolLOF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pncA_lof</td>\n",
       "      <td>LOF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.2523</td>\n",
       "      <td>2.1759</td>\n",
       "      <td>2.3512</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9483</td>\n",
       "      <td>37.8260</td>\n",
       "      <td>49.5805</td>\n",
       "      <td>70.8301</td>\n",
       "      <td>0.8640</td>\n",
       "      <td>0.8746</td>\n",
       "      <td>0.8852</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pncA_p.His57Asp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>2289073</td>\n",
       "      <td>1.7228</td>\n",
       "      <td>1.6757</td>\n",
       "      <td>1.7705</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>121.1180</td>\n",
       "      <td>317.1196</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9564</td>\n",
       "      <td>0.9622</td>\n",
       "      <td>0.9677</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pncA_c.-11A&gt;G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>2289252</td>\n",
       "      <td>1.5396</td>\n",
       "      <td>1.4875</td>\n",
       "      <td>1.5948</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9697</td>\n",
       "      <td>31.1115</td>\n",
       "      <td>53.8721</td>\n",
       "      <td>120.8069</td>\n",
       "      <td>0.9630</td>\n",
       "      <td>0.9683</td>\n",
       "      <td>0.9738</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pncA_p.His51Asp</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>2289091</td>\n",
       "      <td>1.3459</td>\n",
       "      <td>1.3089</td>\n",
       "      <td>1.3795</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>48.2829</td>\n",
       "      <td>175.7530</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9865</td>\n",
       "      <td>0.9895</td>\n",
       "      <td>0.9926</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pncA_p.Val131fs</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>2288850;2288851;2288848;2288852;2288849</td>\n",
       "      <td>1.3420</td>\n",
       "      <td>1.3212</td>\n",
       "      <td>1.3744</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>57.5839</td>\n",
       "      <td>202.4981</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9845</td>\n",
       "      <td>0.9879</td>\n",
       "      <td>0.9910</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2534</th>\n",
       "      <td>rpsA_p.Ile70Leu</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1833749</td>\n",
       "      <td>0.8743</td>\n",
       "      <td>0.8654</td>\n",
       "      <td>0.8848</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2535</th>\n",
       "      <td>Rv3236c_p.Val151Ala</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>3612665</td>\n",
       "      <td>0.8738</td>\n",
       "      <td>0.7786</td>\n",
       "      <td>0.9736</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0329</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2537</th>\n",
       "      <td>Rv3236c_c.546C&gt;T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>3612571</td>\n",
       "      <td>0.8573</td>\n",
       "      <td>0.8438</td>\n",
       "      <td>0.8701</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2538</th>\n",
       "      <td>Rv1258c_c.1029T&gt;C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>1406312</td>\n",
       "      <td>0.8513</td>\n",
       "      <td>0.7757</td>\n",
       "      <td>0.9338</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2539</th>\n",
       "      <td>pncA_c.195C&gt;T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>2289047</td>\n",
       "      <td>0.7373</td>\n",
       "      <td>0.6548</td>\n",
       "      <td>0.8320</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1255</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2536 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 mutation       predicted_effect  \\\n",
       "0                pncA_lof                    LOF   \n",
       "1         pncA_p.His57Asp       missense_variant   \n",
       "2           pncA_c.-11A>G  upstream_gene_variant   \n",
       "5         pncA_p.His51Asp       missense_variant   \n",
       "6         pncA_p.Val131fs             frameshift   \n",
       "...                   ...                    ...   \n",
       "2534      rpsA_p.Ile70Leu       missense_variant   \n",
       "2535  Rv3236c_p.Val151Ala       missense_variant   \n",
       "2537     Rv3236c_c.546C>T     synonymous_variant   \n",
       "2538    Rv1258c_c.1029T>C     synonymous_variant   \n",
       "2539        pncA_c.195C>T     synonymous_variant   \n",
       "\n",
       "                                     position  Odds_Ratio  OR_LB  OR_UB  \\\n",
       "0                                         NaN      2.2523 2.1759 2.3512   \n",
       "1                                     2289073      1.7228 1.6757 1.7705   \n",
       "2                                     2289252      1.5396 1.4875 1.5948   \n",
       "5                                     2289091      1.3459 1.3089 1.3795   \n",
       "6     2288850;2288851;2288848;2288852;2288849      1.3420 1.3212 1.3744   \n",
       "...                                       ...         ...    ...    ...   \n",
       "2534                                  1833749      0.8743 0.8654 0.8848   \n",
       "2535                                  3612665      0.8738 0.7786 0.9736   \n",
       "2537                                  3612571      0.8573 0.8438 0.8701   \n",
       "2538                                  1406312      0.8513 0.7757 0.9338   \n",
       "2539                                  2289047      0.7373 0.6548 0.8320   \n",
       "\n",
       "       pval  BH_pval  Bonferroni_pval confidence_WHO_2021  ...  PPV_UB  \\\n",
       "0    0.0000   0.0000           0.0000                 NaN  ...  0.9483   \n",
       "1    0.0000   0.0000           0.0000        1) Assoc w R  ...  1.0000   \n",
       "2    0.0000   0.0000           0.0000        1) Assoc w R  ...  0.9697   \n",
       "5    0.0000   0.0000           0.0000        1) Assoc w R  ...  1.0000   \n",
       "6    0.0000   0.0000           0.0000                 NaN  ...  1.0000   \n",
       "...     ...      ...              ...                 ...  ...     ...   \n",
       "2534 0.0000   0.0000           0.0000                 NaN  ...     NaN   \n",
       "2535 0.0096   0.0329           1.0000    5) Not assoc w R  ...     NaN   \n",
       "2537 0.0000   0.0000           0.0000                 NaN  ...     NaN   \n",
       "2538 0.0004   0.0027           1.0000                 NaN  ...     NaN   \n",
       "2539 0.0000   0.0000           0.0012                 NaN  ...     NaN   \n",
       "\n",
       "       LR+_LB      LR+   LR+_UB  LR-_LB    LR-  LR-_UB  Tier1  WHO_phenos  \\\n",
       "0     37.8260  49.5805  70.8301  0.8640 0.8746  0.8852      1           1   \n",
       "1    121.1180 317.1196      inf  0.9564 0.9622  0.9677      1           1   \n",
       "2     31.1115  53.8721 120.8069  0.9630 0.9683  0.9738      1           1   \n",
       "5     48.2829 175.7530      inf  0.9865 0.9895  0.9926      1           1   \n",
       "6     57.5839 202.4981      inf  0.9845 0.9879  0.9910      0           0   \n",
       "...       ...      ...      ...     ...    ...     ...    ...         ...   \n",
       "2534      NaN   0.0000      NaN     NaN 1.0030     NaN      0           0   \n",
       "2535      NaN   0.8145      NaN     NaN 1.0052     NaN      0           0   \n",
       "2537      NaN   0.0000      NaN     NaN 1.0066     NaN      0           0   \n",
       "2538      NaN   0.1567      NaN     NaN 1.0202     NaN      0           0   \n",
       "2539      NaN   0.2261      NaN     NaN 1.1255     NaN      0           0   \n",
       "\n",
       "      poolLOF  \n",
       "0           1  \n",
       "1           1  \n",
       "2           1  \n",
       "5           1  \n",
       "6           0  \n",
       "...       ...  \n",
       "2534        1  \n",
       "2535        1  \n",
       "2537        1  \n",
       "2538        1  \n",
       "2539        1  \n",
       "\n",
       "[2536 rows x 34 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_analyses[0].query(\"predicted_effect=='synonymous_variant' & \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8e8a634b-b24c-47dd-8310-4d29cde7ab82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>glpK_lof</td>\n",
       "      <td>0.1000</td>\n",
       "      <td>0.0397</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>0.0065</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1052</td>\n",
       "      <td>1.0405</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0144</td>\n",
       "      <td>0.0229</td>\n",
       "      <td>0.9851</td>\n",
       "      <td>0.9884</td>\n",
       "      <td>1.0648</td>\n",
       "      <td>1.8180</td>\n",
       "      <td>0.9896</td>\n",
       "      <td>0.9991</td>\n",
       "      <td>LOF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>Rv2752c_lof</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>-0.0609</td>\n",
       "      <td>0.0741</td>\n",
       "      <td>0.4522</td>\n",
       "      <td>0.5002</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0042</td>\n",
       "      <td>0.9409</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0186</td>\n",
       "      <td>0.0280</td>\n",
       "      <td>0.9876</td>\n",
       "      <td>0.9904</td>\n",
       "      <td>1.6353</td>\n",
       "      <td>2.6500</td>\n",
       "      <td>0.9827</td>\n",
       "      <td>0.9927</td>\n",
       "      <td>LOF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>gyrA_lof</td>\n",
       "      <td>-0.0003</td>\n",
       "      <td>-0.0010</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>0.1789</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9997</td>\n",
       "      <td>0.9990</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>Rv1129c_lof</td>\n",
       "      <td>-0.0032</td>\n",
       "      <td>-0.0476</td>\n",
       "      <td>0.0212</td>\n",
       "      <td>0.4424</td>\n",
       "      <td>0.4901</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9968</td>\n",
       "      <td>0.9535</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969</th>\n",
       "      <td>Rv2477c_lof</td>\n",
       "      <td>-0.0098</td>\n",
       "      <td>-0.0147</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0397</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9903</td>\n",
       "      <td>0.9854</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mutation    coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "19      glpK_lof  0.1000   0.0397   0.1618 0.0006   0.0065           1.0000   \n",
       "404  Rv2752c_lof  0.0041  -0.0609   0.0741 0.4522   0.5002           1.0000   \n",
       "437     gyrA_lof -0.0003  -0.0010   0.0000 0.1470   0.1789           1.0000   \n",
       "539  Rv1129c_lof -0.0032  -0.0476   0.0212 0.4424   0.4901           1.0000   \n",
       "969  Rv2477c_lof -0.0098  -0.0147   0.0000 0.0397   0.0615           1.0000   \n",
       "\n",
       "    confidence_WHO_2021  Odds_Ratio  OR_LB  ...  Sens_LB  Sens_UB  Spec_LB  \\\n",
       "19                  NaN      1.1052 1.0405  ...   0.0144   0.0229   0.9851   \n",
       "404                 NaN      1.0042 0.9409  ...   0.0186   0.0280   0.9876   \n",
       "437                 NaN      0.9997 0.9990  ...      NaN      NaN      NaN   \n",
       "539                 NaN      0.9968 0.9535  ...      NaN      NaN      NaN   \n",
       "969                 NaN      0.9903 0.9854  ...      NaN      NaN      NaN   \n",
       "\n",
       "     Spec_UB  LR+_LB  LR+_UB  LR-_LB  LR-_UB  predicted_effect  position  \n",
       "19    0.9884  1.0648  1.8180  0.9896  0.9991               LOF       NaN  \n",
       "404   0.9904  1.6353  2.6500  0.9827  0.9927               LOF       NaN  \n",
       "437      NaN     NaN     NaN     NaN     NaN               LOF       NaN  \n",
       "539      NaN     NaN     NaN     NaN     NaN               LOF       NaN  \n",
       "969      NaN     NaN     NaN     NaN     NaN               LOF       NaN  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "moxi_df.loc[pd.isnull(moxi_df[\"position\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a683ba2b-ca3d-4b46-b3fb-eff707fe4cf5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdcae08c-4bc9-4527-8e24-c0c3922beb1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aed4a42c-86c9-4a43-9382-7f9723720ada",
   "metadata": {},
   "source": [
    "# Catalog-Based Method Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "50f78280-0749-4fe9-8cd9-079fa016922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_logReg_summary(out_dir):\n",
    "    \n",
    "    model_summary = pd.read_csv(os.path.join(out_dir, \"logReg_summary.csv\"))\n",
    "    model_analysis = pd.read_csv(os.path.join(out_dir, \"model_analysis.csv\"))\n",
    "    \n",
    "    summary_sens = model_summary.loc[0, \"Sens\"]\n",
    "    summary_spec = model_summary.loc[0, \"Spec\"]\n",
    "    \n",
    "    #max_idx = [np.argmax(model_analysis[\"Sens\"]), np.argmax(model_analysis[\"Spec\"])]\n",
    "    \n",
    "    better_variant = model_analysis.query(\"Sens >= @summary_sens & Spec >= @summary_spec\")\n",
    "    \n",
    "    if len(better_variant) == 0:\n",
    "        print(\"No variants have comparable sensitivity AND specificity\")\n",
    "    else:\n",
    "        print(better_variant)\n",
    "    return model_summary, model_analysis\n",
    "    \n",
    "    #return model_summary, model_analysis.iloc[max_idx, :][[\"orig_variant\", \"coef\", \"Sens\", \"Spec\", \"accuracy\", \"balanced_accuracy\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "20dfc9fa-8f4b-4e2d-81b2-489754985444",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No variants have comparable sensitivity AND specificity\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.8336</td>\n",
       "      <td>0.9084</td>\n",
       "      <td>0.8951</td>\n",
       "      <td>0.8710</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sens   Spec  accuracy  balanced_accuracy\n",
       "0 0.8336 0.9084    0.8951             0.8710"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Levofloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn\"\n",
    "\n",
    "summary, analysis = get_logReg_summary(out_dir)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "63ef135c-6253-4a64-9053-a08410585184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [orig_variant, coef, coef_LB, coef_UB, pval, BH_pval, Bonferroni_pval, genome_index, confidence_WHO_2021, Odds_Ratio, OR_LB, OR_UB, Num_Isolates, Total_Isolates, TP, FP, TN, FN, PPV, Sens, Spec, LR+, LR-, PPV_LB, PPV_UB, Sens_LB, Sens_UB, Spec_LB, Spec_UB, LR+_LB, LR+_UB, LR-_LB, LR-_UB, accuracy, balanced_accuracy]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 35 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.query(\"Sens > 0.9006\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c620b336-0167-4f00-b47b-e4bc031a045d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>PC2</td>\n",
       "      <td>0.1660</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.3034</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0301</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1806</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>PC0</td>\n",
       "      <td>0.1346</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.2630</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>0.0492</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1440</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   orig_variant   coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "35          PC2 0.1660   0.0355   0.3034 0.0078   0.0301           1.0000   \n",
       "57          PC0 0.1346   0.0166   0.2630 0.0140   0.0492           1.0000   \n",
       "\n",
       "    genome_index confidence_WHO_2021  Odds_Ratio  ...  Sens_LB  Sens_UB  \\\n",
       "35           NaN                 NaN      1.1806  ...      NaN      NaN   \n",
       "57           NaN                 NaN      1.1440  ...      NaN      NaN   \n",
       "\n",
       "    Spec_LB  Spec_UB  LR+_LB  LR+_UB  LR-_LB  LR-_UB  accuracy  \\\n",
       "35      NaN      NaN     NaN     NaN     NaN     NaN       NaN   \n",
       "57      NaN      NaN     NaN     NaN     NaN     NaN       NaN   \n",
       "\n",
       "    balanced_accuracy  \n",
       "35                NaN  \n",
       "57                NaN  \n",
       "\n",
       "[2 rows x 35 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.loc[analysis[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "36145380-d850-475c-a6fb-e129a2824e27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pncA_c.-11A&gt;G</td>\n",
       "      <td>0.4088</td>\n",
       "      <td>0.3700</td>\n",
       "      <td>0.4467</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2289252.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.5050</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.0395</td>\n",
       "      <td>0.9993</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>44.4364</td>\n",
       "      <td>284.1349</td>\n",
       "      <td>0.9608</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>0.8278</td>\n",
       "      <td>0.5163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pncA_p.His51Asp</td>\n",
       "      <td>0.3308</td>\n",
       "      <td>0.3054</td>\n",
       "      <td>0.3591</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2289091.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.3921</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.0177</td>\n",
       "      <td>0.9998</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>63.1368</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9823</td>\n",
       "      <td>0.9905</td>\n",
       "      <td>0.8246</td>\n",
       "      <td>0.5067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pncA_p.Val131fs</td>\n",
       "      <td>0.2900</td>\n",
       "      <td>0.2684</td>\n",
       "      <td>0.3154</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0090</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.9998</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>54.4333</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9836</td>\n",
       "      <td>0.9912</td>\n",
       "      <td>0.8245</td>\n",
       "      <td>0.5064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pncA_p.Gln141Pro</td>\n",
       "      <td>0.2869</td>\n",
       "      <td>0.2529</td>\n",
       "      <td>0.3273</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2288820.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.3323</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.0243</td>\n",
       "      <td>0.9995</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>35.3919</td>\n",
       "      <td>344.8754</td>\n",
       "      <td>0.9759</td>\n",
       "      <td>0.9851</td>\n",
       "      <td>0.8256</td>\n",
       "      <td>0.5097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pncA_p.Leu172Pro</td>\n",
       "      <td>0.2779</td>\n",
       "      <td>0.2524</td>\n",
       "      <td>0.3046</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2288727.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.3204</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0129</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9871</td>\n",
       "      <td>0.9938</td>\n",
       "      <td>0.8239</td>\n",
       "      <td>0.5046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>rpsA_p.Ile70Leu</td>\n",
       "      <td>-0.1249</td>\n",
       "      <td>-0.1356</td>\n",
       "      <td>-0.1120</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9966</td>\n",
       "      <td>0.9982</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0018</td>\n",
       "      <td>1.0034</td>\n",
       "      <td>0.1798</td>\n",
       "      <td>0.5013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>855</th>\n",
       "      <td>PPE35_p.Pro670Leu</td>\n",
       "      <td>-0.1287</td>\n",
       "      <td>-0.1817</td>\n",
       "      <td>-0.0860</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>2168604.0000</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.8792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.9800</td>\n",
       "      <td>0.9842</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>1.0145</td>\n",
       "      <td>1.0196</td>\n",
       "      <td>0.1922</td>\n",
       "      <td>0.5084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>856</th>\n",
       "      <td>Rv1258c_p.Gly363Val</td>\n",
       "      <td>-0.1288</td>\n",
       "      <td>-0.1399</td>\n",
       "      <td>-0.1158</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1406253.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.8791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9965</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0019</td>\n",
       "      <td>1.0036</td>\n",
       "      <td>0.1799</td>\n",
       "      <td>0.5013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>Rv3236c_c.546C&gt;T</td>\n",
       "      <td>-0.1577</td>\n",
       "      <td>-0.1859</td>\n",
       "      <td>-0.1337</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9916</td>\n",
       "      <td>0.9942</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0058</td>\n",
       "      <td>1.0085</td>\n",
       "      <td>0.1835</td>\n",
       "      <td>0.5035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>Rv1258c_c.1029T&gt;C</td>\n",
       "      <td>-0.1768</td>\n",
       "      <td>-0.2766</td>\n",
       "      <td>-0.0781</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8379</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>0.9775</td>\n",
       "      <td>0.1052</td>\n",
       "      <td>0.2940</td>\n",
       "      <td>1.0170</td>\n",
       "      <td>1.0244</td>\n",
       "      <td>0.1974</td>\n",
       "      <td>0.5101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>856 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0          pncA_c.-11A>G  0.4088   0.3700   0.4467 0.0000   0.0000   \n",
       "1        pncA_p.His51Asp  0.3308   0.3054   0.3591 0.0000   0.0000   \n",
       "2        pncA_p.Val131fs  0.2900   0.2684   0.3154 0.0000   0.0000   \n",
       "3       pncA_p.Gln141Pro  0.2869   0.2529   0.3273 0.0000   0.0000   \n",
       "4       pncA_p.Leu172Pro  0.2779   0.2524   0.3046 0.0000   0.0000   \n",
       "..                   ...     ...      ...      ...    ...      ...   \n",
       "854      rpsA_p.Ile70Leu -0.1249  -0.1356  -0.1120 0.0000   0.0000   \n",
       "855    PPE35_p.Pro670Leu -0.1287  -0.1817  -0.0860 0.0000   0.0000   \n",
       "856  Rv1258c_p.Gly363Val -0.1288  -0.1399  -0.1158 0.0000   0.0000   \n",
       "857     Rv3236c_c.546C>T -0.1577  -0.1859  -0.1337 0.0000   0.0000   \n",
       "858    Rv1258c_c.1029T>C -0.1768  -0.2766  -0.0781 0.0004   0.0027   \n",
       "\n",
       "     Bonferroni_pval  genome_index        confidence_WHO_2021  Odds_Ratio  \\\n",
       "0             0.0000  2289252.0000               1) Assoc w R      1.5050   \n",
       "1             0.0000  2289091.0000               1) Assoc w R      1.3921   \n",
       "2             0.0000           NaN                        NaN      1.3365   \n",
       "3             0.0000  2288820.0000               1) Assoc w R      1.3323   \n",
       "4             0.0000  2288727.0000               1) Assoc w R      1.3204   \n",
       "..               ...           ...                        ...         ...   \n",
       "854           0.0000           NaN                        NaN      0.8826   \n",
       "855           0.0009  2168604.0000           5) Not assoc w R      0.8792   \n",
       "856           0.0000  1406253.0000  3) Uncertain significance      0.8791   \n",
       "857           0.0000           NaN                        NaN      0.8541   \n",
       "858           1.0000           NaN                        NaN      0.8379   \n",
       "\n",
       "     ...  Sens_LB  Sens_UB  Spec_LB  Spec_UB  LR+_LB   LR+_UB  LR-_LB  LR-_UB  \\\n",
       "0    ...   0.0274   0.0395   0.9993   0.9999 44.4364 284.1349  0.9608  0.9731   \n",
       "1    ...   0.0095   0.0177   0.9998   1.0000 63.1368      inf  0.9823  0.9905   \n",
       "2    ...   0.0090   0.0165   0.9998   1.0000 54.4333      inf  0.9836  0.9912   \n",
       "3    ...   0.0151   0.0243   0.9995   0.9999 35.3919 344.8754  0.9759  0.9851   \n",
       "4    ...   0.0062   0.0129   1.0000   1.0000     inf      inf  0.9871  0.9938   \n",
       "..   ...      ...      ...      ...      ...     ...      ...     ...     ...   \n",
       "854  ...   0.0000   0.0000   0.9966   0.9982  0.0000   0.0000  1.0018  1.0034   \n",
       "855  ...   0.0003   0.0024   0.9800   0.9842  0.0156   0.1377  1.0145  1.0196   \n",
       "856  ...   0.0000   0.0000   0.9965   0.9981  0.0000   0.0000  1.0019  1.0036   \n",
       "857  ...   0.0000   0.0000   0.9916   0.9942  0.0000   0.0000  1.0058  1.0085   \n",
       "858  ...   0.0027   0.0073   0.9726   0.9775  0.1052   0.2940  1.0170  1.0244   \n",
       "\n",
       "     accuracy  balanced_accuracy  \n",
       "0      0.8278             0.5163  \n",
       "1      0.8246             0.5067  \n",
       "2      0.8245             0.5064  \n",
       "3      0.8256             0.5097  \n",
       "4      0.8239             0.5046  \n",
       "..        ...                ...  \n",
       "854    0.1798             0.5013  \n",
       "855    0.1922             0.5084  \n",
       "856    0.1799             0.5013  \n",
       "857    0.1835             0.5035  \n",
       "858    0.1974             0.5101  \n",
       "\n",
       "[856 rows x 35 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.query(\"Spec > 0.9575\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e714d4af-4376-4e03-874e-0d4566d278a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No variants have comparable sensitivity AND specificity\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.9230</td>\n",
       "      <td>0.8945</td>\n",
       "      <td>0.8983</td>\n",
       "      <td>0.9088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sens   Spec  accuracy  balanced_accuracy\n",
       "0 0.9230 0.8945    0.8983             0.9088"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Moxifloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "\n",
    "get_logReg_summary(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9ec5a07e-ae8a-4de2-911f-d4a9ad4e6446",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pza = pd.read_csv(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn/phenos.csv\")\n",
    "\n",
    "# the missing ones might be M. cannettii, most similar to L6 based on the other lineage callers\n",
    "lineages = pd.read_pickle(\"../data/combined_lineage_sample_IDs.pkl\")\n",
    "lineages[\"Lineage\"] = lineages[\"Lineage\"].fillna(\"6\")\n",
    "lineages[\"Lineage_1\"] = lineages[\"Lineage_1\"].fillna(\"6\")\n",
    "\n",
    "lineages = lineages[[\"Sample Name\", \"Sample ID\", \"Lineage_1\"]]\n",
    "lineages[\"Lineage\"] = [str(val).split(\".\")[0] for val in lineages[\"Lineage_1\"].values]\n",
    "lineages.loc[lineages[\"Lineage\"].str.contains(\"BOV\"), \"Lineage\"] = \"M. bovis\"\n",
    "\n",
    "assert len(lineages.loc[pd.isnull(lineages[\"Lineage\"])]) == 0\n",
    "\n",
    "########## KEEP ONLY ISOLATES WITH ALL 3 PIECES OF DATA ##########\n",
    "\n",
    "# get only isolates with data for everyting: SNP matrix, in the model, and lineages\n",
    "combined = lineages.merge(df_pza, left_on=\"Sample ID\", right_on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "56d7652e-ec58-4e0a-88ef-86e80fcc025a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Lineage\n",
       "M. bovis   0.9153\n",
       "2          0.3750\n",
       "1          0.1503\n",
       "6          0.1163\n",
       "4          0.1112\n",
       "5          0.1111\n",
       "3          0.0554\n",
       "7          0.0000\n",
       "Name: phenotype, dtype: float64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.groupby(\"Lineage\")[\"phenotype\"].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f3334fbf-76ee-49e0-a0b9-0dccff1727c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BOV_AFRI', 'BOV'], dtype=object)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.query(\"Lineage == 'M. bovis'\").Lineage_1.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c225d9c-faf1-4141-884c-034c55fa03de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
