{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import sklearn\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "import glob, os, yaml, subprocess, itertools, sparse, pickle\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)\n",
    "samples_summary = pd.read_csv(\"../data/samples_summary.csv\")\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "b6a5a6a1-d7ff-4e89-8765-c19c0311edb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Delamanid\"\n",
    "# final_analysis file with all significant variants for a drug\n",
    "res_df = pd.read_csv(os.path.join(analysis_dir, drug, \"final_analysis.csv\"))\n",
    "# df_phenos = pd.read_csv(os.path.join(analysis_dir, drug, \"phenos_binary.csv\"))\n",
    "\n",
    "# df_genos = pd.read_csv(os.path.join(analysis_dir, drug, \"genos.csv.gz\"), compression=\"gzip\")\n",
    "# df_genos[\"orig_variant\"] = df_genos[\"resolved_symbol\"] + \"_\" + df_genos[\"variant_category\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "483a2c3d-6d83-4c8b-ae72-98e8b22ea6eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig_variant', 'coef', 'coef_LB', 'coef_UB', 'pval', 'BH_pval',\n",
       "       'Bonferroni_pval', 'confidence_WHO_2021', 'Odds_Ratio', 'OR_LB',\n",
       "       'OR_UB', 'Tier', 'Phenos', 'unpooled', 'synonymous', 'Num_Isolates',\n",
       "       'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV', 'Sens', 'Spec', 'LR+',\n",
       "       'LR-', 'PPV_LB', 'PPV_UB', 'Sens_LB', 'Sens_UB', 'Spec_LB', 'Spec_UB',\n",
       "       'LR+_LB', 'LR+_UB', 'LR-_LB', 'LR-_UB', 'predicted_effect'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "95c39875-7ebe-4f20-aa99-3a68e3b694b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>predicted_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_lof</td>\n",
       "      <td>0.2951</td>\n",
       "      <td>0.1698</td>\n",
       "      <td>0.3745</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3432</td>\n",
       "      <td>1.1851</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9444</td>\n",
       "      <td>0.0443</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>0.9991</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>68.8174</td>\n",
       "      <td>770.5864</td>\n",
       "      <td>0.8922</td>\n",
       "      <td>0.9561</td>\n",
       "      <td>lof</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fgd1_lof</td>\n",
       "      <td>0.1563</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.2080</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0332</td>\n",
       "      <td>0.2990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1692</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9782</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>lof</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ddn_p.Val127Phe</td>\n",
       "      <td>0.1547</td>\n",
       "      <td>-0.0273</td>\n",
       "      <td>0.2292</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0299</td>\n",
       "      <td>0.1792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0264</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9738</td>\n",
       "      <td>1.0002</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ddn_p.Leu49Pro</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>0.1530</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.1476</td>\n",
       "      <td>1.1241</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.0717</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>78.2850</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9285</td>\n",
       "      <td>0.9801</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fbiB_p.Lys448Arg</td>\n",
       "      <td>0.1318</td>\n",
       "      <td>-0.0342</td>\n",
       "      <td>0.2785</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>0.1189</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1409</td>\n",
       "      <td>0.9664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1617</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>0.9854</td>\n",
       "      <td>0.9892</td>\n",
       "      <td>3.5155</td>\n",
       "      <td>8.5974</td>\n",
       "      <td>0.9041</td>\n",
       "      <td>0.9667</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>fbiB_p.Leu447Arg</td>\n",
       "      <td>-0.0649</td>\n",
       "      <td>-0.0788</td>\n",
       "      <td>-0.0504</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9371</td>\n",
       "      <td>0.9242</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>fgd1_c.465C&gt;T</td>\n",
       "      <td>-0.0783</td>\n",
       "      <td>-0.0901</td>\n",
       "      <td>-0.0657</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9247</td>\n",
       "      <td>0.9139</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>fbiA_p.Arg304Gln</td>\n",
       "      <td>-0.0799</td>\n",
       "      <td>-0.1325</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0480</td>\n",
       "      <td>0.1129</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.8759</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>ndh_lof</td>\n",
       "      <td>-0.1093</td>\n",
       "      <td>-0.1496</td>\n",
       "      <td>-0.0602</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8964</td>\n",
       "      <td>0.8610</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lof</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>PC0</td>\n",
       "      <td>-0.2291</td>\n",
       "      <td>-0.4017</td>\n",
       "      <td>-0.0775</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.1919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7953</td>\n",
       "      <td>0.6692</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>958 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0             ddn_lof  0.2951   0.1698   0.3745 0.0000   0.0000   \n",
       "1            fgd1_lof  0.1563  -0.0000   0.2080 0.0041   0.0332   \n",
       "2     ddn_p.Val127Phe  0.1547  -0.0273   0.2292 0.0025   0.0299   \n",
       "3      ddn_p.Leu49Pro  0.1377   0.1170   0.1530 0.0000   0.0000   \n",
       "4    fbiB_p.Lys448Arg  0.1318  -0.0342   0.2785 0.0537   0.1189   \n",
       "..                ...     ...      ...      ...    ...      ...   \n",
       "953  fbiB_p.Leu447Arg -0.0649  -0.0788  -0.0504 0.0000   0.0000   \n",
       "954     fgd1_c.465C>T -0.0783  -0.0901  -0.0657 0.0000   0.0000   \n",
       "955  fbiA_p.Arg304Gln -0.0799  -0.1325   0.0000 0.0480   0.1129   \n",
       "956           ndh_lof -0.1093  -0.1496  -0.0602 0.0000   0.0000   \n",
       "957               PC0 -0.2291  -0.4017  -0.0775 0.0026   0.0274   \n",
       "\n",
       "     Bonferroni_pval        confidence_WHO_2021  Odds_Ratio  OR_LB  ...  \\\n",
       "0             0.0000                        NaN      1.3432 1.1851  ...   \n",
       "1             0.2990                        NaN      1.1692 1.0000  ...   \n",
       "2             0.1792                        NaN      1.1673 0.9731  ...   \n",
       "3             0.0000     2) Assoc w R - Interim      1.1476 1.1241  ...   \n",
       "4             1.0000  3) Uncertain significance      1.1409 0.9664  ...   \n",
       "..               ...                        ...         ...    ...  ...   \n",
       "953           0.0000  3) Uncertain significance      0.9371 0.9242  ...   \n",
       "954           0.0000                        NaN      0.9247 0.9139  ...   \n",
       "955           1.0000  3) Uncertain significance      0.9232 0.8759  ...   \n",
       "956           0.0001                        NaN      0.8964 0.8610  ...   \n",
       "957           0.1919                        NaN      0.7953 0.6692  ...   \n",
       "\n",
       "     PPV_UB  Sens_LB Sens_UB  Spec_LB  Spec_UB  LR+_LB   LR+_UB  LR-_LB  \\\n",
       "0    0.9444   0.0443  0.1083   0.9991   0.9999 68.8174 770.5864  0.8922   \n",
       "1    1.0000   0.0000  0.0218   1.0000   1.0000     inf      inf  0.9782   \n",
       "2    1.0000   0.0000  0.0264   0.9996   1.0000  0.0000      inf  0.9738   \n",
       "3    1.0000   0.0201  0.0717   0.9996   1.0000 78.2850      inf  0.9285   \n",
       "4    0.1617   0.0456  0.1083   0.9854   0.9892  3.5155   8.5974  0.9041   \n",
       "..      ...      ...     ...      ...      ...     ...      ...     ...   \n",
       "953     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "954     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "955     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "956     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "957     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "\n",
       "     LR-_UB    predicted_effect  \n",
       "0    0.9561                 lof  \n",
       "1    1.0000                 lof  \n",
       "2    1.0002    missense_variant  \n",
       "3    0.9801    missense_variant  \n",
       "4    0.9667    missense_variant  \n",
       "..      ...                 ...  \n",
       "953     NaN    missense_variant  \n",
       "954     NaN  synonymous_variant  \n",
       "955     NaN    missense_variant  \n",
       "956     NaN                 lof  \n",
       "957     NaN                 NaN  \n",
       "\n",
       "[958 rows x 37 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "b394e2f9-d044-49e8-b17c-1755bfa0b78a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>predicted_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PC4</td>\n",
       "      <td>0.1016</td>\n",
       "      <td>-0.0728</td>\n",
       "      <td>0.2469</td>\n",
       "      <td>0.1077</td>\n",
       "      <td>0.1228</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1069</td>\n",
       "      <td>0.9298</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PC2</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>-0.1190</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2571</td>\n",
       "      <td>0.2844</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0624</td>\n",
       "      <td>0.8878</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>PC3</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>-0.1110</td>\n",
       "      <td>0.1728</td>\n",
       "      <td>0.3431</td>\n",
       "      <td>0.3738</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0305</td>\n",
       "      <td>0.8950</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>PC1</td>\n",
       "      <td>-0.0039</td>\n",
       "      <td>-0.1845</td>\n",
       "      <td>0.1639</td>\n",
       "      <td>0.4829</td>\n",
       "      <td>0.5109</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.8316</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>PC0</td>\n",
       "      <td>-0.2291</td>\n",
       "      <td>-0.4017</td>\n",
       "      <td>-0.0775</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.1919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7953</td>\n",
       "      <td>0.6692</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "11           PC4  0.1016  -0.0728   0.2469 0.1077   0.1228           1.0000   \n",
       "18           PC2  0.0605  -0.1190   0.2391 0.2571   0.2844           1.0000   \n",
       "71           PC3  0.0300  -0.1110   0.1728 0.3431   0.3738           1.0000   \n",
       "88           PC1 -0.0039  -0.1845   0.1639 0.4829   0.5109           1.0000   \n",
       "957          PC0 -0.2291  -0.4017  -0.0775 0.0026   0.0274           0.1919   \n",
       "\n",
       "    confidence_WHO_2021  Odds_Ratio  OR_LB  ...  PPV_UB  Sens_LB Sens_UB  \\\n",
       "11                  NaN      1.1069 0.9298  ...     NaN      NaN     NaN   \n",
       "18                  NaN      1.0624 0.8878  ...     NaN      NaN     NaN   \n",
       "71                  NaN      1.0305 0.8950  ...     NaN      NaN     NaN   \n",
       "88                  NaN      0.9961 0.8316  ...     NaN      NaN     NaN   \n",
       "957                 NaN      0.7953 0.6692  ...     NaN      NaN     NaN   \n",
       "\n",
       "     Spec_LB  Spec_UB  LR+_LB  LR+_UB  LR-_LB  LR-_UB  predicted_effect  \n",
       "11       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "18       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "71       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "88       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "957      NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df.loc[pd.isnull(res_df[\"predicted_effect\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "ad6c6247-caa0-455b-bdab-7aaf1380cd56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig_variant', 'coef', 'coef_LB', 'coef_UB', 'pval', 'BH_pval',\n",
       "       'Bonferroni_pval', 'confidence_WHO_2021', 'Odds_Ratio', 'OR_LB',\n",
       "       'OR_UB', 'Tier', 'Phenos', 'unpooled', 'synonymous', 'Num_Isolates',\n",
       "       'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV', 'Sens', 'Spec', 'LR+',\n",
       "       'LR-', 'PPV_LB', 'PPV_UB', 'Sens_LB', 'Sens_UB', 'Spec_LB', 'Spec_UB',\n",
       "       'LR+_LB', 'LR+_UB', 'LR-_LB', 'LR-_UB', 'predicted_effect'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0f11db9-ec95-4814-919f-7dd19c3fdc36",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c05a0977-5fd4-4421-81f8-29bcd5abfb66",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a2439e0a-1ad9-4471-9e91-ac841f82fba2",
   "metadata": {},
   "source": [
    "# Combined Analysis Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "00729d49-7aeb-457e-af65-74a6bf488b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def final_processing(drug):\n",
    "    '''\n",
    "    Functions for processing outputs before sending to everyone else.\n",
    "    \n",
    "    1. Remove principal components (will describe them separately)\n",
    "    2. Add LOF to the predicted_effect column for pooled LOF mutations\n",
    "    3. Remove genome_index column (should actually do that earlier, but will fix later)\n",
    "    4. Remove the logistic regression coefficient columns (they will prefer to work with odds ratios)\n",
    "    5. Any other column renaming or dropping for clarity\n",
    "    '''\n",
    "    \n",
    "    analysis_df = pd.read_csv(os.path.join(analysis_dir, drug, \"final_analysis.csv\"))\n",
    "    analysis_df.rename(columns={\"orig_variant\": \"mutation\"}, inplace=True)\n",
    "    \n",
    "    # remove logReg coefficients. Keep only odds ratios. Remove the other two columns, which were present mainly for me to see\n",
    "    # if we were picking up many mutations that were in the 2021 mutation catalog\n",
    "    del analysis_df[\"confidence_WHO_2021\"]\n",
    "    analysis_df = analysis_df[analysis_df.columns[~analysis_df.columns.str.contains(\"coef\")]]\n",
    "    \n",
    "    # remove significant principal components and replace the NaNs in the predicted effect column for the gene loss of functions\n",
    "    analysis_df = analysis_df.loc[~analysis_df[\"mutation\"].str.contains(\"PC\")]\n",
    "    \n",
    "    # predicted effect should not be NaN for anything. position is NaN only for the pooled LOF mutations\n",
    "    assert len(analysis_df.loc[pd.isnull(analysis_df[\"predicted_effect\"])]) == 0\n",
    "    \n",
    "    # reorder columns\n",
    "    analysis_df = analysis_df[['mutation', 'predicted_effect', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval', 'Bonferroni_pval',\n",
    "       'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'Sens_LB', 'Sens', 'Sens_UB', 'Spec_LB', 'Spec', 'Spec_UB', 'PPV', 'PPV_LB', 'PPV_UB',\n",
    "       'LR+_LB', 'LR+', 'LR+_UB', 'LR-_LB', 'LR-', 'LR-_UB', 'Tier', 'Phenos', 'unpooled']]\n",
    "            \n",
    "    return analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "9a7cc857-89b9-440d-844d-c5476a780367",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = final_processing(\"Delamanid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "60aed635-6a48-44c8-a9d8-48b26459d133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>predicted_effect</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>Num_Isolates</th>\n",
       "      <th>Total_Isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Phenos</th>\n",
       "      <th>unpooled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_lof</td>\n",
       "      <td>lof</td>\n",
       "      <td>1.3432</td>\n",
       "      <td>1.1851</td>\n",
       "      <td>1.4543</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>24.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>73.3231</td>\n",
       "      <td>174.1817</td>\n",
       "      <td>852.3245</td>\n",
       "      <td>0.8908</td>\n",
       "      <td>0.9250</td>\n",
       "      <td>0.9559</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fgd1_lof</td>\n",
       "      <td>lof</td>\n",
       "      <td>1.1692</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.2312</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0332</td>\n",
       "      <td>0.2990</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9806</td>\n",
       "      <td>0.9921</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ddn_p.Val127Phe</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>1.2576</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0299</td>\n",
       "      <td>0.1792</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>68.7560</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9739</td>\n",
       "      <td>0.9883</td>\n",
       "      <td>1.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ddn_p.Leu49Pro</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1.1476</td>\n",
       "      <td>1.1241</td>\n",
       "      <td>1.1654</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>13.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>75.1657</td>\n",
       "      <td>252.1052</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9278</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.9792</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fbiB_p.Lys448Arg</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1.1409</td>\n",
       "      <td>0.9664</td>\n",
       "      <td>1.3211</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>0.1189</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>165.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1667</td>\n",
       "      <td>3.2246</td>\n",
       "      <td>5.9651</td>\n",
       "      <td>9.1374</td>\n",
       "      <td>0.9028</td>\n",
       "      <td>0.9364</td>\n",
       "      <td>0.9713</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>fbiC_c.967C&gt;T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.9374</td>\n",
       "      <td>0.9270</td>\n",
       "      <td>0.9490</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>35.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>fbiB_p.Leu447Arg</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0.9371</td>\n",
       "      <td>0.9242</td>\n",
       "      <td>0.9509</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>33.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>fgd1_c.465C&gt;T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.9247</td>\n",
       "      <td>0.9139</td>\n",
       "      <td>0.9364</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>71.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>fbiA_p.Arg304Gln</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.8759</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0480</td>\n",
       "      <td>0.1129</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>ndh_lof</td>\n",
       "      <td>lof</td>\n",
       "      <td>0.8964</td>\n",
       "      <td>0.8610</td>\n",
       "      <td>0.9416</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>44.0000</td>\n",
       "      <td>11803.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.1827</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>953 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             mutation    predicted_effect  Odds_Ratio  OR_LB  OR_UB   pval  \\\n",
       "0             ddn_lof                 lof      1.3432 1.1851 1.4543 0.0000   \n",
       "1            fgd1_lof                 lof      1.1692 1.0000 1.2312 0.0041   \n",
       "2     ddn_p.Val127Phe    missense_variant      1.1673 0.9731 1.2576 0.0025   \n",
       "3      ddn_p.Leu49Pro    missense_variant      1.1476 1.1241 1.1654 0.0000   \n",
       "4    fbiB_p.Lys448Arg    missense_variant      1.1409 0.9664 1.3211 0.0537   \n",
       "..                ...                 ...         ...    ...    ...    ...   \n",
       "952     fbiC_c.967C>T  synonymous_variant      0.9374 0.9270 0.9490 0.0000   \n",
       "953  fbiB_p.Leu447Arg    missense_variant      0.9371 0.9242 0.9509 0.0000   \n",
       "954     fgd1_c.465C>T  synonymous_variant      0.9247 0.9139 0.9364 0.0000   \n",
       "955  fbiA_p.Arg304Gln    missense_variant      0.9232 0.8759 1.0000 0.0480   \n",
       "956           ndh_lof                 lof      0.8964 0.8610 0.9416 0.0000   \n",
       "\n",
       "     BH_pval  Bonferroni_pval  Num_Isolates  Total_Isolates  ...  PPV_UB  \\\n",
       "0     0.0000           0.0000       24.0000      11803.0000  ...  0.9500   \n",
       "1     0.0332           0.2990        2.0000      11803.0000  ...  1.0000   \n",
       "2     0.0299           0.1792        5.0000      11803.0000  ...  1.0000   \n",
       "3     0.0000           0.0000       13.0000      11803.0000  ...  1.0000   \n",
       "4     0.1189           1.0000      165.0000      11803.0000  ...  0.1667   \n",
       "..       ...              ...           ...             ...  ...     ...   \n",
       "952   0.0000           0.0000       35.0000      11803.0000  ...     NaN   \n",
       "953   0.0000           0.0000       33.0000      11803.0000  ...     NaN   \n",
       "954   0.0000           0.0000       71.0000      11803.0000  ...     NaN   \n",
       "955   0.1129           1.0000        1.0000      11803.0000  ...     NaN   \n",
       "956   0.0000           0.0001       44.0000      11803.0000  ...     NaN   \n",
       "\n",
       "     LR+_LB      LR+   LR+_UB  LR-_LB    LR-  LR-_UB  Tier  Phenos  unpooled  \n",
       "0   73.3231 174.1817 852.3245  0.8908 0.9250  0.9559     1     WHO         0  \n",
       "1       inf      inf      inf  0.9806 0.9921  1.0000     1     WHO         0  \n",
       "2    0.0000  68.7560      inf  0.9739 0.9883  1.0002     1     WHO         0  \n",
       "3   75.1657 252.1052      inf  0.9278 0.9565  0.9792     2     ALL         1  \n",
       "4    3.2246   5.9651   9.1374  0.9028 0.9364  0.9713     1     WHO         0  \n",
       "..      ...      ...      ...     ...    ...     ...   ...     ...       ...  \n",
       "952     NaN   0.0000      NaN     NaN 1.0030     NaN     2     ALL         0  \n",
       "953     NaN   0.0000      NaN     NaN 1.0029     NaN     2     ALL         1  \n",
       "954     NaN   0.0000      NaN     NaN 1.0062     NaN     2     ALL         0  \n",
       "955     NaN   0.0000      NaN     NaN 1.0001     NaN     1     WHO         0  \n",
       "956     NaN   2.1827      NaN     NaN 0.9957     NaN     1     WHO         0  \n",
       "\n",
       "[953 rows x 32 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "9de0a2e2-a121-4c4a-bc32-678c9b3a16f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/final_analysis.csv because it is not a directory\n",
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1+2/phenos=ALL/encodeAF_noSyn because this model encodes Hets with AFs.\n",
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1+2/phenos=WHO/encodeAF_noSyn because this model encodes Hets with AFs.\n",
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/genos.csv.gz because it is not a directory\n",
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/phenos_binary.csv because it is not a directory\n",
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1/phenos=ALL/encodeAF_noSyn because this model encodes Hets with AFs.\n",
      "Skipping /n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue/Delamanid/tiers=1/phenos=WHO/encodeAF_noSyn because this model encodes Hets with AFs.\n"
     ]
    }
   ],
   "source": [
    "drug = \"Delamanid\"\n",
    "core_model_path = os.path.join(analysis_dir, drug, \"tiers=1/phenos=WHO/dropAF_noSyn\")\n",
    "\n",
    "core_analysis = pd.read_csv(os.path.join(core_model_path, \"model_analysis.csv\"))\n",
    "core_analysis[\"Tier\"] = 1\n",
    "core_analysis[\"Phenos\"] = \"WHO\"\n",
    "core_analysis[\"unpooled\"] = 0\n",
    "core_analysis[\"synonymous\"] = 0\n",
    "\n",
    "# create dictionary of the additional model analyses\n",
    "add_analyses = {}\n",
    "\n",
    "for tier in os.listdir(os.path.join(analysis_dir, drug)):\n",
    "    tiers_path = os.path.join(analysis_dir, drug, tier)\n",
    "    if os.path.isdir(tiers_path):\n",
    "        for pheno in os.listdir(tiers_path):\n",
    "            phenos_path = os.path.join(analysis_dir, drug, tier, pheno)\n",
    "            for model in os.listdir(phenos_path):\n",
    "                model_path = os.path.join(analysis_dir, drug, tier, pheno, model)\n",
    "\n",
    "                analysis_fName = os.path.join(model_path, \"model_analysis.csv\")\n",
    "                if os.path.isfile(analysis_fName):\n",
    "                    \n",
    "                    # only include models where HETs were dropped because we can not compute univariate stats for continuous AFs\n",
    "                    if model_path != core_model_path:\n",
    "                        if (\"encodeAF\" not in model_path):\n",
    "                            add_analysis = pd.read_csv(analysis_fName)\n",
    "                            add_path = model_path.split(os.path.join(analysis_dir, drug))[1].strip(\"/\").replace(\"dropAF_\", \"\")\n",
    "\n",
    "                            add_analysis[\"Tier\"] = [2 if \"+2\" in add_path else 1][0]\n",
    "                            add_analysis[\"Phenos\"] = [\"ALL\" if \"ALL\" in add_path else \"WHO\"][0]\n",
    "                            add_analysis[\"unpooled\"] = int(\"unpooled\" in add_path)\n",
    "                            add_analysis[\"synonymous\"] = int(\"withSyn\" in add_path)\n",
    "\n",
    "                            # keep all significant variants at an FDR threshold of 0.01\n",
    "                            add_analyses[add_path] = add_analysis\n",
    "                        else:\n",
    "                            print(f\"Skipping {model_path} because this model encodes Hets with AFs.\")\n",
    "                else:\n",
    "                    print(f\"No model analysis file in {analysis_fName}\")\n",
    "    else:\n",
    "        print(f\"Skipping {tiers_path} because it is not a directory\")\n",
    "      \n",
    "keys_lst = list(add_analyses.keys())\n",
    "\n",
    "# merge the core analysis dataframe with the first additional dataframe\n",
    "merged_df = pd.concat([core_analysis, add_analyses[keys_lst[0]]], axis=0).drop_duplicates(\"orig_variant\", keep=\"first\")\n",
    "\n",
    "# merge the merged dataframe with the remaining additional analyses, dropping duplicates as before\n",
    "for key in keys_lst[1:]:\n",
    "    merged_df = pd.concat([merged_df, add_analyses[key]], axis=0).drop_duplicates(\"orig_variant\", keep=\"first\")\n",
    "\n",
    "del merged_df[\"genome_index\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "3dc1d0d5-21e8-44fe-9098-e4059a12c2a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Phenos</th>\n",
       "      <th>unpooled</th>\n",
       "      <th>synonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_lof</td>\n",
       "      <td>0.2951</td>\n",
       "      <td>0.1698</td>\n",
       "      <td>0.3745</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3432</td>\n",
       "      <td>1.1851</td>\n",
       "      <td>1.4543</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fgd1_lof</td>\n",
       "      <td>0.1563</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.2080</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0332</td>\n",
       "      <td>0.2990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1692</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.2312</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ddn_p.Val127Phe</td>\n",
       "      <td>0.1547</td>\n",
       "      <td>-0.0273</td>\n",
       "      <td>0.2292</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0299</td>\n",
       "      <td>0.1792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>1.2576</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fbiB_p.Lys448Arg</td>\n",
       "      <td>0.1318</td>\n",
       "      <td>-0.0342</td>\n",
       "      <td>0.2785</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>0.1189</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1409</td>\n",
       "      <td>0.9664</td>\n",
       "      <td>1.3211</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fbiC_lof</td>\n",
       "      <td>0.1286</td>\n",
       "      <td>-0.0551</td>\n",
       "      <td>0.2062</td>\n",
       "      <td>0.0380</td>\n",
       "      <td>0.0926</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1372</td>\n",
       "      <td>0.9464</td>\n",
       "      <td>1.2290</td>\n",
       "      <td>1</td>\n",
       "      <td>WHO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td>fbiC_c.24G&gt;A</td>\n",
       "      <td>-0.0458</td>\n",
       "      <td>-0.0544</td>\n",
       "      <td>-0.0340</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9552</td>\n",
       "      <td>0.9471</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891</th>\n",
       "      <td>fbiC_c.396C&gt;T</td>\n",
       "      <td>-0.0513</td>\n",
       "      <td>-0.0610</td>\n",
       "      <td>-0.0387</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>0.9408</td>\n",
       "      <td>0.9620</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td>fgd1_c.774G&gt;A</td>\n",
       "      <td>-0.0641</td>\n",
       "      <td>-0.0745</td>\n",
       "      <td>-0.0505</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9379</td>\n",
       "      <td>0.9282</td>\n",
       "      <td>0.9507</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td>fbiC_c.967C&gt;T</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>-0.0758</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9374</td>\n",
       "      <td>0.9270</td>\n",
       "      <td>0.9490</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>fgd1_c.465C&gt;T</td>\n",
       "      <td>-0.0783</td>\n",
       "      <td>-0.0901</td>\n",
       "      <td>-0.0657</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9247</td>\n",
       "      <td>0.9139</td>\n",
       "      <td>0.9364</td>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>958 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0             ddn_lof  0.2951   0.1698   0.3745 0.0000   0.0000   \n",
       "1            fgd1_lof  0.1563  -0.0000   0.2080 0.0041   0.0332   \n",
       "2     ddn_p.Val127Phe  0.1547  -0.0273   0.2292 0.0025   0.0299   \n",
       "3    fbiB_p.Lys448Arg  0.1318  -0.0342   0.2785 0.0537   0.1189   \n",
       "4            fbiC_lof  0.1286  -0.0551   0.2062 0.0380   0.0926   \n",
       "..                ...     ...      ...      ...    ...      ...   \n",
       "884      fbiC_c.24G>A -0.0458  -0.0544  -0.0340 0.0000   0.0000   \n",
       "891     fbiC_c.396C>T -0.0513  -0.0610  -0.0387 0.0000   0.0000   \n",
       "896     fgd1_c.774G>A -0.0641  -0.0745  -0.0505 0.0000   0.0000   \n",
       "897     fbiC_c.967C>T -0.0646  -0.0758  -0.0524 0.0000   0.0000   \n",
       "899     fgd1_c.465C>T -0.0783  -0.0901  -0.0657 0.0000   0.0000   \n",
       "\n",
       "     Bonferroni_pval        confidence_WHO_2021  Odds_Ratio  OR_LB  OR_UB  \\\n",
       "0             0.0000                        NaN      1.3432 1.1851 1.4543   \n",
       "1             0.2990                        NaN      1.1692 1.0000 1.2312   \n",
       "2             0.1792                        NaN      1.1673 0.9731 1.2576   \n",
       "3             1.0000  3) Uncertain significance      1.1409 0.9664 1.3211   \n",
       "4             1.0000                        NaN      1.1372 0.9464 1.2290   \n",
       "..               ...                        ...         ...    ...    ...   \n",
       "884           0.0000                        NaN      0.9552 0.9471 0.9666   \n",
       "891           0.0000                        NaN      0.9500 0.9408 0.9620   \n",
       "896           0.0000                        NaN      0.9379 0.9282 0.9507   \n",
       "897           0.0000                        NaN      0.9374 0.9270 0.9490   \n",
       "899           0.0000                        NaN      0.9247 0.9139 0.9364   \n",
       "\n",
       "     Tier Phenos  unpooled  synonymous  \n",
       "0       1    WHO         0           0  \n",
       "1       1    WHO         0           0  \n",
       "2       1    WHO         0           0  \n",
       "3       1    WHO         0           0  \n",
       "4       1    WHO         0           0  \n",
       "..    ...    ...       ...         ...  \n",
       "884     2    ALL         0           1  \n",
       "891     2    ALL         0           1  \n",
       "896     2    ALL         0           1  \n",
       "897     2    ALL         0           1  \n",
       "899     2    ALL         0           1  \n",
       "\n",
       "[958 rows x 15 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75397467-07f9-47ae-81ae-1610c64662a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf93fcb9-3ec6-4938-9d0c-3b88cb2fbd7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "a218179e-bfde-4480-a083-fa2b8986bd75",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_predictive_values(combined_df, return_stats=[]):\n",
    "    '''\n",
    "    Compute positive predictive value. \n",
    "    Compute sensitivity, specificity, and positive and negative likelihood ratios. \n",
    "    \n",
    "    PPV = true_positive / all_positive. NPV = true_negative / all_negative\n",
    "    Sens = true_positive / (true_positive + false_negative)\n",
    "    Spec = true_negative / (true_negative + false_positive)\n",
    "    \n",
    "    Also return the number of isolates with each variant = all_positive\n",
    "    \n",
    "    Positive LR = sens / (1 – spec)\n",
    "    Negative LR = (1 – sens) / spec\n",
    "\n",
    "    '''\n",
    "    # make a copy to keep sample_id in one dataframe\n",
    "    melted = combined_df.melt(id_vars=[\"sample_id\", \"phenotype\"])\n",
    "    melted_2 = melted.copy()\n",
    "    del melted_2[\"sample_id\"]\n",
    "    \n",
    "    # get counts of isolates grouped by phenotype and variant -- so how many isolates have a variant and have a phenotype (all 4 possibilities)\n",
    "    grouped_df = pd.DataFrame(melted_2.groupby([\"phenotype\", \"variable\"]).value_counts()).reset_index()\n",
    "    grouped_df = grouped_df.rename(columns={\"variable\": \"orig_variant\", \"value\": \"variant\", 0:\"count\"})\n",
    "    \n",
    "    # dataframes of the counts of the 4 values\n",
    "    true_pos_df = grouped_df.query(\"variant == 1 & phenotype == 1\").rename(columns={\"count\": \"TP\"})\n",
    "    false_pos_df = grouped_df.query(\"variant == 1 & phenotype == 0\").rename(columns={\"count\": \"FP\"})\n",
    "    true_neg_df = grouped_df.query(\"variant == 0 & phenotype == 0\").rename(columns={\"count\": \"TN\"})\n",
    "    false_neg_df = grouped_df.query(\"variant == 0 & phenotype == 1\").rename(columns={\"count\": \"FN\"})\n",
    "\n",
    "    assert len(true_pos_df) + len(false_pos_df) + len(true_neg_df) + len(false_neg_df) == len(grouped_df)\n",
    "    \n",
    "    # combine the 4 dataframes into a single dataframe (concatenating on axis = 1)\n",
    "    final = true_pos_df[[\"orig_variant\", \"TP\"]].merge(\n",
    "            false_pos_df[[\"orig_variant\", \"FP\"]], on=\"orig_variant\", how=\"outer\").merge(\n",
    "            true_neg_df[[\"orig_variant\", \"TN\"]], on=\"orig_variant\", how=\"outer\").merge(\n",
    "            false_neg_df[[\"orig_variant\", \"FN\"]], on=\"orig_variant\", how=\"outer\").fillna(0)\n",
    "\n",
    "    assert len(final) == len(melted[\"variable\"].unique())\n",
    "    assert len(final) == len(final.drop_duplicates(\"orig_variant\"))\n",
    "        \n",
    "    final[\"Num_Isolates\"] = final[\"TP\"] + final[\"FP\"]\n",
    "    final[\"Total_Isolates\"] = final[\"TP\"] + final[\"FP\"] + final[\"TN\"] + final[\"FN\"]\n",
    "    final[\"PPV\"] = final[\"TP\"] / (final[\"TP\"] + final[\"FP\"])\n",
    "    final[\"Sens\"] = final[\"TP\"] / (final[\"TP\"] + final[\"FN\"])\n",
    "    final[\"Spec\"] = final[\"TN\"] / (final[\"TN\"] + final[\"FP\"])\n",
    "    final[\"LR+\"] = final[\"Sens\"] / (1 - final[\"Spec\"])\n",
    "    final[\"LR-\"] = (1 - final[\"Sens\"]) / final[\"Spec\"]\n",
    "    #final[\"NPV\"] = final[\"TN\"] / (final[\"TN\"] + final[\"FN\"])\n",
    "    \n",
    "    if len(return_stats) == 0:\n",
    "        return final[[\"orig_variant\", \"Num_Isolates\", \"Total_Isolates\", \"TP\", \"FP\", \"TN\", \"FN\", \"PPV\", \"Sens\", \"Spec\", \"LR+\", \"LR-\"]]\n",
    "    else:\n",
    "        return final[return_stats]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "4995cdd4-e678-4303-b6b8-af2cdb4c99a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_6135/3506956419.py:2: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_genos = pd.read_csv(os.path.join(analysis_dir, drug, \"genos.csv.gz\"), compression=\"gzip\")\n"
     ]
    }
   ],
   "source": [
    "df_phenos = pd.read_csv(os.path.join(analysis_dir, drug, \"phenos_binary.csv\"))\n",
    "df_genos = pd.read_csv(os.path.join(analysis_dir, drug, \"genos.csv.gz\"), compression=\"gzip\")\n",
    "df_genos[\"orig_variant\"] = df_genos[\"resolved_symbol\"] + \"_\" + df_genos[\"variant_category\"]\n",
    "df_copy = df_genos.copy()\n",
    "\n",
    "# pool LOF and inframe mutations\n",
    "df_copy.loc[df_copy[\"predicted_effect\"].isin([\"frameshift\", \"start_lost\", \"stop_gained\", \"feature_ablation\"]), [\"variant_category\", \"position\"]] = [\"lof\", np.nan]\n",
    "df_copy.loc[df_copy[\"predicted_effect\"].isin([\"inframe_insertion\", \"inframe_deletion\"]), [\"variant_category\", \"position\"]] = [\"inframe\", np.nan]\n",
    "\n",
    "# update the orig_variant column using the new variant categories (lof and inframe) and combine the pooled and unpooled variants\n",
    "df_copy[\"orig_variant\"] = df_copy[\"resolved_symbol\"] + \"_\" + df_copy[\"variant_category\"]\n",
    "df_pooled = df_copy.query(f\"variant_category in ['lof', 'inframe']\").sort_values(by=[\"variant_binary_status\", \"variant_allele_frequency\"], ascending=False, na_position=\"last\").drop_duplicates(subset=[\"sample_id\", \"orig_variant\"], keep=\"first\")\n",
    "del df_copy\n",
    "df_genos_full = pd.concat([df_genos, df_pooled], axis=0)\n",
    "del df_genos\n",
    "del df_pooled\n",
    "\n",
    "# keep only variants that are in the final_analysis dataframe and drop NaNs (NaNs = either isolate didn't pass QC or it's a Het) \n",
    "# We can't process Hets here because they need to be binary to have univariate statistics\n",
    "df_genos_full = df_genos_full.loc[(df_genos_full[\"orig_variant\"].isin(merged_df[\"orig_variant\"].values))].dropna(subset=\"variant_binary_status\")\n",
    "\n",
    "# check that the only variants that are in res_df but not in df_genos_full are the principal components\n",
    "if sum(~pd.Series(list(set(merged_df[\"orig_variant\"]) - set(df_genos_full[\"orig_variant\"]))).str.contains(\"PC\")) > 0:\n",
    "    raise ValueError(\"Variants are missing from df_genos_full!\")\n",
    "\n",
    "combined = df_genos_full.pivot(index=\"sample_id\", columns=\"orig_variant\", values=\"variant_binary_status\")\n",
    "\n",
    "# predicted effect annotations for later\n",
    "annotated_genos = df_genos_full.query(\"variant_category not in ['lof', 'inframe']\").drop_duplicates([\"orig_variant\", \"predicted_effect\"])[[\"orig_variant\", \"predicted_effect\"]]\n",
    "del df_genos_full\n",
    "combined = combined.merge(df_phenos[[\"sample_id\", \"phenotype\"]], left_index=True, right_on=\"sample_id\").reset_index(drop=True)\n",
    "\n",
    "# get dataframe of predictive values for the non-zero coefficients and add them to the results dataframe\n",
    "full_predict_values = compute_predictive_values(combined)\n",
    "merged_df = merged_df.merge(full_predict_values, on=\"orig_variant\", how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "cc75e982-678c-4e9b-bf2c-2438882d6494",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig_variant', 'coef', 'coef_LB', 'coef_UB', 'pval', 'BH_pval',\n",
       "       'Bonferroni_pval', 'confidence_WHO_2021', 'Odds_Ratio', 'OR_LB',\n",
       "       'OR_UB', 'Tier', 'Phenos', 'unpooled', 'synonymous', 'Num_Isolates',\n",
       "       'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV', 'Sens', 'Spec', 'LR+',\n",
       "       'LR-'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "fd0d9b5c-009d-4bb5-8c76-3dfa2721b1ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "keep_variants = list(merged_df.loc[(~merged_df[\"orig_variant\"].str.contains(\"PC\")) & (merged_df[\"coef\"] > 0)][\"orig_variant\"].values)\n",
    "\n",
    "# Remake this dataframe with fewer features because only going to bootstrap stats for variants with positive coefficients.\n",
    "# check that all samples were preserved. \n",
    "combined_small = combined[[\"sample_id\", \"phenotype\"] + keep_variants]\n",
    "assert len(combined_small) == len(combined)\n",
    "\n",
    "bs_results = pd.DataFrame(columns = keep_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "748db53b-d764-49be-9632-9d96d69a54b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ddn_lof</th>\n",
       "      <th>fgd1_lof</th>\n",
       "      <th>ddn_p.Val127Phe</th>\n",
       "      <th>fbiB_p.Lys448Arg</th>\n",
       "      <th>fbiC_lof</th>\n",
       "      <th>fbiA_lof</th>\n",
       "      <th>fgd1_inframe</th>\n",
       "      <th>fbiB_inframe</th>\n",
       "      <th>ddn_inframe</th>\n",
       "      <th>fbiA_p.Val47Ile</th>\n",
       "      <th>...</th>\n",
       "      <th>ndh_c.228G&gt;A</th>\n",
       "      <th>fbiA_c.900G&gt;A</th>\n",
       "      <th>fbiC_c.1392C&gt;T</th>\n",
       "      <th>fbiB_c.1330T&gt;C</th>\n",
       "      <th>fbiC_c.2235G&gt;A</th>\n",
       "      <th>fbiC_c.2427C&gt;G</th>\n",
       "      <th>fbiB_c.1335G&gt;A</th>\n",
       "      <th>fbiB_c.90C&gt;T</th>\n",
       "      <th>fbiA_c.337C&gt;T</th>\n",
       "      <th>fbiA_c.15T&gt;C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ddn_lof, fgd1_lof, ddn_p.Val127Phe, fbiB_p.Lys448Arg, fbiC_lof, fbiA_lof, fgd1_inframe, fbiB_inframe, ddn_inframe, fbiA_p.Val47Ile, ddn_p.Leu49Pro, ddn_p.Trp139*, ddn_p.Trp88*, ddn_p.Asp108fs, ddn_p.Tyr122_Met129del, fbiB_p.Ala31Thr, ddn_p.Tyr29del, fbiA_p.Arg321Ser, ddn_p.Arg31fs, fbiA_p.Ser35Ala, ddn_p.Gly81Ser, fgd1_p.Lys183Met, ddn_p.Tyr65Ser, fbiC_p.Val581Leu, ddn_p.Leu90fs, fbiC_p.Val713Leu, fbiA_p.Val329fs, fbiC_p.Ala594Ser, fbiC_p.Glu598Asp, fgd1_p.Asn112fs, ddn_p.Lys104fs, ddn_p.Trp88Arg, fbiC_p.Ala852Thr, fbiB_p.Thr268Ile, fgd1_p.Asp190Glu, fgd1_p.Ile174_Glu295del, fbiC_p.Ala623Val, fbiB_p.Val217_Gly378del, ddn_p.Gly38fs, ddn_p.Pro5fs, fbiA_p.Ile53Thr, ndh_p.Tyr342*, fbiA_p.His183Gln, fbiA_p.Ser184Thr, ndh_p.Ile212fs, fbiC_p.Pro362fs, fbiA_p.Cys280Gly, fbiC_p.Met812Ile, fbiC_p.Ala487fs, fbiC_p.Ser443fs, fgd1_deletion, ddn_p.Trp20*, fbiA_p.Asn314Lys, fbiA_p.Leu104Phe, fbiA_p.Ala206Thr, fgd1_p.Val170Met, fgd1_c.960T>C, ddn_c.168C>T, Rv2983_c.216G>A, fgd1_c.666C>T, ndh_c.960G>A, fgd1_c.504G>C, ndh_c.252A>C, fbiA_c.213C>T, ndh_c.561A>C, fbiA_c.744C>A, fbiC_c.1413T>G, Rv2983_c.444C>G, fbiB_c.171G>A, ndh_c.228G>A, fbiA_c.900G>A, fbiC_c.1392C>T, fbiB_c.1330T>C, fbiC_c.2235G>A, fbiC_c.2427C>G, fbiB_c.1335G>A, fbiB_c.90C>T, fbiA_c.337C>T, fbiA_c.15T>C]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 79 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c051dd-4f63-4e6d-a9c2-c057ad504a20",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "26ac6d98-c680-47eb-8a77-0a12f7121517",
   "metadata": {},
   "outputs": [],
   "source": [
    "finished_drugs = []\n",
    "\n",
    "for drug in os.listdir(analysis_dir):\n",
    "    \n",
    "    if os.path.isfile(os.path.join(analysis_dir, drug, \"final_analysis.csv\")):\n",
    "        finished_drugs.append(drug)\n",
    "        # drug_analyses.append(final_processing(drug))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f27f833-b810-4a22-9cd8-1ec10b70a75b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(finished_drugs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f2bfc4-14a1-4afb-9b97-79b58636f547",
   "metadata": {},
   "outputs": [],
   "source": [
    "# those that are actually done\n",
    "finished_drugs = ['Pyrazinamide', 'Amikacin', 'Clofazimine', 'Linezolid', 'Moxifloxacin', 'Kanamycin', 'Bedaquiline', 'Capreomycin', 'Delamanid']\n",
    "\n",
    "drug_analyses = {}\n",
    "for drug in finished_drugs:\n",
    "    drug_analyses[drug] = final_processing(drug)\n",
    "    \n",
    "# write results to an Excel file, where each sheet is named for a drug\n",
    "with pd.ExcelWriter(\"Farhat_logReg_analysis.xlsx\") as file:\n",
    "   \n",
    "    for key, val in drug_analyses.items():\n",
    "        val.to_excel(file, sheet_name=key, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd2ede15-4b01-4a16-8565-531b1925806c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17dbb6e3-d3d8-4427-963b-f2c88919552c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d471f30-5ad3-4545-aaae-53210b0d97a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52ad210e-7627-48c7-84dd-e7d3e365d3fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_analyses[4].query(\"Tier==2&Phenos=='WHO'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f1c384-b661-4f12-b071-ff944fd1ad2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_analyses[0].query(\"Tier==1&Phenos=='ALL'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66beb441-6ac2-4879-bf36-d39e8bc88007",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_analyses[0].query(\"predicted_effect=='synonymous_variant' & \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e8a634b-b24c-47dd-8310-4d29cde7ab82",
   "metadata": {},
   "outputs": [],
   "source": [
    "moxi_df.loc[pd.isnull(moxi_df[\"position\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a683ba2b-ca3d-4b46-b3fb-eff707fe4cf5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdcae08c-4bc9-4527-8e24-c0c3922beb1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aed4a42c-86c9-4a43-9382-7f9723720ada",
   "metadata": {},
   "source": [
    "# Catalog-Based Method Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f78280-0749-4fe9-8cd9-079fa016922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_logReg_summary(out_dir):\n",
    "    \n",
    "    model_summary = pd.read_csv(os.path.join(out_dir, \"logReg_summary.csv\"))\n",
    "    model_analysis = pd.read_csv(os.path.join(out_dir, \"model_analysis.csv\"))\n",
    "    \n",
    "    summary_sens = model_summary.loc[0, \"Sens\"]\n",
    "    summary_spec = model_summary.loc[0, \"Spec\"]\n",
    "    \n",
    "    #max_idx = [np.argmax(model_analysis[\"Sens\"]), np.argmax(model_analysis[\"Spec\"])]\n",
    "    \n",
    "    better_variant = model_analysis.query(\"Sens >= @summary_sens & Spec >= @summary_spec\")\n",
    "    \n",
    "    if len(better_variant) == 0:\n",
    "        print(\"No variants have comparable sensitivity AND specificity\")\n",
    "    else:\n",
    "        print(better_variant)\n",
    "    return model_summary, model_analysis\n",
    "    \n",
    "    #return model_summary, model_analysis.iloc[max_idx, :][[\"orig_variant\", \"coef\", \"Sens\", \"Spec\", \"accuracy\", \"balanced_accuracy\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20dfc9fa-8f4b-4e2d-81b2-489754985444",
   "metadata": {},
   "outputs": [],
   "source": [
    "#out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Levofloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn\"\n",
    "\n",
    "summary, analysis = get_logReg_summary(out_dir)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ef135c-6253-4a64-9053-a08410585184",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.query(\"Sens > 0.9006\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c620b336-0167-4f00-b47b-e4bc031a045d",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.loc[analysis[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36145380-d850-475c-a6fb-e129a2824e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.query(\"Spec > 0.9575\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e714d4af-4376-4e03-874e-0d4566d278a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Moxifloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "\n",
    "get_logReg_summary(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec5a07e-ae8a-4de2-911f-d4a9ad4e6446",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pza = pd.read_csv(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn/phenos.csv\")\n",
    "\n",
    "# the missing ones might be M. cannettii, most similar to L6 based on the other lineage callers\n",
    "lineages = pd.read_pickle(\"../data/combined_lineage_sample_IDs.pkl\")\n",
    "lineages[\"Lineage\"] = lineages[\"Lineage\"].fillna(\"6\")\n",
    "lineages[\"Lineage_1\"] = lineages[\"Lineage_1\"].fillna(\"6\")\n",
    "\n",
    "lineages = lineages[[\"Sample Name\", \"Sample ID\", \"Lineage_1\"]]\n",
    "lineages[\"Lineage\"] = [str(val).split(\".\")[0] for val in lineages[\"Lineage_1\"].values]\n",
    "lineages.loc[lineages[\"Lineage\"].str.contains(\"BOV\"), \"Lineage\"] = \"M. bovis\"\n",
    "\n",
    "assert len(lineages.loc[pd.isnull(lineages[\"Lineage\"])]) == 0\n",
    "\n",
    "########## KEEP ONLY ISOLATES WITH ALL 3 PIECES OF DATA ##########\n",
    "\n",
    "# get only isolates with data for everyting: SNP matrix, in the model, and lineages\n",
    "combined = lineages.merge(df_pza, left_on=\"Sample ID\", right_on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56d7652e-ec58-4e0a-88ef-86e80fcc025a",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.groupby(\"Lineage\")[\"phenotype\"].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3334fbf-76ee-49e0-a0b9-0dccff1727c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.query(\"Lineage == 'M. bovis'\").Lineage_1.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c225d9c-faf1-4141-884c-034c55fa03de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
