{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import sklearn\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "import glob, os, yaml, subprocess, itertools, sparse, pickle\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2439e0a-1ad9-4471-9e91-ac841f82fba2",
   "metadata": {},
   "source": [
    "# Combined Analysis Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "276583f5-2c83-4bf0-bdac-18b6add1c060",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Delamanid\"\n",
    "analysis = pd.read_csv(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/{drug}/final_analysis.csv\")\n",
    "\n",
    "# drug = \"Levofloxacin\"\n",
    "# lev_analysis = pd.read_csv(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/{drug}/final_analysis.csv\")\n",
    "\n",
    "# drug = \"Amikacin\"\n",
    "# ami_analysis = pd.read_csv(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/{drug}/final_analysis.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8727f82b-4186-41e4-a68a-38d4186a1a34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig_variant', 'coef', 'coef_LB', 'coef_UB', 'pval', 'BH_pval',\n",
       "       'Bonferroni_pval', 'genome_index', 'confidence_WHO_2021', 'Odds_Ratio',\n",
       "       'OR_LB', 'OR_UB', 'Tier1_only', 'WHO_phenos', 'poolLOF', 'Syn',\n",
       "       'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV', 'Sens',\n",
       "       'Spec', 'LR+', 'LR-', 'PPV_LB', 'PPV_UB', 'Sens_LB', 'Sens_UB',\n",
       "       'Spec_LB', 'Spec_UB', 'LR+_LB', 'LR+_UB', 'LR-_LB', 'LR-_UB'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cad46c13-f195-425a-8dbf-9dd51b0847e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_LB</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_p.Asp108fs</td>\n",
       "      <td>0.2241</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2830</td>\n",
       "      <td>0.0002</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.2512</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0269</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ddn_p.Val127Phe</td>\n",
       "      <td>0.1563</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.2306</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>0.1742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1692</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0272</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9730</td>\n",
       "      <td>1.0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ddn_p.Leu49Pro</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>0.1171</td>\n",
       "      <td>0.1539</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3986989.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.1476</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6154</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0215</td>\n",
       "      <td>0.0723</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>77.3261</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9278</td>\n",
       "      <td>0.9786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fbiB_p.Lys448Arg</td>\n",
       "      <td>0.1352</td>\n",
       "      <td>-0.0469</td>\n",
       "      <td>0.2926</td>\n",
       "      <td>0.0575</td>\n",
       "      <td>0.1203</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3642877.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1448</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0641</td>\n",
       "      <td>0.1633</td>\n",
       "      <td>0.0376</td>\n",
       "      <td>0.1051</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>0.9902</td>\n",
       "      <td>3.2542</td>\n",
       "      <td>8.9502</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.9734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ddn_p.Trp139*</td>\n",
       "      <td>0.1268</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.1689</td>\n",
       "      <td>0.0354</td>\n",
       "      <td>0.1061</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1352</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0272</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9728</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>fbiC_c.-32A&gt;G</td>\n",
       "      <td>-0.0645</td>\n",
       "      <td>-0.0986</td>\n",
       "      <td>-0.0207</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.0153</td>\n",
       "      <td>0.0306</td>\n",
       "      <td>1302899.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9376</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>938</th>\n",
       "      <td>fbiC_c.967C&gt;T</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>-0.0753</td>\n",
       "      <td>-0.0518</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9374</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>fbiB_p.Leu447Arg</td>\n",
       "      <td>-0.0649</td>\n",
       "      <td>-0.0790</td>\n",
       "      <td>-0.0497</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3642874.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>fgd1_c.465C&gt;T</td>\n",
       "      <td>-0.0783</td>\n",
       "      <td>-0.0907</td>\n",
       "      <td>-0.0649</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9247</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>PC0</td>\n",
       "      <td>-0.2318</td>\n",
       "      <td>-0.4011</td>\n",
       "      <td>-0.0807</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.0420</td>\n",
       "      <td>0.1682</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7931</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>942 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0      ddn_p.Asp108fs  0.2241   0.0000   0.2830 0.0002   0.0151   \n",
       "1     ddn_p.Val127Phe  0.1563  -0.0000   0.2306 0.0025   0.0348   \n",
       "2      ddn_p.Leu49Pro  0.1377   0.1171   0.1539 0.0000   0.0000   \n",
       "3    fbiB_p.Lys448Arg  0.1352  -0.0469   0.2926 0.0575   0.1203   \n",
       "4       ddn_p.Trp139*  0.1268  -0.0000   0.1689 0.0354   0.1061   \n",
       "..                ...     ...      ...      ...    ...      ...   \n",
       "937     fbiC_c.-32A>G -0.0645  -0.0986  -0.0207 0.0004   0.0153   \n",
       "938     fbiC_c.967C>T -0.0646  -0.0753  -0.0518 0.0000   0.0000   \n",
       "939  fbiB_p.Leu447Arg -0.0649  -0.0790  -0.0497 0.0000   0.0000   \n",
       "940     fgd1_c.465C>T -0.0783  -0.0907  -0.0649 0.0000   0.0000   \n",
       "941               PC0 -0.2318  -0.4011  -0.0807 0.0024   0.0420   \n",
       "\n",
       "     Bonferroni_pval  genome_index        confidence_WHO_2021  Odds_Ratio  \\\n",
       "0             0.0151           NaN                        NaN      1.2512   \n",
       "1             0.1742           NaN                        NaN      1.1692   \n",
       "2             0.0000  3986989.0000     2) Assoc w R - Interim      1.1476   \n",
       "3             1.0000  3642877.0000  3) Uncertain significance      1.1448   \n",
       "4             1.0000           NaN                        NaN      1.1352   \n",
       "..               ...           ...                        ...         ...   \n",
       "937           0.0306  1302899.0000  3) Uncertain significance      0.9376   \n",
       "938           0.0000           NaN                        NaN      0.9374   \n",
       "939           0.0000  3642874.0000  3) Uncertain significance      0.9371   \n",
       "940           0.0000           NaN                        NaN      0.9247   \n",
       "941           0.1682           NaN                        NaN      0.7931   \n",
       "\n",
       "     ...  PPV_LB  PPV_UB  Sens_LB  Sens_UB  Spec_LB  Spec_UB  LR+_LB  LR+_UB  \\\n",
       "0    ...  1.0000  1.0000   0.0000   0.0269   1.0000   1.0000     inf     inf   \n",
       "1    ...  0.0000  1.0000   0.0000   0.0272   0.9996   1.0000  0.0000     inf   \n",
       "2    ...  0.6154  1.0000   0.0215   0.0723   0.9996   1.0000 77.3261     inf   \n",
       "3    ...  0.0641  0.1633   0.0376   0.1051   0.9863   0.9902  3.2542  8.9502   \n",
       "4    ...  1.0000  1.0000   0.0000   0.0272   1.0000   1.0000     inf     inf   \n",
       "..   ...     ...     ...      ...      ...      ...      ...     ...     ...   \n",
       "937  ...     NaN     NaN      NaN      NaN      NaN      NaN     NaN     NaN   \n",
       "938  ...     NaN     NaN      NaN      NaN      NaN      NaN     NaN     NaN   \n",
       "939  ...     NaN     NaN      NaN      NaN      NaN      NaN     NaN     NaN   \n",
       "940  ...     NaN     NaN      NaN      NaN      NaN      NaN     NaN     NaN   \n",
       "941  ...     NaN     NaN      NaN      NaN      NaN      NaN     NaN     NaN   \n",
       "\n",
       "     LR-_LB  LR-_UB  \n",
       "0    0.9731  1.0000  \n",
       "1    0.9730  1.0002  \n",
       "2    0.9278  0.9786  \n",
       "3    0.9069  0.9734  \n",
       "4    0.9728  1.0000  \n",
       "..      ...     ...  \n",
       "937     NaN     NaN  \n",
       "938     NaN     NaN  \n",
       "939     NaN     NaN  \n",
       "940     NaN     NaN  \n",
       "941     NaN     NaN  \n",
       "\n",
       "[942 rows x 37 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0de2bb5d-b11d-44ba-ada3-97c1f8911e9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig_variant', 'coef', 'coef_LB', 'coef_UB', 'pval', 'BH_pval',\n",
       "       'Bonferroni_pval', 'genome_index', 'confidence_WHO_2021', 'Odds_Ratio',\n",
       "       'OR_LB', 'OR_UB', 'Tier1_only', 'WHO_phenos', 'poolLOF', 'Syn',\n",
       "       'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'PPV', 'Sens',\n",
       "       'Spec', 'LR+', 'LR-', 'PPV_LB', 'PPV_UB', 'Sens_LB', 'Sens_UB',\n",
       "       'Spec_LB', 'Spec_UB', 'LR+_LB', 'LR+_UB', 'LR-_LB', 'LR-_UB'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70995fda-8a44-4e8b-9be0-adf2d0350584",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8976de2c-05bf-4f27-8fc4-b8a19afef479",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_LB</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ddn_p.Leu49Pro</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>0.1171</td>\n",
       "      <td>0.1539</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3986989.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.1476</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6154</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0215</td>\n",
       "      <td>0.0723</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>77.3261</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9278</td>\n",
       "      <td>0.9786</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      orig_variant   coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "69  ddn_p.Leu49Pro 0.1377   0.1171   0.1539 0.0000   0.0000           0.0000   \n",
       "\n",
       "    genome_index     confidence_WHO_2021  Odds_Ratio  ...  PPV_LB  PPV_UB  \\\n",
       "69  3986989.0000  2) Assoc w R - Interim      1.1476  ...  0.6154  1.0000   \n",
       "\n",
       "    Sens_LB  Sens_UB  Spec_LB  Spec_UB  LR+_LB  LR+_UB  LR-_LB  LR-_UB  \n",
       "69   0.0215   0.0723   0.9996   1.0000 77.3261     inf  0.9278  0.9786  \n",
       "\n",
       "[1 rows x 37 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.query(\"confidence_WHO_2021=='2) Assoc w R - Interim'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ac6d6cbc-bcb9-463a-a223-938fc93c2456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>drug</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence</th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5248</th>\n",
       "      <td>3804</td>\n",
       "      <td>DLM</td>\n",
       "      <td>3986989</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>ddn</td>\n",
       "      <td>ddn_L49P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 drug genome_index              confidence gene   variant\n",
       "5248        3804  DLM      3986989  2) Assoc w R - Interim  ddn  ddn_L49P"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who_variants.query(\"drug=='DLM' & confidence=='2) Assoc w R - Interim'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cef39b9-b447-4949-8482-8bac3b0bea2d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "87719756-60a3-45fd-bb29-9a12d3b5a8b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['3) Uncertain significance', '2) Assoc w R - Interim'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who_variants.query(\"drug=='DLM'\").confidence.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a49372b1-e991-48e1-a8ac-6da194b36f71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "      <th>Tier1_only</th>\n",
       "      <th>WHO_phenos</th>\n",
       "      <th>poolLOF</th>\n",
       "      <th>Syn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>Rv2752c_c.675C&gt;T</td>\n",
       "      <td>0.1101</td>\n",
       "      <td>0.0470</td>\n",
       "      <td>0.1568</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1164</td>\n",
       "      <td>1.0481</td>\n",
       "      <td>1.1698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>glpK_lof</td>\n",
       "      <td>0.1000</td>\n",
       "      <td>0.0384</td>\n",
       "      <td>0.1619</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0085</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1052</td>\n",
       "      <td>1.0392</td>\n",
       "      <td>1.1758</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>gyrA_c.2022C&gt;G</td>\n",
       "      <td>0.0618</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0726</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0637</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0753</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>gyrA_p.Arg392Cys</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.0037</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0634</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0764</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>glpK_c.117G&gt;C</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>-0.0284</td>\n",
       "      <td>0.0948</td>\n",
       "      <td>0.0310</td>\n",
       "      <td>0.0954</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0621</td>\n",
       "      <td>0.9720</td>\n",
       "      <td>1.0995</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2519</th>\n",
       "      <td>glpK_p.Leu228Val</td>\n",
       "      <td>-0.0906</td>\n",
       "      <td>-0.1457</td>\n",
       "      <td>-0.0385</td>\n",
       "      <td>0.0018</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9134</td>\n",
       "      <td>0.8644</td>\n",
       "      <td>0.9622</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2520</th>\n",
       "      <td>Rv2477c_c.1545C&gt;T</td>\n",
       "      <td>-0.1315</td>\n",
       "      <td>-0.2418</td>\n",
       "      <td>-0.0283</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>0.0421</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8767</td>\n",
       "      <td>0.7852</td>\n",
       "      <td>0.9721</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2521</th>\n",
       "      <td>Rv2477c_p.Gly41Glu</td>\n",
       "      <td>-0.1405</td>\n",
       "      <td>-0.2298</td>\n",
       "      <td>-0.0660</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0057</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8689</td>\n",
       "      <td>0.7947</td>\n",
       "      <td>0.9361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2522</th>\n",
       "      <td>glpK_c.1134C&gt;T</td>\n",
       "      <td>-0.1410</td>\n",
       "      <td>-0.2736</td>\n",
       "      <td>-0.0166</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0749</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>0.7606</td>\n",
       "      <td>0.9836</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2523</th>\n",
       "      <td>gyrB_c.1983C&gt;T</td>\n",
       "      <td>-0.1429</td>\n",
       "      <td>-0.2816</td>\n",
       "      <td>-0.0277</td>\n",
       "      <td>0.0197</td>\n",
       "      <td>0.0782</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8668</td>\n",
       "      <td>0.7546</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2154 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "370     Rv2752c_c.675C>T  0.1101   0.0470   0.1568 0.0000   0.0009   \n",
       "371             glpK_lof  0.1000   0.0384   0.1619 0.0008   0.0085   \n",
       "372       gyrA_c.2022C>G  0.0618   0.0000   0.0726 0.0063   0.0339   \n",
       "373     gyrA_p.Arg392Cys  0.0615   0.0000   0.0737 0.0037   0.0278   \n",
       "374        glpK_c.117G>C  0.0603  -0.0284   0.0948 0.0310   0.0954   \n",
       "...                  ...     ...      ...      ...    ...      ...   \n",
       "2519    glpK_p.Leu228Val -0.0906  -0.1457  -0.0385 0.0018   0.0170   \n",
       "2520   Rv2477c_c.1545C>T -0.1315  -0.2418  -0.0283 0.0088   0.0421   \n",
       "2521  Rv2477c_p.Gly41Glu -0.1405  -0.2298  -0.0660 0.0005   0.0057   \n",
       "2522      glpK_c.1134C>T -0.1410  -0.2736  -0.0166 0.0185   0.0749   \n",
       "2523      gyrB_c.1983C>T -0.1429  -0.2816  -0.0277 0.0197   0.0782   \n",
       "\n",
       "      Bonferroni_pval genome_index confidence_WHO_2021  Odds_Ratio  OR_LB  \\\n",
       "370            0.1172          NaN                 NaN      1.1164 1.0481   \n",
       "371            1.0000          NaN                 NaN      1.1052 1.0392   \n",
       "372            1.0000          NaN                 NaN      1.0637 1.0000   \n",
       "373            1.0000          NaN                 NaN      1.0634 1.0000   \n",
       "374            1.0000          NaN                 NaN      1.0621 0.9720   \n",
       "...               ...          ...                 ...         ...    ...   \n",
       "2519           1.0000          NaN                 NaN      0.9134 0.8644   \n",
       "2520           1.0000          NaN                 NaN      0.8767 0.7852   \n",
       "2521           1.0000          NaN                 NaN      0.8689 0.7947   \n",
       "2522           1.0000          NaN                 NaN      0.8685 0.7606   \n",
       "2523           1.0000          NaN                 NaN      0.8668 0.7546   \n",
       "\n",
       "      OR_UB  Tier1_only  WHO_phenos  poolLOF  Syn  \n",
       "370  1.1698           0           0        1    1  \n",
       "371  1.1758           0           0        1    1  \n",
       "372  1.0753           0           0        1    1  \n",
       "373  1.0764           0           0        1    1  \n",
       "374  1.0995           0           0        1    1  \n",
       "...     ...         ...         ...      ...  ...  \n",
       "2519 0.9622           0           0        1    1  \n",
       "2520 0.9721           0           0        1    1  \n",
       "2521 0.9361           0           0        1    1  \n",
       "2522 0.9836           0           0        1    1  \n",
       "2523 0.9726           0           0        1    1  \n",
       "\n",
       "[2154 rows x 16 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxf_analysis.query(\"Tier1_only==0 & WHO_phenos==0 & Syn==1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a683ba2b-ca3d-4b46-b3fb-eff707fe4cf5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e3314b3a-9e9c-4519-a86c-43cd70a994f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    2271\n",
       "1     406\n",
       "Name: Tier1_only, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxf_analysis.Tier1_only.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d5bc2b2e-fd28-48c2-9e9e-46313074a922",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    2307\n",
       "1     370\n",
       "Name: WHO_phenos, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxf_analysis.WHO_phenos.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "56536612-1417-4c1f-8463-42afd2d93e1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    2524\n",
       "0     153\n",
       "Name: poolLOF, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxf_analysis.poolLOF.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c5324e7a-e29b-453d-b36e-9511cbbcc1d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    2176\n",
       "0     501\n",
       "Name: Syn, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mxf_analysis.Syn.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdcae08c-4bc9-4527-8e24-c0c3922beb1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aed4a42c-86c9-4a43-9382-7f9723720ada",
   "metadata": {},
   "source": [
    "# Catalog-Based Method Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "50f78280-0749-4fe9-8cd9-079fa016922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_logReg_summary(out_dir):\n",
    "    \n",
    "    model_summary = pd.read_csv(os.path.join(out_dir, \"logReg_summary.csv\"))\n",
    "    model_analysis = pd.read_csv(os.path.join(out_dir, \"model_analysis.csv\"))\n",
    "    \n",
    "    summary_sens = model_summary.loc[0, \"Sens\"]\n",
    "    summary_spec = model_summary.loc[0, \"Spec\"]\n",
    "    \n",
    "    #max_idx = [np.argmax(model_analysis[\"Sens\"]), np.argmax(model_analysis[\"Spec\"])]\n",
    "    \n",
    "    better_variant = model_analysis.query(\"Sens >= @summary_sens & Spec >= @summary_spec\")\n",
    "    \n",
    "    if len(better_variant) == 0:\n",
    "        print(\"No variants have comparable sensitivity AND specificity\")\n",
    "    else:\n",
    "        print(better_variant)\n",
    "    return model_summary, model_analysis\n",
    "    \n",
    "    #return model_summary, model_analysis.iloc[max_idx, :][[\"orig_variant\", \"coef\", \"Sens\", \"Spec\", \"accuracy\", \"balanced_accuracy\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "20dfc9fa-8f4b-4e2d-81b2-489754985444",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No variants have comparable sensitivity AND specificity\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.8336</td>\n",
       "      <td>0.9084</td>\n",
       "      <td>0.8951</td>\n",
       "      <td>0.8710</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sens   Spec  accuracy  balanced_accuracy\n",
       "0 0.8336 0.9084    0.8951             0.8710"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Levofloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn\"\n",
    "\n",
    "summary, analysis = get_logReg_summary(out_dir)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "63ef135c-6253-4a64-9053-a08410585184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [orig_variant, coef, coef_LB, coef_UB, pval, BH_pval, Bonferroni_pval, genome_index, confidence_WHO_2021, Odds_Ratio, OR_LB, OR_UB, Num_Isolates, Total_Isolates, TP, FP, TN, FN, PPV, Sens, Spec, LR+, LR-, PPV_LB, PPV_UB, Sens_LB, Sens_UB, Spec_LB, Spec_UB, LR+_LB, LR+_UB, LR-_LB, LR-_UB, accuracy, balanced_accuracy]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 35 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.query(\"Sens > 0.9006\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c620b336-0167-4f00-b47b-e4bc031a045d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>PC2</td>\n",
       "      <td>0.1660</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.3034</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0301</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1806</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>PC0</td>\n",
       "      <td>0.1346</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.2630</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>0.0492</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1440</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   orig_variant   coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "35          PC2 0.1660   0.0355   0.3034 0.0078   0.0301           1.0000   \n",
       "57          PC0 0.1346   0.0166   0.2630 0.0140   0.0492           1.0000   \n",
       "\n",
       "    genome_index confidence_WHO_2021  Odds_Ratio  ...  Sens_LB  Sens_UB  \\\n",
       "35           NaN                 NaN      1.1806  ...      NaN      NaN   \n",
       "57           NaN                 NaN      1.1440  ...      NaN      NaN   \n",
       "\n",
       "    Spec_LB  Spec_UB  LR+_LB  LR+_UB  LR-_LB  LR-_UB  accuracy  \\\n",
       "35      NaN      NaN     NaN     NaN     NaN     NaN       NaN   \n",
       "57      NaN      NaN     NaN     NaN     NaN     NaN       NaN   \n",
       "\n",
       "    balanced_accuracy  \n",
       "35                NaN  \n",
       "57                NaN  \n",
       "\n",
       "[2 rows x 35 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.loc[analysis[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "36145380-d850-475c-a6fb-e129a2824e27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pncA_c.-11A&gt;G</td>\n",
       "      <td>0.4088</td>\n",
       "      <td>0.3700</td>\n",
       "      <td>0.4467</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2289252.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.5050</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.0395</td>\n",
       "      <td>0.9993</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>44.4364</td>\n",
       "      <td>284.1349</td>\n",
       "      <td>0.9608</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>0.8278</td>\n",
       "      <td>0.5163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pncA_p.His51Asp</td>\n",
       "      <td>0.3308</td>\n",
       "      <td>0.3054</td>\n",
       "      <td>0.3591</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2289091.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.3921</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.0177</td>\n",
       "      <td>0.9998</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>63.1368</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9823</td>\n",
       "      <td>0.9905</td>\n",
       "      <td>0.8246</td>\n",
       "      <td>0.5067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pncA_p.Val131fs</td>\n",
       "      <td>0.2900</td>\n",
       "      <td>0.2684</td>\n",
       "      <td>0.3154</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0090</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.9998</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>54.4333</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9836</td>\n",
       "      <td>0.9912</td>\n",
       "      <td>0.8245</td>\n",
       "      <td>0.5064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pncA_p.Gln141Pro</td>\n",
       "      <td>0.2869</td>\n",
       "      <td>0.2529</td>\n",
       "      <td>0.3273</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2288820.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.3323</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.0243</td>\n",
       "      <td>0.9995</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>35.3919</td>\n",
       "      <td>344.8754</td>\n",
       "      <td>0.9759</td>\n",
       "      <td>0.9851</td>\n",
       "      <td>0.8256</td>\n",
       "      <td>0.5097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pncA_p.Leu172Pro</td>\n",
       "      <td>0.2779</td>\n",
       "      <td>0.2524</td>\n",
       "      <td>0.3046</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2288727.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.3204</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0129</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9871</td>\n",
       "      <td>0.9938</td>\n",
       "      <td>0.8239</td>\n",
       "      <td>0.5046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>rpsA_p.Ile70Leu</td>\n",
       "      <td>-0.1249</td>\n",
       "      <td>-0.1356</td>\n",
       "      <td>-0.1120</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9966</td>\n",
       "      <td>0.9982</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0018</td>\n",
       "      <td>1.0034</td>\n",
       "      <td>0.1798</td>\n",
       "      <td>0.5013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>855</th>\n",
       "      <td>PPE35_p.Pro670Leu</td>\n",
       "      <td>-0.1287</td>\n",
       "      <td>-0.1817</td>\n",
       "      <td>-0.0860</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>2168604.0000</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.8792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.9800</td>\n",
       "      <td>0.9842</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>1.0145</td>\n",
       "      <td>1.0196</td>\n",
       "      <td>0.1922</td>\n",
       "      <td>0.5084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>856</th>\n",
       "      <td>Rv1258c_p.Gly363Val</td>\n",
       "      <td>-0.1288</td>\n",
       "      <td>-0.1399</td>\n",
       "      <td>-0.1158</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1406253.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.8791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9965</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0019</td>\n",
       "      <td>1.0036</td>\n",
       "      <td>0.1799</td>\n",
       "      <td>0.5013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>Rv3236c_c.546C&gt;T</td>\n",
       "      <td>-0.1577</td>\n",
       "      <td>-0.1859</td>\n",
       "      <td>-0.1337</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9916</td>\n",
       "      <td>0.9942</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0058</td>\n",
       "      <td>1.0085</td>\n",
       "      <td>0.1835</td>\n",
       "      <td>0.5035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>Rv1258c_c.1029T&gt;C</td>\n",
       "      <td>-0.1768</td>\n",
       "      <td>-0.2766</td>\n",
       "      <td>-0.0781</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8379</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>0.9775</td>\n",
       "      <td>0.1052</td>\n",
       "      <td>0.2940</td>\n",
       "      <td>1.0170</td>\n",
       "      <td>1.0244</td>\n",
       "      <td>0.1974</td>\n",
       "      <td>0.5101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>856 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0          pncA_c.-11A>G  0.4088   0.3700   0.4467 0.0000   0.0000   \n",
       "1        pncA_p.His51Asp  0.3308   0.3054   0.3591 0.0000   0.0000   \n",
       "2        pncA_p.Val131fs  0.2900   0.2684   0.3154 0.0000   0.0000   \n",
       "3       pncA_p.Gln141Pro  0.2869   0.2529   0.3273 0.0000   0.0000   \n",
       "4       pncA_p.Leu172Pro  0.2779   0.2524   0.3046 0.0000   0.0000   \n",
       "..                   ...     ...      ...      ...    ...      ...   \n",
       "854      rpsA_p.Ile70Leu -0.1249  -0.1356  -0.1120 0.0000   0.0000   \n",
       "855    PPE35_p.Pro670Leu -0.1287  -0.1817  -0.0860 0.0000   0.0000   \n",
       "856  Rv1258c_p.Gly363Val -0.1288  -0.1399  -0.1158 0.0000   0.0000   \n",
       "857     Rv3236c_c.546C>T -0.1577  -0.1859  -0.1337 0.0000   0.0000   \n",
       "858    Rv1258c_c.1029T>C -0.1768  -0.2766  -0.0781 0.0004   0.0027   \n",
       "\n",
       "     Bonferroni_pval  genome_index        confidence_WHO_2021  Odds_Ratio  \\\n",
       "0             0.0000  2289252.0000               1) Assoc w R      1.5050   \n",
       "1             0.0000  2289091.0000               1) Assoc w R      1.3921   \n",
       "2             0.0000           NaN                        NaN      1.3365   \n",
       "3             0.0000  2288820.0000               1) Assoc w R      1.3323   \n",
       "4             0.0000  2288727.0000               1) Assoc w R      1.3204   \n",
       "..               ...           ...                        ...         ...   \n",
       "854           0.0000           NaN                        NaN      0.8826   \n",
       "855           0.0009  2168604.0000           5) Not assoc w R      0.8792   \n",
       "856           0.0000  1406253.0000  3) Uncertain significance      0.8791   \n",
       "857           0.0000           NaN                        NaN      0.8541   \n",
       "858           1.0000           NaN                        NaN      0.8379   \n",
       "\n",
       "     ...  Sens_LB  Sens_UB  Spec_LB  Spec_UB  LR+_LB   LR+_UB  LR-_LB  LR-_UB  \\\n",
       "0    ...   0.0274   0.0395   0.9993   0.9999 44.4364 284.1349  0.9608  0.9731   \n",
       "1    ...   0.0095   0.0177   0.9998   1.0000 63.1368      inf  0.9823  0.9905   \n",
       "2    ...   0.0090   0.0165   0.9998   1.0000 54.4333      inf  0.9836  0.9912   \n",
       "3    ...   0.0151   0.0243   0.9995   0.9999 35.3919 344.8754  0.9759  0.9851   \n",
       "4    ...   0.0062   0.0129   1.0000   1.0000     inf      inf  0.9871  0.9938   \n",
       "..   ...      ...      ...      ...      ...     ...      ...     ...     ...   \n",
       "854  ...   0.0000   0.0000   0.9966   0.9982  0.0000   0.0000  1.0018  1.0034   \n",
       "855  ...   0.0003   0.0024   0.9800   0.9842  0.0156   0.1377  1.0145  1.0196   \n",
       "856  ...   0.0000   0.0000   0.9965   0.9981  0.0000   0.0000  1.0019  1.0036   \n",
       "857  ...   0.0000   0.0000   0.9916   0.9942  0.0000   0.0000  1.0058  1.0085   \n",
       "858  ...   0.0027   0.0073   0.9726   0.9775  0.1052   0.2940  1.0170  1.0244   \n",
       "\n",
       "     accuracy  balanced_accuracy  \n",
       "0      0.8278             0.5163  \n",
       "1      0.8246             0.5067  \n",
       "2      0.8245             0.5064  \n",
       "3      0.8256             0.5097  \n",
       "4      0.8239             0.5046  \n",
       "..        ...                ...  \n",
       "854    0.1798             0.5013  \n",
       "855    0.1922             0.5084  \n",
       "856    0.1799             0.5013  \n",
       "857    0.1835             0.5035  \n",
       "858    0.1974             0.5101  \n",
       "\n",
       "[856 rows x 35 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis.query(\"Spec > 0.9575\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e714d4af-4376-4e03-874e-0d4566d278a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No variants have comparable sensitivity AND specificity\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sens</th>\n",
       "      <th>Spec</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>balanced_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.9230</td>\n",
       "      <td>0.8945</td>\n",
       "      <td>0.8983</td>\n",
       "      <td>0.9088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sens   Spec  accuracy  balanced_accuracy\n",
       "0 0.9230 0.8945    0.8983             0.9088"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Moxifloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "\n",
    "get_logReg_summary(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9ec5a07e-ae8a-4de2-911f-d4a9ad4e6446",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pza = pd.read_csv(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn/phenos.csv\")\n",
    "\n",
    "# the missing ones might be M. cannettii, most similar to L6 based on the other lineage callers\n",
    "lineages = pd.read_pickle(\"../data/combined_lineage_sample_IDs.pkl\")\n",
    "lineages[\"Lineage\"] = lineages[\"Lineage\"].fillna(\"6\")\n",
    "lineages[\"Lineage_1\"] = lineages[\"Lineage_1\"].fillna(\"6\")\n",
    "\n",
    "lineages = lineages[[\"Sample Name\", \"Sample ID\", \"Lineage_1\"]]\n",
    "lineages[\"Lineage\"] = [str(val).split(\".\")[0] for val in lineages[\"Lineage_1\"].values]\n",
    "lineages.loc[lineages[\"Lineage\"].str.contains(\"BOV\"), \"Lineage\"] = \"M. bovis\"\n",
    "\n",
    "assert len(lineages.loc[pd.isnull(lineages[\"Lineage\"])]) == 0\n",
    "\n",
    "########## KEEP ONLY ISOLATES WITH ALL 3 PIECES OF DATA ##########\n",
    "\n",
    "# get only isolates with data for everyting: SNP matrix, in the model, and lineages\n",
    "combined = lineages.merge(df_pza, left_on=\"Sample ID\", right_on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "56d7652e-ec58-4e0a-88ef-86e80fcc025a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Lineage\n",
       "M. bovis   0.9153\n",
       "2          0.3750\n",
       "1          0.1503\n",
       "6          0.1163\n",
       "4          0.1112\n",
       "5          0.1111\n",
       "3          0.0554\n",
       "7          0.0000\n",
       "Name: phenotype, dtype: float64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.groupby(\"Lineage\")[\"phenotype\"].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f3334fbf-76ee-49e0-a0b9-0dccff1727c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BOV_AFRI', 'BOV'], dtype=object)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.query(\"Lineage == 'M. bovis'\").Lineage_1.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c225d9c-faf1-4141-884c-034c55fa03de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
