{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import sklearn\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "import glob, os, yaml, subprocess, itertools, sparse, pickle\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)\n",
    "samples_summary = pd.read_csv(\"../data/samples_summary.csv\")\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/Sanjana/who-mutation-catalogue'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6a5a6a1-d7ff-4e89-8765-c19c0311edb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Delamanid\"\n",
    "# df_phenos = pd.read_csv(os.path.join(analysis_dir, drug, \"phenos_binary.csv\"))\n",
    "\n",
    "# df_genos = pd.read_csv(os.path.join(analysis_dir, drug, \"genos.csv.gz\"), compression=\"gzip\")\n",
    "# df_genos[\"orig_variant\"] = df_genos[\"resolved_symbol\"] + \"_\" + df_genos[\"variant_category\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "f5c768d6-f93f-4690-adf8-df3d4589f43f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# final_analysis file with all significant variants for a drug\n",
    "res_df = pd.read_csv(os.path.join(analysis_dir, drug, \"final_analysis.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "a81d2859-e450-4fda-b9d6-a28b310afa84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>predicted_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ddn_lof</td>\n",
       "      <td>0.2951</td>\n",
       "      <td>0.1698</td>\n",
       "      <td>0.3745</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3432</td>\n",
       "      <td>1.1851</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>0.0446</td>\n",
       "      <td>0.1092</td>\n",
       "      <td>0.9991</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>73.3231</td>\n",
       "      <td>852.3245</td>\n",
       "      <td>0.8908</td>\n",
       "      <td>0.9559</td>\n",
       "      <td>lof</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fgd1_lof</td>\n",
       "      <td>0.1563</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.2080</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0332</td>\n",
       "      <td>0.2990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1692</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0194</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9806</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>lof</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ddn_p.Val127Phe</td>\n",
       "      <td>0.1547</td>\n",
       "      <td>-0.0273</td>\n",
       "      <td>0.2292</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0299</td>\n",
       "      <td>0.1792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0263</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9739</td>\n",
       "      <td>1.0002</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ddn_p.Leu49Pro</td>\n",
       "      <td>0.1377</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>0.1530</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.1476</td>\n",
       "      <td>1.1241</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.0723</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>75.1657</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.9278</td>\n",
       "      <td>0.9792</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fbiB_p.Lys448Arg</td>\n",
       "      <td>0.1318</td>\n",
       "      <td>-0.0342</td>\n",
       "      <td>0.2785</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>0.1189</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1409</td>\n",
       "      <td>0.9664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1667</td>\n",
       "      <td>0.0410</td>\n",
       "      <td>0.1087</td>\n",
       "      <td>0.9853</td>\n",
       "      <td>0.9895</td>\n",
       "      <td>3.2246</td>\n",
       "      <td>9.1374</td>\n",
       "      <td>0.9028</td>\n",
       "      <td>0.9713</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>fbiB_p.Leu447Arg</td>\n",
       "      <td>-0.0649</td>\n",
       "      <td>-0.0788</td>\n",
       "      <td>-0.0504</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9371</td>\n",
       "      <td>0.9242</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>fgd1_c.465C&gt;T</td>\n",
       "      <td>-0.0783</td>\n",
       "      <td>-0.0901</td>\n",
       "      <td>-0.0657</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9247</td>\n",
       "      <td>0.9139</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>fbiA_p.Arg304Gln</td>\n",
       "      <td>-0.0799</td>\n",
       "      <td>-0.1325</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0480</td>\n",
       "      <td>0.1129</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.8759</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>ndh_lof</td>\n",
       "      <td>-0.1093</td>\n",
       "      <td>-0.1496</td>\n",
       "      <td>-0.0602</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8964</td>\n",
       "      <td>0.8610</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lof</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>PC0</td>\n",
       "      <td>-0.2291</td>\n",
       "      <td>-0.4017</td>\n",
       "      <td>-0.0775</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.1919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7953</td>\n",
       "      <td>0.6692</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>958 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0             ddn_lof  0.2951   0.1698   0.3745 0.0000   0.0000   \n",
       "1            fgd1_lof  0.1563  -0.0000   0.2080 0.0041   0.0332   \n",
       "2     ddn_p.Val127Phe  0.1547  -0.0273   0.2292 0.0025   0.0299   \n",
       "3      ddn_p.Leu49Pro  0.1377   0.1170   0.1530 0.0000   0.0000   \n",
       "4    fbiB_p.Lys448Arg  0.1318  -0.0342   0.2785 0.0537   0.1189   \n",
       "..                ...     ...      ...      ...    ...      ...   \n",
       "953  fbiB_p.Leu447Arg -0.0649  -0.0788  -0.0504 0.0000   0.0000   \n",
       "954     fgd1_c.465C>T -0.0783  -0.0901  -0.0657 0.0000   0.0000   \n",
       "955  fbiA_p.Arg304Gln -0.0799  -0.1325   0.0000 0.0480   0.1129   \n",
       "956           ndh_lof -0.1093  -0.1496  -0.0602 0.0000   0.0000   \n",
       "957               PC0 -0.2291  -0.4017  -0.0775 0.0026   0.0274   \n",
       "\n",
       "     Bonferroni_pval        confidence_WHO_2021  Odds_Ratio  OR_LB  ...  \\\n",
       "0             0.0000                        NaN      1.3432 1.1851  ...   \n",
       "1             0.2990                        NaN      1.1692 1.0000  ...   \n",
       "2             0.1792                        NaN      1.1673 0.9731  ...   \n",
       "3             0.0000     2) Assoc w R - Interim      1.1476 1.1241  ...   \n",
       "4             1.0000  3) Uncertain significance      1.1409 0.9664  ...   \n",
       "..               ...                        ...         ...    ...  ...   \n",
       "953           0.0000  3) Uncertain significance      0.9371 0.9242  ...   \n",
       "954           0.0000                        NaN      0.9247 0.9139  ...   \n",
       "955           1.0000  3) Uncertain significance      0.9232 0.8759  ...   \n",
       "956           0.0001                        NaN      0.8964 0.8610  ...   \n",
       "957           0.1919                        NaN      0.7953 0.6692  ...   \n",
       "\n",
       "     PPV_UB  Sens_LB Sens_UB  Spec_LB  Spec_UB  LR+_LB   LR+_UB  LR-_LB  \\\n",
       "0    0.9500   0.0446  0.1092   0.9991   0.9999 73.3231 852.3245  0.8908   \n",
       "1    1.0000   0.0000  0.0194   1.0000   1.0000     inf      inf  0.9806   \n",
       "2    1.0000   0.0000  0.0263   0.9996   1.0000  0.0000      inf  0.9739   \n",
       "3    1.0000   0.0210  0.0723   0.9996   1.0000 75.1657      inf  0.9278   \n",
       "4    0.1667   0.0410  0.1087   0.9853   0.9895  3.2246   9.1374  0.9028   \n",
       "..      ...      ...     ...      ...      ...     ...      ...     ...   \n",
       "953     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "954     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "955     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "956     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "957     NaN      NaN     NaN      NaN      NaN     NaN      NaN     NaN   \n",
       "\n",
       "     LR-_UB    predicted_effect  \n",
       "0    0.9559                 lof  \n",
       "1    1.0000                 lof  \n",
       "2    1.0002    missense_variant  \n",
       "3    0.9792    missense_variant  \n",
       "4    0.9713    missense_variant  \n",
       "..      ...                 ...  \n",
       "953     NaN    missense_variant  \n",
       "954     NaN  synonymous_variant  \n",
       "955     NaN    missense_variant  \n",
       "956     NaN                 lof  \n",
       "957     NaN                 NaN  \n",
       "\n",
       "[958 rows x 36 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "483a2c3d-6d83-4c8b-ae72-98e8b22ea6eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig_variant', 'coef', 'coef_LB', 'coef_UB', 'pval', 'BH_pval',\n",
       "       'Bonferroni_pval', 'confidence_WHO_2021', 'Odds_Ratio', 'OR_LB',\n",
       "       'OR_UB', 'Tier', 'Phenos', 'unpooled', 'Num_Isolates', 'Total_Isolates',\n",
       "       'TP', 'FP', 'TN', 'FN', 'PPV', 'Sens', 'Spec', 'LR+', 'LR-', 'PPV_LB',\n",
       "       'PPV_UB', 'Sens_LB', 'Sens_UB', 'Spec_LB', 'Spec_UB', 'LR+_LB',\n",
       "       'LR+_UB', 'LR-_LB', 'LR-_UB', 'predicted_effect'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "736c0f79-be89-4c2b-9e5e-97686afdc3d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>...</th>\n",
       "      <th>PPV_UB</th>\n",
       "      <th>Sens_LB</th>\n",
       "      <th>Sens_UB</th>\n",
       "      <th>Spec_LB</th>\n",
       "      <th>Spec_UB</th>\n",
       "      <th>LR+_LB</th>\n",
       "      <th>LR+_UB</th>\n",
       "      <th>LR-_LB</th>\n",
       "      <th>LR-_UB</th>\n",
       "      <th>predicted_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PC4</td>\n",
       "      <td>0.1016</td>\n",
       "      <td>-0.0728</td>\n",
       "      <td>0.2469</td>\n",
       "      <td>0.1077</td>\n",
       "      <td>0.1228</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1069</td>\n",
       "      <td>0.9298</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PC2</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>-0.1190</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2571</td>\n",
       "      <td>0.2844</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0624</td>\n",
       "      <td>0.8878</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>PC3</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>-0.1110</td>\n",
       "      <td>0.1728</td>\n",
       "      <td>0.3431</td>\n",
       "      <td>0.3738</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0305</td>\n",
       "      <td>0.8950</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>PC1</td>\n",
       "      <td>-0.0039</td>\n",
       "      <td>-0.1845</td>\n",
       "      <td>0.1639</td>\n",
       "      <td>0.4829</td>\n",
       "      <td>0.5109</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.8316</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>PC0</td>\n",
       "      <td>-0.2291</td>\n",
       "      <td>-0.4017</td>\n",
       "      <td>-0.0775</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.1919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7953</td>\n",
       "      <td>0.6692</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "11           PC4  0.1016  -0.0728   0.2469 0.1077   0.1228           1.0000   \n",
       "18           PC2  0.0605  -0.1190   0.2391 0.2571   0.2844           1.0000   \n",
       "71           PC3  0.0300  -0.1110   0.1728 0.3431   0.3738           1.0000   \n",
       "88           PC1 -0.0039  -0.1845   0.1639 0.4829   0.5109           1.0000   \n",
       "957          PC0 -0.2291  -0.4017  -0.0775 0.0026   0.0274           0.1919   \n",
       "\n",
       "    confidence_WHO_2021  Odds_Ratio  OR_LB  ...  PPV_UB  Sens_LB Sens_UB  \\\n",
       "11                  NaN      1.1069 0.9298  ...     NaN      NaN     NaN   \n",
       "18                  NaN      1.0624 0.8878  ...     NaN      NaN     NaN   \n",
       "71                  NaN      1.0305 0.8950  ...     NaN      NaN     NaN   \n",
       "88                  NaN      0.9961 0.8316  ...     NaN      NaN     NaN   \n",
       "957                 NaN      0.7953 0.6692  ...     NaN      NaN     NaN   \n",
       "\n",
       "     Spec_LB  Spec_UB  LR+_LB  LR+_UB  LR-_LB  LR-_UB  predicted_effect  \n",
       "11       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "18       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "71       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "88       NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "957      NaN      NaN     NaN     NaN     NaN     NaN               NaN  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df.loc[pd.isnull(res_df[\"Sens\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95c39875-7ebe-4f20-aa99-3a68e3b694b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a2439e0a-1ad9-4471-9e91-ac841f82fba2",
   "metadata": {},
   "source": [
    "# Combined Analysis Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00729d49-7aeb-457e-af65-74a6bf488b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def final_processing(drug):\n",
    "    '''\n",
    "    Functions for processing outputs before sending to everyone else.\n",
    "    \n",
    "    1. Remove principal components (will describe them separately)\n",
    "    2. Add LOF to the predicted_effect column for pooled LOF mutations\n",
    "    3. Remove genome_index column (should actually do that earlier, but will fix later)\n",
    "    4. Remove the logistic regression coefficient columns (they will prefer to work with odds ratios)\n",
    "    5. Any other column renaming or dropping for clarity\n",
    "    '''\n",
    "    \n",
    "    analysis_df = pd.read_csv(f\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/{drug}/final_analysis.csv\")\n",
    "    analysis_df.rename(columns={\"orig_variant\": \"mutation\", \"Tier1_only\": \"Tier\", \"WHO_phenos\": \"Phenos\"}, inplace=True)\n",
    "    \n",
    "    # remove logReg coefficients. Keep only odds ratios. Remove the other two columns, which were present mainly for me to see\n",
    "    # if we were picking up many mutations that were in the 2021 mutation catalog\n",
    "    del analysis_df[\"genome_index\"]\n",
    "    del analysis_df[\"confidence_WHO_2021\"]\n",
    "    analysis_df = analysis_df[analysis_df.columns[~analysis_df.columns.str.contains(\"coef\")]]\n",
    "    \n",
    "    # remove significant principal components and replace the NaNs in the predicted effect column for the gene loss of functions\n",
    "    analysis_df = analysis_df.loc[~analysis_df[\"mutation\"].str.contains(\"PC\")]\n",
    "    analysis_df.loc[analysis_df[\"mutation\"].str.contains(\"lof\"), \"predicted_effect\"] = \"LOF\"\n",
    "    \n",
    "    # predicted effect should not be NaN for anything. position is NaN only for the pooled LOF mutations\n",
    "    assert len(analysis_df.loc[pd.isnull(analysis_df[\"predicted_effect\"])]) == 0\n",
    "    assert len(analysis_df.loc[(~analysis_df[\"mutation\"].str.contains(\"lof\")) & (pd.isnull(analysis_df[\"position\"]))]) == 0\n",
    "    \n",
    "    analysis_df[\"Tier\"] = analysis_df[\"Tier\"].map({1: 1, 0: 2})\n",
    "    analysis_df[\"Phenos\"] = analysis_df[\"Phenos\"].map({1: \"WHO\", 0: \"ALL\"})\n",
    "    \n",
    "    # reorder columns\n",
    "    analysis_df = analysis_df[['mutation', 'predicted_effect', 'position', 'Odds_Ratio', 'OR_LB', 'OR_UB', 'pval', 'BH_pval', 'Bonferroni_pval',\n",
    "       'Num_Isolates', 'Total_Isolates', 'TP', 'FP', 'TN', 'FN', 'Sens_LB', 'Sens', 'Sens_UB', 'Spec_LB', 'Spec', 'Spec_UB', 'PPV', 'PPV_LB', 'PPV_UB',\n",
    "       'LR+_LB', 'LR+', 'LR+_UB', 'LR-_LB', 'LR-', 'LR-_UB', 'Tier', 'Phenos', 'poolLOF', 'Syn']]\n",
    "    \n",
    "    analysis_df[['poolLOF', \"Syn\"]] = analysis_df[['poolLOF', \"Syn\"]].astype(int)\n",
    "        \n",
    "    return analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ac6d98-c680-47eb-8a77-0a12f7121517",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis\"\n",
    "finished_drugs = []\n",
    "\n",
    "for drug in os.listdir(analysis_dir):\n",
    "    \n",
    "    if os.path.isfile(os.path.join(analysis_dir, drug, \"final_analysis.csv\")):\n",
    "        finished_drugs.append(drug)\n",
    "        # drug_analyses.append(final_processing(drug))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f27f833-b810-4a22-9cd8-1ec10b70a75b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(finished_drugs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f2bfc4-14a1-4afb-9b97-79b58636f547",
   "metadata": {},
   "outputs": [],
   "source": [
    "# those that are actually done\n",
    "finished_drugs = ['Pyrazinamide', 'Amikacin', 'Clofazimine', 'Linezolid', 'Moxifloxacin', 'Kanamycin', 'Bedaquiline', 'Capreomycin', 'Delamanid']\n",
    "\n",
    "drug_analyses = {}\n",
    "for drug in finished_drugs:\n",
    "    drug_analyses[drug] = final_processing(drug)\n",
    "    \n",
    "# write results to an Excel file, where each sheet is named for a drug\n",
    "with pd.ExcelWriter(\"Farhat_logReg_analysis.xlsx\") as file:\n",
    "   \n",
    "    for key, val in drug_analyses.items():\n",
    "        val.to_excel(file, sheet_name=key, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd2ede15-4b01-4a16-8565-531b1925806c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17dbb6e3-d3d8-4427-963b-f2c88919552c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d471f30-5ad3-4545-aaae-53210b0d97a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52ad210e-7627-48c7-84dd-e7d3e365d3fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_analyses[4].query(\"Tier==2&Phenos=='WHO'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f1c384-b661-4f12-b071-ff944fd1ad2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_analyses[0].query(\"Tier==1&Phenos=='ALL'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66beb441-6ac2-4879-bf36-d39e8bc88007",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_analyses[0].query(\"predicted_effect=='synonymous_variant' & \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e8a634b-b24c-47dd-8310-4d29cde7ab82",
   "metadata": {},
   "outputs": [],
   "source": [
    "moxi_df.loc[pd.isnull(moxi_df[\"position\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a683ba2b-ca3d-4b46-b3fb-eff707fe4cf5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdcae08c-4bc9-4527-8e24-c0c3922beb1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aed4a42c-86c9-4a43-9382-7f9723720ada",
   "metadata": {},
   "source": [
    "# Catalog-Based Method Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f78280-0749-4fe9-8cd9-079fa016922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_logReg_summary(out_dir):\n",
    "    \n",
    "    model_summary = pd.read_csv(os.path.join(out_dir, \"logReg_summary.csv\"))\n",
    "    model_analysis = pd.read_csv(os.path.join(out_dir, \"model_analysis.csv\"))\n",
    "    \n",
    "    summary_sens = model_summary.loc[0, \"Sens\"]\n",
    "    summary_spec = model_summary.loc[0, \"Spec\"]\n",
    "    \n",
    "    #max_idx = [np.argmax(model_analysis[\"Sens\"]), np.argmax(model_analysis[\"Spec\"])]\n",
    "    \n",
    "    better_variant = model_analysis.query(\"Sens >= @summary_sens & Spec >= @summary_spec\")\n",
    "    \n",
    "    if len(better_variant) == 0:\n",
    "        print(\"No variants have comparable sensitivity AND specificity\")\n",
    "    else:\n",
    "        print(better_variant)\n",
    "    return model_summary, model_analysis\n",
    "    \n",
    "    #return model_summary, model_analysis.iloc[max_idx, :][[\"orig_variant\", \"coef\", \"Sens\", \"Spec\", \"accuracy\", \"balanced_accuracy\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20dfc9fa-8f4b-4e2d-81b2-489754985444",
   "metadata": {},
   "outputs": [],
   "source": [
    "#out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Levofloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn\"\n",
    "\n",
    "summary, analysis = get_logReg_summary(out_dir)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ef135c-6253-4a64-9053-a08410585184",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.query(\"Sens > 0.9006\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c620b336-0167-4f00-b47b-e4bc031a045d",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.loc[analysis[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36145380-d850-475c-a6fb-e129a2824e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis.query(\"Spec > 0.9575\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e714d4af-4376-4e03-874e-0d4566d278a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Moxifloxacin/tiers=1+2/phenos=ALL/dropAF_withSyn\"\n",
    "\n",
    "get_logReg_summary(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec5a07e-ae8a-4de2-911f-d4a9ad4e6446",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pza = pd.read_csv(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Pyrazinamide/tiers=1+2/phenos=WHO/dropAF_withSyn/phenos.csv\")\n",
    "\n",
    "# the missing ones might be M. cannettii, most similar to L6 based on the other lineage callers\n",
    "lineages = pd.read_pickle(\"../data/combined_lineage_sample_IDs.pkl\")\n",
    "lineages[\"Lineage\"] = lineages[\"Lineage\"].fillna(\"6\")\n",
    "lineages[\"Lineage_1\"] = lineages[\"Lineage_1\"].fillna(\"6\")\n",
    "\n",
    "lineages = lineages[[\"Sample Name\", \"Sample ID\", \"Lineage_1\"]]\n",
    "lineages[\"Lineage\"] = [str(val).split(\".\")[0] for val in lineages[\"Lineage_1\"].values]\n",
    "lineages.loc[lineages[\"Lineage\"].str.contains(\"BOV\"), \"Lineage\"] = \"M. bovis\"\n",
    "\n",
    "assert len(lineages.loc[pd.isnull(lineages[\"Lineage\"])]) == 0\n",
    "\n",
    "########## KEEP ONLY ISOLATES WITH ALL 3 PIECES OF DATA ##########\n",
    "\n",
    "# get only isolates with data for everyting: SNP matrix, in the model, and lineages\n",
    "combined = lineages.merge(df_pza, left_on=\"Sample ID\", right_on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56d7652e-ec58-4e0a-88ef-86e80fcc025a",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.groupby(\"Lineage\")[\"phenotype\"].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3334fbf-76ee-49e0-a0b9-0dccff1727c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.query(\"Lineage == 'M. bovis'\").Lineage_1.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c225d9c-faf1-4141-884c-034c55fa03de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
