{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a4647438-032f-45bc-acab-e16b6955dbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.dpi'] = 125\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, Seq\n",
    "import scipy.stats as st\n",
    "import sklearn\n",
    "from sklearn.linear_model import LogisticRegression, LogisticRegressionCV\n",
    "import glob, os, yaml, subprocess, itertools, sparse\n",
    "\n",
    "who_variants = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/MIC_data/WHO_resistance_variants_all.csv\")\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11db1965-af99-46c0-b3d3-48957e1019bc",
   "metadata": {},
   "source": [
    "# Check for analysis files to see if anything needs to be rerun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b48bf339-1f4b-4820-acd3-4c5cb104b54c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 drugs with phenotypes and genotypes\n"
     ]
    }
   ],
   "source": [
    "genos_dir = '/n/data1/hms/dbmi/farhat/ye12/who/full_genotypes'\n",
    "phenos_dir = '/n/data1/hms/dbmi/farhat/ye12/who/phenotypes'\n",
    "analysis_dir = '/n/data1/hms/dbmi/farhat/ye12/who/analysis'\n",
    "\n",
    "pheno_drugs = os.listdir(phenos_dir)\n",
    "geno_drugs = os.listdir(genos_dir)\n",
    "\n",
    "drugs_for_analysis = list(set(geno_drugs).intersection(set(pheno_drugs)))\n",
    "drugs_for_analysis = [drug.split(\"=\")[1] for drug in drugs_for_analysis]\n",
    "print(len(drugs_for_analysis), \"drugs with phenotypes and genotypes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "bcf33e66-f184-4582-8a36-3d3fb73820c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis\"\n",
    "mid_dir = \"tiers=1+2/phenos=WHO\"\n",
    "#model_prefix = \"dropAF_withSyn_poolLOF\"\n",
    "model_prefix = \"dropAF_withSyn\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b54327da-a3b2-4069-9542-4f6afdd4bd59",
   "metadata": {},
   "outputs": [],
   "source": [
    "for drug in drugs_for_analysis:\n",
    "    analysis_dir = os.path.join(out_dir, drug, mid_dir, model_prefix)\n",
    "    \n",
    "    try:\n",
    "        fNames = os.listdir(analysis_dir)\n",
    "        if 'model_analysis.csv' not in fNames:\n",
    "            print(drug)\n",
    "    except:\n",
    "        if \"LOF\" in model_prefix:\n",
    "            print(f\"{drug} did not have a unique model\")\n",
    "        else:\n",
    "            print(f\"{drug} needs all 3 scripts to be re-run\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1a27ad-74e6-4fa0-8293-4a2b8c5fd805",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "25652646-b13a-4cb8-85b1-0222a5777573",
   "metadata": {},
   "outputs": [],
   "source": [
    "mid_dirs = [\"tiers=1/phenos=WHO\",  \"tiers=1/phenos=ALL\", \"tiers=1+2/phenos=WHO\", \"tiers=1+2/phenos=ALL\"]\n",
    "model_prefixes = [\"dropAF_noSyn\", \"dropAF_noSyn_poolLOF\", \"dropAF_withSyn\", \"dropAF_withSyn_poolLOF\"]\n",
    "\n",
    "combos = list(itertools.product(mid_dirs, model_prefixes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "52244c7e-3e54-4643-b62f-a76265e8f735",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('tiers=1/phenos=WHO', 'dropAF_noSyn'),\n",
       " ('tiers=1/phenos=WHO', 'dropAF_noSyn_poolLOF'),\n",
       " ('tiers=1/phenos=WHO', 'dropAF_withSyn'),\n",
       " ('tiers=1/phenos=WHO', 'dropAF_withSyn_poolLOF'),\n",
       " ('tiers=1/phenos=ALL', 'dropAF_noSyn'),\n",
       " ('tiers=1/phenos=ALL', 'dropAF_noSyn_poolLOF'),\n",
       " ('tiers=1/phenos=ALL', 'dropAF_withSyn'),\n",
       " ('tiers=1/phenos=ALL', 'dropAF_withSyn_poolLOF'),\n",
       " ('tiers=1+2/phenos=WHO', 'dropAF_noSyn'),\n",
       " ('tiers=1+2/phenos=WHO', 'dropAF_noSyn_poolLOF'),\n",
       " ('tiers=1+2/phenos=WHO', 'dropAF_withSyn'),\n",
       " ('tiers=1+2/phenos=WHO', 'dropAF_withSyn_poolLOF'),\n",
       " ('tiers=1+2/phenos=ALL', 'dropAF_noSyn'),\n",
       " ('tiers=1+2/phenos=ALL', 'dropAF_noSyn_poolLOF'),\n",
       " ('tiers=1+2/phenos=ALL', 'dropAF_withSyn'),\n",
       " ('tiers=1+2/phenos=ALL', 'dropAF_withSyn_poolLOF')]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "fd194fea-3c18-4f02-8be8-1ce58c61af73",
   "metadata": {},
   "outputs": [],
   "source": [
    "combos = [('tiers=1/phenos=WHO', 'dropAF_noSyn'),\n",
    " ('tiers=1/phenos=WHO', 'dropAF_withSyn'),\n",
    " ('tiers=1/phenos=WHO', 'dropAF_withSyn_poolLOF'),\n",
    " ('tiers=1/phenos=ALL', 'dropAF_noSyn'),\n",
    " ('tiers=1/phenos=ALL', 'dropAF_withSyn'),\n",
    " ('tiers=1+2/phenos=WHO', 'dropAF_noSyn'),\n",
    " ('tiers=1+2/phenos=WHO', 'dropAF_noSyn_poolLOF'),\n",
    " ('tiers=1+2/phenos=WHO', 'dropAF_withSyn'),\n",
    " ('tiers=1+2/phenos=WHO', 'dropAF_withSyn_poolLOF')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "9e69fddd-e965-42d2-aeeb-5b466db2358e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Capreomycin/tiers=1+2/phenos=WHO/dropAF_noSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Capreomycin/tiers=1+2/phenos=WHO/dropAF_noSyn_poolLOF/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Ethambutol/tiers=1+2/phenos=WHO/dropAF_withSyn_poolLOF/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Isoniazid/tiers=1+2/phenos=WHO/dropAF_withSyn_poolLOF/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Rifampicin/tiers=1+2/phenos=WHO/dropAF_withSyn_poolLOF/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Streptomycin/tiers=1/phenos=ALL/dropAF_withSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Amikacin/tiers=1/phenos=WHO/dropAF_withSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Amikacin/tiers=1/phenos=ALL/dropAF_noSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Amikacin/tiers=1/phenos=ALL/dropAF_withSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Kanamycin/tiers=1/phenos=WHO/dropAF_noSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Kanamycin/tiers=1/phenos=WHO/dropAF_withSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Kanamycin/tiers=1/phenos=WHO/dropAF_withSyn_poolLOF/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Kanamycin/tiers=1/phenos=ALL/dropAF_noSyn/model_analysis.csv\n",
      "/n/data1/hms/dbmi/farhat/ye12/who/analysis/Kanamycin/tiers=1/phenos=ALL/dropAF_withSyn/model_analysis.csv\n"
     ]
    }
   ],
   "source": [
    "for drug in drugs_for_analysis:\n",
    "    \n",
    "    for mid_dir, model_prefix in combos:\n",
    "        \n",
    "        if os.path.isdir(os.path.join(out_dir, drug, mid_dir, model_prefix)):\n",
    "            analysis_file = os.path.join(out_dir, drug, mid_dir, model_prefix, \"model_analysis.csv\")\n",
    "            if not os.path.isfile(analysis_file):\n",
    "#                 analysis = pd.read_csv(analysis_file)\n",
    "\n",
    "#                 if sum(analysis[\"coef_LB\"] > analysis[\"coef\"]) != 0:\n",
    "#                     print(f\"    {analysis_file}, LB\")\n",
    "#                 if sum(analysis[\"coef_UB\"] < analysis[\"coef\"]) != 0:\n",
    "#                     print(f\"    {analysis_file}, UB\")\n",
    "                print(analysis_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "219e0649-7a61-40b9-a664-5565f7765ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_file = os.path.join(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Capreomycin/tiers=1+2/phenos=WHO/dropAF_withSyn/model_analysis.csv\")\n",
    "analysis = pd.read_csv(analysis_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba59488-09f1-4e9e-acd1-e524f366a0b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a5b0c16-d97b-4cbc-84af-91b22a636c90",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "083deeed-d3ab-4991-8869-f8bebd2e3eea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9daf0d89-bc7e-4576-b62e-bcc3282d7f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_models_population_structure_correction(pop_corr, no_pop_corr, plot=False, alpha=0.05, diff_thresh=0):\n",
    "    '''\n",
    "    Comparison to determine if population structure determination is necessary. Things to compare:\n",
    "    \n",
    "    1. Do all the same variants have non-zero coefficients (the returned dataframes are only non-zero coefficients)?\n",
    "    2. Compute Pearson correlation between coefficients.\n",
    "    3. Do features have p-values that are on the same side of the threshold (0.05 or 0.01)?\n",
    "    '''\n",
    "    \n",
    "    # merge outer to include anything (in case some features have 0 or non-zero coefficients, depending on whether PCs are included)\n",
    "    plot_df = pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\"]].merge(no_pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\"]], how=\"outer\", on=\"orig_variant\")    \n",
    "    \n",
    "    # then drop the principal components\n",
    "    plot_df = plot_df.loc[~plot_df[\"orig_variant\"].str.contains(\"PC\")]\n",
    "    \n",
    "    # plot_df_missing = plot_df.loc[(pd.isnull(plot_df[\"coef_x\"])) | (pd.isnull(plot_df[\"coef_y\"]))]\n",
    "    # print(f\"{len(plot_df_missing)} coefficients are zero in one model and non-zero in the other\")\n",
    "    # plot_df_nominal_sig = plot_df.query(\"pval_x < @alpha & pval_y < @alpha\")\n",
    "    # print(\"   \", len(plot_df_nominal_sig.query(\"(coef_x < 0 & coef_y > 0) | (coef_x > 0 & coef_y < 0)\")), \"nominally significant features have conflicting coefficients\")\n",
    "    \n",
    "    plot_df_sig = plot_df.query(\"BH_pval_x < @alpha & BH_pval_y < @alpha\")\n",
    "    print(\"   \", len(plot_df_sig.query(\"(coef_x < 0 & coef_y > 0) | (coef_x > 0 & coef_y < 0)\")), \"significant features have conflicting coefficients\")    \n",
    "    \n",
    "    # NaNs are 0 coefficients, fill them in for plotting and correlation computations\n",
    "    plot_df[[\"coef_x\", \"coef_y\"]] = plot_df[[\"coef_x\", \"coef_y\"]].fillna(0)\n",
    "    \n",
    "    # coefficient without population structure correction - coefficient with correction\n",
    "    plot_df[\"diff\"] = plot_df[\"coef_y\"] - plot_df[\"coef_x\"]\n",
    "    plot_df[\"norm_diff\"] = plot_df[\"diff\"] / plot_df[\"coef_x\"]\n",
    "    \n",
    "    pearson = st.pearsonr(plot_df['coef_x'], plot_df['coef_y'])[0]\n",
    "\n",
    "    if plot:\n",
    "        fig, ax = plt.subplots(1, 2, figsize=(9, 3))\n",
    "        sns.scatterplot(data=plot_df, x=\"coef_x\", y=\"coef_y\", ax=ax[0], alpha=0.5)\n",
    "        ax[0].set_xlabel(\"Coef, Population Structure Correction\")\n",
    "        ax[0].set_ylabel(\"Coef, No Correction\")\n",
    "        ax[0].set_title(np.round(pearson, 4), \"Pearson R\")\n",
    "\n",
    "        sns.scatterplot(data=plot_df, x=\"coef_x\", y=\"norm_diff\", ax=ax[1], alpha=0.5)\n",
    "        ax[1].set_xlabel(\"Coef, Population Structure Correction\")\n",
    "        ax[1].set_ylabel(\"\")\n",
    "        ax[1].set_title(\"Norm. Difference in Coefs.\")\n",
    "\n",
    "        sns.despine()\n",
    "        plt.show()\n",
    "    else:\n",
    "        print(\"   \", np.round(pearson, 4), \"Pearson R\")\n",
    "    \n",
    "    # get features whose coefficients differ by more than some threshold (0 to 100) for percent difference\n",
    "    # (difference between coefficients without and with correction is greater than the coefficient with population structure correction\n",
    "    #large_diff = plot_df.query(\"percent_diff >= @diff_thresh | percent_diff <= -@diff_thresh\")\n",
    "    \n",
    "    # get the number of features with confidence intervals that lie on different sides of 0\n",
    "    print('   ', len(plot_df.query(\"(coef_LB_x > 0 & coef_UB_x > 0 & coef_LB_y < 0 & coef_UB_y < 0) | (coef_LB_x < 0 & coef_UB_x < 0 & coef_LB_y > 0 & coef_UB_y > 0)\")), 'features have confidence intervals on different sides of 0')\n",
    "    #print(f'{len(large_diff.query(\"coef_LB_x > 0 & coef_UB_x > 0 & coef_LB_y < 0 & coef_UB_y < 0\"))} features with very different coefficients have confidence intervals on different sides of 0')\n",
    "    \n",
    "    # TO-DO: COMPARISON OF CONFIDENCE INTERVALS\n",
    "    return pearson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "37b0349c-af23-4a72-b246-9b637512536b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 drugs with phenotypes and genotypes\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8de26051-5622-40f6-bb40-cbf9a791faec",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Ethambutol\"\n",
    "out_dir = os.path.join(analysis_dir, drug, \"tiers=1/phenos=WHO\")\n",
    "corr_prefix = \"dropAF_noSyn\"\n",
    "no_corr_prefix = \"dropAF_noSyn_noPopCorr\"\n",
    "    \n",
    "pop_corr = pd.read_pickle(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\"))\n",
    "no_pop_corr = pd.read_pickle(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\"))\n",
    "\n",
    "# number of principal components with positive coefficients (OR > 1)\n",
    "sig_PC = pop_corr.loc[(pop_corr[\"orig_variant\"].str.contains(\"PC\")) & \n",
    "                      (((pop_corr[\"coef_LB\"] > 0) & (pop_corr[\"coef_UB\"] > 0)) |\n",
    "                      ((pop_corr[\"coef_LB\"] < 0) & (pop_corr[\"coef_UB\"] < 0)))\n",
    "                      ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "34fa06c4-f8d0-4f15-893b-12db753c7cee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Levofloxacin\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9998 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Moxifloxacin\n",
      "   ['PC1' 'PC2' 'PC4'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9985 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Ethionamide\n",
      "   ['PC3' 'PC4'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9989 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Clofazimine\n",
      "   ['PC1'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9945 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Capreomycin\n",
      "   ['PC0' 'PC1' 'PC2'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9998 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Streptomycin\n",
      "   ['PC0' 'PC1' 'PC2' 'PC3'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9987 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Rifampicin\n",
      "   ['PC0' 'PC1' 'PC2' 'PC3' 'PC4'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9995 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Ethambutol\n",
      "   ['PC0' 'PC2'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9921 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Kanamycin\n",
      "   ['PC1' 'PC3' 'PC4'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9992 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Isoniazid\n",
      "   ['PC0' 'PC1' 'PC3'] have significant coefficients\n",
      "    1 significant features have conflicting coefficients\n",
      "    0.9938 Pearson R\n",
      "    1 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Delamanid\n",
      "   ['PC0'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9787 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Pyrazinamide\n",
      "   ['PC0' 'PC1' 'PC2' 'PC3'] have significant coefficients\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9952 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n",
      "Amikacin\n",
      "    0 significant features have conflicting coefficients\n",
      "    0.9998 Pearson R\n",
      "    0 features have confidence intervals on different sides of 0\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pearson_df = pd.DataFrame(columns={\"Drug\", \"R\"})\n",
    "\n",
    "for drug in drugs_for_analysis:\n",
    "    \n",
    "    drug = drug.split(\"=\")[1]\n",
    "    out_dir = os.path.join(analysis_dir, drug, \"tiers=1/phenos=WHO\")\n",
    "    corr_prefix = \"dropAF_noSyn\"\n",
    "    no_corr_prefix = \"dropAF_noSyn_noPopCorr\"\n",
    "    \n",
    "    if os.path.isfile(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\")) and os.path.isfile(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\")):\n",
    "        \n",
    "        print(drug)\n",
    "        pop_corr = pd.read_pickle(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\"))\n",
    "        no_pop_corr = pd.read_pickle(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\"))\n",
    "\n",
    "        # number of principal components with positive coefficients (OR > 1)\n",
    "        sig_PC = pop_corr.loc[(pop_corr[\"orig_variant\"].str.contains(\"PC\")) & \n",
    "                              (((pop_corr[\"coef_LB\"] > 0) & (pop_corr[\"coef_UB\"] > 0)) |\n",
    "                              ((pop_corr[\"coef_LB\"] < 0) & (pop_corr[\"coef_UB\"] < 0)))\n",
    "                              ]\n",
    "        \n",
    "        if len(sig_PC) > 0:\n",
    "            print(\"  \", np.sort(sig_PC[\"orig_variant\"].values), \"have significant coefficients\")\n",
    "\n",
    "        # variants in moxi_res that are not in the results without population structure correction\n",
    "        overlap_variants = set(pop_corr.loc[~pop_corr[\"orig_variant\"].str.contains(\"PC\")][\"orig_variant\"]).intersection(set(no_pop_corr[\"orig_variant\"]))\n",
    "        if len(overlap_variants) == len(pop_corr.loc[~pop_corr[\"orig_variant\"].str.contains(\"PC\")]) == len(no_pop_corr):\n",
    "            pearson = compare_models_population_structure_correction(pop_corr, no_pop_corr, plot=False, alpha=0.05, diff_thresh=0)\n",
    "            pearson_df = pd.concat([pearson_df, pd.DataFrame({\"Drug\": drug, \"R\": pearson}, index=[0])], axis=0)\n",
    "        else:\n",
    "            print(\"    Differing non-zero coefficients\")        \n",
    "        print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e590a4f-0b52-4338-844c-3522cb5660c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#who_variants.loc[(who_variants.drug == 'PZA') & (who_variants.confidence.str.contains(\"|\".join([\"1\"])))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a536f2b-7d48-4eec-8b1d-f4a119409765",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Capreomycin</td>\n",
       "      <td>0.9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Amikacin</td>\n",
       "      <td>0.9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Levofloxacin</td>\n",
       "      <td>0.9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>0.9995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Kanamycin</td>\n",
       "      <td>0.9992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ethionamide</td>\n",
       "      <td>0.9989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Streptomycin</td>\n",
       "      <td>0.9987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Moxifloxacin</td>\n",
       "      <td>0.9985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pyrazinamide</td>\n",
       "      <td>0.9952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clofazimine</td>\n",
       "      <td>0.9945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Isoniazid</td>\n",
       "      <td>0.9938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>0.9921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delamanid</td>\n",
       "      <td>0.9787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Drug      R\n",
       "0   Capreomycin 0.9998\n",
       "0      Amikacin 0.9998\n",
       "0  Levofloxacin 0.9998\n",
       "0    Rifampicin 0.9995\n",
       "0     Kanamycin 0.9992\n",
       "0   Ethionamide 0.9989\n",
       "0  Streptomycin 0.9987\n",
       "0  Moxifloxacin 0.9985\n",
       "0  Pyrazinamide 0.9952\n",
       "0   Clofazimine 0.9945\n",
       "0     Isoniazid 0.9938\n",
       "0    Ethambutol 0.9921\n",
       "0     Delamanid 0.9787"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pearson_df.sort_values(\"R\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7726ce8-605b-4e29-aa53-ecdd1f260e37",
   "metadata": {},
   "source": [
    "# 1 variant in INH has positive or negative coefficients in the 2 different models\n",
    "\n",
    "## It's a category 5 mutation, and the confidence intervals are significant\n",
    "\n",
    "## The other 2 variants do not have significant p-values, and the confidence intervals are not fully above or below 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7205f478-ec9e-4d70-a175-c2cf982bde88",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug = \"Isoniazid\"\n",
    "out_dir = os.path.join(analysis_dir, drug, \"tiers=1/phenos=WHO\")\n",
    "corr_prefix = \"dropAF_noSyn\"\n",
    "no_corr_prefix = \"dropAF_noSyn_noPopCorr\"\n",
    "\n",
    "pop_corr = pd.read_pickle(os.path.join(out_dir, corr_prefix, \"model_analysis.pkl\"))\n",
    "no_pop_corr = pd.read_pickle(os.path.join(out_dir, no_corr_prefix, \"model_analysis.pkl\"))\n",
    "\n",
    "plot_df = pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\", \"confidence_WHO_2021\"]].merge(no_pop_corr[[\"orig_variant\", \"coef\", \"pval\", \"BH_pval\", \"coef_LB\", \"coef_UB\", \"confidence_WHO_2021\"]], how=\"outer\", on=\"orig_variant\")    \n",
    "    \n",
    "# then drop the principal components\n",
    "plot_df = plot_df.loc[~plot_df[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "613bdb38-1bc4-4f12-8ab9-3c1ccb53be54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef_x</th>\n",
       "      <th>BH_pval_x</th>\n",
       "      <th>coef_LB_x</th>\n",
       "      <th>coef_UB_x</th>\n",
       "      <th>coef_y</th>\n",
       "      <th>BH_pval_y</th>\n",
       "      <th>coef_LB_y</th>\n",
       "      <th>coef_UB_y</th>\n",
       "      <th>confidence_WHO_2021_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>katG_p.Arg463Leu</td>\n",
       "      <td>-0.1853</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>-0.2719</td>\n",
       "      <td>-0.0975</td>\n",
       "      <td>0.1827</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1364</td>\n",
       "      <td>0.2290</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          orig_variant  coef_x  BH_pval_x  coef_LB_x  coef_UB_x  coef_y  \\\n",
       "1317  katG_p.Arg463Leu -0.1853     0.0005    -0.2719    -0.0975  0.1827   \n",
       "\n",
       "      BH_pval_y  coef_LB_y  coef_UB_y confidence_WHO_2021_y  \n",
       "1317     0.0000     0.1364     0.2290      5) Not assoc w R  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_cols = [\"orig_variant\", \"coef_x\", \"BH_pval_x\", \"coef_LB_x\", \"coef_UB_x\", \"coef_y\", \"BH_pval_y\", \"coef_LB_y\", \"coef_UB_y\", \"confidence_WHO_2021_y\"]\n",
    "plot_df.query(\"(coef_LB_x > 0 & coef_UB_x > 0 & coef_LB_y < 0 & coef_UB_y < 0) | (coef_LB_x < 0 & coef_UB_x < 0 & coef_LB_y > 0 & coef_UB_y > 0)\")[show_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d51ec463-379d-4747-9408-a1c2245927ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Moxifloxacin/tiers=1+2/phenos=WHO/encodeAF_noSyn/model_analysis.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7770e76f-9e7e-4f8c-9658-58934177d140",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>PC0</td>\n",
       "      <td>0.0362</td>\n",
       "      <td>-0.0177</td>\n",
       "      <td>0.0939</td>\n",
       "      <td>0.1195</td>\n",
       "      <td>0.2875</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0368</td>\n",
       "      <td>0.9824</td>\n",
       "      <td>1.0984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PC3</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>-0.0485</td>\n",
       "      <td>0.0840</td>\n",
       "      <td>0.2941</td>\n",
       "      <td>0.6729</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0185</td>\n",
       "      <td>0.9527</td>\n",
       "      <td>1.0876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>PC1</td>\n",
       "      <td>-0.0889</td>\n",
       "      <td>-0.1317</td>\n",
       "      <td>-0.0515</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9150</td>\n",
       "      <td>0.8766</td>\n",
       "      <td>0.9498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>PC2</td>\n",
       "      <td>-0.1073</td>\n",
       "      <td>-0.1743</td>\n",
       "      <td>-0.0352</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>0.0342</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8983</td>\n",
       "      <td>0.8400</td>\n",
       "      <td>0.9654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>PC4</td>\n",
       "      <td>-0.1635</td>\n",
       "      <td>-0.2228</td>\n",
       "      <td>-0.0962</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8492</td>\n",
       "      <td>0.8003</td>\n",
       "      <td>0.9083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  Bonferroni_pval  \\\n",
       "79           PC0  0.0362  -0.0177   0.0939 0.1195   0.2875           1.0000   \n",
       "119          PC3  0.0184  -0.0485   0.0840 0.2941   0.6729           1.0000   \n",
       "989          PC1 -0.0889  -0.1317  -0.0515 0.0000   0.0003           0.0086   \n",
       "992          PC2 -0.1073  -0.1743  -0.0352 0.0012   0.0342           1.0000   \n",
       "993          PC4 -0.1635  -0.2228  -0.0962 0.0000   0.0000           0.0007   \n",
       "\n",
       "    genome_index confidence_WHO_2021  Odds_Ratio  OR_LB  OR_UB  \n",
       "79           NaN                 NaN      1.0368 0.9824 1.0984  \n",
       "119          NaN                 NaN      1.0185 0.9527 1.0876  \n",
       "989          NaN                 NaN      0.9150 0.8766 0.9498  \n",
       "992          NaN                 NaN      0.8983 0.8400 0.9654  \n",
       "993          NaN                 NaN      0.8492 0.8003 0.9083  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df[\"orig_variant\"].str.contains(\"PC\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed4a42c-86c9-4a43-9382-7f9723720ada",
   "metadata": {},
   "source": [
    "# Catalog-Based Method Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0d61fa59-e346-4cdb-a538-ef2ba57c919f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34623,)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_dir = \"/n/data1/hms/dbmi/farhat/ye12/who/analysis/Rifampicin/tiers=1/phenos=WHO/dropAF_withSyn\"\n",
    "\n",
    "model = pd.read_pickle(os.path.join(out_dir, 'logReg_model'))\n",
    "\n",
    "model_matrix = pd.read_pickle(os.path.join(out_dir, \"model_matrix.pkl\"))\n",
    "eigenvec_df = pd.read_pickle(os.path.join(out_dir, \"model_eigenvecs.pkl\"))\n",
    "X = model_matrix.merge(eigenvec_df, left_index=True, right_index=True).values\n",
    "\n",
    "df_phenos = pd.read_csv(os.path.join(out_dir, \"phenos.csv\"))\n",
    "y = df_phenos.sort_values(\"sample_id\").phenotype.values\n",
    "y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e2bd0cfa-4c46-4aa0-a0f2-c84ee82c6666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.818268831600007"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sklearn.metrics.roc_auc_score(y, model.predict(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "227b6aa9-f203-401e-aa0e-adae4bfcba05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8182688316000071"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sklearn.metrics.balanced_accuracy_score(y, model.predict(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fc67041d-8231-41d2-a661-5721dc18dea7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8836322675678018"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sklearn.metrics.accuracy_score(y, model.predict(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9d4d7929-726f-4851-ba2a-7398718e31bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.019201214743343"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sklearn.metrics.log_loss(y, model.predict(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "10fe136c-6c5a-4111-9355-46715884ce5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[0;31mSignature:\u001b[0m\n",
       "\u001b[0msklearn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetrics\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog_loss\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0meps\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1e-15\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mnormalize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0msample_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mlabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mDocstring:\u001b[0m\n",
       "Log loss, aka logistic loss or cross-entropy loss.\n",
       "\n",
       "This is the loss function used in (multinomial) logistic regression\n",
       "and extensions of it such as neural networks, defined as the negative\n",
       "log-likelihood of a logistic model that returns ``y_pred`` probabilities\n",
       "for its training data ``y_true``.\n",
       "The log loss is only defined for two or more labels.\n",
       "For a single sample with true label :math:`y \\in \\{0,1\\}` and\n",
       "a probability estimate :math:`p = \\operatorname{Pr}(y = 1)`, the log\n",
       "loss is:\n",
       "\n",
       ".. math::\n",
       "    L_{\\log}(y, p) = -(y \\log (p) + (1 - y) \\log (1 - p))\n",
       "\n",
       "Read more in the :ref:`User Guide <log_loss>`.\n",
       "\n",
       "Parameters\n",
       "----------\n",
       "y_true : array-like or label indicator matrix\n",
       "    Ground truth (correct) labels for n_samples samples.\n",
       "\n",
       "y_pred : array-like of float, shape = (n_samples, n_classes) or (n_samples,)\n",
       "    Predicted probabilities, as returned by a classifier's\n",
       "    predict_proba method. If ``y_pred.shape = (n_samples,)``\n",
       "    the probabilities provided are assumed to be that of the\n",
       "    positive class. The labels in ``y_pred`` are assumed to be\n",
       "    ordered alphabetically, as done by\n",
       "    :class:`preprocessing.LabelBinarizer`.\n",
       "\n",
       "eps : float, default=1e-15\n",
       "    Log loss is undefined for p=0 or p=1, so probabilities are\n",
       "    clipped to max(eps, min(1 - eps, p)).\n",
       "\n",
       "normalize : bool, default=True\n",
       "    If true, return the mean loss per sample.\n",
       "    Otherwise, return the sum of the per-sample losses.\n",
       "\n",
       "sample_weight : array-like of shape (n_samples,), default=None\n",
       "    Sample weights.\n",
       "\n",
       "labels : array-like, default=None\n",
       "    If not provided, labels will be inferred from y_true. If ``labels``\n",
       "    is ``None`` and ``y_pred`` has shape (n_samples,) the labels are\n",
       "    assumed to be binary and are inferred from ``y_true``.\n",
       "\n",
       "    .. versionadded:: 0.18\n",
       "\n",
       "Returns\n",
       "-------\n",
       "loss : float\n",
       "\n",
       "Notes\n",
       "-----\n",
       "The logarithm used is the natural logarithm (base-e).\n",
       "\n",
       "Examples\n",
       "--------\n",
       ">>> from sklearn.metrics import log_loss\n",
       ">>> log_loss([\"spam\", \"ham\", \"ham\", \"spam\"],\n",
       "...          [[.1, .9], [.9, .1], [.8, .2], [.35, .65]])\n",
       "0.21616...\n",
       "\n",
       "References\n",
       "----------\n",
       "C.M. Bishop (2006). Pattern Recognition and Machine Learning. Springer,\n",
       "p. 209.\n",
       "\u001b[0;31mFile:\u001b[0m      ~/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py\n",
       "\u001b[0;31mType:\u001b[0m      function\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sklearn.metrics.log_loss?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "af25fe49-cce0-41e4-bc6c-355f2ffdb817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(model.predict(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9b766456-67a2-40ec-85b9-d7b9c6c56465",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "68c497f7-842f-43e9-838f-62a3ec96c9d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_analysis = pd.read_csv(os.path.join(out_dir, \"model_analysis.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bd95e33c-ffff-410a-9a3d-902ff14f21d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_variant</th>\n",
       "      <th>coef</th>\n",
       "      <th>coef_LB</th>\n",
       "      <th>coef_UB</th>\n",
       "      <th>pval</th>\n",
       "      <th>BH_pval</th>\n",
       "      <th>Bonferroni_pval</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>confidence_WHO_2021</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>OR_LB</th>\n",
       "      <th>OR_UB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gyrA_p.Asp94Gly</td>\n",
       "      <td>0.9749</td>\n",
       "      <td>0.9196</td>\n",
       "      <td>1.0283</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7582.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>2.6508</td>\n",
       "      <td>2.5083</td>\n",
       "      <td>2.7963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gyrA_p.Ala90Val</td>\n",
       "      <td>0.5845</td>\n",
       "      <td>0.5296</td>\n",
       "      <td>0.6370</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7570.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.7941</td>\n",
       "      <td>1.6982</td>\n",
       "      <td>1.8907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gyrA_p.Asp94Asn</td>\n",
       "      <td>0.4584</td>\n",
       "      <td>0.4125</td>\n",
       "      <td>0.5094</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7581.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.5815</td>\n",
       "      <td>1.5107</td>\n",
       "      <td>1.6643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gyrA_p.Asp94Ala</td>\n",
       "      <td>0.4115</td>\n",
       "      <td>0.3586</td>\n",
       "      <td>0.4617</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7582.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.5090</td>\n",
       "      <td>1.4313</td>\n",
       "      <td>1.5867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gyrA_p.Asp94Tyr</td>\n",
       "      <td>0.3677</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.4117</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7581.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.4444</td>\n",
       "      <td>1.3792</td>\n",
       "      <td>1.5094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gyrA_p.Ser91Pro</td>\n",
       "      <td>0.3581</td>\n",
       "      <td>0.3115</td>\n",
       "      <td>0.4047</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7572.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.4307</td>\n",
       "      <td>1.3655</td>\n",
       "      <td>1.4988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gyrA_p.Gly668Asp</td>\n",
       "      <td>0.2678</td>\n",
       "      <td>0.2020</td>\n",
       "      <td>0.3297</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>9304.0000</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>1.3071</td>\n",
       "      <td>1.2239</td>\n",
       "      <td>1.3905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gyrB_p.Glu501Asp</td>\n",
       "      <td>0.2435</td>\n",
       "      <td>0.1886</td>\n",
       "      <td>0.2989</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>6742.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.2756</td>\n",
       "      <td>1.2075</td>\n",
       "      <td>1.3483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gyrA_p.Asp94His</td>\n",
       "      <td>0.2419</td>\n",
       "      <td>0.1962</td>\n",
       "      <td>0.2932</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7581.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.2737</td>\n",
       "      <td>1.2168</td>\n",
       "      <td>1.3407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gyrA_p.Asp89Asn</td>\n",
       "      <td>0.2002</td>\n",
       "      <td>0.1556</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7566.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.2217</td>\n",
       "      <td>1.1684</td>\n",
       "      <td>1.2802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gyrA_p.Gly88Cys</td>\n",
       "      <td>0.1965</td>\n",
       "      <td>0.1656</td>\n",
       "      <td>0.2146</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7563.0000</td>\n",
       "      <td>1) Assoc w R</td>\n",
       "      <td>1.2171</td>\n",
       "      <td>1.1801</td>\n",
       "      <td>1.2394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>gyrB_p.Ala504Val</td>\n",
       "      <td>0.1345</td>\n",
       "      <td>0.0788</td>\n",
       "      <td>0.1630</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>6750.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>1.1440</td>\n",
       "      <td>1.0820</td>\n",
       "      <td>1.1770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gyrB_p.Thr500Asn</td>\n",
       "      <td>0.1200</td>\n",
       "      <td>0.0792</td>\n",
       "      <td>0.1644</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>6738.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1275</td>\n",
       "      <td>1.0825</td>\n",
       "      <td>1.1787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>gyrB_p.Gly512Arg</td>\n",
       "      <td>0.1184</td>\n",
       "      <td>0.0514</td>\n",
       "      <td>0.1465</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>6773.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>1.1256</td>\n",
       "      <td>1.0527</td>\n",
       "      <td>1.1578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gyrB_c.-60_-59insG</td>\n",
       "      <td>0.0798</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.1003</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.5253</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0831</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.1055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gyrA_c.1440C&gt;T</td>\n",
       "      <td>-0.0069</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>-0.0023</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0195</td>\n",
       "      <td>0.7996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9932</td>\n",
       "      <td>0.9897</td>\n",
       "      <td>0.9977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>gyrA_c.474C&gt;T</td>\n",
       "      <td>-0.0076</td>\n",
       "      <td>-0.0106</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9924</td>\n",
       "      <td>0.9895</td>\n",
       "      <td>0.9962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>gyrB_p.Ile93Val</td>\n",
       "      <td>-0.0083</td>\n",
       "      <td>-0.0107</td>\n",
       "      <td>-0.0049</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>5516.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9918</td>\n",
       "      <td>0.9893</td>\n",
       "      <td>0.9951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gyrB_c.870G&gt;A</td>\n",
       "      <td>-0.0083</td>\n",
       "      <td>-0.0107</td>\n",
       "      <td>-0.0049</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9918</td>\n",
       "      <td>0.9893</td>\n",
       "      <td>0.9951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gyrB_p.Ile493Thr</td>\n",
       "      <td>-0.0083</td>\n",
       "      <td>-0.0107</td>\n",
       "      <td>-0.0049</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>6717.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9918</td>\n",
       "      <td>0.9893</td>\n",
       "      <td>0.9951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>gyrA_c.984C&gt;T</td>\n",
       "      <td>-0.0105</td>\n",
       "      <td>-0.0139</td>\n",
       "      <td>-0.0059</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9895</td>\n",
       "      <td>0.9862</td>\n",
       "      <td>0.9941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>gyrB_c.1167C&gt;T</td>\n",
       "      <td>-0.0110</td>\n",
       "      <td>-0.0141</td>\n",
       "      <td>-0.0063</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9890</td>\n",
       "      <td>0.9860</td>\n",
       "      <td>0.9937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>gyrB_c.513G&gt;A</td>\n",
       "      <td>-0.0110</td>\n",
       "      <td>-0.0141</td>\n",
       "      <td>-0.0063</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9890</td>\n",
       "      <td>0.9860</td>\n",
       "      <td>0.9937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>gyrA_c.1629C&gt;T</td>\n",
       "      <td>-0.0123</td>\n",
       "      <td>-0.0199</td>\n",
       "      <td>-0.0056</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>0.0253</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9878</td>\n",
       "      <td>0.9803</td>\n",
       "      <td>0.9944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>gyrA_p.Gly729Arg</td>\n",
       "      <td>-0.0131</td>\n",
       "      <td>-0.0194</td>\n",
       "      <td>-0.0056</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.2277</td>\n",
       "      <td>9486.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9870</td>\n",
       "      <td>0.9808</td>\n",
       "      <td>0.9944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>gyrB_p.Thr393Ser</td>\n",
       "      <td>-0.0131</td>\n",
       "      <td>-0.0194</td>\n",
       "      <td>-0.0056</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.2277</td>\n",
       "      <td>6417.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9870</td>\n",
       "      <td>0.9808</td>\n",
       "      <td>0.9944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gyrB_c.291C&gt;T</td>\n",
       "      <td>-0.0132</td>\n",
       "      <td>-0.0205</td>\n",
       "      <td>-0.0036</td>\n",
       "      <td>0.0011</td>\n",
       "      <td>0.0377</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9869</td>\n",
       "      <td>0.9797</td>\n",
       "      <td>0.9964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>gyrA_p.Leu296Pro</td>\n",
       "      <td>-0.0132</td>\n",
       "      <td>-0.0206</td>\n",
       "      <td>-0.0036</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0453</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>8188.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9869</td>\n",
       "      <td>0.9796</td>\n",
       "      <td>0.9964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>gyrB_p.Ala403Ser</td>\n",
       "      <td>-0.0150</td>\n",
       "      <td>-0.0182</td>\n",
       "      <td>-0.0103</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>6446.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9851</td>\n",
       "      <td>0.9820</td>\n",
       "      <td>0.9898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>gyrA_p.Ala456Val</td>\n",
       "      <td>-0.0168</td>\n",
       "      <td>-0.0274</td>\n",
       "      <td>-0.0057</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0455</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>8668.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9833</td>\n",
       "      <td>0.9729</td>\n",
       "      <td>0.9943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>gyrB_p.Thr378Ser</td>\n",
       "      <td>-0.0176</td>\n",
       "      <td>-0.0267</td>\n",
       "      <td>-0.0060</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>6372.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9826</td>\n",
       "      <td>0.9737</td>\n",
       "      <td>0.9940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>gyrB_p.Ile84Val</td>\n",
       "      <td>-0.0189</td>\n",
       "      <td>-0.0276</td>\n",
       "      <td>-0.0064</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.0186</td>\n",
       "      <td>0.7253</td>\n",
       "      <td>5489.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9813</td>\n",
       "      <td>0.9728</td>\n",
       "      <td>0.9936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>gyrA_p.Phe152Leu</td>\n",
       "      <td>-0.0193</td>\n",
       "      <td>-0.0288</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0454</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>7757.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9809</td>\n",
       "      <td>0.9716</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>gyrB_p.Pro254Leu</td>\n",
       "      <td>-0.0198</td>\n",
       "      <td>-0.0287</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>6000.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9804</td>\n",
       "      <td>0.9717</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>gyrA_c.195C&gt;G</td>\n",
       "      <td>-0.0200</td>\n",
       "      <td>-0.0295</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0442</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9802</td>\n",
       "      <td>0.9709</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>gyrA_p.Thr836Arg</td>\n",
       "      <td>-0.0205</td>\n",
       "      <td>-0.0295</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0437</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>9808.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9797</td>\n",
       "      <td>0.9710</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>gyrB_p.Met58Ile</td>\n",
       "      <td>-0.0217</td>\n",
       "      <td>-0.0316</td>\n",
       "      <td>-0.0083</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0195</td>\n",
       "      <td>0.8184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9785</td>\n",
       "      <td>0.9689</td>\n",
       "      <td>0.9917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>gyrA_c.2295G&gt;T</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>-0.0301</td>\n",
       "      <td>-0.0101</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>0.0215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9784</td>\n",
       "      <td>0.9704</td>\n",
       "      <td>0.9900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>gyrB_c.285C&gt;A</td>\n",
       "      <td>-0.0220</td>\n",
       "      <td>-0.0314</td>\n",
       "      <td>-0.0082</td>\n",
       "      <td>0.0002</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.3499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9782</td>\n",
       "      <td>0.9691</td>\n",
       "      <td>0.9918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>gyrB_p.Ala593Val</td>\n",
       "      <td>-0.0229</td>\n",
       "      <td>-0.0326</td>\n",
       "      <td>-0.0090</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.2217</td>\n",
       "      <td>7017.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9773</td>\n",
       "      <td>0.9679</td>\n",
       "      <td>0.9910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>gyrB_c.-1C&gt;G</td>\n",
       "      <td>-0.0231</td>\n",
       "      <td>-0.0319</td>\n",
       "      <td>-0.0110</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0314</td>\n",
       "      <td>5239.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9772</td>\n",
       "      <td>0.9687</td>\n",
       "      <td>0.9890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>gyrB_c.1119C&gt;T</td>\n",
       "      <td>-0.0232</td>\n",
       "      <td>-0.0331</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.8660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9771</td>\n",
       "      <td>0.9675</td>\n",
       "      <td>0.9947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>gyrA_c.1494C&gt;T</td>\n",
       "      <td>-0.0233</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>-0.0000</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>0.0416</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9769</td>\n",
       "      <td>0.9658</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>gyrA_c.1113C&gt;T</td>\n",
       "      <td>-0.0249</td>\n",
       "      <td>-0.0339</td>\n",
       "      <td>-0.0136</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9754</td>\n",
       "      <td>0.9667</td>\n",
       "      <td>0.9865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>gyrB_p.Asp461Asn</td>\n",
       "      <td>-0.0250</td>\n",
       "      <td>-0.0343</td>\n",
       "      <td>-0.0118</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>6620.0000</td>\n",
       "      <td>2) Assoc w R - Interim</td>\n",
       "      <td>0.9754</td>\n",
       "      <td>0.9663</td>\n",
       "      <td>0.9882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>gyrA_p.Pro472Ser</td>\n",
       "      <td>-0.0270</td>\n",
       "      <td>-0.0362</td>\n",
       "      <td>-0.0134</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>8715.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9733</td>\n",
       "      <td>0.9645</td>\n",
       "      <td>0.9867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>gyrA_p.Arg254Cys</td>\n",
       "      <td>-0.0293</td>\n",
       "      <td>-0.0388</td>\n",
       "      <td>-0.0143</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>8061.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9711</td>\n",
       "      <td>0.9619</td>\n",
       "      <td>0.9858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>gyrA_p.Leu711Met</td>\n",
       "      <td>-0.0294</td>\n",
       "      <td>-0.0394</td>\n",
       "      <td>-0.0152</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>9432.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9710</td>\n",
       "      <td>0.9614</td>\n",
       "      <td>0.9849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>gyrA_c.2142G&gt;A</td>\n",
       "      <td>-0.0299</td>\n",
       "      <td>-0.0415</td>\n",
       "      <td>-0.0186</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9705</td>\n",
       "      <td>0.9594</td>\n",
       "      <td>0.9815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>gyrA_c.1746C&gt;T</td>\n",
       "      <td>-0.0367</td>\n",
       "      <td>-0.0496</td>\n",
       "      <td>-0.0224</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9640</td>\n",
       "      <td>0.9516</td>\n",
       "      <td>0.9778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>gyrA_p.Pro8Ala</td>\n",
       "      <td>-0.0384</td>\n",
       "      <td>-0.0489</td>\n",
       "      <td>-0.0253</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7323.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9624</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>0.9750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>gyrA_c.1677C&gt;T</td>\n",
       "      <td>-0.0446</td>\n",
       "      <td>-0.0566</td>\n",
       "      <td>-0.0309</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9563</td>\n",
       "      <td>0.9449</td>\n",
       "      <td>0.9696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>gyrA_p.Arg252Leu</td>\n",
       "      <td>-0.0545</td>\n",
       "      <td>-0.0747</td>\n",
       "      <td>-0.0315</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>8056.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9469</td>\n",
       "      <td>0.9280</td>\n",
       "      <td>0.9689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>gyrB_p.Arg575His</td>\n",
       "      <td>-0.0555</td>\n",
       "      <td>-0.0830</td>\n",
       "      <td>-0.0201</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>6963.0000</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.9460</td>\n",
       "      <td>0.9203</td>\n",
       "      <td>0.9801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>gyrA_c.2457C&gt;T</td>\n",
       "      <td>-0.0564</td>\n",
       "      <td>-0.0709</td>\n",
       "      <td>-0.0291</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9451</td>\n",
       "      <td>0.9316</td>\n",
       "      <td>0.9713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>gyrA_c.2487C&gt;T</td>\n",
       "      <td>-0.0637</td>\n",
       "      <td>-0.0856</td>\n",
       "      <td>-0.0109</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>0.0269</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9383</td>\n",
       "      <td>0.9179</td>\n",
       "      <td>0.9891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>gyrA_p.Ala384Val</td>\n",
       "      <td>-0.0807</td>\n",
       "      <td>-0.1340</td>\n",
       "      <td>-0.0403</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.7417</td>\n",
       "      <td>8452.0000</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.9225</td>\n",
       "      <td>0.8745</td>\n",
       "      <td>0.9605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>PC4</td>\n",
       "      <td>-0.2325</td>\n",
       "      <td>-0.2994</td>\n",
       "      <td>-0.1600</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7926</td>\n",
       "      <td>0.7413</td>\n",
       "      <td>0.8522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>gyrA_p.Ser95Thr</td>\n",
       "      <td>-0.5224</td>\n",
       "      <td>-0.6298</td>\n",
       "      <td>-0.4004</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7585.0000</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>0.5931</td>\n",
       "      <td>0.5327</td>\n",
       "      <td>0.6701</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          orig_variant    coef  coef_LB  coef_UB   pval  BH_pval  \\\n",
       "0      gyrA_p.Asp94Gly  0.9749   0.9196   1.0283 0.0000   0.0000   \n",
       "1      gyrA_p.Ala90Val  0.5845   0.5296   0.6370 0.0000   0.0000   \n",
       "2      gyrA_p.Asp94Asn  0.4584   0.4125   0.5094 0.0000   0.0000   \n",
       "3      gyrA_p.Asp94Ala  0.4115   0.3586   0.4617 0.0000   0.0000   \n",
       "4      gyrA_p.Asp94Tyr  0.3677   0.3215   0.4117 0.0000   0.0000   \n",
       "5      gyrA_p.Ser91Pro  0.3581   0.3115   0.4047 0.0000   0.0000   \n",
       "6     gyrA_p.Gly668Asp  0.2678   0.2020   0.3297 0.0000   0.0000   \n",
       "7     gyrB_p.Glu501Asp  0.2435   0.1886   0.2989 0.0000   0.0000   \n",
       "8      gyrA_p.Asp94His  0.2419   0.1962   0.2932 0.0000   0.0000   \n",
       "9      gyrA_p.Asp89Asn  0.2002   0.1556   0.2470 0.0000   0.0000   \n",
       "10     gyrA_p.Gly88Cys  0.1965   0.1656   0.2146 0.0000   0.0000   \n",
       "11    gyrB_p.Ala504Val  0.1345   0.0788   0.1630 0.0000   0.0000   \n",
       "12    gyrB_p.Thr500Asn  0.1200   0.0792   0.1644 0.0000   0.0000   \n",
       "13    gyrB_p.Gly512Arg  0.1184   0.0514   0.1465 0.0000   0.0005   \n",
       "14  gyrB_c.-60_-59insG  0.0798  -0.0000   0.1003 0.0003   0.0138   \n",
       "15      gyrA_c.1440C>T -0.0069  -0.0104  -0.0023 0.0005   0.0195   \n",
       "16       gyrA_c.474C>T -0.0076  -0.0106  -0.0038 0.0000   0.0003   \n",
       "17     gyrB_p.Ile93Val -0.0083  -0.0107  -0.0049 0.0000   0.0000   \n",
       "18       gyrB_c.870G>A -0.0083  -0.0107  -0.0049 0.0000   0.0000   \n",
       "19    gyrB_p.Ile493Thr -0.0083  -0.0107  -0.0049 0.0000   0.0000   \n",
       "20       gyrA_c.984C>T -0.0105  -0.0139  -0.0059 0.0000   0.0000   \n",
       "21      gyrB_c.1167C>T -0.0110  -0.0141  -0.0063 0.0000   0.0000   \n",
       "22       gyrB_c.513G>A -0.0110  -0.0141  -0.0063 0.0000   0.0000   \n",
       "23      gyrA_c.1629C>T -0.0123  -0.0199  -0.0056 0.0007   0.0253   \n",
       "24    gyrA_p.Gly729Arg -0.0131  -0.0194  -0.0056 0.0001   0.0063   \n",
       "25    gyrB_p.Thr393Ser -0.0131  -0.0194  -0.0056 0.0001   0.0063   \n",
       "26       gyrB_c.291C>T -0.0132  -0.0205  -0.0036 0.0011   0.0377   \n",
       "27    gyrA_p.Leu296Pro -0.0132  -0.0206  -0.0036 0.0015   0.0453   \n",
       "28    gyrB_p.Ala403Ser -0.0150  -0.0182  -0.0103 0.0000   0.0000   \n",
       "29    gyrA_p.Ala456Val -0.0168  -0.0274  -0.0057 0.0015   0.0455   \n",
       "30    gyrB_p.Thr378Ser -0.0176  -0.0267  -0.0060 0.0009   0.0300   \n",
       "31     gyrB_p.Ile84Val -0.0189  -0.0276  -0.0064 0.0004   0.0186   \n",
       "32    gyrA_p.Phe152Leu -0.0193  -0.0288  -0.0000 0.0015   0.0454   \n",
       "33    gyrB_p.Pro254Leu -0.0198  -0.0287  -0.0000 0.0014   0.0456   \n",
       "34       gyrA_c.195C>G -0.0200  -0.0295  -0.0000 0.0013   0.0442   \n",
       "35    gyrA_p.Thr836Arg -0.0205  -0.0295  -0.0000 0.0014   0.0437   \n",
       "36     gyrB_p.Met58Ile -0.0217  -0.0316  -0.0083 0.0005   0.0195   \n",
       "37      gyrA_c.2295G>T -0.0219  -0.0301  -0.0101 0.0000   0.0007   \n",
       "38       gyrB_c.285C>A -0.0220  -0.0314  -0.0082 0.0002   0.0095   \n",
       "39    gyrB_p.Ala593Val -0.0229  -0.0326  -0.0090 0.0001   0.0063   \n",
       "40        gyrB_c.-1C>G -0.0231  -0.0319  -0.0110 0.0000   0.0009   \n",
       "41      gyrB_c.1119C>T -0.0232  -0.0331  -0.0054 0.0005   0.0201   \n",
       "42      gyrA_c.1494C>T -0.0233  -0.0348  -0.0000 0.0012   0.0416   \n",
       "43      gyrA_c.1113C>T -0.0249  -0.0339  -0.0136 0.0000   0.0001   \n",
       "44    gyrB_p.Asp461Asn -0.0250  -0.0343  -0.0118 0.0000   0.0003   \n",
       "45    gyrA_p.Pro472Ser -0.0270  -0.0362  -0.0134 0.0000   0.0003   \n",
       "46    gyrA_p.Arg254Cys -0.0293  -0.0388  -0.0143 0.0000   0.0000   \n",
       "47    gyrA_p.Leu711Met -0.0294  -0.0394  -0.0152 0.0000   0.0001   \n",
       "48      gyrA_c.2142G>A -0.0299  -0.0415  -0.0186 0.0000   0.0000   \n",
       "49      gyrA_c.1746C>T -0.0367  -0.0496  -0.0224 0.0000   0.0000   \n",
       "50      gyrA_p.Pro8Ala -0.0384  -0.0489  -0.0253 0.0000   0.0000   \n",
       "51      gyrA_c.1677C>T -0.0446  -0.0566  -0.0309 0.0000   0.0000   \n",
       "52    gyrA_p.Arg252Leu -0.0545  -0.0747  -0.0315 0.0000   0.0000   \n",
       "53    gyrB_p.Arg575His -0.0555  -0.0830  -0.0201 0.0008   0.0292   \n",
       "54      gyrA_c.2457C>T -0.0564  -0.0709  -0.0291 0.0000   0.0000   \n",
       "55      gyrA_c.2487C>T -0.0637  -0.0856  -0.0109 0.0007   0.0269   \n",
       "56    gyrA_p.Ala384Val -0.0807  -0.1340  -0.0403 0.0005   0.0185   \n",
       "57                 PC4 -0.2325  -0.2994  -0.1600 0.0000   0.0000   \n",
       "58     gyrA_p.Ser95Thr -0.5224  -0.6298  -0.4004 0.0000   0.0000   \n",
       "\n",
       "    Bonferroni_pval  genome_index        confidence_WHO_2021  Odds_Ratio  \\\n",
       "0            0.0000     7582.0000               1) Assoc w R      2.6508   \n",
       "1            0.0000     7570.0000               1) Assoc w R      1.7941   \n",
       "2            0.0000     7581.0000               1) Assoc w R      1.5815   \n",
       "3            0.0000     7582.0000               1) Assoc w R      1.5090   \n",
       "4            0.0000     7581.0000               1) Assoc w R      1.4444   \n",
       "5            0.0000     7572.0000               1) Assoc w R      1.4307   \n",
       "6            0.0000     9304.0000           5) Not assoc w R      1.3071   \n",
       "7            0.0000     6742.0000               1) Assoc w R      1.2756   \n",
       "8            0.0000     7581.0000               1) Assoc w R      1.2737   \n",
       "9            0.0000     7566.0000  3) Uncertain significance      1.2217   \n",
       "10           0.0000     7563.0000               1) Assoc w R      1.2171   \n",
       "11           0.0001     6750.0000     2) Assoc w R - Interim      1.1440   \n",
       "12           0.0001     6738.0000  3) Uncertain significance      1.1275   \n",
       "13           0.0160     6773.0000  3) Uncertain significance      1.1256   \n",
       "14           0.5253           NaN                        NaN      1.0831   \n",
       "15           0.7996           NaN                        NaN      0.9932   \n",
       "16           0.0094           NaN                        NaN      0.9924   \n",
       "17           0.0001     5516.0000  3) Uncertain significance      0.9918   \n",
       "18           0.0001           NaN                        NaN      0.9918   \n",
       "19           0.0001     6717.0000  3) Uncertain significance      0.9918   \n",
       "20           0.0004           NaN                        NaN      0.9895   \n",
       "21           0.0000           NaN                        NaN      0.9890   \n",
       "22           0.0000           NaN                        NaN      0.9890   \n",
       "23           1.0000           NaN                        NaN      0.9878   \n",
       "24           0.2277     9486.0000  3) Uncertain significance      0.9870   \n",
       "25           0.2277     6417.0000  3) Uncertain significance      0.9870   \n",
       "26           1.0000           NaN                        NaN      0.9869   \n",
       "27           1.0000     8188.0000  3) Uncertain significance      0.9869   \n",
       "28           0.0000     6446.0000  3) Uncertain significance      0.9851   \n",
       "29           1.0000     8668.0000  3) Uncertain significance      0.9833   \n",
       "30           1.0000     6372.0000  3) Uncertain significance      0.9826   \n",
       "31           0.7253     5489.0000  3) Uncertain significance      0.9813   \n",
       "32           1.0000     7757.0000  3) Uncertain significance      0.9809   \n",
       "33           1.0000     6000.0000  3) Uncertain significance      0.9804   \n",
       "34           1.0000           NaN                        NaN      0.9802   \n",
       "35           1.0000     9808.0000  3) Uncertain significance      0.9797   \n",
       "36           0.8184           NaN                        NaN      0.9785   \n",
       "37           0.0215           NaN                        NaN      0.9784   \n",
       "38           0.3499           NaN                        NaN      0.9782   \n",
       "39           0.2217     7017.0000  3) Uncertain significance      0.9773   \n",
       "40           0.0314     5239.0000  3) Uncertain significance      0.9772   \n",
       "41           0.8660           NaN                        NaN      0.9771   \n",
       "42           1.0000           NaN                        NaN      0.9769   \n",
       "43           0.0031           NaN                        NaN      0.9754   \n",
       "44           0.0098     6620.0000     2) Assoc w R - Interim      0.9754   \n",
       "45           0.0078     8715.0000  3) Uncertain significance      0.9733   \n",
       "46           0.0013     8061.0000  3) Uncertain significance      0.9711   \n",
       "47           0.0040     9432.0000  3) Uncertain significance      0.9710   \n",
       "48           0.0003           NaN                        NaN      0.9705   \n",
       "49           0.0001           NaN                        NaN      0.9640   \n",
       "50           0.0000     7323.0000  3) Uncertain significance      0.9624   \n",
       "51           0.0000           NaN                        NaN      0.9563   \n",
       "52           0.0007     8056.0000  3) Uncertain significance      0.9469   \n",
       "53           1.0000     6963.0000  3) Uncertain significance      0.9460   \n",
       "54           0.0009           NaN                        NaN      0.9451   \n",
       "55           1.0000           NaN                        NaN      0.9383   \n",
       "56           0.7417     8452.0000           5) Not assoc w R      0.9225   \n",
       "57           0.0000           NaN                        NaN      0.7926   \n",
       "58           0.0000     7585.0000           5) Not assoc w R      0.5931   \n",
       "\n",
       "    OR_LB  OR_UB  \n",
       "0  2.5083 2.7963  \n",
       "1  1.6982 1.8907  \n",
       "2  1.5107 1.6643  \n",
       "3  1.4313 1.5867  \n",
       "4  1.3792 1.5094  \n",
       "5  1.3655 1.4988  \n",
       "6  1.2239 1.3905  \n",
       "7  1.2075 1.3483  \n",
       "8  1.2168 1.3407  \n",
       "9  1.1684 1.2802  \n",
       "10 1.1801 1.2394  \n",
       "11 1.0820 1.1770  \n",
       "12 1.0825 1.1787  \n",
       "13 1.0527 1.1578  \n",
       "14 1.0000 1.1055  \n",
       "15 0.9897 0.9977  \n",
       "16 0.9895 0.9962  \n",
       "17 0.9893 0.9951  \n",
       "18 0.9893 0.9951  \n",
       "19 0.9893 0.9951  \n",
       "20 0.9862 0.9941  \n",
       "21 0.9860 0.9937  \n",
       "22 0.9860 0.9937  \n",
       "23 0.9803 0.9944  \n",
       "24 0.9808 0.9944  \n",
       "25 0.9808 0.9944  \n",
       "26 0.9797 0.9964  \n",
       "27 0.9796 0.9964  \n",
       "28 0.9820 0.9898  \n",
       "29 0.9729 0.9943  \n",
       "30 0.9737 0.9940  \n",
       "31 0.9728 0.9936  \n",
       "32 0.9716 1.0000  \n",
       "33 0.9717 1.0000  \n",
       "34 0.9709 1.0000  \n",
       "35 0.9710 1.0000  \n",
       "36 0.9689 0.9917  \n",
       "37 0.9704 0.9900  \n",
       "38 0.9691 0.9918  \n",
       "39 0.9679 0.9910  \n",
       "40 0.9687 0.9890  \n",
       "41 0.9675 0.9947  \n",
       "42 0.9658 1.0000  \n",
       "43 0.9667 0.9865  \n",
       "44 0.9663 0.9882  \n",
       "45 0.9645 0.9867  \n",
       "46 0.9619 0.9858  \n",
       "47 0.9614 0.9849  \n",
       "48 0.9594 0.9815  \n",
       "49 0.9516 0.9778  \n",
       "50 0.9523 0.9750  \n",
       "51 0.9449 0.9696  \n",
       "52 0.9280 0.9689  \n",
       "53 0.9203 0.9801  \n",
       "54 0.9316 0.9713  \n",
       "55 0.9179 0.9891  \n",
       "56 0.8745 0.9605  \n",
       "57 0.7413 0.8522  \n",
       "58 0.5327 0.6701  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4183190-ff95-42ad-bc47-e239c23854ba",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
